// thirdweb/dist/cjs/react/web/wallets/in-app/CountrySelector.js
"use strict";
"use client";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CountrySelector = CountrySelector;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_query_1 = require("@tanstack/react-query");
const react_1 = require("react");
const CustomThemeProvider_js_1 = require("../../../core/design-system/CustomThemeProvider.js");
const index_js_1 = require("../../../core/design-system/index.js");
const elements_js_1 = require("../../ui/design-system/elements.js");
function CountrySelector({ countryCode, setCountryCode, }) {
    const selectRef = (0, react_1.useRef)(null);
    const { data: supportedCountries } = (0, react_query_1.useQuery)({
        queryKey: ["supported-sms-countries"],
        queryFn: async () => {
            const { supportedSmsCountries } = await Promise.resolve().then(() => require("./supported-sms-countries.js"));
            return supportedSmsCountries;
        },
    });
    const supportedCountriesForSms = supportedCountries ?? [
        {
            countryIsoCode: "US",
            countryName: "United States",
            phoneNumberCode: 1,
        },
    ];
    return ((0, jsx_runtime_1.jsx)(jsx_runtime_1.Fragment, { children: (0, jsx_runtime_1.jsxs)(Select, { ref: selectRef, name: "countries", id: "countries", value: countryCode, onChange: (e) => {
                setCountryCode(e.target.value);
            }, style: {
                padding: `${index_js_1.spacing.sm} ${index_js_1.spacing.md}`,
            }, children: [(0, jsx_runtime_1.jsx)(Option, { style: {
                        display: "none",
                    }, value: countryCode, children: countryCode }), supportedCountriesForSms.map((country) => {
                    return ((0, jsx_runtime_1.jsxs)(Option, { value: `${country.countryIsoCode} +${country.phoneNumberCode}`, children: [country.countryName, " +", country.phoneNumberCode] }, country.countryIsoCode));
                })] }) }));
}
const Option = /* @__PURE__ */ (0, elements_js_1.StyledOption)(() => {
    const theme = (0, CustomThemeProvider_js_1.useCustomTheme)();
    return {
        color: theme.colors.primaryText,
        background: theme.colors.modalBg,
        transition: "background 0.3s ease",
        "&:hover": {
            background: theme.colors.tertiaryBg,
        },
    };
});
const Select = /* @__PURE__ */ (0, elements_js_1.StyledSelect)((_) => {
    const theme = (0, CustomThemeProvider_js_1.useCustomTheme)();
    return {
        display: "block",
        padding: index_js_1.spacing.sm,
        boxSizing: "border-box",
        outline: "none",
        border: "none",
        borderRadius: index_js_1.radius.lg,
        color: theme.colors.primaryText,
        WebkitAppearance: "none",
        appearance: "none",
        cursor: "pointer",
        background: "transparent",
        "&::placeholder": {
            color: theme.colors.secondaryText,
        },
        "&[disabled]": {
            cursor: "not-allowed",
        },
        minWidth: "0px",
        maxWidth: "90px",
        textOverflow: "ellipsis",
        overflow: "hidden",
        whiteSpace: "nowrap",
    };
});
