// thirdweb/dist/esm/extensions/modules/__generated__/OwnableRoles/write/renounceRoles.js
import { prepareContractCall } from "../../../../../transaction/prepare-contract-call.js";
import { encodeAbiParameters } from "../../../../../utils/abi/encodeAbiParameters.js";
import { once } from "../../../../../utils/promise/once.js";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0x183a4f6e";
const FN_INPUTS = [
    {
        type: "uint256",
        name: "roles",
    },
];
const FN_OUTPUTS = [];
/**
 * Checks if the `renounceRoles` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `renounceRoles` method is supported.
 * @extension MODULES
 * @example
 * ```ts
 * import { isRenounceRolesSupported } from "thirdweb/extensions/modules";
 *
 * const supported = isRenounceRolesSupported(["0x..."]);
 * ```
 */
export function isRenounceRolesSupported(availableSelectors) {
    return detectMethod({
        availableSelectors,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "renounceRoles" function.
 * @param options - The options for the renounceRoles function.
 * @returns The encoded ABI parameters.
 * @extension MODULES
 * @example
 * ```ts
 * import { encodeRenounceRolesParams } from "thirdweb/extensions/modules";
 * const result = encodeRenounceRolesParams({
 *  roles: ...,
 * });
 * ```
 */
export function encodeRenounceRolesParams(options) {
    return encodeAbiParameters(FN_INPUTS, [options.roles]);
}
/**
 * Encodes the "renounceRoles" function into a Hex string with its parameters.
 * @param options - The options for the renounceRoles function.
 * @returns The encoded hexadecimal string.
 * @extension MODULES
 * @example
 * ```ts
 * import { encodeRenounceRoles } from "thirdweb/extensions/modules";
 * const result = encodeRenounceRoles({
 *  roles: ...,
 * });
 * ```
 */
export function encodeRenounceRoles(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (FN_SELECTOR +
        encodeRenounceRolesParams(options).slice(2));
}
/**
 * Prepares a transaction to call the "renounceRoles" function on the contract.
 * @param options - The options for the "renounceRoles" function.
 * @returns A prepared transaction object.
 * @extension MODULES
 * @example
 * ```ts
 * import { sendTransaction } from "thirdweb";
 * import { renounceRoles } from "thirdweb/extensions/modules";
 *
 * const transaction = renounceRoles({
 *  contract,
 *  roles: ...,
 *  overrides: {
 *    ...
 *  }
 * });
 *
 * // Send the transaction
 * await sendTransaction({ transaction, account });
 * ```
 */
export function renounceRoles(options) {
    const asyncOptions = once(async () => {
        return "asyncParams" in options ? await options.asyncParams() : options;
    });
    return prepareContractCall({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: async () => {
            const resolvedOptions = await asyncOptions();
            return [resolvedOptions.roles];
        },
        value: async () => (await asyncOptions()).overrides?.value,
        accessList: async () => (await asyncOptions()).overrides?.accessList,
        gas: async () => (await asyncOptions()).overrides?.gas,
        gasPrice: async () => (await asyncOptions()).overrides?.gasPrice,
        maxFeePerGas: async () => (await asyncOptions()).overrides?.maxFeePerGas,
        maxPriorityFeePerGas: async () => (await asyncOptions()).overrides?.maxPriorityFeePerGas,
        nonce: async () => (await asyncOptions()).overrides?.nonce,
        extraGas: async () => (await asyncOptions()).overrides?.extraGas,
        erc20Value: async () => (await asyncOptions()).overrides?.erc20Value,
    });
}
