// thirdweb/dist/esm/extensions/farcaster/__generated__/IKeyRegistry/read/keysOf.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { encodeAbiParameters } from "../../../../../utils/abi/encodeAbiParameters.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0x1f64222f";
const FN_INPUTS = [
    {
        type: "uint256",
        name: "fid",
    },
    {
        type: "uint8",
        name: "state",
    },
];
const FN_OUTPUTS = [
    {
        type: "bytes[]",
    },
];
/**
 * Checks if the `keysOf` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `keysOf` method is supported.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { isKeysOfSupported } from "thirdweb/extensions/farcaster";
 * const supported = isKeysOfSupported(["0x..."]);
 * ```
 */
export function isKeysOfSupported(availableSelectors) {
    return detectMethod({
        availableSelectors,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "keysOf" function.
 * @param options - The options for the keysOf function.
 * @returns The encoded ABI parameters.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { encodeKeysOfParams } from "thirdweb/extensions/farcaster";
 * const result = encodeKeysOfParams({
 *  fid: ...,
 *  state: ...,
 * });
 * ```
 */
export function encodeKeysOfParams(options) {
    return encodeAbiParameters(FN_INPUTS, [options.fid, options.state]);
}
/**
 * Encodes the "keysOf" function into a Hex string with its parameters.
 * @param options - The options for the keysOf function.
 * @returns The encoded hexadecimal string.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { encodeKeysOf } from "thirdweb/extensions/farcaster";
 * const result = encodeKeysOf({
 *  fid: ...,
 *  state: ...,
 * });
 * ```
 */
export function encodeKeysOf(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (FN_SELECTOR +
        encodeKeysOfParams(options).slice(2));
}
/**
 * Decodes the result of the keysOf function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { decodeKeysOfResult } from "thirdweb/extensions/farcaster";
 * const result = decodeKeysOfResultResult("...");
 * ```
 */
export function decodeKeysOfResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "keysOf" function on the contract.
 * @param options - The options for the keysOf function.
 * @returns The parsed result of the function call.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { keysOf } from "thirdweb/extensions/farcaster";
 *
 * const result = await keysOf({
 *  contract,
 *  fid: ...,
 *  state: ...,
 * });
 *
 * ```
 */
export async function keysOf(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.fid, options.state],
    });
}
