// thirdweb/dist/esm/extensions/farcaster/__generated__/IKeyRegistry/read/maxKeysPerFid.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0xe33acf38";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `maxKeysPerFid` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `maxKeysPerFid` method is supported.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { isMaxKeysPerFidSupported } from "thirdweb/extensions/farcaster";
 * const supported = isMaxKeysPerFidSupported(["0x..."]);
 * ```
 */
export function isMaxKeysPerFidSupported(availableSelectors) {
    return detectMethod({
        availableSelectors,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the maxKeysPerFid function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { decodeMaxKeysPerFidResult } from "thirdweb/extensions/farcaster";
 * const result = decodeMaxKeysPerFidResultResult("...");
 * ```
 */
export function decodeMaxKeysPerFidResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "maxKeysPerFid" function on the contract.
 * @param options - The options for the maxKeysPerFid function.
 * @returns The parsed result of the function call.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { maxKeysPerFid } from "thirdweb/extensions/farcaster";
 *
 * const result = await maxKeysPerFid({
 *  contract,
 * });
 *
 * ```
 */
export async function maxKeysPerFid(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
