// thirdweb/dist/cjs/extensions/farcaster/__generated__/IKeyRegistry/read/keyDataOf.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isKeyDataOfSupported = isKeyDataOfSupported;
exports.encodeKeyDataOfParams = encodeKeyDataOfParams;
exports.encodeKeyDataOf = encodeKeyDataOf;
exports.decodeKeyDataOfResult = decodeKeyDataOfResult;
exports.keyDataOf = keyDataOf;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xac34cc5a";
const FN_INPUTS = [
    {
        type: "uint256",
        name: "fid",
    },
    {
        type: "bytes",
        name: "key",
    },
];
const FN_OUTPUTS = [
    {
        type: "address",
    },
];
/**
 * Checks if the `keyDataOf` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `keyDataOf` method is supported.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { isKeyDataOfSupported } from "thirdweb/extensions/farcaster";
 * const supported = isKeyDataOfSupported(["0x..."]);
 * ```
 */
function isKeyDataOfSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "keyDataOf" function.
 * @param options - The options for the keyDataOf function.
 * @returns The encoded ABI parameters.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { encodeKeyDataOfParams } from "thirdweb/extensions/farcaster";
 * const result = encodeKeyDataOfParams({
 *  fid: ...,
 *  key: ...,
 * });
 * ```
 */
function encodeKeyDataOfParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.fid, options.key]);
}
/**
 * Encodes the "keyDataOf" function into a Hex string with its parameters.
 * @param options - The options for the keyDataOf function.
 * @returns The encoded hexadecimal string.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { encodeKeyDataOf } from "thirdweb/extensions/farcaster";
 * const result = encodeKeyDataOf({
 *  fid: ...,
 *  key: ...,
 * });
 * ```
 */
function encodeKeyDataOf(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeKeyDataOfParams(options).slice(2));
}
/**
 * Decodes the result of the keyDataOf function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { decodeKeyDataOfResult } from "thirdweb/extensions/farcaster";
 * const result = decodeKeyDataOfResultResult("...");
 * ```
 */
function decodeKeyDataOfResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "keyDataOf" function on the contract.
 * @param options - The options for the keyDataOf function.
 * @returns The parsed result of the function call.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { keyDataOf } from "thirdweb/extensions/farcaster";
 *
 * const result = await keyDataOf({
 *  contract,
 *  fid: ...,
 *  key: ...,
 * });
 *
 * ```
 */
async function keyDataOf(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.fid, options.key],
    });
}
