// thirdweb/dist/cjs/extensions/erc721/__generated__/IDelayedReveal/read/encryptedData.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isEncryptedDataSupported = isEncryptedDataSupported;
exports.encodeEncryptedDataParams = encodeEncryptedDataParams;
exports.encodeEncryptedData = encodeEncryptedData;
exports.decodeEncryptedDataResult = decodeEncryptedDataResult;
exports.encryptedData = encryptedData;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xa05112fc";
const FN_INPUTS = [
    {
        type: "uint256",
        name: "index",
    },
];
const FN_OUTPUTS = [
    {
        type: "bytes",
    },
];
/**
 * Checks if the `encryptedData` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `encryptedData` method is supported.
 * @extension ERC721
 * @example
 * ```ts
 * import { isEncryptedDataSupported } from "thirdweb/extensions/erc721";
 * const supported = isEncryptedDataSupported(["0x..."]);
 * ```
 */
function isEncryptedDataSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "encryptedData" function.
 * @param options - The options for the encryptedData function.
 * @returns The encoded ABI parameters.
 * @extension ERC721
 * @example
 * ```ts
 * import { encodeEncryptedDataParams } from "thirdweb/extensions/erc721";
 * const result = encodeEncryptedDataParams({
 *  index: ...,
 * });
 * ```
 */
function encodeEncryptedDataParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.index]);
}
/**
 * Encodes the "encryptedData" function into a Hex string with its parameters.
 * @param options - The options for the encryptedData function.
 * @returns The encoded hexadecimal string.
 * @extension ERC721
 * @example
 * ```ts
 * import { encodeEncryptedData } from "thirdweb/extensions/erc721";
 * const result = encodeEncryptedData({
 *  index: ...,
 * });
 * ```
 */
function encodeEncryptedData(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeEncryptedDataParams(options).slice(2));
}
/**
 * Decodes the result of the encryptedData function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC721
 * @example
 * ```ts
 * import { decodeEncryptedDataResult } from "thirdweb/extensions/erc721";
 * const result = decodeEncryptedDataResultResult("...");
 * ```
 */
function decodeEncryptedDataResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "encryptedData" function on the contract.
 * @param options - The options for the encryptedData function.
 * @returns The parsed result of the function call.
 * @extension ERC721
 * @example
 * ```ts
 * import { encryptedData } from "thirdweb/extensions/erc721";
 *
 * const result = await encryptedData({
 *  contract,
 *  index: ...,
 * });
 *
 * ```
 */
async function encryptedData(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.index],
    });
}
