// thirdweb/dist/esm/extensions/modules/__generated__/IModularCore/read/getSupportedCallbackFunctions.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0xf147db8a";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "tuple[]",
        components: [
            {
                type: "bytes4",
                name: "selector",
            },
            {
                type: "uint8",
                name: "mode",
            },
        ],
    },
];
/**
 * Checks if the `getSupportedCallbackFunctions` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `getSupportedCallbackFunctions` method is supported.
 * @extension MODULES
 * @example
 * ```ts
 * import { isGetSupportedCallbackFunctionsSupported } from "thirdweb/extensions/modules";
 * const supported = isGetSupportedCallbackFunctionsSupported(["0x..."]);
 * ```
 */
export function isGetSupportedCallbackFunctionsSupported(availableSelectors) {
    return detectMethod({
        availableSelectors,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the getSupportedCallbackFunctions function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension MODULES
 * @example
 * ```ts
 * import { decodeGetSupportedCallbackFunctionsResult } from "thirdweb/extensions/modules";
 * const result = decodeGetSupportedCallbackFunctionsResultResult("...");
 * ```
 */
export function decodeGetSupportedCallbackFunctionsResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getSupportedCallbackFunctions" function on the contract.
 * @param options - The options for the getSupportedCallbackFunctions function.
 * @returns The parsed result of the function call.
 * @extension MODULES
 * @example
 * ```ts
 * import { getSupportedCallbackFunctions } from "thirdweb/extensions/modules";
 *
 * const result = await getSupportedCallbackFunctions({
 *  contract,
 * });
 *
 * ```
 */
export async function getSupportedCallbackFunctions(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
