// thirdweb/dist/cjs/stories/ConnectButton/hideButtons.stories.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HideSendReceiveBuy = exports.HideReceiveBuy = exports.HideSendBuy = exports.HideSendReceive = exports.HideBuy = exports.HideReceive = exports.HideSend = exports.Default = void 0;
const ConnectButton_js_1 = require("../../react/web/ui/ConnectWallet/ConnectButton.js");
const utils_js_1 = require("../utils.js");
const meta = {
    title: "Connect/ConnectButton/hide buttons",
    component: ConnectButton_js_1.ConnectButton,
    parameters: {
        layout: "centered",
    },
    args: {
        client: utils_js_1.storyClient,
    },
};
exports.Default = {
    args: {},
};
exports.HideSend = {
    args: {
        detailsModal: {
            hideSendFunds: true,
        },
    },
};
exports.HideReceive = {
    args: {
        detailsModal: {
            hideReceiveFunds: true,
        },
    },
};
exports.HideBuy = {
    args: {
        detailsModal: {
            hideBuyFunds: true,
        },
    },
};
exports.HideSendReceive = {
    args: {
        detailsModal: {
            hideSendFunds: true,
            hideReceiveFunds: true,
        },
    },
};
exports.HideSendBuy = {
    args: {
        detailsModal: {
            hideSendFunds: true,
            hideBuyFunds: true,
        },
    },
};
exports.HideReceiveBuy = {
    args: {
        detailsModal: {
            hideReceiveFunds: true,
            hideBuyFunds: true,
        },
    },
};
exports.HideSendReceiveBuy = {
    args: {
        detailsModal: {
            hideSendFunds: true,
            hideReceiveFunds: true,
            hideBuyFunds: true,
        },
    },
};
exports.default = meta;
