// thirdweb/dist/cjs/wallets/smart/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getEntryPointVersion = exports.getDefaultBundlerUrl = exports.getDefaultAccountFactory = exports.MANAGED_ACCOUNT_GAS_BUFFER = exports.ENTRYPOINT_ADDRESS_v0_7 = exports.ENTRYPOINT_ADDRESS_v0_6 = exports.DEFAULT_ACCOUNT_FACTORY_V0_7 = exports.DEFAULT_ACCOUNT_FACTORY_V0_6 = exports.DUMMY_SIGNATURE = exports.DEBUG = void 0;
const address_js_1 = require("../../../utils/address.js");
const domains_js_1 = require("../../../utils/domains.js");
// dev only
exports.DEBUG = false;
exports.DUMMY_SIGNATURE = "0xfffffffffffffffffffffffffffffff0000000000000000000000000000000007aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa1c";
exports.DEFAULT_ACCOUNT_FACTORY_V0_6 = "0x85e23b94e7F5E9cC1fF78BCe78cfb15B81f0DF00";
exports.DEFAULT_ACCOUNT_FACTORY_V0_7 = "0x4be0ddfebca9a5a4a617dee4dece99e7c862dceb";
exports.ENTRYPOINT_ADDRESS_v0_6 = "0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789"; // v0.6
exports.ENTRYPOINT_ADDRESS_v0_7 = "0x0000000071727De22E5E9d8BAf0edAc6f37da032"; // v0.7
exports.MANAGED_ACCOUNT_GAS_BUFFER = 50000n;
/*
 * @internal
 */
const getDefaultAccountFactory = (entryPointAddress) => {
    const version = (0, exports.getEntryPointVersion)(entryPointAddress || exports.ENTRYPOINT_ADDRESS_v0_6);
    if (version === "v0.7") {
        return exports.DEFAULT_ACCOUNT_FACTORY_V0_7;
    }
    return exports.DEFAULT_ACCOUNT_FACTORY_V0_6;
};
exports.getDefaultAccountFactory = getDefaultAccountFactory;
/**
 * @internal
 */
const getDefaultBundlerUrl = (chain, version) => {
    const domain = (0, domains_js_1.getThirdwebDomains)().bundler;
    if (domain.startsWith("localhost:")) {
        return `http://${domain}/${version ?? "v1"}?chain=${chain.id}`;
    }
    return `https://${chain.id}.${domain}/${version ?? "v1"}`;
};
exports.getDefaultBundlerUrl = getDefaultBundlerUrl;
const getEntryPointVersion = (address) => {
    const checksummedAddress = (0, address_js_1.getAddress)(address);
    if (checksummedAddress === exports.ENTRYPOINT_ADDRESS_v0_6) {
        return "v0.6";
    }
    if (checksummedAddress === exports.ENTRYPOINT_ADDRESS_v0_7) {
        return "v0.7";
    }
    throw new Error("Unknown paymaster version");
};
exports.getEntryPointVersion = getEntryPointVersion;
