// thirdweb/dist/cjs/wallets/smart/lib/packUserOp.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPackedUserOperation = void 0;
exports.getInitCode = getInitCode;
exports.getAccountGasLimits = getAccountGasLimits;
exports.unpackAccountGasLimits = unpackAccountGasLimits;
exports.getGasLimits = getGasLimits;
exports.unpackGasLimits = unpackGasLimits;
exports.getPaymasterAndData = getPaymasterAndData;
const viem_1 = require("viem");
function getInitCode(unpackedUserOperation) {
    return unpackedUserOperation.factory
        ? (0, viem_1.concat)([
            unpackedUserOperation.factory,
            unpackedUserOperation.factoryData || "0x",
        ])
        : "0x";
}
function getAccountGasLimits(unpackedUserOperation) {
    return (0, viem_1.concat)([
        (0, viem_1.pad)((0, viem_1.toHex)(unpackedUserOperation.verificationGasLimit), {
            size: 16,
        }),
        (0, viem_1.pad)((0, viem_1.toHex)(unpackedUserOperation.callGasLimit), { size: 16 }),
    ]);
}
function unpackAccountGasLimits(accountGasLimits) {
    return {
        verificationGasLimit: BigInt((0, viem_1.slice)(accountGasLimits, 0, 16)),
        callGasLimit: BigInt((0, viem_1.slice)(accountGasLimits, 16)),
    };
}
function getGasLimits(unpackedUserOperation) {
    return (0, viem_1.concat)([
        (0, viem_1.pad)((0, viem_1.toHex)(unpackedUserOperation.maxPriorityFeePerGas), {
            size: 16,
        }),
        (0, viem_1.pad)((0, viem_1.toHex)(unpackedUserOperation.maxFeePerGas), { size: 16 }),
    ]);
}
function unpackGasLimits(gasLimits) {
    return {
        maxPriorityFeePerGas: BigInt((0, viem_1.slice)(gasLimits, 0, 16)),
        maxFeePerGas: BigInt((0, viem_1.slice)(gasLimits, 16)),
    };
}
function getPaymasterAndData(unpackedUserOperation) {
    return unpackedUserOperation.paymaster
        ? (0, viem_1.concat)([
            unpackedUserOperation.paymaster,
            (0, viem_1.pad)((0, viem_1.toHex)(unpackedUserOperation.paymasterVerificationGasLimit || BigInt(0)), {
                size: 16,
            }),
            (0, viem_1.pad)((0, viem_1.toHex)(unpackedUserOperation.paymasterPostOpGasLimit || BigInt(0)), {
                size: 16,
            }),
            unpackedUserOperation.paymasterData || "0x",
        ])
        : "0x";
}
const getPackedUserOperation = (userOperation) => {
    return {
        sender: userOperation.sender,
        nonce: userOperation.nonce,
        initCode: getInitCode(userOperation),
        callData: userOperation.callData,
        accountGasLimits: getAccountGasLimits(userOperation),
        preVerificationGas: userOperation.preVerificationGas,
        gasFees: getGasLimits(userOperation),
        paymasterAndData: getPaymasterAndData(userOperation),
        signature: userOperation.signature,
    };
};
exports.getPackedUserOperation = getPackedUserOperation;
