// thirdweb/dist/esm/extensions/vote/__generated__/Vote/write/relay.js
import { prepareContractCall } from "../../../../../transaction/prepare-contract-call.js";
import { encodeAbiParameters } from "../../../../../utils/abi/encodeAbiParameters.js";
import { once } from "../../../../../utils/promise/once.js";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0xc28bc2fa";
const FN_INPUTS = [
    {
        type: "address",
        name: "target",
    },
    {
        type: "uint256",
        name: "value",
    },
    {
        type: "bytes",
        name: "data",
    },
];
const FN_OUTPUTS = [];
/**
 * Checks if the `relay` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `relay` method is supported.
 * @extension VOTE
 * @example
 * ```ts
 * import { isRelaySupported } from "thirdweb/extensions/vote";
 *
 * const supported = isRelaySupported(["0x..."]);
 * ```
 */
export function isRelaySupported(availableSelectors) {
    return detectMethod({
        availableSelectors,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "relay" function.
 * @param options - The options for the relay function.
 * @returns The encoded ABI parameters.
 * @extension VOTE
 * @example
 * ```ts
 * import { encodeRelayParams } from "thirdweb/extensions/vote";
 * const result = encodeRelayParams({
 *  target: ...,
 *  value: ...,
 *  data: ...,
 * });
 * ```
 */
export function encodeRelayParams(options) {
    return encodeAbiParameters(FN_INPUTS, [
        options.target,
        options.value,
        options.data,
    ]);
}
/**
 * Encodes the "relay" function into a Hex string with its parameters.
 * @param options - The options for the relay function.
 * @returns The encoded hexadecimal string.
 * @extension VOTE
 * @example
 * ```ts
 * import { encodeRelay } from "thirdweb/extensions/vote";
 * const result = encodeRelay({
 *  target: ...,
 *  value: ...,
 *  data: ...,
 * });
 * ```
 */
export function encodeRelay(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (FN_SELECTOR +
        encodeRelayParams(options).slice(2));
}
/**
 * Prepares a transaction to call the "relay" function on the contract.
 * @param options - The options for the "relay" function.
 * @returns A prepared transaction object.
 * @extension VOTE
 * @example
 * ```ts
 * import { sendTransaction } from "thirdweb";
 * import { relay } from "thirdweb/extensions/vote";
 *
 * const transaction = relay({
 *  contract,
 *  target: ...,
 *  value: ...,
 *  data: ...,
 *  overrides: {
 *    ...
 *  }
 * });
 *
 * // Send the transaction
 * await sendTransaction({ transaction, account });
 * ```
 */
export function relay(options) {
    const asyncOptions = once(async () => {
        return "asyncParams" in options ? await options.asyncParams() : options;
    });
    return prepareContractCall({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: async () => {
            const resolvedOptions = await asyncOptions();
            return [
                resolvedOptions.target,
                resolvedOptions.value,
                resolvedOptions.data,
            ];
        },
        value: async () => (await asyncOptions()).overrides?.value,
        accessList: async () => (await asyncOptions()).overrides?.accessList,
        gas: async () => (await asyncOptions()).overrides?.gas,
        gasPrice: async () => (await asyncOptions()).overrides?.gasPrice,
        maxFeePerGas: async () => (await asyncOptions()).overrides?.maxFeePerGas,
        maxPriorityFeePerGas: async () => (await asyncOptions()).overrides?.maxPriorityFeePerGas,
        nonce: async () => (await asyncOptions()).overrides?.nonce,
        extraGas: async () => (await asyncOptions()).overrides?.extraGas,
        erc20Value: async () => (await asyncOptions()).overrides?.erc20Value,
    });
}
