// thirdweb/dist/cjs/extensions/modules/common/uninstallPublishedModule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.uninstallPublishedModule = uninstallPublishedModule;
const infra_js_1 = require("../../../contract/deployment/utils/infra.js");
const uninstallModule_js_1 = require("../__generated__/IModularCore/write/uninstallModule.js");
/**
 * Uninstall a published module
 * @param options - The options for uninstalling a published module
 * @returns A prepared transaction to send
 * @modules
 * @example
 * ```ts
 * import { uninstallPublishedModule } from "thirdweb/modules";
 * const transaction = uninstallPublishedModule({
 *  client,
 *  chain,
 *  contract,
 *  moduleName: "MyModule",
 *  publisherAddress: "0x...",
 * });
 * await sendTransaction({ transaction, account });
 * ```
 */
function uninstallPublishedModule(options) {
    const { contract, moduleName, publisherAddress, moduleData } = options;
    return (0, uninstallModule_js_1.uninstallModule)({
        contract,
        asyncParams: async () => {
            const deployedModule = await (0, infra_js_1.getDeployedInfraContract)({
                chain: contract.chain,
                client: contract.client,
                contractId: moduleName,
                publisher: publisherAddress,
            });
            return {
                moduleContract: deployedModule?.address,
                data: moduleData || "0x",
            };
        },
    });
}
