// thirdweb/dist/cjs/extensions/farcaster/__generated__/IIdGateway/read/REGISTER_TYPEHASH.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isREGISTER_TYPEHASHSupported = isREGISTER_TYPEHASHSupported;
exports.decodeREGISTER_TYPEHASHResult = decodeREGISTER_TYPEHASHResult;
exports.REGISTER_TYPEHASH = REGISTER_TYPEHASH;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x6a5306a3";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "bytes32",
    },
];
/**
 * Checks if the `REGISTER_TYPEHASH` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `REGISTER_TYPEHASH` method is supported.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { isREGISTER_TYPEHASHSupported } from "thirdweb/extensions/farcaster";
 * const supported = isREGISTER_TYPEHASHSupported(["0x..."]);
 * ```
 */
function isREGISTER_TYPEHASHSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the REGISTER_TYPEHASH function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { decodeREGISTER_TYPEHASHResult } from "thirdweb/extensions/farcaster";
 * const result = decodeREGISTER_TYPEHASHResultResult("...");
 * ```
 */
function decodeREGISTER_TYPEHASHResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "REGISTER_TYPEHASH" function on the contract.
 * @param options - The options for the REGISTER_TYPEHASH function.
 * @returns The parsed result of the function call.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { REGISTER_TYPEHASH } from "thirdweb/extensions/farcaster";
 *
 * const result = await REGISTER_TYPEHASH({
 *  contract,
 * });
 *
 * ```
 */
async function REGISTER_TYPEHASH(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
