// thirdweb/dist/cjs/extensions/airdrop/write/airdropERC721WithSignature.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.airdropERC721WithSignature = void 0;
exports.generateAirdropSignatureERC721 = generateAirdropSignatureERC721;
const date_js_1 = require("../../../utils/date.js");
const random_js_1 = require("../../../utils/random.js");
const airdropERC721WithSignature_js_1 = require("../__generated__/Airdrop/write/airdropERC721WithSignature.js");
/**
 * Airdrops ERC721 tokens to a list of recipients, with the request signed by admin
 * @param options - The transaction options.
 * @example
 * ```ts
 * import { airdropERC721WithSignature, generateAirdropSignatureERC721 } from "thirdweb/extensions/airdrop";
 * import { sendTransaction } from "thirdweb";
 *
 * const { req, signature } = await generateAirdropSignatureERC721(...)
 *
 * const transaction = airdropERC721WithSignature({
 *   contract,
 *   req,
 *   signature,
 * });
 * await sendTransaction({ transaction, account });
 * ```
 * @extension AIRDROP
 * @returns A promise that resolves to the transaction result.
 */
exports.airdropERC721WithSignature = airdropERC721WithSignature_js_1.airdropERC721WithSignature;
/**
 * Generates the req and signature for sending ERC721 airdrop.
 * @param options - The options for the airdrop.
 * @example
 * ```ts
 * import { airdropERC721WithSignature, generateAirdropSignatureERC721 } from "thirdweb/extensions/airdrop";
 *
 * // list of recipients and tokenIds to airdrop for each recipient
 * const contents = [
 *    { recipient: "0x...", tokenId: 0 },
 *    { recipient: "0x...", tokenId: 1 },
 *    { recipient: "0x...", tokenId: 2 },
 *  ];
 *
 * const { req, signature } = await generateAirdropSignatureERC721({
 *   account,
 *   contract,
 *   airdropRequest: {
 *     tokenAddress: "0x...", // address of the ERC721 token to airdrop
 *     contents
 *   },
 * });
 *
 * const transaction = airdropERC721WithSignature({
 *   contract,
 *   req,
 *   signature,
 * });
 * await sendTransaction({ transaction, account });
 * ```
 * @extension AIRDROP
 * @returns A promise that resolves to the req and signature.
 */
async function generateAirdropSignatureERC721(options) {
    const { airdropRequest, account, contract } = options;
    const tokenAddress = airdropRequest.tokenAddress;
    const contents = airdropRequest.contents;
    const uid = airdropRequest.uid || (await (0, random_js_1.randomBytesHex)());
    const endTime = airdropRequest.expirationTimestamp || (0, date_js_1.tenYearsFromNow)();
    const req = {
        uid,
        tokenAddress,
        expirationTimestamp: (0, date_js_1.dateToSeconds)(endTime),
        contents,
    };
    const signature = await account.signTypedData({
        domain: {
            name: "Airdrop",
            version: "1",
            chainId: contract.chain.id,
            verifyingContract: contract.address,
        },
        types: { AirdropRequestERC721, AirdropContentERC721 },
        primaryType: "AirdropRequestERC721",
        message: req,
    });
    return { req, signature };
}
const AirdropContentERC721 = [
    { name: "recipient", type: "address" },
    { name: "tokenId", type: "uint256" },
];
const AirdropRequestERC721 = [
    { name: "uid", type: "bytes32" },
    { name: "tokenAddress", type: "address" },
    { name: "expirationTimestamp", type: "uint256" },
    { name: "contents", type: "AirdropContentERC721[]" },
];
