// thirdweb/dist/cjs/wallets/create-wallet.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createWallet = createWallet;
exports.walletConnect = walletConnect;
const track_js_1 = require("../analytics/track.js");
const utils_js_1 = require("../chains/utils.js");
const webStorage_js_1 = require("../utils/storage/webStorage.js");
const isMobile_js_1 = require("../utils/web/isMobile.js");
const openWindow_js_1 = require("../utils/web/openWindow.js");
const coinbase_wallet_js_1 = require("./coinbase/coinbase-wallet.js");
const coinbaseWebSDK_js_1 = require("./coinbase/coinbaseWebSDK.js");
const constants_js_1 = require("./constants.js");
const is_ecosystem_wallet_js_1 = require("./ecosystem/is-ecosystem-wallet.js");
const ecosystem_js_1 = require("./in-app/web/ecosystem.js");
const in_app_js_1 = require("./in-app/web/in-app.js");
const smart_wallet_js_1 = require("./smart/smart-wallet.js");
const wallet_emitter_js_1 = require("./wallet-emitter.js");
// TODO: figure out how to define the type without tuple args type and using function overloads
/**
 * Creates a wallet based on the provided ID and arguments.
 *
 * - Supports 350+ wallets
 * - Handles both injected browser wallets and WalletConnect sessions
 *
 * [View all available wallets](https://portal.thirdweb.com/typescript/v5/supported-wallets)
 *
 * @param args - The arguments for creating the wallet.
 * @param args.id - The ID of the wallet to create, this will be autocompleted by your IDE.
 * [View all available wallets](https://portal.thirdweb.com/typescript/v5/supported-wallets)
 * @param args.createOptions - The options for creating the wallet.
 * The arguments are different for each wallet type.
 * Refer to the [WalletCreationOptions](https://portal.thirdweb.com/references/typescript/v5/WalletCreationOptions) type for more details.
 * @returns - The created wallet.
 * @example
 *
 * ## Connecting the wallet
 *
 * Once created, you can connect the wallet to your app by calling the `connect` method.
 *
 * The `connect` method returns a promise that resolves to the connected account.
 *
 * Each wallet type can have different connect options. [View the different connect options](https://portal.thirdweb.com/references/typescript/v5/WalletConnectionOption)
 *
 * ## Connecting to an injected wallet
 *
 * ```ts
 * import { createWallet } from "thirdweb/wallets";
 *
 * const metamaskWallet = createWallet("io.metamask");
 *
 * const account = await metamaskWallet.connect({
 *  client,
 * });
 * ```
 *
 * You can check if a wallet is installed by calling the [injectedProvider](https://portal.thirdweb.com/references/typescript/v5/injectedProvider) method.
 *
 * ## Connecting via WalletConnect modal
 *
 * ```ts
 * import { createWallet } from "thirdweb/wallets";
 *
 * const metamaskWallet = createWallet("io.metamask");
 *
 * await metamask.connect({
 *   client,
 *   walletConnect: {
 *     projectId: "YOUR_PROJECT_ID",
 *     showQrModal: true,
 *     appMetadata: {
 *       name: "My App",
 *       url: "https://my-app.com",
 *       description: "my app description",
 *       logoUrl: "https://path/to/my-app/logo.svg",
 *     },
 *   },
 * });
 * ```
 * [View ConnectWallet connection options](https://portal.thirdweb.com/references/typescript/v5/WCConnectOptions)
 *
 * ## Connecting with coinbase wallet
 *
 * ```ts
 * import { createWallet } from "thirdweb/wallets";
 *
 * const cbWallet = createWallet("com.coinbase.wallet", {
 *   appMetadata: {
 *     name: "My App",
 *     url: "https://my-app.com",
 *     description: "my app description",
 *     logoUrl: "https://path/to/my-app/logo.svg",
 *   },
 *   walletConfig: {
 *     // options: 'all' | 'smartWalletOnly' | 'eoaOnly'
 *     options: 'all',
 *   },
 * });
 *
 * const account = await cbWallet.connect({
 *  client,
 * });
 * ```
 *
 * [View Coinbase wallet creation options](https://portal.thirdweb.com/references/typescript/v5/CoinbaseWalletCreationOptions)
 *
 * @wallet
 */
function createWallet(...args) {
    const [id, creationOptions] = args;
    switch (true) {
        /**
         * SMART WALLET
         */
        case id === "smart": {
            return (0, smart_wallet_js_1.smartWallet)(creationOptions);
        }
        /**
         * IN-APP WALLET
         */
        case id === "embedded" || id === "inApp": {
            return (0, in_app_js_1.inAppWallet)(creationOptions);
        }
        /**
         * COINBASE WALLET VIA SDK
         * -> if no injected coinbase found, we'll use the coinbase SDK
         */
        case id === constants_js_1.COINBASE: {
            const options = creationOptions;
            return (0, coinbase_wallet_js_1.coinbaseWalletSDK)({
                createOptions: options,
                providerFactory: () => (0, coinbaseWebSDK_js_1.getCoinbaseWebProvider)(options),
                onConnectRequested: async (provider) => {
                    // on the web, make sure to show the coinbase popup IMMEDIATELY on connection requested
                    // otherwise the popup might get blocked in safari
                    // TODO awaiting the provider is fast only thanks to preloading that happens in our components
                    // these probably need to actually imported / created synchronously to be used headless properly
                    const { showCoinbasePopup } = await Promise.resolve().then(() => require("./coinbase/utils.js"));
                    return showCoinbasePopup(provider);
                },
            });
        }
        case (0, is_ecosystem_wallet_js_1.isEcosystemWallet)(id):
            return (0, ecosystem_js_1.ecosystemWallet)(...args);
        /**
         * WALLET CONNECT AND INJECTED WALLETS + walletConnect standalone
         */
        default: {
            const emitter = (0, wallet_emitter_js_1.createWalletEmitter)();
            let account = undefined;
            let chain = undefined;
            const unsubscribeChain = emitter.subscribe("chainChanged", (newChain) => {
                chain = newChain;
            });
            function reset() {
                account = undefined;
                chain = undefined;
            }
            let handleDisconnect = async () => { };
            const unsubscribeDisconnect = emitter.subscribe("disconnect", () => {
                reset();
                unsubscribeChain();
                unsubscribeDisconnect();
            });
            emitter.subscribe("accountChanged", (_account) => {
                account = _account;
            });
            let handleSwitchChain = async () => {
                throw new Error("Not implemented yet");
            };
            // on mobile, deeplink to the wallet app for session handling
            const sessionHandler = (0, isMobile_js_1.isMobile)()
                ? (uri) => (0, openWindow_js_1.openWindow)(uri)
                : undefined;
            const wallet = {
                id,
                subscribe: emitter.subscribe,
                getConfig: () => args[1],
                getChain() {
                    if (!chain) {
                        return undefined;
                    }
                    chain = (0, utils_js_1.getCachedChainIfExists)(chain.id) || chain;
                    return chain;
                },
                getAccount: () => account,
                autoConnect: async (options) => {
                    const { injectedProvider } = await Promise.resolve().then(() => require("./injected/mipdStore.js"));
                    // injected wallet priority for autoConnect
                    if (id !== "walletConnect" && injectedProvider(id)) {
                        const { autoConnectInjectedWallet } = await Promise.resolve().then(() => require("./injected/index.js"));
                        const [connectedAccount, connectedChain, doDisconnect, doSwitchChain,] = await autoConnectInjectedWallet(id, emitter, options.chain);
                        // set the states
                        account = connectedAccount;
                        chain = connectedChain;
                        handleDisconnect = doDisconnect;
                        handleSwitchChain = doSwitchChain;
                        (0, track_js_1.trackConnect)({
                            client: options.client,
                            walletType: id,
                            walletAddress: account.address,
                        });
                        // return account
                        return account;
                    }
                    if (options && "client" in options) {
                        const { autoConnectWC } = await Promise.resolve().then(() => require("./wallet-connect/controller.js"));
                        const [connectedAccount, connectedChain, doDisconnect, doSwitchChain,] = await autoConnectWC(options, emitter, wallet.id, webStorage_js_1.webLocalStorage, sessionHandler);
                        // set the states
                        account = connectedAccount;
                        chain = connectedChain;
                        handleDisconnect = doDisconnect;
                        handleSwitchChain = doSwitchChain;
                        (0, track_js_1.trackConnect)({
                            client: options.client,
                            walletType: id,
                            walletAddress: account.address,
                        });
                        // return account
                        return account;
                    }
                    throw new Error("Failed to auto connect");
                },
                connect: async (options) => {
                    async function wcConnect(wcOptions) {
                        const { connectWC } = await Promise.resolve().then(() => require("./wallet-connect/controller.js"));
                        const [connectedAccount, connectedChain, doDisconnect, doSwitchChain,] = await connectWC(wcOptions, emitter, wallet.id, webStorage_js_1.webLocalStorage, sessionHandler);
                        // set the states
                        account = connectedAccount;
                        chain = connectedChain;
                        handleDisconnect = doDisconnect;
                        handleSwitchChain = doSwitchChain;
                        (0, track_js_1.trackConnect)({
                            client: wcOptions.client,
                            walletType: id,
                            walletAddress: account.address,
                        });
                        return account;
                    }
                    if (id === "walletConnect") {
                        const { client, chain: _chain, ...walletConnectOptions } = options;
                        return wcConnect({
                            client,
                            chain: _chain,
                            walletConnect: {
                                ...walletConnectOptions,
                            },
                        });
                    }
                    // prefer walletconnect over injected if explicitely passing walletConnect options
                    const forceWalletConnectOption = options && "walletConnect" in options;
                    const { injectedProvider } = await Promise.resolve().then(() => require("./injected/mipdStore.js"));
                    if (injectedProvider(id) && !forceWalletConnectOption) {
                        const { connectInjectedWallet } = await Promise.resolve().then(() => require("./injected/index.js"));
                        const [connectedAccount, connectedChain, doDisconnect, doSwitchChain,] = await connectInjectedWallet(id, options, emitter);
                        // set the states
                        account = connectedAccount;
                        chain = connectedChain;
                        handleDisconnect = doDisconnect;
                        handleSwitchChain = doSwitchChain;
                        (0, track_js_1.trackConnect)({
                            client: options.client,
                            walletType: id,
                            walletAddress: account.address,
                        });
                        // return account
                        return account;
                    }
                    if (options && "client" in options) {
                        return wcConnect(options);
                    }
                    throw new Error("Failed to connect");
                },
                // these get overridden in connect and autoConnect
                disconnect: async () => {
                    reset();
                    await handleDisconnect();
                },
                switchChain: (c) => handleSwitchChain(c),
            };
            return wallet;
        }
    }
}
/**
 * Creates a wallet that allows connecting to any wallet that supports the WalletConnect protocol.
 * @returns The created smart wallet.
 * @example
 * ```ts
 * import { walletConnect } from "thirdweb/wallets";
 *
 * const wallet = walletConnect();
 *
 * const account = await wallet.connect({
 *  client
 * });
 * ```
 * @wallet
 */
function walletConnect() {
    return createWallet("walletConnect");
}
