// thirdweb/dist/cjs/extensions/modules/__generated__/MintableERC1155/module/install.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.module = module;
exports.install = install;
exports.encodeInstall = encodeInstall;
const encodeBytesOnInstall_js_1 = require("../encode/encodeBytesOnInstall.js");
const getOrDeployModule_js_1 = require("../../../common/getOrDeployModule.js");
const installPublishedModule_js_1 = require("../../../common/installPublishedModule.js");
const contractId = "MintableERC1155";
/**
 * Convenience function to add the MintableERC1155 module as a default module on a core contract.
 * @param params - The parameters for the module.
 * @returns - The module function.
 * @example
 * ```ts
 * import { MintableERC1155, deployModularContract } from "thirdweb/modules";
 *
 * const deployed = deployModularContract({
 *   client,
 *   chain,
 *   account,
 *   core: "ERC1155",
 *   params: {
 *     name: "My Modular Contract",
 *   },
 *   modules: [
 *     MintableERC1155.module({
 *        primarySaleRecipient: ...,
 *     }),
 *   ],
 * });
 * ```
 * @modules MintableERC1155
 */
function module(params) {
    return async (args) => {
        // deploys if needed
        const moduleContract = await (0, getOrDeployModule_js_1.getOrDeployModule)({
            account: args.account,
            chain: args.chain,
            client: args.client,
            contractId,
            publisher: params?.publisher,
        });
        return {
            module: moduleContract.address,
            data: encodeInstall(params),
        };
    };
}
/**
 * Installs the MintableERC1155 module on a core contract.
 * @param options
 * @returns the transaction to install the module
 * @example
 * ```ts
 * import { MintableERC1155 } from "thirdweb/modules";
 *
 * const transaction  = MintableERC1155.install({
 *  contract: coreContract,
 *  account: account,
 *  params: {
 *     primarySaleRecipient: ...,
 *  },
 * });
 *
 * await sendTransaction({
 *  transaction,
 *  account,
 * });
 * ```
 * @modules MintableERC1155
 */
function install(options) {
    return (0, installPublishedModule_js_1.installPublishedModule)({
        account: options.account,
        contract: options.contract,
        moduleName: contractId,
        moduleData: encodeInstall(options.params),
        publisher: options.params?.publisher,
    });
}
/**
 * Encodes the install data for the MintableERC1155 module.
 * @param params - The parameters for the module.
 * @returns - The encoded data.
 * @modules MintableERC1155
 */
function encodeInstall(params) {
    return (0, encodeBytesOnInstall_js_1.encodeBytesOnInstallParams)(params);
}
