// thirdweb/dist/esm/extensions/lens/__generated__/TokenHandleRegistry/read/getDefaultHandle.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { encodeAbiParameters } from "../../../../../utils/abi/encodeAbiParameters.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0xe524488d";
const FN_INPUTS = [
    {
        type: "uint256",
        name: "profileId",
    },
];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `getDefaultHandle` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `getDefaultHandle` method is supported.
 * @extension LENS
 * @example
 * ```ts
 * import { isGetDefaultHandleSupported } from "thirdweb/extensions/lens";
 * const supported = isGetDefaultHandleSupported(["0x..."]);
 * ```
 */
export function isGetDefaultHandleSupported(availableSelectors) {
    return detectMethod({
        availableSelectors,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "getDefaultHandle" function.
 * @param options - The options for the getDefaultHandle function.
 * @returns The encoded ABI parameters.
 * @extension LENS
 * @example
 * ```ts
 * import { encodeGetDefaultHandleParams } from "thirdweb/extensions/lens";
 * const result = encodeGetDefaultHandleParams({
 *  profileId: ...,
 * });
 * ```
 */
export function encodeGetDefaultHandleParams(options) {
    return encodeAbiParameters(FN_INPUTS, [options.profileId]);
}
/**
 * Encodes the "getDefaultHandle" function into a Hex string with its parameters.
 * @param options - The options for the getDefaultHandle function.
 * @returns The encoded hexadecimal string.
 * @extension LENS
 * @example
 * ```ts
 * import { encodeGetDefaultHandle } from "thirdweb/extensions/lens";
 * const result = encodeGetDefaultHandle({
 *  profileId: ...,
 * });
 * ```
 */
export function encodeGetDefaultHandle(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (FN_SELECTOR +
        encodeGetDefaultHandleParams(options).slice(2));
}
/**
 * Decodes the result of the getDefaultHandle function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension LENS
 * @example
 * ```ts
 * import { decodeGetDefaultHandleResult } from "thirdweb/extensions/lens";
 * const result = decodeGetDefaultHandleResultResult("...");
 * ```
 */
export function decodeGetDefaultHandleResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getDefaultHandle" function on the contract.
 * @param options - The options for the getDefaultHandle function.
 * @returns The parsed result of the function call.
 * @extension LENS
 * @example
 * ```ts
 * import { getDefaultHandle } from "thirdweb/extensions/lens";
 *
 * const result = await getDefaultHandle({
 *  contract,
 *  profileId: ...,
 * });
 *
 * ```
 */
export async function getDefaultHandle(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.profileId],
    });
}
