// thirdweb/dist/esm/extensions/lens/__generated__/TokenHandleRegistry/read/nonces.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { encodeAbiParameters } from "../../../../../utils/abi/encodeAbiParameters.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0x7ecebe00";
const FN_INPUTS = [
    {
        type: "address",
        name: "signer",
    },
];
const FN_OUTPUTS = [
    {
        type: "uint256",
        name: "nonce",
    },
];
/**
 * Checks if the `nonces` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `nonces` method is supported.
 * @extension LENS
 * @example
 * ```ts
 * import { isNoncesSupported } from "thirdweb/extensions/lens";
 * const supported = isNoncesSupported(["0x..."]);
 * ```
 */
export function isNoncesSupported(availableSelectors) {
    return detectMethod({
        availableSelectors,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "nonces" function.
 * @param options - The options for the nonces function.
 * @returns The encoded ABI parameters.
 * @extension LENS
 * @example
 * ```ts
 * import { encodeNoncesParams } from "thirdweb/extensions/lens";
 * const result = encodeNoncesParams({
 *  signer: ...,
 * });
 * ```
 */
export function encodeNoncesParams(options) {
    return encodeAbiParameters(FN_INPUTS, [options.signer]);
}
/**
 * Encodes the "nonces" function into a Hex string with its parameters.
 * @param options - The options for the nonces function.
 * @returns The encoded hexadecimal string.
 * @extension LENS
 * @example
 * ```ts
 * import { encodeNonces } from "thirdweb/extensions/lens";
 * const result = encodeNonces({
 *  signer: ...,
 * });
 * ```
 */
export function encodeNonces(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (FN_SELECTOR +
        encodeNoncesParams(options).slice(2));
}
/**
 * Decodes the result of the nonces function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension LENS
 * @example
 * ```ts
 * import { decodeNoncesResult } from "thirdweb/extensions/lens";
 * const result = decodeNoncesResultResult("...");
 * ```
 */
export function decodeNoncesResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "nonces" function on the contract.
 * @param options - The options for the nonces function.
 * @returns The parsed result of the function call.
 * @extension LENS
 * @example
 * ```ts
 * import { nonces } from "thirdweb/extensions/lens";
 *
 * const result = await nonces({
 *  contract,
 *  signer: ...,
 * });
 *
 * ```
 */
export async function nonces(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.signer],
    });
}
