// thirdweb/dist/cjs/react/native/ui/connect/ReceiveScreen.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReceiveScreen = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_native_1 = require("react-native");
const address_js_1 = require("../../../../utils/address.js");
const index_js_1 = require("../../design-system/index.js");
const Address_js_1 = require("../components/Address.js");
const Header_js_1 = require("../components/Header.js");
const WalletImage_js_1 = require("../components/WalletImage.js");
const spacer_js_1 = require("../components/spacer.js");
const text_js_1 = require("../components/text.js");
const ReceiveScreen = (props) => {
    const { wallet, account, theme, onClose, onBack, containerType } = props;
    return ((0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [(0, jsx_runtime_1.jsx)(Header_js_1.Header, { theme: theme, onClose: onClose, onBack: onBack, containerType: containerType, title: "Receive Funds" }), (0, jsx_runtime_1.jsxs)(react_native_1.View, { style: styles.container, children: [(0, jsx_runtime_1.jsx)(WalletImage_js_1.WalletImage, { theme: theme, wallet: wallet, size: 80 }), (0, jsx_runtime_1.jsx)(spacer_js_1.Spacer, { size: "lg" }), (0, jsx_runtime_1.jsx)(react_native_1.View, { style: [
                            styles.addressContainer,
                            { borderColor: theme.colors.borderColor },
                        ], children: (0, jsx_runtime_1.jsx)(Address_js_1.Address, { account: account, addressOrENS: (0, address_js_1.shortenAddress)(account?.address, 8), theme: theme }) }), (0, jsx_runtime_1.jsx)(text_js_1.ThemedText, { theme: theme, type: "subtext", style: { textAlign: "center" }, children: "Copy your address to send funds to this wallet" })] })] }));
};
exports.ReceiveScreen = ReceiveScreen;
const styles = react_native_1.StyleSheet.create({
    container: {
        paddingHorizontal: index_js_1.spacing.lg,
        paddingVertical: index_js_1.spacing.xxl,
        flexDirection: "column",
        gap: index_js_1.spacing.md,
        alignItems: "center",
        justifyContent: "center",
        flex: 1,
    },
    addressContainer: {
        width: "100%",
        flexDirection: "row",
        gap: index_js_1.spacing.sm,
        alignItems: "center",
        justifyContent: "center",
        borderWidth: 1,
        padding: index_js_1.spacing.md,
        borderRadius: index_js_1.spacing.lg,
    },
});
