// thirdweb/dist/cjs/extensions/marketplace/english-auctions/write/cancelAuction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cancelAuction = cancelAuction;
exports.isCancelAuctionSupported = isCancelAuctionSupported;
const CancelAuction = require("../../__generated__/IEnglishAuctions/write/cancelAuction.js");
const getWinningBid_js_1 = require("../read/getWinningBid.js");
/**
 * Cancels an auction by providing the necessary options.
 *
 * @param options - The options for canceling the auction.
 * @returns A transaction that can be sent to cancel the auction.
 * @throws An error when sending the transaction if the auction has an existing bid and cannot be canceled.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { cancelAuction } from "thirdweb/extensions/marketplace";
 * import { sendTransaction } from "thirdweb";
 *
 * const transaction = cancelAuction({
 *  contract,
 *  auctionId: 0n,
 * });
 *
 * await sendTransaction({ transaction, account });
 * ```
 */
function cancelAuction(options) {
    return CancelAuction.cancelAuction({
        contract: options.contract,
        asyncParams: async () => {
            const winningBid = await (0, getWinningBid_js_1.getWinningBid)({
                contract: options.contract,
                auctionId: options.auctionId,
            });
            if (winningBid) {
                throw new Error("Cannot cancel an auction with an existing bid");
            }
            return {
                auctionId: options.auctionId,
                overrides: {
                    extraGas: 50000n, // add extra gas to account for router call
                },
            };
        },
    });
}
/**
 * Checks if the `cancelAuction` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `cancelAuction` method is supported.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { isCancelAuctionSupported } from "thirdweb/extensions/marketplace";
 *
 * const supported = isCancelAuctionSupported(["0x..."]);
 * ```
 */
function isCancelAuctionSupported(availableSelectors) {
    return (CancelAuction.isCancelAuctionSupported(availableSelectors) &&
        (0, getWinningBid_js_1.isGetWinningBidSupported)(availableSelectors));
}
