// thirdweb/dist/cjs/extensions/erc6551/__generated__/IERC6551Account/read/isValidSigner.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isIsValidSignerSupported = isIsValidSignerSupported;
exports.encodeIsValidSignerParams = encodeIsValidSignerParams;
exports.encodeIsValidSigner = encodeIsValidSigner;
exports.decodeIsValidSignerResult = decodeIsValidSignerResult;
exports.isValidSigner = isValidSigner;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x523e3260";
const FN_INPUTS = [
    {
        type: "address",
        name: "signer",
    },
    {
        type: "bytes",
        name: "context",
    },
];
const FN_OUTPUTS = [
    {
        type: "bytes4",
        name: "magicValue",
    },
];
/**
 * Checks if the `isValidSigner` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `isValidSigner` method is supported.
 * @extension ERC6551
 * @example
 * ```ts
 * import { isIsValidSignerSupported } from "thirdweb/extensions/erc6551";
 * const supported = isIsValidSignerSupported(["0x..."]);
 * ```
 */
function isIsValidSignerSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "isValidSigner" function.
 * @param options - The options for the isValidSigner function.
 * @returns The encoded ABI parameters.
 * @extension ERC6551
 * @example
 * ```ts
 * import { encodeIsValidSignerParams } from "thirdweb/extensions/erc6551";
 * const result = encodeIsValidSignerParams({
 *  signer: ...,
 *  context: ...,
 * });
 * ```
 */
function encodeIsValidSignerParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.signer, options.context]);
}
/**
 * Encodes the "isValidSigner" function into a Hex string with its parameters.
 * @param options - The options for the isValidSigner function.
 * @returns The encoded hexadecimal string.
 * @extension ERC6551
 * @example
 * ```ts
 * import { encodeIsValidSigner } from "thirdweb/extensions/erc6551";
 * const result = encodeIsValidSigner({
 *  signer: ...,
 *  context: ...,
 * });
 * ```
 */
function encodeIsValidSigner(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeIsValidSignerParams(options).slice(2));
}
/**
 * Decodes the result of the isValidSigner function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC6551
 * @example
 * ```ts
 * import { decodeIsValidSignerResult } from "thirdweb/extensions/erc6551";
 * const result = decodeIsValidSignerResultResult("...");
 * ```
 */
function decodeIsValidSignerResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "isValidSigner" function on the contract.
 * @param options - The options for the isValidSigner function.
 * @returns The parsed result of the function call.
 * @extension ERC6551
 * @example
 * ```ts
 * import { isValidSigner } from "thirdweb/extensions/erc6551";
 *
 * const result = await isValidSigner({
 *  contract,
 *  signer: ...,
 *  context: ...,
 * });
 *
 * ```
 */
async function isValidSigner(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.signer, options.context],
    });
}
