// thirdweb/dist/esm/extensions/vote/__generated__/Vote/read/getVotesWithParams.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { encodeAbiParameters } from "../../../../../utils/abi/encodeAbiParameters.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0x9a802a6d";
const FN_INPUTS = [
    {
        type: "address",
        name: "account",
    },
    {
        type: "uint256",
        name: "blockNumber",
    },
    {
        type: "bytes",
        name: "params",
    },
];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `getVotesWithParams` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `getVotesWithParams` method is supported.
 * @extension VOTE
 * @example
 * ```ts
 * import { isGetVotesWithParamsSupported } from "thirdweb/extensions/vote";
 * const supported = isGetVotesWithParamsSupported(["0x..."]);
 * ```
 */
export function isGetVotesWithParamsSupported(availableSelectors) {
    return detectMethod({
        availableSelectors,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "getVotesWithParams" function.
 * @param options - The options for the getVotesWithParams function.
 * @returns The encoded ABI parameters.
 * @extension VOTE
 * @example
 * ```ts
 * import { encodeGetVotesWithParamsParams } from "thirdweb/extensions/vote";
 * const result = encodeGetVotesWithParamsParams({
 *  account: ...,
 *  blockNumber: ...,
 *  params: ...,
 * });
 * ```
 */
export function encodeGetVotesWithParamsParams(options) {
    return encodeAbiParameters(FN_INPUTS, [
        options.account,
        options.blockNumber,
        options.params,
    ]);
}
/**
 * Encodes the "getVotesWithParams" function into a Hex string with its parameters.
 * @param options - The options for the getVotesWithParams function.
 * @returns The encoded hexadecimal string.
 * @extension VOTE
 * @example
 * ```ts
 * import { encodeGetVotesWithParams } from "thirdweb/extensions/vote";
 * const result = encodeGetVotesWithParams({
 *  account: ...,
 *  blockNumber: ...,
 *  params: ...,
 * });
 * ```
 */
export function encodeGetVotesWithParams(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (FN_SELECTOR +
        encodeGetVotesWithParamsParams(options).slice(2));
}
/**
 * Decodes the result of the getVotesWithParams function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension VOTE
 * @example
 * ```ts
 * import { decodeGetVotesWithParamsResult } from "thirdweb/extensions/vote";
 * const result = decodeGetVotesWithParamsResultResult("...");
 * ```
 */
export function decodeGetVotesWithParamsResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getVotesWithParams" function on the contract.
 * @param options - The options for the getVotesWithParams function.
 * @returns The parsed result of the function call.
 * @extension VOTE
 * @example
 * ```ts
 * import { getVotesWithParams } from "thirdweb/extensions/vote";
 *
 * const result = await getVotesWithParams({
 *  contract,
 *  account: ...,
 *  blockNumber: ...,
 *  params: ...,
 * });
 *
 * ```
 */
export async function getVotesWithParams(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.account, options.blockNumber, options.params],
    });
}
