// thirdweb/dist/esm/extensions/vote/__generated__/Vote/read/proposals.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { encodeAbiParameters } from "../../../../../utils/abi/encodeAbiParameters.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0x013cf08b";
const FN_INPUTS = [
    {
        type: "uint256",
        name: "key",
    },
];
const FN_OUTPUTS = [
    {
        type: "uint256",
        name: "proposalId",
    },
    {
        type: "address",
        name: "proposer",
    },
    {
        type: "uint256",
        name: "startBlock",
    },
    {
        type: "uint256",
        name: "endBlock",
    },
    {
        type: "string",
        name: "description",
    },
];
/**
 * Checks if the `proposals` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `proposals` method is supported.
 * @extension VOTE
 * @example
 * ```ts
 * import { isProposalsSupported } from "thirdweb/extensions/vote";
 * const supported = isProposalsSupported(["0x..."]);
 * ```
 */
export function isProposalsSupported(availableSelectors) {
    return detectMethod({
        availableSelectors,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "proposals" function.
 * @param options - The options for the proposals function.
 * @returns The encoded ABI parameters.
 * @extension VOTE
 * @example
 * ```ts
 * import { encodeProposalsParams } from "thirdweb/extensions/vote";
 * const result = encodeProposalsParams({
 *  key: ...,
 * });
 * ```
 */
export function encodeProposalsParams(options) {
    return encodeAbiParameters(FN_INPUTS, [options.key]);
}
/**
 * Encodes the "proposals" function into a Hex string with its parameters.
 * @param options - The options for the proposals function.
 * @returns The encoded hexadecimal string.
 * @extension VOTE
 * @example
 * ```ts
 * import { encodeProposals } from "thirdweb/extensions/vote";
 * const result = encodeProposals({
 *  key: ...,
 * });
 * ```
 */
export function encodeProposals(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (FN_SELECTOR +
        encodeProposalsParams(options).slice(2));
}
/**
 * Decodes the result of the proposals function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension VOTE
 * @example
 * ```ts
 * import { decodeProposalsResult } from "thirdweb/extensions/vote";
 * const result = decodeProposalsResultResult("...");
 * ```
 */
export function decodeProposalsResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result);
}
/**
 * Calls the "proposals" function on the contract.
 * @param options - The options for the proposals function.
 * @returns The parsed result of the function call.
 * @extension VOTE
 * @example
 * ```ts
 * import { proposals } from "thirdweb/extensions/vote";
 *
 * const result = await proposals({
 *  contract,
 *  key: ...,
 * });
 *
 * ```
 */
export async function proposals(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.key],
    });
}
