// thirdweb/dist/esm/extensions/vote/__generated__/Vote/read/proposalVotes.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { encodeAbiParameters } from "../../../../../utils/abi/encodeAbiParameters.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0x544ffc9c";
const FN_INPUTS = [
    {
        type: "uint256",
        name: "proposalId",
    },
];
const FN_OUTPUTS = [
    {
        type: "uint256",
        name: "againstVotes",
    },
    {
        type: "uint256",
        name: "forVotes",
    },
    {
        type: "uint256",
        name: "abstainVotes",
    },
];
/**
 * Checks if the `proposalVotes` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `proposalVotes` method is supported.
 * @extension VOTE
 * @example
 * ```ts
 * import { isProposalVotesSupported } from "thirdweb/extensions/vote";
 * const supported = isProposalVotesSupported(["0x..."]);
 * ```
 */
export function isProposalVotesSupported(availableSelectors) {
    return detectMethod({
        availableSelectors,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "proposalVotes" function.
 * @param options - The options for the proposalVotes function.
 * @returns The encoded ABI parameters.
 * @extension VOTE
 * @example
 * ```ts
 * import { encodeProposalVotesParams } from "thirdweb/extensions/vote";
 * const result = encodeProposalVotesParams({
 *  proposalId: ...,
 * });
 * ```
 */
export function encodeProposalVotesParams(options) {
    return encodeAbiParameters(FN_INPUTS, [options.proposalId]);
}
/**
 * Encodes the "proposalVotes" function into a Hex string with its parameters.
 * @param options - The options for the proposalVotes function.
 * @returns The encoded hexadecimal string.
 * @extension VOTE
 * @example
 * ```ts
 * import { encodeProposalVotes } from "thirdweb/extensions/vote";
 * const result = encodeProposalVotes({
 *  proposalId: ...,
 * });
 * ```
 */
export function encodeProposalVotes(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (FN_SELECTOR +
        encodeProposalVotesParams(options).slice(2));
}
/**
 * Decodes the result of the proposalVotes function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension VOTE
 * @example
 * ```ts
 * import { decodeProposalVotesResult } from "thirdweb/extensions/vote";
 * const result = decodeProposalVotesResultResult("...");
 * ```
 */
export function decodeProposalVotesResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result);
}
/**
 * Calls the "proposalVotes" function on the contract.
 * @param options - The options for the proposalVotes function.
 * @returns The parsed result of the function call.
 * @extension VOTE
 * @example
 * ```ts
 * import { proposalVotes } from "thirdweb/extensions/vote";
 *
 * const result = await proposalVotes({
 *  contract,
 *  proposalId: ...,
 * });
 *
 * ```
 */
export async function proposalVotes(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.proposalId],
    });
}
