// thirdweb/dist/cjs/extensions/erc721/drops/write/updateMetadata.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateMetadata = updateMetadata;
exports.isUpdateMetadataSupported = isUpdateMetadataSupported;
const BatchBaseURI = require("../../__generated__/DropERC721/write/updateBatchBaseURI.js");
const BaseURICount = require("../../__generated__/IBatchMintMetadata/read/getBaseURICount.js");
const BatchAtIndex = require("../../__generated__/IBatchMintMetadata/read/getBatchIdAtIndex.js");
const GetNFT = require("../../read/getNFT.js");
/**
 * @internal
 */
async function getUpdateMetadataParams(options) {
    const { contract, targetTokenId, newMetadata } = options;
    const batchCount = await BaseURICount.getBaseURICount(options);
    if (batchCount === 0n) {
        throw new Error("No base URI set. Please set a base URI before updating metadata");
    }
    // Look for the batchId & determine the start + end tokenId of the batch
    let startTokenId = 0n;
    let endTokenId = 0n;
    let batchIndex = 0n;
    for (let i = 0n; i < batchCount; i += 1n) {
        batchIndex = i;
        endTokenId = await BatchAtIndex.getBatchIdAtIndex({
            contract,
            index: batchIndex,
        });
        if (endTokenId > targetTokenId) {
            break;
        }
        startTokenId = endTokenId;
    }
    const range = Array.from({ length: Number(endTokenId - startTokenId) }, (_, k) => BigInt(k) + startTokenId);
    const currentMetadatas = await Promise.all(range.map((id) => GetNFT.getNFT({ contract, tokenId: id, includeOwner: false })));
    // Abort if any of the items failed to load
    if (currentMetadatas.some((item) => item === undefined || !item.tokenURI)) {
        throw new Error(`Failed to load all ${range.length} items from batchIndex: ${batchIndex}`);
    }
    const newMetadatas = [];
    for (let i = 0; i < currentMetadatas.length; i++) {
        const { id, metadata } = currentMetadatas[i];
        if (targetTokenId === id) {
            newMetadatas.push(newMetadata);
        }
        else {
            newMetadatas.push(metadata);
        }
    }
    const { uploadOrExtractURIs } = await Promise.resolve().then(() => require("../../../../utils/ipfs.js"));
    const batchOfUris = await uploadOrExtractURIs(newMetadatas, contract.client, Number(startTokenId));
    if (!batchOfUris || !batchOfUris.length || !batchOfUris[0]) {
        throw new Error("Failed to upload batch of new metadatas");
    }
    const baseUri = batchOfUris[0].substring(0, batchOfUris[0].lastIndexOf("/"));
    // IMPORTANT: The new ipfs URI must have the trailing slash at the end
    // this is required by the Drop contract
    const uri = `${baseUri}/`;
    return { index: batchIndex, uri };
}
/**
 * Update the metadata of the single token in an NFT Drop (DropERC721) collection
 * For NFT Collection, please use `setTokenURI`
 * @param options
 * @returns the prepared transaction
 * @extension ERC721
 * @example
 * ```ts
 * import { updateMetadata } from "thirdweb/extensions/erc721";
 *
 * const transaction = updateMetadata({
 *  contract,
 *  targetTokenId: 0n,
 *  client: thirdwebClient,
 *  newMetadata: {
 *    name: "this is the new nft name",
 *    description: "...",
 *    image: "new image uri"
 *    // ...
 *  }
 * });
 * ```
 */
function updateMetadata(options) {
    const { contract } = options;
    return BatchBaseURI.updateBatchBaseURI({
        contract,
        asyncParams: async () => getUpdateMetadataParams(options),
    });
}
/**
 * Checks if the `updateMetadata` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `updateMetadata` method is supported.
 * @extension ERC721
 * @example
 * ```ts
 * import { isUpdateMetadataSupported } from "thirdweb/extensions/erc721";
 *
 * const supported = isUpdateMetadataSupported(["0x..."]);
 * ```
 */
function isUpdateMetadataSupported(availableSelectors) {
    return (BaseURICount.isGetBaseURICountSupported(availableSelectors) &&
        BatchBaseURI.isUpdateBatchBaseURISupported(availableSelectors) &&
        BatchAtIndex.isGetBatchIdAtIndexSupported(availableSelectors) &&
        GetNFT.isGetNFTSupported(availableSelectors));
}
