// thirdweb/dist/cjs/react/web/ui/components/DragNDrop.js
"use strict";
"use client";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DragNDrop = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const styled_1 = require("@emotion/styled");
const react_icons_1 = require("@radix-ui/react-icons");
const react_1 = require("react");
const isMobile_js_1 = require("../../../../utils/web/isMobile.js");
const CustomThemeProvider_js_1 = require("../../../core/design-system/CustomThemeProvider.js");
const index_js_1 = require("../../../core/design-system/index.js");
const elements_js_1 = require("../design-system/elements.js");
const Spacer_js_1 = require("./Spacer.js");
const basic_js_1 = require("./basic.js");
const text_js_1 = require("./text.js");
/**
 * @internal
 */
const DragNDrop = (props) => {
    const [error, setError] = (0, react_1.useState)(false);
    const [uploaded, setUploaded] = (0, react_1.useState)();
    const [isDragging, setIsDragging] = (0, react_1.useState)(false);
    const dragIn = (e) => {
        setError(false);
        setUploaded(undefined);
        setIsDragging(true);
        e.preventDefault();
        e.stopPropagation();
    };
    const dragOut = (e) => {
        setIsDragging(false);
        e.preventDefault();
        e.stopPropagation();
    };
    const handleFileUpload = (file) => {
        if (file.type !== props.accept) {
            setError(true);
        }
        else {
            setUploaded(file);
            props.onUpload(file);
        }
    };
    const drop = (e) => {
        setIsDragging(false);
        e.preventDefault();
        e.stopPropagation();
        if (e.dataTransfer.files && e.dataTransfer.files.length > 0) {
            const fileContent = e.dataTransfer.files[0];
            if (fileContent) {
                handleFileUpload(fileContent);
                e.dataTransfer.clearData();
            }
        }
    };
    const message = (0, isMobile_js_1.isMobile)()
        ? "Click to Upload"
        : "Drop your file here or click to upload";
    return ((0, jsx_runtime_1.jsx)("div", { onDragEnter: dragIn, onDragLeave: dragOut, onDragOver: (e) => {
            setIsDragging(true);
            e.preventDefault();
            e.stopPropagation();
        }, onMouseDown: () => {
            setError(false);
        }, onDrop: drop, style: {
            cursor: "pointer",
        }, children: (0, jsx_runtime_1.jsxs)("label", { htmlFor: "file-upload", children: [(0, jsx_runtime_1.jsx)("input", { id: "file-upload", type: "file", accept: props.accept, multiple: false, style: {
                        display: "none",
                    }, onChange: (e) => {
                        if (e.target.files?.[0]) {
                            handleFileUpload(e.target.files[0]);
                        }
                    } }), (0, jsx_runtime_1.jsx)(DropContainer, { "data-error": error, "data-is-dragging": isDragging, children: !uploaded ? ((0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [" ", (0, jsx_runtime_1.jsx)(UploadIconSecondary, { width: index_js_1.iconSize.lg, height: index_js_1.iconSize.lg }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "md" }), (0, jsx_runtime_1.jsxs)(text_js_1.Text, { color: "primaryText", weight: 600, center: true, multiline: true, children: [" ", message] }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "lg" }), error ? ((0, jsx_runtime_1.jsx)(text_js_1.Text, { color: "danger", size: "sm", children: props.locale.wrongFileError })) : ((0, jsx_runtime_1.jsxs)(text_js_1.Text, { size: "sm", children: [" ", props.extension, " "] }))] })) : ((0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [(0, jsx_runtime_1.jsxs)(text_js_1.Text, { weight: 600, color: "primaryText", center: true, multiline: true, children: [uploaded.name, " ", props.locale.uploadedSuccessfully] }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "md" }), (0, jsx_runtime_1.jsx)(basic_js_1.Container, { color: "success", children: (0, jsx_runtime_1.jsx)(CheckCircleIcon, { size: index_js_1.iconSize.xl }) })] })) })] }) }));
};
exports.DragNDrop = DragNDrop;
const UploadIconSecondary = /* @__PURE__ */ (0, styled_1.default)(react_icons_1.UploadIcon)(() => {
    const theme = (0, CustomThemeProvider_js_1.useCustomTheme)();
    return {
        color: theme.colors.secondaryIconColor,
        transition: "transform 200ms ease, color 200ms ease",
    };
});
const DropContainer = /* @__PURE__ */ (0, elements_js_1.StyledDiv)((_) => {
    const theme = (0, CustomThemeProvider_js_1.useCustomTheme)();
    return {
        border: `2px solid ${theme.colors.borderColor}`,
        borderRadius: index_js_1.radius.md,
        padding: `${index_js_1.spacing.xl} ${index_js_1.spacing.md}`,
        display: "flex",
        alignItems: "center",
        flexDirection: "column",
        cursor: "pointer",
        transition: "border-color 200ms ease",
        "&:hover, &[data-is-dragging='true']": {
            borderColor: theme.colors.accentText,
            svg: {
                color: theme.colors.accentText,
            },
        },
        "&[data-error='true']": {
            borderColor: theme.colors.danger,
        },
    };
});
/**
 * @internal
 */
const CheckCircleIcon = (props) => ((0, jsx_runtime_1.jsxs)("svg", { width: props.size, height: props.size, viewBox: "0 0 38 38", fill: "none", xmlns: "http://www.w3.org/2000/svg", role: "presentation", children: [(0, jsx_runtime_1.jsx)("path", { d: "M35.6666 17.4671V19.0004C35.6645 22.5945 34.5008 26.0916 32.3488 28.9701C30.1969 31.8487 27.1721 33.9546 23.7255 34.9736C20.279 35.9926 16.5954 35.8703 13.224 34.6247C9.85272 33.3792 6.97434 31.0773 5.01819 28.0622C3.06203 25.0472 2.1329 21.4805 2.36938 17.8943C2.60586 14.308 3.99526 10.8943 6.33039 8.16221C8.66551 5.43012 11.8212 3.52606 15.3269 2.734C18.8326 1.94194 22.5004 2.30432 25.7833 3.76709", stroke: "currentColor", strokeWidth: "3.33333", strokeLinecap: "round", strokeLinejoin: "round" }), (0, jsx_runtime_1.jsx)("path", { d: "M35.6667 5.66699L19 22.3503L14 17.3503", stroke: "currentColor", strokeWidth: "3.33333", strokeLinecap: "round", strokeLinejoin: "round" })] }));
