// thirdweb/dist/cjs/react/web/ui/components/Drawer.js
"use strict";
"use client";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DrawerOverlay = exports.Drawer = void 0;
exports.useDrawer = useDrawer;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("@emotion/react");
const react_icons_1 = require("@radix-ui/react-icons");
const react_2 = require("react");
const CustomThemeProvider_js_1 = require("../../../core/design-system/CustomThemeProvider.js");
const index_js_1 = require("../../../core/design-system/index.js");
const animations_js_1 = require("../design-system/animations.js");
const elements_js_1 = require("../design-system/elements.js");
const DynamicHeight_js_1 = require("./DynamicHeight.js");
const Modal_js_1 = require("./Modal.js");
const basic_js_1 = require("./basic.js");
const buttons_js_1 = require("./buttons.js");
/**
 *
 * @internal
 */
exports.Drawer = (0, react_2.forwardRef)(function Drawer_(props, ref) {
    return ((0, jsx_runtime_1.jsx)(DrawerContainer, { ref: ref, children: (0, jsx_runtime_1.jsx)(DynamicHeight_js_1.DynamicHeight, { children: (0, jsx_runtime_1.jsxs)(basic_js_1.Container, { p: "lg", children: [(0, jsx_runtime_1.jsx)(Modal_js_1.CrossContainer, { children: (0, jsx_runtime_1.jsx)(buttons_js_1.IconButton, { type: "button", "aria-label": "Close", onClick: props.close, children: (0, jsx_runtime_1.jsx)(react_icons_1.Cross2Icon, { width: index_js_1.iconSize.md, height: index_js_1.iconSize.md, style: {
                                    color: "inherit",
                                } }) }) }), props.children] }) }) }));
});
const DrawerContainer = /* @__PURE__ */ (0, elements_js_1.StyledDiv)((_) => {
    const theme = (0, CustomThemeProvider_js_1.useCustomTheme)();
    return {
        zIndex: 10000,
        borderTopLeftRadius: index_js_1.radius.xl,
        borderTopRightRadius: index_js_1.radius.xl,
        background: theme.colors.modalBg,
        position: "absolute",
        bottom: 0,
        left: 0,
        right: 0,
        animation: `${drawerOpenAnimation} 0.3s cubic-bezier(0.175, 0.885, 0.32, 1.1)`,
        borderTop: `1px solid ${theme.colors.borderColor}`,
    };
});
const drawerOpenAnimation = (0, react_1.keyframes) `
  from {
    opacity: 0;
    transform: translateY(100px);
  }
  to {
    opacity: 1;
    transform: translateY(0);
  }
`;
exports.DrawerOverlay = (0, elements_js_1.StyledDiv)((_) => {
    const theme = (0, CustomThemeProvider_js_1.useCustomTheme)();
    return {
        backgroundColor: theme.colors.modalOverlayBg,
        zIndex: 9999,
        position: "absolute",
        inset: 0,
        animation: `${animations_js_1.fadeInAnimation} 400ms cubic-bezier(0.16, 1, 0.3, 1)`,
    };
});
/**
 *
 * @internal
 */
function useDrawer() {
    const [isOpen, _setIsOpen] = (0, react_2.useState)(false);
    const drawerRef = (0, react_2.useRef)(null);
    const drawerOverlayRef = (0, react_2.useRef)(null);
    const closeDrawerAnimation = (0, react_2.useCallback)(() => {
        return new Promise((resolve) => {
            if (drawerRef.current) {
                const animOptions = {
                    easing: "cubic-bezier(0.175, 0.885, 0.32, 1.1)",
                    fill: "forwards",
                    duration: 300,
                };
                const closeAnimation = drawerRef.current.animate([{ transform: "translateY(100%)", opacity: 0 }], animOptions);
                drawerOverlayRef.current?.animate([{ opacity: 0 }], animOptions);
                closeAnimation.onfinish = () => resolve();
            }
            else {
                resolve();
            }
        });
    }, []);
    const setIsOpen = (0, react_2.useCallback)(async (value) => {
        if (value) {
            _setIsOpen(true);
        }
        else {
            await closeDrawerAnimation();
            _setIsOpen(false);
        }
    }, [closeDrawerAnimation]);
    // close on outside click
    (0, react_2.useLayoutEffect)(() => {
        if (!isOpen) {
            return;
        }
        const handleClick = (event) => {
            if (drawerRef.current &&
                event.target instanceof Node &&
                !drawerRef.current.contains(event.target)) {
                setIsOpen(false);
            }
        };
        // avoid listening to the click event that opened the drawer by adding a frame delay
        requestAnimationFrame(() => {
            document.addEventListener("click", handleClick);
        });
        return () => {
            document.removeEventListener("click", handleClick);
        };
    }, [isOpen, setIsOpen]);
    return {
        drawerRef,
        drawerOverlayRef,
        setIsOpen,
        isOpen,
    };
}
