// thirdweb/dist/cjs/react/web/ui/components/Img.js
"use strict";
"use client";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Img = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const ipfs_js_1 = require("../../../../utils/ipfs.js");
const Skeleton_js_1 = require("./Skeleton.js");
/**
 * @internal
 */
const Img = (props) => {
    const [isLoaded, setIsLoaded] = (0, react_1.useState)(false);
    const propSrc = props.src;
    const widthPx = `${props.width}px`;
    const heightPx = `${props.height || props.width}px`;
    if (propSrc === undefined) {
        return (0, jsx_runtime_1.jsx)(Skeleton_js_1.Skeleton, { width: widthPx, height: heightPx });
    }
    const getSrc = () => {
        try {
            return (0, ipfs_js_1.resolveScheme)({
                uri: propSrc,
                client: props.client,
            });
        }
        catch {
            return props.src;
        }
    };
    const src = getSrc();
    return ((0, jsx_runtime_1.jsxs)("div", { style: {
            position: "relative",
            display: "inline-flex",
            flexShrink: 0,
            alignItems: "center",
            justifyItems: "center",
        }, children: [!isLoaded && (0, jsx_runtime_1.jsx)(Skeleton_js_1.Skeleton, { width: widthPx, height: heightPx }), (0, jsx_runtime_1.jsx)("img", { onLoad: () => {
                    setIsLoaded(true);
                }, width: props.width, height: props.height, src: src, alt: props.alt || "", loading: props.loading, decoding: "async", style: {
                    objectFit: "contain",
                    height: !isLoaded
                        ? 0
                        : props.height
                            ? `${props.height}px`
                            : undefined,
                    width: !isLoaded ? 0 : props.width ? `${props.width}px` : undefined,
                    userSelect: "none",
                    visibility: isLoaded ? "visible" : "hidden",
                    opacity: isLoaded ? 1 : 0,
                    transition: "opacity 0.4s ease",
                    ...props.style,
                }, draggable: false, className: props.className, onError: (e) => {
                    if (props.fallbackImage &&
                        e.currentTarget.src !== props.fallbackImage) {
                        e.currentTarget.src = props.fallbackImage;
                    }
                } }, src)] }));
};
exports.Img = Img;
