// thirdweb/dist/cjs/react/web/ui/components/QRCode.js
"use strict";
"use client";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QRCode = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("@emotion/react");
const react_2 = require("react");
const CustomThemeProvider_js_1 = require("../../../core/design-system/CustomThemeProvider.js");
const index_js_1 = require("../../../core/design-system/index.js");
const animations_js_1 = require("../design-system/animations.js");
const elements_js_1 = require("../design-system/elements.js");
const QRCodeRenderer = /* @__PURE__ */ (0, react_2.lazy)(() => Promise.resolve().then(() => require("./QRCode/QRCodeRenderer.js")));
/**
 * @internal
 */
const QRCode = (props) => {
    const size = props.size || 310;
    const placeholder = ((0, jsx_runtime_1.jsxs)(QRPlaceholder, { style: {
            width: `${size}px`,
            height: `${size}px`,
        }, children: [(0, jsx_runtime_1.jsx)("span", { "data-v1": true }), (0, jsx_runtime_1.jsx)("span", { "data-v2": true }), (0, jsx_runtime_1.jsx)("span", { "data-v3": true }), (0, jsx_runtime_1.jsx)("div", {})] }));
    return ((0, jsx_runtime_1.jsxs)("div", { style: {
            position: "relative",
            display: "flex",
            justifyContent: "center",
        }, children: [props.qrCodeUri ? ((0, jsx_runtime_1.jsx)(react_2.Suspense, { fallback: placeholder, children: (0, jsx_runtime_1.jsx)(QRCodeContainer, { children: (0, jsx_runtime_1.jsx)(QRCodeRenderer, { uri: props.qrCodeUri, size: size + 20, ecl: "M", clearSize: props.QRIcon ? 70 : undefined }) }) })) : (placeholder), props.QRIcon && (0, jsx_runtime_1.jsx)(IconContainer, { children: props.QRIcon })] }));
};
exports.QRCode = QRCode;
const IconContainer = /* @__PURE__ */ (0, elements_js_1.StyledDiv)({
    position: "absolute",
    top: "50%",
    left: "50%",
    transform: "translate(-50%, -50%)",
    display: "flex",
    justifyContent: "center",
    alignContent: "center",
    zIndex: 1000,
});
const QRCodeContainer = /* @__PURE__ */ (0, elements_js_1.StyledDiv)(() => {
    const theme = (0, CustomThemeProvider_js_1.useCustomTheme)();
    return {
        animation: `${animations_js_1.fadeInAnimation} 600ms ease`,
        "--ck-qr-dot-color": theme.colors.primaryText,
        "--ck-body-background": theme.colors.modalBg,
        "--ck-qr-background": theme.colors.modalBg,
    };
});
const PlaceholderKeyframes = (0, react_1.keyframes) `
  0%{ background-position: 100% 0; }
  100%{ background-position: -100% 0; }
`;
const QRPlaceholder = /* @__PURE__ */ (0, elements_js_1.StyledDiv)((_) => {
    const theme = (0, CustomThemeProvider_js_1.useCustomTheme)();
    return {
        "--color": theme.colors.skeletonBg,
        "--bg": theme.colors.modalBg,
        overflow: "hidden",
        position: "relative",
        display: "flex",
        alignItems: "center",
        justifyContent: "center",
        borderRadius: index_js_1.radius.md,
        "> div": {
            zIndex: 4,
            position: "relative",
            width: "28%",
            height: "28%",
            borderRadius: "5px",
            background: "var(--bg)",
            boxShadow: "0 0 0 7px var(--bg)",
        },
        "> span": {
            zIndex: 4,
            position: "absolute",
            background: "var(--color)",
            borderRadius: "12px",
            width: "13.25%",
            height: "13.25%",
            boxShadow: "0 0 0 4px var(--bg)",
            "&:before": {
                content: '""',
                position: "absolute",
                inset: "9px",
                borderRadius: "3px",
                boxShadow: "0 0 0 4px var(--bg)",
            },
            "&[data-v1]": {
                top: 0,
                left: 0,
            },
            "&[data-v2]": {
                top: 0,
                right: 0,
            },
            "&[data-v3]": {
                bottom: 0,
                left: 0,
            },
        },
        "&:before": {
            zIndex: 3,
            content: '""',
            position: "absolute",
            inset: 0,
            background: "repeat",
            backgroundSize: "1.888% 1.888%",
            backgroundImage: "radial-gradient(var(--color) 41%, transparent 41%)",
        },
        "&::after": {
            zIndex: 100,
            content: '""',
            position: "absolute",
            inset: 0,
            transform: "scale(1.5) rotate(45deg)",
            background: "linear-gradient(90deg, transparent 50%, var(--color), transparent)",
            backgroundSize: "200% 100%",
            animation: `${PlaceholderKeyframes} 1000ms linear infinite both`,
        },
    };
});
