// thirdweb/dist/cjs/react/web/ui/components/Skeleton.js
"use strict";
"use client";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Skeleton = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("@emotion/react");
const CustomThemeProvider_js_1 = require("../../../core/design-system/CustomThemeProvider.js");
const index_js_1 = require("../../../core/design-system/index.js");
const elements_js_1 = require("../design-system/elements.js");
/**
 * @internal
 */
const Skeleton = (props) => {
    return ((0, jsx_runtime_1.jsx)(SkeletonDiv, { color: props.color, style: {
            height: props.height,
            width: props.width || "auto",
        } }));
};
exports.Skeleton = Skeleton;
const skeletonAnimation = (0, react_1.keyframes) `
0% {
    opacity: 0.5;
  }
  100% {
    opacity: 1;
  }
`;
const SkeletonDiv = /* @__PURE__ */ (0, elements_js_1.StyledDiv)((props) => {
    const theme = (0, CustomThemeProvider_js_1.useCustomTheme)();
    return {
        backgroundSize: "200% 200%",
        backgroundColor: theme.colors[props.color || "skeletonBg"],
        animation: `${skeletonAnimation} 500ms ease-in-out infinite alternate`,
        borderRadius: index_js_1.radius.sm,
    };
});
