// thirdweb/dist/cjs/react/web/ui/components/Spinner.js
"use strict";
"use client";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Spinner = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("@emotion/react");
const CustomThemeProvider_js_1 = require("../../../core/design-system/CustomThemeProvider.js");
const index_js_1 = require("../../../core/design-system/index.js");
const elements_js_1 = require("../design-system/elements.js");
/**
 * @internal
 */
const Spinner = (props) => {
    const theme = (0, CustomThemeProvider_js_1.useCustomTheme)();
    return ((0, jsx_runtime_1.jsx)(Svg, { style: {
            width: `${index_js_1.iconSize[props.size]}px`,
            height: `${index_js_1.iconSize[props.size]}px`,
        }, viewBox: "0 0 50 50", children: (0, jsx_runtime_1.jsx)(Circle, { cx: "25", cy: "25", r: "20", fill: "none", stroke: theme.colors[props.color], strokeWidth: Number(index_js_1.iconSize[props.size]) > 64 ? "2" : "4" }) }));
};
exports.Spinner = Spinner;
// animations
const dashAnimation = (0, react_1.keyframes) `
  0% {
    stroke-dasharray: 1, 150;
    stroke-dashoffset: 0;
  }
  50% {
    stroke-dasharray: 90, 150;
    stroke-dashoffset: -35;
  }
  100% {
    stroke-dasharray: 90, 150;
    stroke-dashoffset: -124;
  }
`;
const rotateAnimation = (0, react_1.keyframes) `
  100% {
    transform: rotate(360deg);
  }
`;
const Svg = /* @__PURE__ */ (0, elements_js_1.StyledSvg)({
    animation: `${rotateAnimation} 2s linear infinite`,
    width: "1em",
    height: "1em",
});
const Circle = /* @__PURE__ */ (0, elements_js_1.StyledCircle)({
    strokeLinecap: "round",
    animation: `${dashAnimation} 1.5s ease-in-out infinite`,
});
