// thirdweb/dist/cjs/react/web/ui/components/Tabs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = Tabs;
const jsx_runtime_1 = require("react/jsx-runtime");
const CustomThemeProvider_js_1 = require("../../../core/design-system/CustomThemeProvider.js");
const index_js_1 = require("../../../core/design-system/index.js");
const text_js_1 = require("../components/text.js");
const Spacer_js_1 = require("./Spacer.js");
const basic_js_1 = require("./basic.js");
const buttons_js_1 = require("./buttons.js");
function Tabs({ selected, onSelect, options, children, }) {
    const theme = (0, CustomThemeProvider_js_1.useCustomTheme)();
    return ((0, jsx_runtime_1.jsxs)("div", { children: [(0, jsx_runtime_1.jsx)(basic_js_1.Container, { flex: "row", center: "y", style: { width: "100%", borderRadius: index_js_1.radius.lg }, p: "xxs", bg: "secondaryButtonBg", children: options.map((option) => ((0, jsx_runtime_1.jsx)(buttons_js_1.Button, { variant: "accent", type: "button", onClick: () => onSelect(option.value), style: {
                        flex: 1,
                        paddingBlock: index_js_1.spacing.sm,
                        display: "flex",
                        alignItems: "center",
                        justifyContent: "center",
                        position: "relative",
                        borderRadius: index_js_1.radius.md,
                        backgroundColor: option.value === selected
                            ? theme.colors.modalBg
                            : "transparent",
                    }, children: (0, jsx_runtime_1.jsx)(text_js_1.Text, { color: option.value === selected ? "primaryText" : "secondaryText", style: { textAlign: "center" }, size: "sm", children: option.label }) }, option.value))) }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "sm" }), children] }));
}
