// thirdweb/dist/cjs/react/web/ui/components/TextDivider.js
"use strict";
"use client";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TextDivider = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const CustomThemeProvider_js_1 = require("../../../core/design-system/CustomThemeProvider.js");
const index_js_1 = require("../../../core/design-system/index.js");
const elements_js_1 = require("../design-system/elements.js");
/**
 *
 * @internal
 */
const TextDivider = (props) => {
    return ((0, jsx_runtime_1.jsx)(TextDividerEl, { style: {
            paddingBlock: props.py ? index_js_1.spacing[props.py] : 0,
        }, children: (0, jsx_runtime_1.jsxs)("span", { children: [" ", props.text] }) }));
};
exports.TextDivider = TextDivider;
const TextDividerEl = /* @__PURE__ */ (0, elements_js_1.StyledDiv)(() => {
    const theme = (0, CustomThemeProvider_js_1.useCustomTheme)();
    return {
        display: "flex",
        alignItems: "center",
        color: theme.colors.secondaryText,
        fontSize: index_js_1.fontSize.sm,
        "&::before, &::after": {
            content: '""',
            flex: 1,
            borderBottom: `1px solid ${theme.colors.separatorLine}`,
        },
        span: {
            margin: "0 16px",
        },
    };
});
