// thirdweb/dist/cjs/react/web/ui/components/text.js
"use strict";
"use client";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Link = exports.Text = void 0;
const CustomThemeProvider_js_1 = require("../../../core/design-system/CustomThemeProvider.js");
const index_js_1 = require("../../../core/design-system/index.js");
const elements_js_1 = require("../design-system/elements.js");
exports.Text = (0, elements_js_1.StyledSpan)((p) => {
    const theme = (0, CustomThemeProvider_js_1.useCustomTheme)();
    return {
        fontSize: index_js_1.fontSize[p.size || "md"],
        color: theme.colors[p.color || "secondaryText"],
        margin: 0,
        display: p.inline ? "inline" : "block",
        fontWeight: p.weight || 500,
        lineHeight: p.multiline ? 1.5 : "normal",
        textAlign: p.center ? "center" : "left",
        textWrap: p.balance ? "balance" : "inherit",
        maxWidth: "100%",
        overflow: "hidden",
        textOverflow: "ellipsis",
    };
});
exports.Link = (0, elements_js_1.StyledAnchor)((p) => {
    const theme = (0, CustomThemeProvider_js_1.useCustomTheme)();
    return {
        all: "unset",
        cursor: "pointer",
        color: theme.colors[p.color || "accentText"],
        fontSize: index_js_1.fontSize[p.size || "md"],
        textDecoration: "none",
        textAlign: p.center ? "center" : "left",
        display: p.inline ? "inline" : "block",
        fontWeight: p.weight || 500,
        lineHeight: "normal",
        transition: "color 0.2s ease",
        "&:hover": {
            color: theme.colors[p.hoverColor || "primaryText"],
            textDecoration: "none",
        },
    };
});
