// thirdweb/dist/cjs/extensions/modules/TransferableERC1155/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setTransferableFor = exports.setTransferable = exports.isTransferEnabledFor = exports.isTransferEnabled = exports.encodeInstall = exports.module = exports.install = void 0;
// generated
var install_js_1 = require("../__generated__/TransferableERC1155/module/install.js");
Object.defineProperty(exports, "install", { enumerable: true, get: function () { return install_js_1.install; } });
Object.defineProperty(exports, "module", { enumerable: true, get: function () { return install_js_1.module; } });
Object.defineProperty(exports, "encodeInstall", { enumerable: true, get: function () { return install_js_1.encodeInstall; } });
// read
var isTransferEnabled_js_1 = require("../__generated__/TransferableERC1155/read/isTransferEnabled.js");
Object.defineProperty(exports, "isTransferEnabled", { enumerable: true, get: function () { return isTransferEnabled_js_1.isTransferEnabled; } });
var isTransferEnabledFor_js_1 = require("../__generated__/TransferableERC1155/read/isTransferEnabledFor.js");
Object.defineProperty(exports, "isTransferEnabledFor", { enumerable: true, get: function () { return isTransferEnabledFor_js_1.isTransferEnabledFor; } });
// write
var setTransferable_js_1 = require("../__generated__/TransferableERC1155/write/setTransferable.js");
Object.defineProperty(exports, "setTransferable", { enumerable: true, get: function () { return setTransferable_js_1.setTransferable; } });
var setTransferableFor_js_1 = require("../__generated__/TransferableERC1155/write/setTransferableFor.js");
Object.defineProperty(exports, "setTransferableFor", { enumerable: true, get: function () { return setTransferableFor_js_1.setTransferableFor; } });
