// thirdweb/dist/esm/extensions/ens/__generated__/L2Resolver/read/name.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { encodeAbiParameters } from "../../../../../utils/abi/encodeAbiParameters.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0x691f3431";
const FN_INPUTS = [
    {
        type: "bytes32",
        name: "node",
    },
];
const FN_OUTPUTS = [
    {
        type: "string",
    },
];
/**
 * Checks if the `name` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `name` method is supported.
 * @extension ENS
 * @example
 * ```ts
 * import { isNameSupported } from "thirdweb/extensions/ens";
 * const supported = isNameSupported(["0x..."]);
 * ```
 */
export function isNameSupported(availableSelectors) {
    return detectMethod({
        availableSelectors,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "name" function.
 * @param options - The options for the name function.
 * @returns The encoded ABI parameters.
 * @extension ENS
 * @example
 * ```ts
 * import { encodeNameParams } from "thirdweb/extensions/ens";
 * const result = encodeNameParams({
 *  node: ...,
 * });
 * ```
 */
export function encodeNameParams(options) {
    return encodeAbiParameters(FN_INPUTS, [options.node]);
}
/**
 * Encodes the "name" function into a Hex string with its parameters.
 * @param options - The options for the name function.
 * @returns The encoded hexadecimal string.
 * @extension ENS
 * @example
 * ```ts
 * import { encodeName } from "thirdweb/extensions/ens";
 * const result = encodeName({
 *  node: ...,
 * });
 * ```
 */
export function encodeName(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (FN_SELECTOR +
        encodeNameParams(options).slice(2));
}
/**
 * Decodes the result of the name function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ENS
 * @example
 * ```ts
 * import { decodeNameResult } from "thirdweb/extensions/ens";
 * const result = decodeNameResultResult("...");
 * ```
 */
export function decodeNameResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "name" function on the contract.
 * @param options - The options for the name function.
 * @returns The parsed result of the function call.
 * @extension ENS
 * @example
 * ```ts
 * import { name } from "thirdweb/extensions/ens";
 *
 * const result = await name({
 *  contract,
 *  node: ...,
 * });
 *
 * ```
 */
export async function name(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.node],
    });
}
