// thirdweb/dist/cjs/extensions/vote/__generated__/Vote/read/BALLOT_TYPEHASH.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isBALLOT_TYPEHASHSupported = isBALLOT_TYPEHASHSupported;
exports.decodeBALLOT_TYPEHASHResult = decodeBALLOT_TYPEHASHResult;
exports.BALLOT_TYPEHASH = BALLOT_TYPEHASH;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xdeaaa7cc";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "bytes32",
    },
];
/**
 * Checks if the `BALLOT_TYPEHASH` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `BALLOT_TYPEHASH` method is supported.
 * @extension VOTE
 * @example
 * ```ts
 * import { isBALLOT_TYPEHASHSupported } from "thirdweb/extensions/vote";
 * const supported = isBALLOT_TYPEHASHSupported(["0x..."]);
 * ```
 */
function isBALLOT_TYPEHASHSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the BALLOT_TYPEHASH function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension VOTE
 * @example
 * ```ts
 * import { decodeBALLOT_TYPEHASHResult } from "thirdweb/extensions/vote";
 * const result = decodeBALLOT_TYPEHASHResultResult("...");
 * ```
 */
function decodeBALLOT_TYPEHASHResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "BALLOT_TYPEHASH" function on the contract.
 * @param options - The options for the BALLOT_TYPEHASH function.
 * @returns The parsed result of the function call.
 * @extension VOTE
 * @example
 * ```ts
 * import { BALLOT_TYPEHASH } from "thirdweb/extensions/vote";
 *
 * const result = await BALLOT_TYPEHASH({
 *  contract,
 * });
 *
 * ```
 */
async function BALLOT_TYPEHASH(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
