// thirdweb/dist/cjs/extensions/vote/__generated__/Vote/read/COUNTING_MODE.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isCOUNTING_MODESupported = isCOUNTING_MODESupported;
exports.decodeCOUNTING_MODEResult = decodeCOUNTING_MODEResult;
exports.COUNTING_MODE = COUNTING_MODE;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xdd4e2ba5";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "string",
    },
];
/**
 * Checks if the `COUNTING_MODE` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `COUNTING_MODE` method is supported.
 * @extension VOTE
 * @example
 * ```ts
 * import { isCOUNTING_MODESupported } from "thirdweb/extensions/vote";
 * const supported = isCOUNTING_MODESupported(["0x..."]);
 * ```
 */
function isCOUNTING_MODESupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the COUNTING_MODE function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension VOTE
 * @example
 * ```ts
 * import { decodeCOUNTING_MODEResult } from "thirdweb/extensions/vote";
 * const result = decodeCOUNTING_MODEResultResult("...");
 * ```
 */
function decodeCOUNTING_MODEResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "COUNTING_MODE" function on the contract.
 * @param options - The options for the COUNTING_MODE function.
 * @returns The parsed result of the function call.
 * @extension VOTE
 * @example
 * ```ts
 * import { COUNTING_MODE } from "thirdweb/extensions/vote";
 *
 * const result = await COUNTING_MODE({
 *  contract,
 * });
 *
 * ```
 */
async function COUNTING_MODE(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
