// thirdweb/dist/cjs/extensions/vote/__generated__/Vote/read/proposalDeadline.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isProposalDeadlineSupported = isProposalDeadlineSupported;
exports.encodeProposalDeadlineParams = encodeProposalDeadlineParams;
exports.encodeProposalDeadline = encodeProposalDeadline;
exports.decodeProposalDeadlineResult = decodeProposalDeadlineResult;
exports.proposalDeadline = proposalDeadline;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xc01f9e37";
const FN_INPUTS = [
    {
        type: "uint256",
        name: "proposalId",
    },
];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `proposalDeadline` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `proposalDeadline` method is supported.
 * @extension VOTE
 * @example
 * ```ts
 * import { isProposalDeadlineSupported } from "thirdweb/extensions/vote";
 * const supported = isProposalDeadlineSupported(["0x..."]);
 * ```
 */
function isProposalDeadlineSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "proposalDeadline" function.
 * @param options - The options for the proposalDeadline function.
 * @returns The encoded ABI parameters.
 * @extension VOTE
 * @example
 * ```ts
 * import { encodeProposalDeadlineParams } from "thirdweb/extensions/vote";
 * const result = encodeProposalDeadlineParams({
 *  proposalId: ...,
 * });
 * ```
 */
function encodeProposalDeadlineParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.proposalId]);
}
/**
 * Encodes the "proposalDeadline" function into a Hex string with its parameters.
 * @param options - The options for the proposalDeadline function.
 * @returns The encoded hexadecimal string.
 * @extension VOTE
 * @example
 * ```ts
 * import { encodeProposalDeadline } from "thirdweb/extensions/vote";
 * const result = encodeProposalDeadline({
 *  proposalId: ...,
 * });
 * ```
 */
function encodeProposalDeadline(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeProposalDeadlineParams(options).slice(2));
}
/**
 * Decodes the result of the proposalDeadline function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension VOTE
 * @example
 * ```ts
 * import { decodeProposalDeadlineResult } from "thirdweb/extensions/vote";
 * const result = decodeProposalDeadlineResultResult("...");
 * ```
 */
function decodeProposalDeadlineResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "proposalDeadline" function on the contract.
 * @param options - The options for the proposalDeadline function.
 * @returns The parsed result of the function call.
 * @extension VOTE
 * @example
 * ```ts
 * import { proposalDeadline } from "thirdweb/extensions/vote";
 *
 * const result = await proposalDeadline({
 *  contract,
 *  proposalId: ...,
 * });
 *
 * ```
 */
async function proposalDeadline(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.proposalId],
    });
}
