// thirdweb/dist/cjs/extensions/vote/__generated__/Vote/read/proposalSnapshot.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isProposalSnapshotSupported = isProposalSnapshotSupported;
exports.encodeProposalSnapshotParams = encodeProposalSnapshotParams;
exports.encodeProposalSnapshot = encodeProposalSnapshot;
exports.decodeProposalSnapshotResult = decodeProposalSnapshotResult;
exports.proposalSnapshot = proposalSnapshot;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x2d63f693";
const FN_INPUTS = [
    {
        type: "uint256",
        name: "proposalId",
    },
];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `proposalSnapshot` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `proposalSnapshot` method is supported.
 * @extension VOTE
 * @example
 * ```ts
 * import { isProposalSnapshotSupported } from "thirdweb/extensions/vote";
 * const supported = isProposalSnapshotSupported(["0x..."]);
 * ```
 */
function isProposalSnapshotSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "proposalSnapshot" function.
 * @param options - The options for the proposalSnapshot function.
 * @returns The encoded ABI parameters.
 * @extension VOTE
 * @example
 * ```ts
 * import { encodeProposalSnapshotParams } from "thirdweb/extensions/vote";
 * const result = encodeProposalSnapshotParams({
 *  proposalId: ...,
 * });
 * ```
 */
function encodeProposalSnapshotParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.proposalId]);
}
/**
 * Encodes the "proposalSnapshot" function into a Hex string with its parameters.
 * @param options - The options for the proposalSnapshot function.
 * @returns The encoded hexadecimal string.
 * @extension VOTE
 * @example
 * ```ts
 * import { encodeProposalSnapshot } from "thirdweb/extensions/vote";
 * const result = encodeProposalSnapshot({
 *  proposalId: ...,
 * });
 * ```
 */
function encodeProposalSnapshot(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeProposalSnapshotParams(options).slice(2));
}
/**
 * Decodes the result of the proposalSnapshot function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension VOTE
 * @example
 * ```ts
 * import { decodeProposalSnapshotResult } from "thirdweb/extensions/vote";
 * const result = decodeProposalSnapshotResultResult("...");
 * ```
 */
function decodeProposalSnapshotResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "proposalSnapshot" function on the contract.
 * @param options - The options for the proposalSnapshot function.
 * @returns The parsed result of the function call.
 * @extension VOTE
 * @example
 * ```ts
 * import { proposalSnapshot } from "thirdweb/extensions/vote";
 *
 * const result = await proposalSnapshot({
 *  contract,
 *  proposalId: ...,
 * });
 *
 * ```
 */
async function proposalSnapshot(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.proposalId],
    });
}
