// thirdweb/dist/cjs/extensions/vote/__generated__/Vote/read/quorumDenominator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isQuorumDenominatorSupported = isQuorumDenominatorSupported;
exports.decodeQuorumDenominatorResult = decodeQuorumDenominatorResult;
exports.quorumDenominator = quorumDenominator;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x97c3d334";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `quorumDenominator` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `quorumDenominator` method is supported.
 * @extension VOTE
 * @example
 * ```ts
 * import { isQuorumDenominatorSupported } from "thirdweb/extensions/vote";
 * const supported = isQuorumDenominatorSupported(["0x..."]);
 * ```
 */
function isQuorumDenominatorSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the quorumDenominator function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension VOTE
 * @example
 * ```ts
 * import { decodeQuorumDenominatorResult } from "thirdweb/extensions/vote";
 * const result = decodeQuorumDenominatorResultResult("...");
 * ```
 */
function decodeQuorumDenominatorResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "quorumDenominator" function on the contract.
 * @param options - The options for the quorumDenominator function.
 * @returns The parsed result of the function call.
 * @extension VOTE
 * @example
 * ```ts
 * import { quorumDenominator } from "thirdweb/extensions/vote";
 *
 * const result = await quorumDenominator({
 *  contract,
 * });
 *
 * ```
 */
async function quorumDenominator(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
