// thirdweb/dist/cjs/extensions/vote/__generated__/Vote/read/votingDelay.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isVotingDelaySupported = isVotingDelaySupported;
exports.decodeVotingDelayResult = decodeVotingDelayResult;
exports.votingDelay = votingDelay;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x3932abb1";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `votingDelay` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `votingDelay` method is supported.
 * @extension VOTE
 * @example
 * ```ts
 * import { isVotingDelaySupported } from "thirdweb/extensions/vote";
 * const supported = isVotingDelaySupported(["0x..."]);
 * ```
 */
function isVotingDelaySupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the votingDelay function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension VOTE
 * @example
 * ```ts
 * import { decodeVotingDelayResult } from "thirdweb/extensions/vote";
 * const result = decodeVotingDelayResultResult("...");
 * ```
 */
function decodeVotingDelayResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "votingDelay" function on the contract.
 * @param options - The options for the votingDelay function.
 * @returns The parsed result of the function call.
 * @extension VOTE
 * @example
 * ```ts
 * import { votingDelay } from "thirdweb/extensions/vote";
 *
 * const result = await votingDelay({
 *  contract,
 * });
 *
 * ```
 */
async function votingDelay(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
