// thirdweb/dist/cjs/extensions/airdrop/__generated__/Airdrop/read/tokenConditionId.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isTokenConditionIdSupported = isTokenConditionIdSupported;
exports.encodeTokenConditionIdParams = encodeTokenConditionIdParams;
exports.encodeTokenConditionId = encodeTokenConditionId;
exports.decodeTokenConditionIdResult = decodeTokenConditionIdResult;
exports.tokenConditionId = tokenConditionId;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x3dc28d49";
const FN_INPUTS = [
    {
        type: "address",
        name: "tokenAddress",
    },
];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `tokenConditionId` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `tokenConditionId` method is supported.
 * @extension AIRDROP
 * @example
 * ```ts
 * import { isTokenConditionIdSupported } from "thirdweb/extensions/airdrop";
 * const supported = isTokenConditionIdSupported(["0x..."]);
 * ```
 */
function isTokenConditionIdSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "tokenConditionId" function.
 * @param options - The options for the tokenConditionId function.
 * @returns The encoded ABI parameters.
 * @extension AIRDROP
 * @example
 * ```ts
 * import { encodeTokenConditionIdParams } from "thirdweb/extensions/airdrop";
 * const result = encodeTokenConditionIdParams({
 *  tokenAddress: ...,
 * });
 * ```
 */
function encodeTokenConditionIdParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.tokenAddress]);
}
/**
 * Encodes the "tokenConditionId" function into a Hex string with its parameters.
 * @param options - The options for the tokenConditionId function.
 * @returns The encoded hexadecimal string.
 * @extension AIRDROP
 * @example
 * ```ts
 * import { encodeTokenConditionId } from "thirdweb/extensions/airdrop";
 * const result = encodeTokenConditionId({
 *  tokenAddress: ...,
 * });
 * ```
 */
function encodeTokenConditionId(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeTokenConditionIdParams(options).slice(2));
}
/**
 * Decodes the result of the tokenConditionId function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension AIRDROP
 * @example
 * ```ts
 * import { decodeTokenConditionIdResult } from "thirdweb/extensions/airdrop";
 * const result = decodeTokenConditionIdResultResult("...");
 * ```
 */
function decodeTokenConditionIdResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "tokenConditionId" function on the contract.
 * @param options - The options for the tokenConditionId function.
 * @returns The parsed result of the function call.
 * @extension AIRDROP
 * @example
 * ```ts
 * import { tokenConditionId } from "thirdweb/extensions/airdrop";
 *
 * const result = await tokenConditionId({
 *  contract,
 *  tokenAddress: ...,
 * });
 *
 * ```
 */
async function tokenConditionId(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.tokenAddress],
    });
}
