// thirdweb/dist/esm/react/web/ui/components/TextDivider.js
"use client";
import { jsxs as _jsxs, jsx as _jsx } from "react/jsx-runtime";
import { useCustomTheme } from "../../../core/design-system/CustomThemeProvider.js";
import { fontSize, spacing } from "../../../core/design-system/index.js";
import { StyledDiv } from "../design-system/elements.js";
/**
 *
 * @internal
 */
export const TextDivider = (props) => {
    return (_jsx(TextDividerEl, { style: {
            paddingBlock: props.py ? spacing[props.py] : 0,
        }, children: _jsxs("span", { children: [" ", props.text] }) }));
};
const TextDividerEl = /* @__PURE__ */ StyledDiv(() => {
    const theme = useCustomTheme();
    return {
        display: "flex",
        alignItems: "center",
        color: theme.colors.secondaryText,
        fontSize: fontSize.sm,
        "&::before, &::after": {
            content: '""',
            flex: 1,
            borderBottom: `1px solid ${theme.colors.separatorLine}`,
        },
        span: {
            margin: "0 16px",
        },
    };
});
