// thirdweb/dist/esm/extensions/modules/__generated__/TransferableERC1155/read/isTransferEnabledFor.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { encodeAbiParameters } from "../../../../../utils/abi/encodeAbiParameters.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0x735d0538";
const FN_INPUTS = [
    {
        type: "address",
        name: "target",
    },
];
const FN_OUTPUTS = [
    {
        type: "bool",
    },
];
/**
 * Checks if the `isTransferEnabledFor` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `isTransferEnabledFor` method is supported.
 * @modules TransferableERC1155
 * @example
 * ```ts
 * import { TransferableERC1155 } from "thirdweb/modules";
 * const supported = TransferableERC1155.isIsTransferEnabledForSupported(["0x..."]);
 * ```
 */
export function isIsTransferEnabledForSupported(availableSelectors) {
    return detectMethod({
        availableSelectors,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "isTransferEnabledFor" function.
 * @param options - The options for the isTransferEnabledFor function.
 * @returns The encoded ABI parameters.
 * @modules TransferableERC1155
 * @example
 * ```ts
 * import { TransferableERC1155 } from "thirdweb/modules";
 * const result = TransferableERC1155.encodeIsTransferEnabledForParams({
 *  target: ...,
 * });
 * ```
 */
export function encodeIsTransferEnabledForParams(options) {
    return encodeAbiParameters(FN_INPUTS, [options.target]);
}
/**
 * Encodes the "isTransferEnabledFor" function into a Hex string with its parameters.
 * @param options - The options for the isTransferEnabledFor function.
 * @returns The encoded hexadecimal string.
 * @modules TransferableERC1155
 * @example
 * ```ts
 * import { TransferableERC1155 } from "thirdweb/modules";
 * const result = TransferableERC1155.encodeIsTransferEnabledFor({
 *  target: ...,
 * });
 * ```
 */
export function encodeIsTransferEnabledFor(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (FN_SELECTOR +
        encodeIsTransferEnabledForParams(options).slice(2));
}
/**
 * Decodes the result of the isTransferEnabledFor function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @modules TransferableERC1155
 * @example
 * ```ts
 * import { TransferableERC1155 } from "thirdweb/modules";
 * const result = TransferableERC1155.decodeIsTransferEnabledForResultResult("...");
 * ```
 */
export function decodeIsTransferEnabledForResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "isTransferEnabledFor" function on the contract.
 * @param options - The options for the isTransferEnabledFor function.
 * @returns The parsed result of the function call.
 * @modules TransferableERC1155
 * @example
 * ```ts
 * import { TransferableERC1155 } from "thirdweb/modules";
 *
 * const result = await TransferableERC1155.isTransferEnabledFor({
 *  contract,
 *  target: ...,
 * });
 *
 * ```
 */
export async function isTransferEnabledFor(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.target],
    });
}
