// thirdweb/dist/cjs/react/web/ui/ConnectWallet/Details.js
"use strict";
"use client";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConnectedWalletDetails = void 0;
exports.useWalletDetailsModal = useWalletDetailsModal;
const jsx_runtime_1 = require("react/jsx-runtime");
const styled_1 = require("@emotion/styled");
const react_icons_1 = require("@radix-ui/react-icons");
const react_query_1 = require("@tanstack/react-query");
const react_1 = require("react");
const contract_js_1 = require("../../../../contract/contract.js");
const is_contract_deployed_js_1 = require("../../../../utils/bytecode/is-contract-deployed.js");
const formatNumber_js_1 = require("../../../../utils/formatNumber.js");
const CustomThemeProvider_js_1 = require("../../../core/design-system/CustomThemeProvider.js");
const index_js_1 = require("../../../core/design-system/index.js");
const useChainQuery_js_1 = require("../../../core/hooks/others/useChainQuery.js");
const useActiveAccount_js_1 = require("../../../core/hooks/wallets/useActiveAccount.js");
const useActiveWallet_js_1 = require("../../../core/hooks/wallets/useActiveWallet.js");
const useActiveWalletChain_js_1 = require("../../../core/hooks/wallets/useActiveWalletChain.js");
const useDisconnect_js_1 = require("../../../core/hooks/wallets/useDisconnect.js");
const useSwitchActiveWalletChain_js_1 = require("../../../core/hooks/wallets/useSwitchActiveWalletChain.js");
const RootElementContext_js_1 = require("../../../core/providers/RootElementContext.js");
const isSmartWallet_js_1 = require("../../../core/utils/isSmartWallet.js");
const wallet_js_1 = require("../../../core/utils/wallet.js");
const wallet_ui_states_provider_js_1 = require("../../providers/wallet-ui-states-provider.js");
const ChainIcon_js_1 = require("../components/ChainIcon.js");
const CopyIcon_js_1 = require("../components/CopyIcon.js");
const IconContainer_js_1 = require("../components/IconContainer.js");
const Modal_js_1 = require("../components/Modal.js");
const Skeleton_js_1 = require("../components/Skeleton.js");
const Spacer_js_1 = require("../components/Spacer.js");
const Spinner_js_1 = require("../components/Spinner.js");
const Tooltip_js_1 = require("../components/Tooltip.js");
const WalletImage_js_1 = require("../components/WalletImage.js");
const basic_js_1 = require("../components/basic.js");
const buttons_js_1 = require("../components/buttons.js");
const text_js_1 = require("../components/text.js");
const animations_js_1 = require("../design-system/animations.js");
const elements_js_1 = require("../design-system/elements.js");
const Blobbie_js_1 = require("./Blobbie.js");
const MenuButton_js_1 = require("./MenuButton.js");
const screen_js_1 = require("./Modal/screen.js");
const NetworkSelector_js_1 = require("./NetworkSelector.js");
const TransactionsScreen_js_1 = require("./TransactionsScreen.js");
const constants_js_1 = require("./constants.js");
const CoinsIcon_js_1 = require("./icons/CoinsIcon.js");
const FundsIcon_js_1 = require("./icons/FundsIcon.js");
const OutlineWalletIcon_js_1 = require("./icons/OutlineWalletIcon.js");
const getConnectLocale_js_1 = require("./locale/getConnectLocale.js");
const LazyBuyScreen_js_1 = require("./screens/Buy/LazyBuyScreen.js");
const WalletManagerScreen_js_1 = require("./screens/Details/WalletManagerScreen.js");
const LinkProfileScreen_js_1 = require("./screens/LinkProfileScreen.js");
const LinkedProfilesScreen_js_1 = require("./screens/LinkedProfilesScreen.js");
const ManageWalletScreen_js_1 = require("./screens/ManageWalletScreen.js");
const PrivateKey_js_1 = require("./screens/PrivateKey.js");
const ReceiveFunds_js_1 = require("./screens/ReceiveFunds.js");
const SendFunds_js_1 = require("./screens/SendFunds.js");
const ViewAssets_js_1 = require("./screens/ViewAssets.js");
const ViewNFTs_js_1 = require("./screens/ViewNFTs.js");
const ViewTokens_js_1 = require("./screens/ViewTokens.js");
const WalletConnectReceiverScreen_js_1 = require("./screens/WalletConnectReceiverScreen.js");
const TW_CONNECTED_WALLET = "tw-connected-wallet";
const LocalhostChainId = 1337;
/**
 * @internal
 */
const ConnectedWalletDetails = (props) => {
    const { connectLocale: locale, client } = props;
    const setRootEl = (0, react_1.useContext)(RootElementContext_js_1.SetRootElementContext);
    const activeAccount = (0, useActiveAccount_js_1.useActiveAccount)();
    const walletChain = (0, useActiveWalletChain_js_1.useActiveWalletChain)();
    const { pfp, name, balanceQuery } = (0, wallet_js_1.useConnectedWalletDetails)(client, walletChain, activeAccount, props.detailsButton?.displayBalanceToken);
    function closeModal() {
        setRootEl(null);
    }
    function openModal() {
        setRootEl((0, jsx_runtime_1.jsx)(DetailsModal, { client: client, locale: locale, detailsModal: props.detailsModal, theme: props.theme, supportedTokens: props.supportedTokens, supportedNFTs: props.supportedNFTs, closeModal: closeModal, onDisconnect: props.onDisconnect, chains: props.chains, displayBalanceToken: props.detailsButton?.displayBalanceToken, connectOptions: props.connectOptions }));
    }
    const isNetworkMismatch = props.chain && walletChain && walletChain.id !== props.chain.id;
    if (props.detailsButton?.render) {
        return (
        // biome-ignore lint/a11y/useKeyWithClickEvents: ok
        (0, jsx_runtime_1.jsx)("div", { onClick: openModal, children: (0, jsx_runtime_1.jsx)(props.detailsButton.render, {}) }));
    }
    if (props.chain && isNetworkMismatch) {
        return ((0, jsx_runtime_1.jsx)(SwitchNetworkButton, { style: props.switchButton?.style, className: props.switchButton?.className, switchNetworkBtnTitle: props.switchButton?.label, targetChain: props.chain, connectLocale: locale }));
    }
    const avatarSrc = props.detailsButton?.connectedAccountAvatarUrl || pfp;
    return ((0, jsx_runtime_1.jsxs)(WalletInfoButton, { type: "button", className: `${TW_CONNECTED_WALLET} ${props.detailsButton?.className || ""}`, style: props.detailsButton?.style, "data-test": "connected-wallet-details", onClick: openModal, children: [(0, jsx_runtime_1.jsx)(basic_js_1.Container, { style: {
                    borderRadius: "100%",
                    overflow: "hidden",
                    width: "35px",
                    height: "35px",
                }, children: avatarSrc ? ((0, jsx_runtime_1.jsx)("img", { alt: "", src: avatarSrc, style: {
                        width: "100%",
                        height: "100%",
                    } })) : (activeAccount && (0, jsx_runtime_1.jsx)(Blobbie_js_1.Blobbie, { address: activeAccount.address, size: 35 })) }), (0, jsx_runtime_1.jsxs)(basic_js_1.Container, { flex: "column", gap: "4xs", style: {
                    textOverflow: "ellipsis",
                    whiteSpace: "nowrap",
                    borderRadius: `0 ${index_js_1.radius.md} ${index_js_1.radius.md} 0`,
                }, children: [(0, jsx_runtime_1.jsx)(text_js_1.Text, { size: "xs", color: "primaryText", weight: 500, className: `${TW_CONNECTED_WALLET}__address`, children: props.detailsButton?.connectedAccountName ?? name }), balanceQuery.data ? ((0, jsx_runtime_1.jsxs)(text_js_1.Text, { className: `${TW_CONNECTED_WALLET}__balance`, size: "xs", color: "secondaryText", weight: 400, children: [formatBalanceOnButton(Number(balanceQuery.data.displayValue)), " ", balanceQuery.data.symbol] })) : ((0, jsx_runtime_1.jsx)(Skeleton_js_1.Skeleton, { height: index_js_1.fontSize.xs, width: "70px" }))] })] }));
};
exports.ConnectedWalletDetails = ConnectedWalletDetails;
function DetailsModal(props) {
    const [screen, setScreen] = (0, react_1.useState)("main");
    const { disconnect } = (0, useDisconnect_js_1.useDisconnect)();
    const [isOpen, setIsOpen] = (0, react_1.useState)(true);
    const { client, locale } = props;
    const walletChain = (0, useActiveWalletChain_js_1.useActiveWalletChain)();
    const activeAccount = (0, useActiveAccount_js_1.useActiveAccount)();
    const { pfp, name, balanceQuery } = (0, wallet_js_1.useConnectedWalletDetails)(client, walletChain, activeAccount, props.displayBalanceToken);
    const theme = (0, CustomThemeProvider_js_1.parseTheme)(props.theme);
    const activeWallet = (0, useActiveWallet_js_1.useActiveWallet)();
    const chainIconQuery = (0, useChainQuery_js_1.useChainIconUrl)(walletChain);
    const chainNameQuery = (0, useChainQuery_js_1.useChainName)(walletChain);
    const chainFaucetsQuery = (0, useChainQuery_js_1.useChainFaucets)(walletChain);
    const disableSwitchChain = !activeWallet?.switchChain;
    const screenSetup = (0, screen_js_1.useSetupScreen)({
        size: "compact",
        welcomeScreen: undefined,
        wallets: activeWallet ? [activeWallet] : [],
    });
    function closeModal() {
        setIsOpen(false);
        (0, constants_js_1.onModalUnmount)(() => {
            props.closeModal();
        });
    }
    function handleDisconnect(info) {
        setIsOpen(false);
        props.closeModal();
        props.onDisconnect(info);
    }
    const networkSwitcherButton = ((0, jsx_runtime_1.jsxs)(MenuButton_js_1.MenuButton, { type: "button", disabled: disableSwitchChain, onClick: () => {
            setScreen("network-switcher");
        }, "data-variant": "primary", children: [(0, jsx_runtime_1.jsx)("div", { style: {
                    display: "flex",
                    alignItems: "center",
                    position: "relative",
                }, children: !chainIconQuery.isLoading ? ((0, jsx_runtime_1.jsx)(ChainIcon_js_1.ChainIcon, { chainIconUrl: chainIconQuery.url, size: index_js_1.iconSize.md, active: true, client: client })) : ((0, jsx_runtime_1.jsx)(Skeleton_js_1.Skeleton, { height: `${index_js_1.iconSize.md}px`, width: `${index_js_1.iconSize.md}px` })) }), chainNameQuery.isLoading ? ((0, jsx_runtime_1.jsx)(Skeleton_js_1.Skeleton, { height: "16px", width: "150px" })) : ((0, jsx_runtime_1.jsxs)(text_js_1.Text, { color: "primaryText", size: "md", multiline: true, children: [chainNameQuery.name || `Unknown chain #${walletChain?.id}`, (0, jsx_runtime_1.jsxs)(text_js_1.Text, { color: "secondaryText", size: "xs", children: [balanceQuery.data ? ((0, formatNumber_js_1.formatNumber)(Number(balanceQuery.data.displayValue), 5)) : ((0, jsx_runtime_1.jsx)(Skeleton_js_1.Skeleton, { height: "1em", width: "100px" })), " ", balanceQuery.data?.symbol] })] })), (0, jsx_runtime_1.jsx)(StyledChevronRightIcon, { width: index_js_1.iconSize.sm, height: index_js_1.iconSize.sm, style: {
                    flexShrink: 0,
                    marginLeft: "auto",
                } })] }));
    const avatarSrc = props.detailsModal?.connectedAccountAvatarUrl ?? pfp;
    const { hideSendFunds, hideReceiveFunds, hideBuyFunds } = props.detailsModal || {};
    const hideAllButtons = hideSendFunds && hideReceiveFunds && hideBuyFunds;
    const avatarContent = ((0, jsx_runtime_1.jsxs)(basic_js_1.Container, { style: {
            position: "relative",
            height: `${index_js_1.iconSize.xl}px`,
            width: `${index_js_1.iconSize.xl}px`,
        }, children: [(0, jsx_runtime_1.jsx)(basic_js_1.Container, { style: {
                    width: "100%",
                    height: "100%",
                    borderRadius: "100%",
                    overflow: "hidden",
                }, children: avatarSrc ? ((0, jsx_runtime_1.jsx)("img", { src: avatarSrc, style: {
                        width: index_js_1.iconSize.xxl,
                        height: index_js_1.iconSize.xxl,
                    }, alt: "" })) : (activeAccount && ((0, jsx_runtime_1.jsx)(Blobbie_js_1.Blobbie, { address: activeAccount.address, size: Number(index_js_1.iconSize.xxl) }))) }), !props.detailsModal?.hideSwitchWallet ? ((0, jsx_runtime_1.jsx)(basic_js_1.Container, { style: {
                    position: "absolute",
                    bottom: -2,
                    right: -2,
                }, children: (0, jsx_runtime_1.jsx)(IconContainer_js_1.IconContainer, { style: {
                        background: theme.colors.modalBg,
                    }, padding: "4px", children: activeWallet && ((0, jsx_runtime_1.jsx)(WalletImage_js_1.WalletImage, { style: { borderRadius: 0 }, id: activeWallet.id, client: client, size: "12" })) }) })) : null] }));
    let content = ((0, jsx_runtime_1.jsxs)("div", { children: [(0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "xs" }), (0, jsx_runtime_1.jsxs)(basic_js_1.Container, { px: "lg", gap: "sm", flex: "row", center: "y", style: {
                    paddingTop: index_js_1.spacing.lg,
                    paddingBottom: hideAllButtons ? index_js_1.spacing.md : index_js_1.spacing.lg,
                }, children: [props.detailsModal?.hideSwitchWallet ? (avatarContent) : ((0, jsx_runtime_1.jsx)(Tooltip_js_1.ToolTip, { tip: "Switch wallet", children: (0, jsx_runtime_1.jsx)("div", { style: {
                                cursor: "pointer",
                            }, onKeyDown: (e) => {
                                if (e.key === "w") {
                                    setScreen("wallet-manager");
                                }
                            }, onClick: () => {
                                setScreen("wallet-manager");
                            }, children: avatarContent }) })), (0, jsx_runtime_1.jsxs)(basic_js_1.Container, { flex: "column", gap: "3xs", children: [(0, jsx_runtime_1.jsxs)("div", { style: {
                                    display: "flex",
                                    gap: index_js_1.spacing.xxs,
                                    alignItems: "center",
                                }, children: [(0, jsx_runtime_1.jsx)(text_js_1.Text, { color: "primaryText", weight: 500, size: "md", children: props.detailsModal?.connectedAccountName ?? name }), (0, jsx_runtime_1.jsx)(buttons_js_1.IconButton, { children: (0, jsx_runtime_1.jsx)(CopyIcon_js_1.CopyIcon, { text: activeAccount?.address || "", tip: locale.copyAddress }) })] }), (0, jsx_runtime_1.jsx)(InAppWalletUserInfo, { client: client, locale: locale })] })] }), !hideAllButtons && ((0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [(0, jsx_runtime_1.jsx)(basic_js_1.Container, { px: "lg", children: (0, jsx_runtime_1.jsxs)(basic_js_1.Container, { style: {
                                display: "flex",
                                gap: index_js_1.spacing.xs,
                            }, children: [!hideSendFunds && ((0, jsx_runtime_1.jsxs)(buttons_js_1.Button, { variant: "outline", style: {
                                        fontSize: index_js_1.fontSize.sm,
                                        display: "flex",
                                        gap: index_js_1.spacing.xs,
                                        alignItems: "center",
                                        padding: index_js_1.spacing.sm,
                                        flex: 1,
                                    }, onClick: () => {
                                        setScreen("send");
                                    }, children: [(0, jsx_runtime_1.jsx)(basic_js_1.Container, { color: "secondaryText", flex: "row", center: "both", children: (0, jsx_runtime_1.jsx)(react_icons_1.PaperPlaneIcon, { width: index_js_1.iconSize.sm, height: index_js_1.iconSize.sm, style: {
                                                    transform: "translateY(-10%) rotate(-45deg) ",
                                                } }) }), locale.send] })), !hideReceiveFunds && ((0, jsx_runtime_1.jsxs)(buttons_js_1.Button, { variant: "outline", style: {
                                        fontSize: index_js_1.fontSize.sm,
                                        display: "flex",
                                        gap: index_js_1.spacing.xs,
                                        alignItems: "center",
                                        padding: index_js_1.spacing.sm,
                                        flex: 1,
                                    }, onClick: () => {
                                        setScreen("receive");
                                    }, children: [(0, jsx_runtime_1.jsxs)(basic_js_1.Container, { color: "secondaryText", flex: "row", center: "both", children: [(0, jsx_runtime_1.jsx)(react_icons_1.PinBottomIcon, { width: index_js_1.iconSize.sm, height: index_js_1.iconSize.sm }), " "] }), locale.receive, " "] })), !hideBuyFunds && ((0, jsx_runtime_1.jsxs)(buttons_js_1.Button, { variant: "outline", style: {
                                        fontSize: index_js_1.fontSize.sm,
                                        display: "flex",
                                        gap: index_js_1.spacing.xs,
                                        alignItems: "center",
                                        padding: index_js_1.spacing.sm,
                                        flex: 1,
                                    }, onClick: () => {
                                        setScreen("buy");
                                    }, children: [(0, jsx_runtime_1.jsx)(basic_js_1.Container, { color: "secondaryText", flex: "row", center: "both", children: (0, jsx_runtime_1.jsx)(react_icons_1.PlusIcon, { width: index_js_1.iconSize.sm, height: index_js_1.iconSize.sm }) }), locale.buy] }))] }) }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "md" })] })), (0, jsx_runtime_1.jsxs)(basic_js_1.Container, { px: "md", children: [(0, jsx_runtime_1.jsxs)(basic_js_1.Container, { flex: "column", style: {
                            gap: "1px",
                        }, children: [networkSwitcherButton, (0, jsx_runtime_1.jsxs)(MenuButton_js_1.MenuButton, { onClick: () => {
                                    setScreen("transactions");
                                }, style: {
                                    fontSize: index_js_1.fontSize.sm,
                                }, children: [(0, jsx_runtime_1.jsx)(react_icons_1.TextAlignJustifyIcon, { width: index_js_1.iconSize.md, height: index_js_1.iconSize.md }), (0, jsx_runtime_1.jsx)(basic_js_1.Container, { flex: "row", gap: "xs", center: "y", children: (0, jsx_runtime_1.jsx)(text_js_1.Text, { color: "primaryText", children: locale.transactions }) })] }), (0, jsx_runtime_1.jsxs)(MenuButton_js_1.MenuButton, { onClick: () => {
                                    setScreen("view-assets");
                                }, style: {
                                    fontSize: index_js_1.fontSize.sm,
                                }, children: [(0, jsx_runtime_1.jsx)(CoinsIcon_js_1.CoinsIcon, { size: index_js_1.iconSize.md }), (0, jsx_runtime_1.jsx)(text_js_1.Text, { color: "primaryText", children: props.supportedNFTs
                                            ? locale.viewFunds.viewAssets
                                            : locale.viewFunds.title })] }), (0, jsx_runtime_1.jsxs)(MenuButton_js_1.MenuButton, { onClick: () => {
                                    setScreen("manage-wallet");
                                }, style: {
                                    fontSize: index_js_1.fontSize.sm,
                                }, children: [(0, jsx_runtime_1.jsx)(OutlineWalletIcon_js_1.OutlineWalletIcon, { size: index_js_1.iconSize.md }), (0, jsx_runtime_1.jsx)(text_js_1.Text, { color: "primaryText", children: props.locale.manageWallet.title })] }), (props.detailsModal?.showTestnetFaucet ?? false) &&
                                (chainFaucetsQuery.faucets.length > 0 ||
                                    walletChain?.id === LocalhostChainId) && ((0, jsx_runtime_1.jsxs)(MenuButton_js_1.MenuLink, { href: chainFaucetsQuery.faucets ? chainFaucetsQuery.faucets[0] : "#", target: "_blank", as: "a", style: {
                                    textDecoration: "none",
                                    color: "inherit",
                                }, children: [(0, jsx_runtime_1.jsx)(basic_js_1.Container, { flex: "row", center: "both", color: "secondaryText", children: (0, jsx_runtime_1.jsx)(FundsIcon_js_1.FundsIcon, { size: index_js_1.iconSize.md }) }), locale.requestTestnetFunds] })), props.detailsModal?.footer && ((0, jsx_runtime_1.jsx)(props.detailsModal.footer, { close: closeModal }))] }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "md" })] }), props.detailsModal?.hideDisconnect !== true && ((0, jsx_runtime_1.jsxs)(basic_js_1.Container, { children: [(0, jsx_runtime_1.jsx)(basic_js_1.Line, {}), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "sm" }), (0, jsx_runtime_1.jsx)(basic_js_1.Container, { px: "md", children: (0, jsx_runtime_1.jsxs)(MenuButton_js_1.MenuButton, { "data-variant": "danger", type: "button", onClick: () => {
                                if (activeWallet && activeAccount) {
                                    disconnect(activeWallet);
                                    handleDisconnect({
                                        account: activeAccount,
                                        wallet: activeWallet,
                                    });
                                }
                            }, children: [(0, jsx_runtime_1.jsx)(react_icons_1.ExitIcon, { width: index_js_1.iconSize.md, height: index_js_1.iconSize.md }), (0, jsx_runtime_1.jsx)(text_js_1.Text, { color: "primaryText", children: locale.disconnectWallet })] }) }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "sm" })] }))] }));
    if (screen === "transactions") {
        content = ((0, jsx_runtime_1.jsx)(TransactionsScreen_js_1.TransactionsScreen, { title: locale.buy, onBack: () => setScreen("main"), closeModal: closeModal, locale: locale, setScreen: setScreen, client: client }));
    }
    if (screen === "wallet-manager" &&
        activeAccount &&
        walletChain &&
        activeWallet) {
        content = ((0, jsx_runtime_1.jsx)(WalletManagerScreen_js_1.WalletManagerScreen, { onBack: () => setScreen("main"), accountAbstraction: props.connectOptions?.accountAbstraction, appMetadata: props.connectOptions?.appMetadata, chain: props.connectOptions?.chain, chains: props.connectOptions?.chains, client: client, connectLocale: locale, recommendedWallets: props.connectOptions?.recommendedWallets, showAllWallets: !!props.connectOptions?.showAllWallets, walletConnect: props.connectOptions?.walletConnect, wallets: props.connectOptions?.wallets, activeAccount: activeAccount, activeChain: walletChain, activeWallet: activeWallet }));
    }
    if (screen === "network-switcher") {
        content = ((0, jsx_runtime_1.jsx)(NetworkSelector_js_1.NetworkSelectorContent
        // add currently connected chain to the list of chains if it's not already in the list
        , { 
            // add currently connected chain to the list of chains if it's not already in the list
            chains: walletChain &&
                props.chains.find((c) => c.id === walletChain.id) === undefined
                ? [walletChain, ...props.chains]
                : props.chains, closeModal: closeModal, networkSelector: props.detailsModal?.networkSelector, onBack: () => {
                setScreen("main");
            }, connectLocale: locale, client: client }));
    }
    else if (screen === "view-assets") {
        if (props.supportedNFTs) {
            content = ((0, jsx_runtime_1.jsx)(ViewAssets_js_1.ViewAssets, { supportedTokens: props.supportedTokens, supportedNFTs: props.supportedNFTs, onBack: () => {
                    setScreen("main");
                }, theme: props.theme, setScreen: setScreen, client: client, connectLocale: locale }));
        }
        else {
            // Always show tokens (has the native token at least)
            content = ((0, jsx_runtime_1.jsx)(ViewTokens_js_1.ViewTokens, { supportedTokens: props.supportedTokens, onBack: () => {
                    setScreen("main");
                }, client: client, connectLocale: locale }));
        }
    }
    else if (screen === "view-nfts") {
        content = ((0, jsx_runtime_1.jsx)(ViewNFTs_js_1.ViewNFTs, { theme: props.theme, supportedNFTs: props.supportedNFTs, onBack: () => {
                setScreen("main");
            }, client: client, connectLocale: locale }));
    }
    else if (screen === "view-tokens") {
        content = ((0, jsx_runtime_1.jsx)(ViewTokens_js_1.ViewTokens, { supportedTokens: props.supportedTokens, onBack: () => {
                setScreen("main");
            }, client: client, connectLocale: locale }));
    }
    else if (screen === "private-key") {
        content = ((0, jsx_runtime_1.jsx)(PrivateKey_js_1.PrivateKey, { theme: props.theme, onBack: () => {
                setScreen("manage-wallet");
            }, wallet: activeWallet, client: client }));
    }
    else if (screen === "manage-wallet") {
        content = ((0, jsx_runtime_1.jsx)(ManageWalletScreen_js_1.ManageWalletScreen, { onBack: () => {
                setScreen("main");
            }, locale: locale, closeModal: closeModal, client: client, setScreen: setScreen }));
    }
    else if (screen === "wallet-connect-receiver") {
        content = ((0, jsx_runtime_1.jsx)(WalletConnectReceiverScreen_js_1.WalletConnectReceiverScreen, { onBack: () => {
                setScreen("manage-wallet");
            }, chains: props.chains, client: client }));
    }
    else if (screen === "linked-profiles") {
        content = ((0, jsx_runtime_1.jsx)(LinkedProfilesScreen_js_1.LinkedProfilesScreen, { onBack: () => setScreen("manage-wallet"), client: client, locale: locale, setScreen: setScreen }));
    }
    else if (screen === "link-profile") {
        content = ((0, jsx_runtime_1.jsx)(LinkProfileScreen_js_1.LinkProfileScreen, { onBack: () => {
                setScreen("linked-profiles");
            }, client: client, locale: locale, walletConnect: props.connectOptions?.walletConnect }));
    }
    // send funds
    else if (screen === "send") {
        content = ((0, jsx_runtime_1.jsx)(SendFunds_js_1.SendFunds, { supportedTokens: props.supportedTokens, onBack: () => {
                setScreen("main");
            }, client: client, connectLocale: locale }));
    }
    // receive funds
    else if (screen === "receive") {
        content = ((0, jsx_runtime_1.jsx)(ReceiveFunds_js_1.ReceiveFunds, { walletId: activeWallet?.id, onBack: () => {
                setScreen("main");
            }, client: client, connectLocale: locale }));
    }
    // thirdweb pay
    else if (screen === "buy") {
        content = ((0, jsx_runtime_1.jsx)(LazyBuyScreen_js_1.LazyBuyScreen, { title: locale.buy, isEmbed: false, client: client, onBack: () => setScreen("main"), supportedTokens: props.supportedTokens, connectLocale: locale, payOptions: props.detailsModal?.payOptions || {
                mode: "fund_wallet",
            }, theme: typeof props.theme === "string" ? props.theme : props.theme.type, onDone: closeModal, connectOptions: undefined }));
    }
    return ((0, jsx_runtime_1.jsx)(CustomThemeProvider_js_1.CustomThemeProvider, { theme: props.theme, children: (0, jsx_runtime_1.jsx)(wallet_ui_states_provider_js_1.WalletUIStatesProvider, { theme: props.theme, isOpen: false, children: (0, jsx_runtime_1.jsx)(screen_js_1.ScreenSetupContext.Provider, { value: screenSetup, children: (0, jsx_runtime_1.jsx)(Modal_js_1.Modal, { size: "compact", open: isOpen, setOpen: (_open) => {
                        if (!_open) {
                            closeModal();
                        }
                    }, children: content }) }) }) }));
}
function formatBalanceOnButton(num) {
    return (0, formatNumber_js_1.formatNumber)(num, num < 1 ? 5 : 4);
}
const WalletInfoButton = /* @__PURE__ */ (0, elements_js_1.StyledButton)((_) => {
    const theme = (0, CustomThemeProvider_js_1.useCustomTheme)();
    return {
        all: "unset",
        background: theme.colors.connectedButtonBg,
        overflow: "hidden",
        borderRadius: index_js_1.radius.md,
        cursor: "pointer",
        display: "inline-flex",
        gap: index_js_1.spacing.xs,
        padding: index_js_1.spacing.xs,
        alignItems: "center",
        minWidth: "165px",
        height: "50px",
        boxSizing: "border-box",
        border: `1px solid ${theme.colors.borderColor}`,
        WebkitTapHighlightColor: "transparent",
        lineHeight: "normal",
        animation: `${animations_js_1.fadeInAnimation} 300ms ease`,
        "&:hover": {
            transition: "background 250ms ease",
            background: theme.colors.connectedButtonBgHover,
        },
    };
});
const StyledChevronRightIcon = /* @__PURE__ */ (0, styled_1.default)(
/* @__PURE__ */ react_icons_1.ChevronRightIcon)(() => {
    const theme = (0, CustomThemeProvider_js_1.useCustomTheme)();
    return {
        color: theme.colors.secondaryText,
    };
});
function ConnectedToSmartWallet(props) {
    const activeAccount = (0, useActiveAccount_js_1.useActiveAccount)();
    const activeWallet = (0, useActiveWallet_js_1.useActiveWallet)();
    const isSmartWallet = (0, isSmartWallet_js_1.hasSmartAccount)(activeWallet);
    const chain = (0, useActiveWalletChain_js_1.useActiveWalletChain)();
    const { client, connectLocale: locale } = props;
    const [isSmartWalletDeployed, setIsSmartWalletDeployed] = (0, react_1.useState)(false);
    (0, react_1.useEffect)(() => {
        if (activeAccount && isSmartWallet && activeAccount.address && chain) {
            const contract = (0, contract_js_1.getContract)({
                address: activeAccount.address,
                chain,
                client,
            });
            (0, is_contract_deployed_js_1.isContractDeployed)(contract).then((isDeployed) => {
                setIsSmartWalletDeployed(isDeployed);
            });
        }
        else {
            setIsSmartWalletDeployed(false);
        }
    }, [activeAccount, chain, client, isSmartWallet]);
    const content = ((0, jsx_runtime_1.jsx)(basic_js_1.Container, { flex: "row", gap: "3xs", center: "y", children: (0, jsx_runtime_1.jsx)(text_js_1.Text, { size: "xs", weight: 400, color: "secondaryText", children: locale.connectedToSmartWallet }) }));
    if (chain && activeAccount && isSmartWallet) {
        return ((0, jsx_runtime_1.jsx)(jsx_runtime_1.Fragment, { children: isSmartWalletDeployed ? ((0, jsx_runtime_1.jsx)(text_js_1.Link, { color: "secondaryText", hoverColor: "primaryText", href: `https://thirdweb.com/${chain.id}/${activeAccount.address}/account`, target: "_blank", size: "sm", children: content })) : ((0, jsx_runtime_1.jsxs)(text_js_1.Text, { size: "sm", children: [" ", content] })) }));
    }
    return null;
}
function InAppWalletUserInfo(props) {
    const { client, locale } = props;
    const account = (0, useActiveAccount_js_1.useActiveAccount)();
    const activeWallet = (0, useActiveWallet_js_1.useActiveWallet)();
    const { data: walletInfo } = (0, wallet_js_1.useWalletInfo)(activeWallet?.id);
    const isSmartWallet = (0, isSmartWallet_js_1.hasSmartAccount)(activeWallet);
    const userInfoQuery = (0, react_query_1.useQuery)({
        queryKey: ["in-app-wallet-user", client, account?.address],
        queryFn: async () => {
            const { getUserEmail, getUserPhoneNumber } = await Promise.resolve().then(() => require("../../../../wallets/in-app/web/lib/auth/index.js"));
            const [email, phone] = await Promise.all([
                getUserEmail({
                    client: client,
                }),
                getUserPhoneNumber({
                    client: client,
                }),
            ]);
            return email || phone || null;
        },
        enabled: !isSmartWallet,
    });
    if (isSmartWallet) {
        return (0, jsx_runtime_1.jsx)(ConnectedToSmartWallet, { client: client, connectLocale: locale });
    }
    if (userInfoQuery.data || walletInfo) {
        return ((0, jsx_runtime_1.jsx)(text_js_1.Text, { size: "xs", weight: 400, children: userInfoQuery.data || walletInfo?.name }));
    }
    return (0, jsx_runtime_1.jsx)(Skeleton_js_1.Skeleton, { width: "50px", height: "10px" });
}
/**
 * @internal
 */
function SwitchNetworkButton(props) {
    const switchChain = (0, useSwitchActiveWalletChain_js_1.useSwitchActiveWalletChain)();
    const [switching, setSwitching] = (0, react_1.useState)(false);
    const locale = props.connectLocale;
    const switchNetworkBtnTitle = props.switchNetworkBtnTitle ?? locale.switchNetwork;
    return ((0, jsx_runtime_1.jsx)(buttons_js_1.Button, { className: `tw-connect-wallet--switch-network ${props.className || ""}`, variant: "primary", type: "button", "data-is-loading": switching, "data-test": "switch-network-button", disabled: switching, onClick: async () => {
            setSwitching(true);
            try {
                await switchChain(props.targetChain);
            }
            catch (e) {
                console.error(e);
            }
            setSwitching(false);
        }, style: {
            minWidth: "140px",
            ...props.style,
        }, "aria-label": switching ? locale.switchingNetwork : undefined, children: switching ? ((0, jsx_runtime_1.jsx)(Spinner_js_1.Spinner, { size: "sm", color: "primaryButtonText" })) : (switchNetworkBtnTitle) }));
}
/**
 * Hook to open the Wallet Details Modal that shows various information about the connected wallet and allows users to perform various actions like sending funds, receiving funds, switching networks, Buying tokens, etc.
 *
 * @example
 * ```tsx
 * import { createThirdwebClient } from "thirdweb";
 * import { useWalletDetailsModal } from "thirdweb/react";
 *
 * const client = createThirdwebClient({
 *  clientId: "<your_client_id>",
 * });
 *
 * function Example() {
 *   const detailsModal = useWalletDetailsModal();
 *
 *   function handleClick() {
 *      detailsModal.open({ client, theme: 'light' });
 *   }
 *
 *   return <button onClick={handleClick}> Show Wallet Details </button>
 * }
 * ```
 * @wallet
 */
function useWalletDetailsModal() {
    const account = (0, useActiveAccount_js_1.useActiveAccount)();
    const setRootEl = (0, react_1.useContext)(RootElementContext_js_1.SetRootElementContext);
    function closeModal() {
        setRootEl(null);
    }
    function openModal(props) {
        if (!account) {
            throw new Error("Wallet is not connected.");
        }
        (0, getConnectLocale_js_1.getConnectLocale)(props.locale || "en_US")
            .then((locale) => {
            setRootEl((0, jsx_runtime_1.jsx)(DetailsModal, { client: props.client, locale: locale, detailsModal: {
                    footer: props.footer,
                    hideDisconnect: props.hideDisconnect,
                    hideSwitchWallet: props.hideSwitchWallet,
                    networkSelector: props.networkSelector,
                    payOptions: props.payOptions,
                    showTestnetFaucet: props.showTestnetFaucet,
                    connectedAccountName: props.connectedAccountName,
                    connectedAccountAvatarUrl: props.connectedAccountAvatarUrl,
                    hideBuyFunds: props.hideBuyFunds,
                    hideReceiveFunds: props.hideReceiveFunds,
                    hideSendFunds: props.hideSendFunds,
                }, displayBalanceToken: props.displayBalanceToken, theme: props.theme || "dark", supportedTokens: props.supportedTokens, supportedNFTs: props.supportedNFTs, closeModal: closeModal, onDisconnect: (info) => {
                    props.onDisconnect?.(info);
                    closeModal();
                }, chains: props.chains || [], connectOptions: props.connectOptions }));
        })
            .catch(() => {
            closeModal();
        });
    }
    return {
        open: openModal,
    };
}
