// thirdweb/dist/esm/exports/wallets.js
// --- KEEEP IN SYNC with exports/wallets.native.ts ---
export { createWallet, walletConnect, } from "../wallets/create-wallet.js";
export { inAppWallet, 
/**
 * @deprecated use inAppWallet instead
 */
inAppWallet as embeddedWallet, } from "../wallets/in-app/web/in-app.js";
export { ecosystemWallet } from "../wallets/in-app/web/ecosystem.js";
export { smartWallet } from "../wallets/smart/smart-wallet.js";
// utils
export { getWalletBalance, } from "../wallets/utils/getWalletBalance.js";
export { generateAccount, } from "../wallets/utils/generateAccount.js";
// private-key
export { privateKeyToAccount, 
/**
 * @internal
 * @deprecated - use {@link privateKeyToAccount} instead
 */
privateKeyToAccount as privateKeyAccount, } from "../wallets/private-key.js";
export { getProfiles, linkProfile, } from "../wallets/in-app/core/wallet/profiles.js";
export { getAllWalletsList } from "../wallets/getAllWalletsList.js";
export { getWalletInfo } from "../wallets/__generated__/getWalletInfo.js";
export {} from "../wallets/wallet-info.js";
export { createWalletAdapter } from "../adapters/wallet-adapter.js";
// wallet connect
export { createWalletConnectClient, createWalletConnectSession, disconnectWalletConnectSession, getActiveWalletConnectSessions, DefaultWalletConnectRequestHandlers, } from "../wallets/wallet-connect/receiver/index.js";
// WEB ONLY EXPORTS
// injected
export { injectedProvider } from "../wallets/injected/mipdStore.js";
