// thirdweb/dist/cjs/exports/wallets/in-app.native.js
"use strict";
// --- KEEEP IN SYNC with exports/wallets/in-app.ts ---
Object.defineProperty(exports, "__esModule", { value: true });
exports.hasStoredPasskey = exports.getUserPhoneNumber = exports.getUserEmail = exports.authenticate = exports.preAuthenticate = exports.inAppWallet = void 0;
var in_app_js_1 = require("../../wallets/in-app/native/in-app.js");
Object.defineProperty(exports, "inAppWallet", { enumerable: true, get: function () { return in_app_js_1.inAppWallet; } });
var index_js_1 = require("../../wallets/in-app/native/auth/index.js");
Object.defineProperty(exports, "preAuthenticate", { enumerable: true, get: function () { return index_js_1.preAuthenticate; } });
Object.defineProperty(exports, "authenticate", { enumerable: true, get: function () { return index_js_1.authenticate; } });
Object.defineProperty(exports, "getUserEmail", { enumerable: true, get: function () { return index_js_1.getUserEmail; } });
Object.defineProperty(exports, "getUserPhoneNumber", { enumerable: true, get: function () { return index_js_1.getUserPhoneNumber; } });
var passkeys_js_1 = require("../../wallets/in-app/native/auth/passkeys.js");
Object.defineProperty(exports, "hasStoredPasskey", { enumerable: true, get: function () { return passkeys_js_1.hasStoredPasskey; } });
