// thirdweb/dist/cjs/extensions/modules/__generated__/BatchMetadataERC721/read/getAllMetadataBatches.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isGetAllMetadataBatchesSupported = isGetAllMetadataBatchesSupported;
exports.decodeGetAllMetadataBatchesResult = decodeGetAllMetadataBatchesResult;
exports.getAllMetadataBatches = getAllMetadataBatches;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xe6c23512";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "tuple[]",
        components: [
            {
                type: "uint256",
                name: "startTokenIdInclusive",
            },
            {
                type: "uint256",
                name: "endTokenIdInclusive",
            },
            {
                type: "string",
                name: "baseURI",
            },
        ],
    },
];
/**
 * Checks if the `getAllMetadataBatches` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `getAllMetadataBatches` method is supported.
 * @modules BatchMetadataERC721
 * @example
 * ```ts
 * import { BatchMetadataERC721 } from "thirdweb/modules";
 * const supported = BatchMetadataERC721.isGetAllMetadataBatchesSupported(["0x..."]);
 * ```
 */
function isGetAllMetadataBatchesSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the getAllMetadataBatches function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @modules BatchMetadataERC721
 * @example
 * ```ts
 * import { BatchMetadataERC721 } from "thirdweb/modules";
 * const result = BatchMetadataERC721.decodeGetAllMetadataBatchesResultResult("...");
 * ```
 */
function decodeGetAllMetadataBatchesResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getAllMetadataBatches" function on the contract.
 * @param options - The options for the getAllMetadataBatches function.
 * @returns The parsed result of the function call.
 * @modules BatchMetadataERC721
 * @example
 * ```ts
 * import { BatchMetadataERC721 } from "thirdweb/modules";
 *
 * const result = await BatchMetadataERC721.getAllMetadataBatches({
 *  contract,
 * });
 *
 * ```
 */
async function getAllMetadataBatches(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
