// thirdweb/dist/cjs/wallets/in-app/core/constants/settings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEVICE_SHARE_LOCAL_STORAGE_NAME_DEPRECATED = exports.WALLET_CONNECT_SESSIONS_LOCAL_STORAGE_NAME = exports.DEVICE_SHARE_LOCAL_STORAGE_NAME = exports.PASSKEY_CREDENTIAL_ID_LOCAL_STORAGE_NAME = exports.AUTH_TOKEN_LOCAL_STORAGE_NAME = exports.WALLET_USER_ID_LOCAL_STORAGE_NAME = exports.WALLET_USER_DETAILS_LOCAL_STORAGE_NAME = exports.BASE_IN_APP_WALLET_URL = exports.HEADLESS_GOOGLE_OAUTH_ROUTE = exports.IN_APP_WALLET_PATH = void 0;
/**
 * @internal
 */
exports.IN_APP_WALLET_PATH = "/sdk/2022-08-12/embedded-wallet";
/**
 * @internal
 */
exports.HEADLESS_GOOGLE_OAUTH_ROUTE = "/auth/headless-google-login-managed";
/**
 * @internal
 */
exports.BASE_IN_APP_WALLET_URL = "embedded-wallet.thirdweb.com";
// STORAGE
/**
 * @internal
 */
const WALLET_USER_DETAILS_LOCAL_STORAGE_NAME = (key) => `thirdwebEwsWalletUserDetails-${key}`;
exports.WALLET_USER_DETAILS_LOCAL_STORAGE_NAME = WALLET_USER_DETAILS_LOCAL_STORAGE_NAME;
/**
 * @internal
 */
const WALLET_USER_ID_LOCAL_STORAGE_NAME = (cliekeytId) => `thirdwebEwsWalletUserId-${cliekeytId}`;
exports.WALLET_USER_ID_LOCAL_STORAGE_NAME = WALLET_USER_ID_LOCAL_STORAGE_NAME;
/**
 * @internal
 */
const AUTH_TOKEN_LOCAL_STORAGE_PREFIX = "walletToken";
/**
 * @internal
 */
const AUTH_TOKEN_LOCAL_STORAGE_NAME = (key) => {
    return `${AUTH_TOKEN_LOCAL_STORAGE_PREFIX}-${key}`;
};
exports.AUTH_TOKEN_LOCAL_STORAGE_NAME = AUTH_TOKEN_LOCAL_STORAGE_NAME;
/**
 * @internal
 */
const PASSKEY_CREDENTIAL_ID_LOCAL_STORAGE_NAME = (key) => {
    return `passkey-credential-id-${key}`;
};
exports.PASSKEY_CREDENTIAL_ID_LOCAL_STORAGE_NAME = PASSKEY_CREDENTIAL_ID_LOCAL_STORAGE_NAME;
/**
 * @internal
 */
const DEVICE_SHARE_LOCAL_STORAGE_PREFIX = "a";
/**
 * @internal
 */
const DEVICE_SHARE_LOCAL_STORAGE_NAME = (key, userId) => `${DEVICE_SHARE_LOCAL_STORAGE_PREFIX}-${key}-${userId}`;
exports.DEVICE_SHARE_LOCAL_STORAGE_NAME = DEVICE_SHARE_LOCAL_STORAGE_NAME;
/**
 * @internal
 */
const WALLET_CONNECT_SESSIONS_LOCAL_STORAGE_NAME = (key) => `walletConnectSessions-${key}`;
exports.WALLET_CONNECT_SESSIONS_LOCAL_STORAGE_NAME = WALLET_CONNECT_SESSIONS_LOCAL_STORAGE_NAME;
/**
 * @internal
 */
const DEVICE_SHARE_LOCAL_STORAGE_NAME_DEPRECATED = (key) => `${DEVICE_SHARE_LOCAL_STORAGE_PREFIX}-${key}`;
exports.DEVICE_SHARE_LOCAL_STORAGE_NAME_DEPRECATED = DEVICE_SHARE_LOCAL_STORAGE_NAME_DEPRECATED;
