// thirdweb/dist/cjs/extensions/split/__generated__/Split/read/released.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isReleasedSupported = isReleasedSupported;
exports.encodeReleasedParams = encodeReleasedParams;
exports.encodeReleased = encodeReleased;
exports.decodeReleasedResult = decodeReleasedResult;
exports.released = released;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x9852595c";
const FN_INPUTS = [
    {
        type: "address",
        name: "account",
    },
];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `released` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `released` method is supported.
 * @extension SPLIT
 * @example
 * ```ts
 * import { isReleasedSupported } from "thirdweb/extensions/split";
 * const supported = isReleasedSupported(["0x..."]);
 * ```
 */
function isReleasedSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "released" function.
 * @param options - The options for the released function.
 * @returns The encoded ABI parameters.
 * @extension SPLIT
 * @example
 * ```ts
 * import { encodeReleasedParams } from "thirdweb/extensions/split";
 * const result = encodeReleasedParams({
 *  account: ...,
 * });
 * ```
 */
function encodeReleasedParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.account]);
}
/**
 * Encodes the "released" function into a Hex string with its parameters.
 * @param options - The options for the released function.
 * @returns The encoded hexadecimal string.
 * @extension SPLIT
 * @example
 * ```ts
 * import { encodeReleased } from "thirdweb/extensions/split";
 * const result = encodeReleased({
 *  account: ...,
 * });
 * ```
 */
function encodeReleased(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeReleasedParams(options).slice(2));
}
/**
 * Decodes the result of the released function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension SPLIT
 * @example
 * ```ts
 * import { decodeReleasedResult } from "thirdweb/extensions/split";
 * const result = decodeReleasedResultResult("...");
 * ```
 */
function decodeReleasedResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "released" function on the contract.
 * @param options - The options for the released function.
 * @returns The parsed result of the function call.
 * @extension SPLIT
 * @example
 * ```ts
 * import { released } from "thirdweb/extensions/split";
 *
 * const result = await released({
 *  contract,
 *  account: ...,
 * });
 *
 * ```
 */
async function released(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.account],
    });
}
