// thirdweb/dist/cjs/extensions/split/__generated__/Split/read/totalShares.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isTotalSharesSupported = isTotalSharesSupported;
exports.decodeTotalSharesResult = decodeTotalSharesResult;
exports.totalShares = totalShares;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x3a98ef39";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `totalShares` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `totalShares` method is supported.
 * @extension SPLIT
 * @example
 * ```ts
 * import { isTotalSharesSupported } from "thirdweb/extensions/split";
 * const supported = isTotalSharesSupported(["0x..."]);
 * ```
 */
function isTotalSharesSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the totalShares function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension SPLIT
 * @example
 * ```ts
 * import { decodeTotalSharesResult } from "thirdweb/extensions/split";
 * const result = decodeTotalSharesResultResult("...");
 * ```
 */
function decodeTotalSharesResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "totalShares" function on the contract.
 * @param options - The options for the totalShares function.
 * @returns The parsed result of the function call.
 * @extension SPLIT
 * @example
 * ```ts
 * import { totalShares } from "thirdweb/extensions/split";
 *
 * const result = await totalShares({
 *  contract,
 * });
 *
 * ```
 */
async function totalShares(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
