// thirdweb/dist/cjs/extensions/lens/__generated__/ModuleRegistry/read/isErc20CurrencyRegistered.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isIsErc20CurrencyRegisteredSupported = isIsErc20CurrencyRegisteredSupported;
exports.encodeIsErc20CurrencyRegisteredParams = encodeIsErc20CurrencyRegisteredParams;
exports.encodeIsErc20CurrencyRegistered = encodeIsErc20CurrencyRegistered;
exports.decodeIsErc20CurrencyRegisteredResult = decodeIsErc20CurrencyRegisteredResult;
exports.isErc20CurrencyRegistered = isErc20CurrencyRegistered;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xf21b24d7";
const FN_INPUTS = [
    {
        type: "address",
        name: "currencyAddress",
    },
];
const FN_OUTPUTS = [
    {
        type: "bool",
    },
];
/**
 * Checks if the `isErc20CurrencyRegistered` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `isErc20CurrencyRegistered` method is supported.
 * @extension LENS
 * @example
 * ```ts
 * import { isIsErc20CurrencyRegisteredSupported } from "thirdweb/extensions/lens";
 * const supported = isIsErc20CurrencyRegisteredSupported(["0x..."]);
 * ```
 */
function isIsErc20CurrencyRegisteredSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "isErc20CurrencyRegistered" function.
 * @param options - The options for the isErc20CurrencyRegistered function.
 * @returns The encoded ABI parameters.
 * @extension LENS
 * @example
 * ```ts
 * import { encodeIsErc20CurrencyRegisteredParams } from "thirdweb/extensions/lens";
 * const result = encodeIsErc20CurrencyRegisteredParams({
 *  currencyAddress: ...,
 * });
 * ```
 */
function encodeIsErc20CurrencyRegisteredParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.currencyAddress]);
}
/**
 * Encodes the "isErc20CurrencyRegistered" function into a Hex string with its parameters.
 * @param options - The options for the isErc20CurrencyRegistered function.
 * @returns The encoded hexadecimal string.
 * @extension LENS
 * @example
 * ```ts
 * import { encodeIsErc20CurrencyRegistered } from "thirdweb/extensions/lens";
 * const result = encodeIsErc20CurrencyRegistered({
 *  currencyAddress: ...,
 * });
 * ```
 */
function encodeIsErc20CurrencyRegistered(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeIsErc20CurrencyRegisteredParams(options).slice(2));
}
/**
 * Decodes the result of the isErc20CurrencyRegistered function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension LENS
 * @example
 * ```ts
 * import { decodeIsErc20CurrencyRegisteredResult } from "thirdweb/extensions/lens";
 * const result = decodeIsErc20CurrencyRegisteredResultResult("...");
 * ```
 */
function decodeIsErc20CurrencyRegisteredResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "isErc20CurrencyRegistered" function on the contract.
 * @param options - The options for the isErc20CurrencyRegistered function.
 * @returns The parsed result of the function call.
 * @extension LENS
 * @example
 * ```ts
 * import { isErc20CurrencyRegistered } from "thirdweb/extensions/lens";
 *
 * const result = await isErc20CurrencyRegistered({
 *  contract,
 *  currencyAddress: ...,
 * });
 *
 * ```
 */
async function isErc20CurrencyRegistered(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.currencyAddress],
    });
}
