// thirdweb/dist/cjs/extensions/lens/__generated__/ModuleRegistry/read/isModuleRegisteredAs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isIsModuleRegisteredAsSupported = isIsModuleRegisteredAsSupported;
exports.encodeIsModuleRegisteredAsParams = encodeIsModuleRegisteredAsParams;
exports.encodeIsModuleRegisteredAs = encodeIsModuleRegisteredAs;
exports.decodeIsModuleRegisteredAsResult = decodeIsModuleRegisteredAsResult;
exports.isModuleRegisteredAs = isModuleRegisteredAs;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xc2b62fdd";
const FN_INPUTS = [
    {
        type: "address",
        name: "moduleAddress",
    },
    {
        type: "uint256",
        name: "moduleType",
    },
];
const FN_OUTPUTS = [
    {
        type: "bool",
    },
];
/**
 * Checks if the `isModuleRegisteredAs` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `isModuleRegisteredAs` method is supported.
 * @extension LENS
 * @example
 * ```ts
 * import { isIsModuleRegisteredAsSupported } from "thirdweb/extensions/lens";
 * const supported = isIsModuleRegisteredAsSupported(["0x..."]);
 * ```
 */
function isIsModuleRegisteredAsSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "isModuleRegisteredAs" function.
 * @param options - The options for the isModuleRegisteredAs function.
 * @returns The encoded ABI parameters.
 * @extension LENS
 * @example
 * ```ts
 * import { encodeIsModuleRegisteredAsParams } from "thirdweb/extensions/lens";
 * const result = encodeIsModuleRegisteredAsParams({
 *  moduleAddress: ...,
 *  moduleType: ...,
 * });
 * ```
 */
function encodeIsModuleRegisteredAsParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [
        options.moduleAddress,
        options.moduleType,
    ]);
}
/**
 * Encodes the "isModuleRegisteredAs" function into a Hex string with its parameters.
 * @param options - The options for the isModuleRegisteredAs function.
 * @returns The encoded hexadecimal string.
 * @extension LENS
 * @example
 * ```ts
 * import { encodeIsModuleRegisteredAs } from "thirdweb/extensions/lens";
 * const result = encodeIsModuleRegisteredAs({
 *  moduleAddress: ...,
 *  moduleType: ...,
 * });
 * ```
 */
function encodeIsModuleRegisteredAs(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeIsModuleRegisteredAsParams(options).slice(2));
}
/**
 * Decodes the result of the isModuleRegisteredAs function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension LENS
 * @example
 * ```ts
 * import { decodeIsModuleRegisteredAsResult } from "thirdweb/extensions/lens";
 * const result = decodeIsModuleRegisteredAsResultResult("...");
 * ```
 */
function decodeIsModuleRegisteredAsResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "isModuleRegisteredAs" function on the contract.
 * @param options - The options for the isModuleRegisteredAs function.
 * @returns The parsed result of the function call.
 * @extension LENS
 * @example
 * ```ts
 * import { isModuleRegisteredAs } from "thirdweb/extensions/lens";
 *
 * const result = await isModuleRegisteredAs({
 *  contract,
 *  moduleAddress: ...,
 *  moduleType: ...,
 * });
 *
 * ```
 */
async function isModuleRegisteredAs(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.moduleAddress, options.moduleType],
    });
}
