// thirdweb/dist/esm/extensions/split/__generated__/Split/read/payeeCount.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0x00dbe109";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `payeeCount` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `payeeCount` method is supported.
 * @extension SPLIT
 * @example
 * ```ts
 * import { isPayeeCountSupported } from "thirdweb/extensions/split";
 * const supported = isPayeeCountSupported(["0x..."]);
 * ```
 */
export function isPayeeCountSupported(availableSelectors) {
    return detectMethod({
        availableSelectors,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the payeeCount function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension SPLIT
 * @example
 * ```ts
 * import { decodePayeeCountResult } from "thirdweb/extensions/split";
 * const result = decodePayeeCountResultResult("...");
 * ```
 */
export function decodePayeeCountResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "payeeCount" function on the contract.
 * @param options - The options for the payeeCount function.
 * @returns The parsed result of the function call.
 * @extension SPLIT
 * @example
 * ```ts
 * import { payeeCount } from "thirdweb/extensions/split";
 *
 * const result = await payeeCount({
 *  contract,
 * });
 *
 * ```
 */
export async function payeeCount(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
