// thirdweb/dist/cjs/event/actions/get-events.js
"use strict";
/**
 * 1. blockTime + contract (with abi) + no events -> logs with types and parsing *if* contract has abi defined
 * 2. blockTime + contract (no abi) + no events -> logs with NO types but *with* parsing
 * 3. blockTime + no contract + events -> logs with types and parsing (across all "addresses") (no contract filter)
 * 4. blockTime + contract + events -> logs with types and parsing (filtered by contract address +  event topics)
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getContractEvents = getContractEvents;
const resolve_abi_js_1 = require("../../contract/actions/resolve-abi.js");
const eth_blockNumber_js_1 = require("../../rpc/actions/eth_blockNumber.js");
const eth_getLogs_js_1 = require("../../rpc/actions/eth_getLogs.js");
const rpc_js_1 = require("../../rpc/rpc.js");
const prepare_event_js_1 = require("../prepare-event.js");
const utils_js_1 = require("../utils.js");
const parse_logs_js_1 = require("./parse-logs.js");
/**
 * Retrieves events from a contract based on the provided options.
 * @param options - The options for retrieving events.
 * @returns A promise that resolves to an array of parsed event logs.
 * Note: toBlock and fromBlock are both inclusive.
 * @example
 * ```ts
 * import { getContractEvents } from "thirdweb";
 * const events = await getContractEvents({
 *  contract: myContract,
 *  fromBlock: 123456n,
 *  toBlock: 123456n,
 *  events: [preparedEvent, preparedEvent2],
 * });
 * ```
 * @example
 * Optionally specify a blockRange as the number of blocks to retrieve. toBlock will default to the current block number.
 * ```ts
 * import { getContractEvents } from "thirdweb";
 * const events = await getContractEvents({
 *  contract: myContract,
 *  blockRange: 123456n,
 *  events: [preparedEvent, preparedEvent2],
 * });
 * ```
 * @example
 * Use fromBlock with blockRange for pagination.
 * ```ts
 * import { getContractEvents } from "thirdweb";
 * const events = await getContractEvents({
 *  contract: myContract,
 *  fromBlock: lastBlockFetched,
 *  blockRange: 123456n,
 *  events: [preparedEvent, preparedEvent2],
 * });
 * ```
 * @example
 * Retrieve events for a specific block hash.
 * ```ts
 * import { getContractEvents } from "thirdweb";
 * const events = await getContractEvents({
 *  contract: myContract,
 *  blockHash: "0x...",
 *  events: [preparedEvent, preparedEvent2],
 * });
 * ```
 * @contract
 */
async function getContractEvents(options) {
    const { contract, events, blockRange, ...restParams } = options;
    const rpcRequest = (0, rpc_js_1.getRpcClient)(contract);
    if (restParams.blockHash &&
        (blockRange || restParams.fromBlock || restParams.toBlock)) {
        throw new Error("Cannot specify blockHash and range simultaneously,");
    }
    const latestBlockNumber = await (0, eth_blockNumber_js_1.eth_blockNumber)(rpcRequest);
    // Compute toBlock and fromBlock if blockRange was passed
    if (blockRange) {
        const { fromBlock, toBlock } = restParams;
        // Make sure the inputs were properly defined
        if (fromBlock &&
            toBlock &&
            BigInt(toBlock) - BigInt(fromBlock) !== BigInt(blockRange)) {
            throw new Error("Incompatible blockRange with specified fromBlock and toBlock. Please only define fromBlock or toBlock when specifying blockRange.");
        }
        if (fromBlock) {
            restParams.toBlock = BigInt(fromBlock) + BigInt(blockRange) - 1n; // Subtract one because toBlock is inclusive
        }
        else if (toBlock) {
            restParams.fromBlock = BigInt(toBlock) - BigInt(blockRange) + 1n; // Add one because fromBlock is inclusive
        }
        else {
            // If no from or to block specified, use the latest block as the to block
            restParams.toBlock = latestBlockNumber;
            restParams.fromBlock = latestBlockNumber - BigInt(blockRange) + 1n; // Add one because fromBlock is inclusive
        }
    }
    let resolvedEvents = events ?? [];
    // if we have an abi on the contract, we can encode the topics with it
    if (!events?.length && !!contract) {
        // if we have a contract *WITH* an abi we can use that
        if (contract.abi?.length) {
            // @ts-expect-error - we can't make typescript happy here, but we know this is an abi event
            resolvedEvents = contract.abi
                .filter(utils_js_1.isAbiEvent)
                .map((abiEvent) => (0, prepare_event_js_1.prepareEvent)({ signature: abiEvent }));
        }
        else {
            const runtimeAbi = await (0, resolve_abi_js_1.resolveContractAbi)(contract);
            // @ts-expect-error - we can't make typescript happy here, but we know this is an abi event
            resolvedEvents = runtimeAbi
                .filter(utils_js_1.isAbiEvent)
                .map((abiEvent) => (0, prepare_event_js_1.prepareEvent)({ signature: abiEvent }));
        }
    }
    const logsParams = events && events.length > 0
        ? // if we have events passed in then we use those
            events.map((e) => ({
                ...restParams,
                address: contract?.address,
                topics: e.topics,
            }))
        : // otherwise we want "all" events (aka not pass any topics at all)
            [{ ...restParams, address: contract?.address }];
    const logs = await Promise.all(logsParams.map((ethLogParams) => (0, eth_getLogs_js_1.eth_getLogs)(rpcRequest, ethLogParams)));
    const flattenLogs = logs
        .flat()
        .sort((a, b) => Number((a.blockNumber ?? 0n) - (b.blockNumber ?? 0n)));
    return (0, parse_logs_js_1.parseEventLogs)({
        logs: flattenLogs,
        events: resolvedEvents,
    });
}
