// thirdweb/dist/cjs/extensions/permissions/__generated__/IPermissionsEnumerable/read/getRoleMemberCount.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isGetRoleMemberCountSupported = isGetRoleMemberCountSupported;
exports.encodeGetRoleMemberCountParams = encodeGetRoleMemberCountParams;
exports.encodeGetRoleMemberCount = encodeGetRoleMemberCount;
exports.decodeGetRoleMemberCountResult = decodeGetRoleMemberCountResult;
exports.getRoleMemberCount = getRoleMemberCount;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xca15c873";
const FN_INPUTS = [
    {
        type: "bytes32",
        name: "role",
    },
];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `getRoleMemberCount` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `getRoleMemberCount` method is supported.
 * @extension PERMISSIONS
 * @example
 * ```ts
 * import { isGetRoleMemberCountSupported } from "thirdweb/extensions/permissions";
 * const supported = isGetRoleMemberCountSupported(["0x..."]);
 * ```
 */
function isGetRoleMemberCountSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "getRoleMemberCount" function.
 * @param options - The options for the getRoleMemberCount function.
 * @returns The encoded ABI parameters.
 * @extension PERMISSIONS
 * @example
 * ```ts
 * import { encodeGetRoleMemberCountParams } from "thirdweb/extensions/permissions";
 * const result = encodeGetRoleMemberCountParams({
 *  role: ...,
 * });
 * ```
 */
function encodeGetRoleMemberCountParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.role]);
}
/**
 * Encodes the "getRoleMemberCount" function into a Hex string with its parameters.
 * @param options - The options for the getRoleMemberCount function.
 * @returns The encoded hexadecimal string.
 * @extension PERMISSIONS
 * @example
 * ```ts
 * import { encodeGetRoleMemberCount } from "thirdweb/extensions/permissions";
 * const result = encodeGetRoleMemberCount({
 *  role: ...,
 * });
 * ```
 */
function encodeGetRoleMemberCount(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeGetRoleMemberCountParams(options).slice(2));
}
/**
 * Decodes the result of the getRoleMemberCount function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension PERMISSIONS
 * @example
 * ```ts
 * import { decodeGetRoleMemberCountResult } from "thirdweb/extensions/permissions";
 * const result = decodeGetRoleMemberCountResultResult("...");
 * ```
 */
function decodeGetRoleMemberCountResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getRoleMemberCount" function on the contract.
 * @param options - The options for the getRoleMemberCount function.
 * @returns The parsed result of the function call.
 * @extension PERMISSIONS
 * @example
 * ```ts
 * import { getRoleMemberCount } from "thirdweb/extensions/permissions";
 *
 * const result = await getRoleMemberCount({
 *  contract,
 *  role: ...,
 * });
 *
 * ```
 */
async function getRoleMemberCount(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.role],
    });
}
