// thirdweb/dist/cjs/react/web/ui/ConnectWallet/screens/Buy/main/useBuyTxStates.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useTransactionCostAndData = useTransactionCostAndData;
const react_query_1 = require("@tanstack/react-query");
const react_1 = require("react");
const utils_js_1 = require("../../../../../../../chains/utils.js");
const addresses_js_1 = require("../../../../../../../constants/addresses.js");
const contract_js_1 = require("../../../../../../../contract/contract.js");
const getCurrencyMetadata_js_1 = require("../../../../../../../extensions/erc20/read/getCurrencyMetadata.js");
const get_gas_price_js_1 = require("../../../../../../../gas/get-gas-price.js");
const encode_js_1 = require("../../../../../../../transaction/actions/encode.js");
const estimate_gas_cost_js_1 = require("../../../../../../../transaction/actions/estimate-gas-cost.js");
const resolve_promised_value_js_1 = require("../../../../../../../utils/promise/resolve-promised-value.js");
const getWalletBalance_js_1 = require("../../../../../../../wallets/utils/getWalletBalance.js");
function useTransactionCostAndData(args) {
    const { transaction, account, supportedDestinations } = args;
    // Compute query key of the transaction first
    const [txQueryKey, setTxQueryKey] = (0, react_1.useState)();
    (0, react_1.useEffect)(() => {
        Promise.all([
            (0, resolve_promised_value_js_1.resolvePromisedValue)(transaction.value),
            (0, resolve_promised_value_js_1.resolvePromisedValue)(transaction.erc20Value),
            (0, resolve_promised_value_js_1.resolvePromisedValue)(transaction.to),
            (0, encode_js_1.encode)(transaction),
        ]).then(([value, erc20Value, to, data]) => {
            setTxQueryKey({
                value: value?.toString(),
                erc20Value: erc20Value?.amountWei?.toString(),
                erc20Currency: erc20Value?.tokenAddress,
                to,
                data,
            });
        });
    }, [transaction]);
    return (0, react_query_1.useQuery)({
        queryKey: [
            "transaction-cost",
            transaction.chain.id,
            account?.address,
            txQueryKey,
        ],
        queryFn: async () => {
            if (!account) {
                throw new Error("No account");
            }
            const erc20Value = await (0, resolve_promised_value_js_1.resolvePromisedValue)(transaction.erc20Value);
            if (erc20Value) {
                const [tokenBalance, tokenMeta, gasCostWei] = await Promise.all([
                    (0, getWalletBalance_js_1.getWalletBalance)({
                        address: account.address,
                        chain: transaction.chain,
                        client: transaction.client,
                        tokenAddress: erc20Value.tokenAddress,
                    }),
                    (0, getCurrencyMetadata_js_1.getCurrencyMetadata)({
                        contract: (0, contract_js_1.getContract)({
                            address: erc20Value.tokenAddress,
                            chain: transaction.chain,
                            client: transaction.client,
                        }),
                    }),
                    getTransactionGasCost(transaction, account?.address),
                ]);
                const transactionValueWei = erc20Value.amountWei;
                const walletBalance = tokenBalance;
                const currency = {
                    address: erc20Value.tokenAddress,
                    name: tokenMeta.name,
                    symbol: tokenMeta.symbol,
                    icon: supportedDestinations
                        .find((c) => c.chain.id === transaction.chain.id)
                        ?.tokens.find((t) => t.address.toLowerCase() ===
                        erc20Value.tokenAddress.toLowerCase())?.icon,
                };
                return {
                    token: currency,
                    decimals: tokenMeta.decimals,
                    walletBalance,
                    gasCostWei,
                    transactionValueWei,
                };
            }
            const [nativeWalletBalance, chainMetadata, gasCostWei] = await Promise.all([
                (0, getWalletBalance_js_1.getWalletBalance)({
                    address: account.address,
                    chain: transaction.chain,
                    client: transaction.client,
                }),
                (0, utils_js_1.getChainMetadata)(transaction.chain),
                getTransactionGasCost(transaction, account?.address),
            ]);
            const walletBalance = nativeWalletBalance;
            const transactionValueWei = (await (0, resolve_promised_value_js_1.resolvePromisedValue)(transaction.value)) || 0n;
            return {
                token: {
                    address: addresses_js_1.NATIVE_TOKEN_ADDRESS,
                    name: chainMetadata.nativeCurrency.name,
                    symbol: chainMetadata.nativeCurrency.symbol,
                    icon: chainMetadata.icon?.url,
                },
                decimals: 18,
                walletBalance,
                gasCostWei,
                transactionValueWei,
            };
        },
        enabled: !!transaction && !!account && !!txQueryKey,
        refetchInterval: () => {
            if (transaction.erc20Value) {
                // if erc20 value is set, we don't need to poll
                return undefined;
            }
            return 30_000;
        },
    });
}
async function getTransactionGasCost(tx, from) {
    try {
        const gasCost = await (0, estimate_gas_cost_js_1.estimateGasCost)({
            transaction: tx,
            from,
        });
        const bufferCost = gasCost.wei / 10n;
        // Note: get tx.value AFTER estimateGasCost
        // add 10% extra gas cost to the estimate to ensure user buys enough to cover the tx cost
        return gasCost.wei + bufferCost;
    }
    catch {
        if (from) {
            // try again without passing from
            return await getTransactionGasCost(tx);
        }
        // fallback if both fail, use the tx value + 2M * gas price
        const gasPrice = await (0, get_gas_price_js_1.getGasPrice)({
            client: tx.client,
            chain: tx.chain,
        });
        return 2000000n * gasPrice;
    }
}
