// thirdweb/dist/cjs/analytics/track.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.trackConnect = trackConnect;
exports.trackPayEvent = trackPayEvent;
const fetch_js_1 = require("../utils/fetch.js");
const ANALYTICS_ENDPOINT = "https://c.thirdweb.com/event";
/**
 * @internal
 */
function trackConnect(args) {
    const { client, walletType, walletAddress } = args;
    track(client, {
        source: "connectWallet",
        action: "connect",
        walletType,
        walletAddress,
    });
}
/**
 * @internal
 */
function trackPayEvent(args) {
    track(args.client, {
        source: "pay",
        action: args.event,
        clientId: args.client.clientId,
        chainId: args.chainId,
        walletAddress: args.walletAddress,
        walletType: args.walletType,
        tokenAddress: args.fromToken,
        amountWei: args.fromAmount,
        dstTokenAddress: args.toToken,
        dstChainId: args.chainId,
    });
}
/**
 * @internal
 */
function track(client, data) {
    const fetch = (0, fetch_js_1.getClientFetch)(client);
    fetch(ANALYTICS_ENDPOINT, {
        method: "POST",
        body: JSON.stringify(data, (_key, value) => {
            if (typeof value === "bigint") {
                return value.toString();
            }
            return value;
        }),
    });
}
