// thirdweb/dist/cjs/extensions/marketplace/english-auctions/read/getWinningBid.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getWinningBid = getWinningBid;
exports.isGetWinningBidSupported = isGetWinningBidSupported;
const addresses_js_1 = require("../../../../constants/addresses.js");
const resolve_currency_value_js_1 = require("../../../../utils/extensions/resolve-currency-value.js");
const GetWinningBid = require("../../__generated__/IEnglishAuctions/read/getWinningBid.js");
/**
 * Retrieves the winning bid information for a given auction.
 *
 * @param options - The options for retrieving the winning bid.
 * @returns The winning bid information, or undefined if there is no winning bid.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { getWinningBid } from "thirdweb/extensions/marketplace";
 *
 * const winningBid = await getWinningBid({
 *  contract,
 *  auctionId: 0n,
 * });
 * ```
 */
async function getWinningBid(options) {
    const [bidderAddress, currencyAddress, bidAmountWei] = await GetWinningBid.getWinningBid(options);
    if (bidderAddress === addresses_js_1.ZERO_ADDRESS) {
        return undefined;
    }
    return {
        bidderAddress,
        currencyAddress,
        bidAmountWei,
        currencyValue: await (0, resolve_currency_value_js_1.resolveCurrencyValue)({
            chain: options.contract.chain,
            client: options.contract.client,
            currencyAddress,
            wei: bidAmountWei,
        }),
    };
}
/**
 * Checks if the `getWinningBid` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `getWinningBid` method is supported.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { isGetWinningBidSupported } from "thirdweb/extensions/marketplace";
 *
 * const supported = isGetWinningBidSupported(["0x..."]);
 * ```
 */
function isGetWinningBidSupported(availableSelectors) {
    return GetWinningBid.isGetWinningBidSupported(availableSelectors);
}
