// thirdweb/dist/cjs/extensions/erc1155/__generated__/IPackVRFDirect/read/canClaimRewards.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isCanClaimRewardsSupported = isCanClaimRewardsSupported;
exports.encodeCanClaimRewardsParams = encodeCanClaimRewardsParams;
exports.encodeCanClaimRewards = encodeCanClaimRewards;
exports.decodeCanClaimRewardsResult = decodeCanClaimRewardsResult;
exports.canClaimRewards = canClaimRewards;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xa9b47a66";
const FN_INPUTS = [
    {
        type: "address",
        name: "_opener",
    },
];
const FN_OUTPUTS = [
    {
        type: "bool",
    },
];
/**
 * Checks if the `canClaimRewards` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `canClaimRewards` method is supported.
 * @extension ERC1155
 * @example
 * ```ts
 * import { isCanClaimRewardsSupported } from "thirdweb/extensions/erc1155";
 * const supported = isCanClaimRewardsSupported(["0x..."]);
 * ```
 */
function isCanClaimRewardsSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "canClaimRewards" function.
 * @param options - The options for the canClaimRewards function.
 * @returns The encoded ABI parameters.
 * @extension ERC1155
 * @example
 * ```ts
 * import { encodeCanClaimRewardsParams } from "thirdweb/extensions/erc1155";
 * const result = encodeCanClaimRewardsParams({
 *  opener: ...,
 * });
 * ```
 */
function encodeCanClaimRewardsParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.opener]);
}
/**
 * Encodes the "canClaimRewards" function into a Hex string with its parameters.
 * @param options - The options for the canClaimRewards function.
 * @returns The encoded hexadecimal string.
 * @extension ERC1155
 * @example
 * ```ts
 * import { encodeCanClaimRewards } from "thirdweb/extensions/erc1155";
 * const result = encodeCanClaimRewards({
 *  opener: ...,
 * });
 * ```
 */
function encodeCanClaimRewards(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeCanClaimRewardsParams(options).slice(2));
}
/**
 * Decodes the result of the canClaimRewards function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC1155
 * @example
 * ```ts
 * import { decodeCanClaimRewardsResult } from "thirdweb/extensions/erc1155";
 * const result = decodeCanClaimRewardsResultResult("...");
 * ```
 */
function decodeCanClaimRewardsResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "canClaimRewards" function on the contract.
 * @param options - The options for the canClaimRewards function.
 * @returns The parsed result of the function call.
 * @extension ERC1155
 * @example
 * ```ts
 * import { canClaimRewards } from "thirdweb/extensions/erc1155";
 *
 * const result = await canClaimRewards({
 *  contract,
 *  opener: ...,
 * });
 *
 * ```
 */
async function canClaimRewards(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.opener],
    });
}
