// thirdweb/dist/esm/react/web/ui/ConnectWallet/screens/Buy/EstimatedTimeAndFees.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { ClockIcon } from "@radix-ui/react-icons";
import { fontSize, iconSize, radius, } from "../../../../../core/design-system/index.js";
import { Skeleton } from "../../../components/Skeleton.js";
import { Container } from "../../../components/basic.js";
import { Button } from "../../../components/buttons.js";
import { Text } from "../../../components/text.js";
import { formatSeconds } from "./swap/formatSeconds.js";
export function EstimatedTimeAndFees(props) {
    const { estimatedSeconds, quoteIsLoading } = props;
    return (_jsxs(Container, { bg: "tertiaryBg", flex: "row", borderColor: "borderColor", style: {
            borderRadius: radius.md,
            borderTopLeftRadius: 0,
            borderTopRightRadius: 0,
            justifyContent: "space-between",
            alignItems: "center",
            borderWidth: "1px",
            borderStyle: "solid",
        }, children: [_jsxs(Container, { flex: "row", center: "y", gap: "xxs", color: "accentText", p: "sm", children: [_jsx(ClockIcon, { width: iconSize.sm, height: iconSize.sm }), quoteIsLoading ? (_jsx(Skeleton, { height: fontSize.xs, width: "50px", color: "borderColor" })) : (_jsx(Text, { size: "xs", color: "secondaryText", children: estimatedSeconds !== undefined
                            ? `~${formatSeconds(estimatedSeconds)}`
                            : "--" }))] }), _jsxs(Button, { variant: "ghost", onClick: props.onViewFees, gap: "xs", children: [_jsx(Container, { color: "accentText", flex: "row", center: "both", children: _jsx(ViewFeeIcon, { size: iconSize.sm }) }), _jsx(Text, { size: "xs", color: "secondaryText", children: "View Fees" })] })] }));
}
const ViewFeeIcon = (props) => {
    return (_jsxs("svg", { width: props.size, height: props.size, viewBox: "0 0 12 12", fill: "none", xmlns: "http://www.w3.org/2000/svg", "aria-hidden": "true", children: [_jsx("path", { d: "M9.5 1.5H2.5C1.94772 1.5 1.5 1.94772 1.5 2.5V9.5C1.5 10.0523 1.94772 10.5 2.5 10.5H9.5C10.0523 10.5 10.5 10.0523 10.5 9.5V2.5C10.5 1.94772 10.0523 1.5 9.5 1.5Z", stroke: "currentColor", strokeLinecap: "round", strokeLinejoin: "round" }), _jsx("path", { d: "M4.5 7.5L7.5 4.5", stroke: "currentColor", strokeLinecap: "round", strokeLinejoin: "round" })] }));
};
