// thirdweb/dist/cjs/wallets/utils/getTokenBalance.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTokenBalance = getTokenBalance;
const utils_js_1 = require("../../chains/utils.js");
const contract_js_1 = require("../../contract/contract.js");
const eth_getBalance_js_1 = require("../../rpc/actions/eth_getBalance.js");
const rpc_js_1 = require("../../rpc/rpc.js");
const units_js_1 = require("../../utils/units.js");
/**
 * Retrieves the balance of a token for a given wallet.
 * @param options - The options for retrieving the token balance.
 * @returns A promise that resolves to the token balance result.
 * @example
 * ```ts
 * import { getTokenBalance } from "thirdweb/wallets";
 * const balance = await getTokenBalance({ account, client, chain, tokenAddress });
 * ```
 */
async function getTokenBalance(options) {
    const { account, client, chain, tokenAddress } = options;
    // erc20 case
    if (tokenAddress) {
        // load balanceOf dynamically to avoid circular dependency
        const { getBalance } = await Promise.resolve().then(() => require("../../extensions/erc20/read/getBalance.js"));
        return getBalance({
            contract: (0, contract_js_1.getContract)({ client, chain, address: tokenAddress }),
            address: account.address,
        });
    }
    // native token case
    const rpcRequest = (0, rpc_js_1.getRpcClient)({ client, chain });
    const [nativeSymbol, nativeDecimals, nativeName, nativeBalance] = await Promise.all([
        (0, utils_js_1.getChainSymbol)(chain),
        (0, utils_js_1.getChainDecimals)(chain),
        (0, utils_js_1.getChainNativeCurrencyName)(chain),
        (0, eth_getBalance_js_1.eth_getBalance)(rpcRequest, { address: account.address }),
    ]);
    return {
        value: nativeBalance,
        decimals: nativeDecimals,
        displayValue: (0, units_js_1.toTokens)(nativeBalance, nativeDecimals),
        symbol: nativeSymbol,
        name: nativeName,
    };
}
