// thirdweb/dist/cjs/wallets/ecosystem/get-ecosystem-wallet-info.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getEcosystemWalletInfo = getEcosystemWalletInfo;
const domains_js_1 = require("../../utils/domains.js");
/**
 * Fetches metadata for a given ecosystem wallet.
 *
 * @param {EcosystemWalletId} walletId - The ecosystem wallet ID.
 * @returns {Promise<{ imageUrl: string | undefined, name: string | undefined }>} A promise that resolves to an object containing the wallet's image URL and name.
 * @throws {Error} Throws an error if no partner ID is provided in the wallet configuration.
 * @internal
 */
async function getEcosystemWalletInfo(walletId) {
    const res = await fetch(`${(0, domains_js_1.getThirdwebBaseUrl)("inAppWallet")}/api/2024-05-05/ecosystem-wallet`, {
        headers: {
            "x-ecosystem-id": walletId,
        },
    });
    const data = await res.json();
    if (!data || data.code === "UNAUTHORIZED") {
        throw new Error(data.message ||
            `Could not find ecosystem wallet with id ${walletId}, please check your ecosystem wallet configuration.`);
    }
    return {
        id: walletId,
        name: data.name,
        image_id: data.imageUrl,
        homepage: data.homepage,
        rdns: null,
        app: {
            browser: null,
            ios: null,
            android: null,
            mac: null,
            windows: null,
            linux: null,
            opera: null,
            chrome: null,
            firefox: null,
            safari: null,
            edge: null,
        },
        mobile: {
            native: null,
            universal: null,
        },
        desktop: {
            native: null,
            universal: null,
        },
    };
}
