// thirdweb/dist/cjs/extensions/erc20/drops/write/resetClaimEligibility.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.resetClaimEligibility = resetClaimEligibility;
exports.isResetClaimEligibilitySupported = isResetClaimEligibilitySupported;
const setClaimConditions_js_1 = require("../../__generated__/IDropERC20/write/setClaimConditions.js");
const getClaimConditions_js_1 = require("../read/getClaimConditions.js");
/**
 * Reset the claim eligibility for all users.
 * @param options
 * @returns the prepared transaction
 * @extension ERC20
 * @example
 * ```ts
 * import { resetClaimEligibility } from "thirdweb/extensions/erc20";
 * import { sendTransaction } from "thirdweb";
 *
 * const transaction = resetClaimEligibility({
 *  contract,
 * });
 *
 * await sendTransaction({ transaction, account });
 * ```
 */
function resetClaimEligibility(options) {
    // download existing conditions
    return (0, setClaimConditions_js_1.setClaimConditions)({
        contract: options.contract,
        asyncParams: async () => {
            // get existing conditions
            const existingConditions = await (0, getClaimConditions_js_1.getClaimConditions)(options);
            // then simply return the exact same ones, but with the resetClaimEligibility flag set to true
            return {
                // type is necessary because of viem hex shenanigans (strict vs non-strict `0x` prefix string)
                phases: existingConditions,
                resetClaimEligibility: true,
            };
        },
    });
}
/**
 * Checks if the `resetClaimEligibility` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `resetClaimEligibility` method is supported.
 * @extension ERC20
 * @example
 * ```ts
 * import { isResetClaimEligibilitySupported } from "thirdweb/extensions/erc20";
 *
 * const supported = isResetClaimEligibilitySupported(["0x..."]);
 * ```
 */
function isResetClaimEligibilitySupported(availableSelectors) {
    return ((0, getClaimConditions_js_1.isGetClaimConditionsSupported)(availableSelectors) &&
        (0, setClaimConditions_js_1.isSetClaimConditionsSupported)(availableSelectors));
}
