// thirdweb/dist/cjs/extensions/farcaster/__generated__/IKeyGateway/read/nonces.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isNoncesSupported = isNoncesSupported;
exports.encodeNoncesParams = encodeNoncesParams;
exports.encodeNonces = encodeNonces;
exports.decodeNoncesResult = decodeNoncesResult;
exports.nonces = nonces;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x7ecebe00";
const FN_INPUTS = [
    {
        type: "address",
        name: "account",
    },
];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `nonces` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `nonces` method is supported.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { isNoncesSupported } from "thirdweb/extensions/farcaster";
 * const supported = isNoncesSupported(["0x..."]);
 * ```
 */
function isNoncesSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "nonces" function.
 * @param options - The options for the nonces function.
 * @returns The encoded ABI parameters.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { encodeNoncesParams } from "thirdweb/extensions/farcaster";
 * const result = encodeNoncesParams({
 *  account: ...,
 * });
 * ```
 */
function encodeNoncesParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.account]);
}
/**
 * Encodes the "nonces" function into a Hex string with its parameters.
 * @param options - The options for the nonces function.
 * @returns The encoded hexadecimal string.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { encodeNonces } from "thirdweb/extensions/farcaster";
 * const result = encodeNonces({
 *  account: ...,
 * });
 * ```
 */
function encodeNonces(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeNoncesParams(options).slice(2));
}
/**
 * Decodes the result of the nonces function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { decodeNoncesResult } from "thirdweb/extensions/farcaster";
 * const result = decodeNoncesResultResult("...");
 * ```
 */
function decodeNoncesResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "nonces" function on the contract.
 * @param options - The options for the nonces function.
 * @returns The parsed result of the function call.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { nonces } from "thirdweb/extensions/farcaster";
 *
 * const result = await nonces({
 *  contract,
 *  account: ...,
 * });
 *
 * ```
 */
async function nonces(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.account],
    });
}
