// thirdweb/dist/esm/extensions/modules/__generated__/RoyaltyERC1155/read/getTransferValidationFunction.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0x0d705df6";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "bytes4",
        name: "functionSignature",
    },
    {
        type: "bool",
        name: "isViewFunction",
    },
];
/**
 * Checks if the `getTransferValidationFunction` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `getTransferValidationFunction` method is supported.
 * @modules RoyaltyERC1155
 * @example
 * ```ts
 * import { RoyaltyERC1155 } from "thirdweb/modules";
 * const supported = RoyaltyERC1155.isGetTransferValidationFunctionSupported(["0x..."]);
 * ```
 */
export function isGetTransferValidationFunctionSupported(availableSelectors) {
    return detectMethod({
        availableSelectors,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the getTransferValidationFunction function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @modules RoyaltyERC1155
 * @example
 * ```ts
 * import { RoyaltyERC1155 } from "thirdweb/modules";
 * const result = RoyaltyERC1155.decodeGetTransferValidationFunctionResultResult("...");
 * ```
 */
export function decodeGetTransferValidationFunctionResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result);
}
/**
 * Calls the "getTransferValidationFunction" function on the contract.
 * @param options - The options for the getTransferValidationFunction function.
 * @returns The parsed result of the function call.
 * @modules RoyaltyERC1155
 * @example
 * ```ts
 * import { RoyaltyERC1155 } from "thirdweb/modules";
 *
 * const result = await RoyaltyERC1155.getTransferValidationFunction({
 *  contract,
 * });
 *
 * ```
 */
export async function getTransferValidationFunction(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
