// thirdweb/dist/cjs/contract/deployment/deploy-via-autofactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.prepareAutoFactoryDeployTransaction = prepareAutoFactoryDeployTransaction;
exports.deployViaAutoFactory = deployViaAutoFactory;
const parse_logs_js_1 = require("../../event/actions/parse-logs.js");
const ProxyDeployed_js_1 = require("../../extensions/thirdweb/__generated__/IContractFactory/events/ProxyDeployed.js");
const deployProxyByImplementation_js_1 = require("../../extensions/thirdweb/__generated__/IContractFactory/write/deployProxyByImplementation.js");
const eth_blockNumber_js_1 = require("../../rpc/actions/eth_blockNumber.js");
const rpc_js_1 = require("../../rpc/rpc.js");
const encode_js_1 = require("../../transaction/actions/encode.js");
const send_and_confirm_transaction_js_1 = require("../../transaction/actions/send-and-confirm-transaction.js");
const keccak_id_js_1 = require("../../utils/any-evm/keccak-id.js");
const isZkSyncChain_js_1 = require("../../utils/any-evm/zksync/isZkSyncChain.js");
const hex_js_1 = require("../../utils/encoding/hex.js");
const resolve_promised_value_js_1 = require("../../utils/promise/resolve-promised-value.js");
const zkDeployProxy_js_1 = require("./zksync/zkDeployProxy.js");
/**
 * @internal
 */
function prepareAutoFactoryDeployTransaction(args) {
    return (0, deployProxyByImplementation_js_1.deployProxyByImplementation)({
        contract: args.cloneFactoryContract,
        async asyncParams() {
            const rpcRequest = (0, rpc_js_1.getRpcClient)({
                ...args,
            });
            const blockNumber = await (0, eth_blockNumber_js_1.eth_blockNumber)(rpcRequest);
            const salt = args.salt
                ? (0, keccak_id_js_1.keccakId)(args.salt)
                : (0, hex_js_1.toHex)(blockNumber, {
                    size: 32,
                });
            const implementation = await (0, resolve_promised_value_js_1.resolvePromisedValue)(args.initializeTransaction.to);
            if (!implementation) {
                throw new Error("initializeTransaction must have a 'to' field set");
            }
            return {
                data: await (0, encode_js_1.encode)(args.initializeTransaction),
                implementation,
                salt,
            };
        },
    });
}
/**
 * @internal
 */
async function deployViaAutoFactory(options) {
    const { chain, client, account, cloneFactoryContract, initializeTransaction, } = options;
    if ((0, isZkSyncChain_js_1.isZkSyncChain)(chain)) {
        return (0, zkDeployProxy_js_1.zkDeployProxy)({
            chain,
            client,
            account,
            cloneFactoryContract,
            initializeTransaction,
        });
    }
    const tx = prepareAutoFactoryDeployTransaction({
        chain,
        client,
        cloneFactoryContract,
        initializeTransaction,
    });
    const receipt = await (0, send_and_confirm_transaction_js_1.sendAndConfirmTransaction)({
        transaction: tx,
        account,
    });
    const decodedEvent = (0, parse_logs_js_1.parseEventLogs)({
        events: [(0, ProxyDeployed_js_1.proxyDeployedEvent)()],
        logs: receipt.logs,
    });
    if (decodedEvent.length === 0 || !decodedEvent[0]) {
        throw new Error(`No ProxyDeployed event found in transaction: ${receipt.transactionHash}`);
    }
    return decodedEvent[0]?.args.proxy;
}
