// thirdweb/dist/cjs/gas/estimate-l1-fee.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.estimateL1Fee = estimateL1Fee;
const viem_1 = require("viem");
const contract_js_1 = require("../contract/contract.js");
const to_serializable_transaction_js_1 = require("../transaction/actions/to-serializable-transaction.js");
const read_contract_js_1 = require("../transaction/read-contract.js");
const OPStackGasPriceOracleAddress = "0x420000000000000000000000000000000000000F";
/**
 * @internal
 */
async function estimateL1Fee(options) {
    const { transaction, gasPriceOracleAddress } = options;
    const oracleContract = (0, contract_js_1.getContract)({
        client: transaction.client,
        address: gasPriceOracleAddress || OPStackGasPriceOracleAddress,
        chain: transaction.chain,
    });
    // purposefully remove gasPrice from the transaction
    const { gasPrice, ...serializableTx } = await (0, to_serializable_transaction_js_1.toSerializableTransaction)({
        transaction,
    });
    const serialized = (0, viem_1.serializeTransaction)({
        ...serializableTx,
        type: "eip1559",
    });
    //serializeTransaction(transaction);
    return (0, read_contract_js_1.readContract)({
        contract: oracleContract,
        method: "function getL1Fee(bytes memory _data) view returns (uint256)",
        params: [serialized],
    });
}
