// thirdweb/dist/cjs/wallets/in-app/web/lib/actions/sign-message.enclave.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.signMessage = signMessage;
const domains_js_1 = require("../../../../../utils/domains.js");
const fetch_js_1 = require("../../../../../utils/fetch.js");
const json_js_1 = require("../../../../../utils/json.js");
const get_auth_token_js_1 = require("../get-auth-token.js");
async function signMessage({ client, ecosystem, payload: { message, isRaw }, }) {
    const clientFetch = (0, fetch_js_1.getClientFetch)(client, ecosystem);
    const authToken = await (0, get_auth_token_js_1.getAuthToken)(client, ecosystem);
    const response = await clientFetch(`${(0, domains_js_1.getThirdwebBaseUrl)("inAppWallet")}/api/v1/enclave-wallet/sign-message`, {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
            "x-thirdweb-client-id": client.clientId,
            Authorization: `Bearer embedded-wallet-token:${authToken}`,
        },
        body: (0, json_js_1.stringify)({
            messagePayload: {
                message,
                isRaw,
            },
        }),
    });
    if (!response.ok) {
        throw new Error("Failed to sign message");
    }
    const signedMessage = (await response.json());
    return signedMessage;
}
