// thirdweb/dist/esm/extensions/marketplace/__generated__/IMarketplace/read/contractURI.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0xe8a3d485";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "string",
    },
];
/**
 * Checks if the `contractURI` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `contractURI` method is supported.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { isContractURISupported } from "thirdweb/extensions/marketplace";
 * const supported = isContractURISupported(["0x..."]);
 * ```
 */
export function isContractURISupported(availableSelectors) {
    return detectMethod({
        availableSelectors,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the contractURI function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { decodeContractURIResult } from "thirdweb/extensions/marketplace";
 * const result = decodeContractURIResultResult("...");
 * ```
 */
export function decodeContractURIResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "contractURI" function on the contract.
 * @param options - The options for the contractURI function.
 * @returns The parsed result of the function call.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { contractURI } from "thirdweb/extensions/marketplace";
 *
 * const result = await contractURI({
 *  contract,
 * });
 *
 * ```
 */
export async function contractURI(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
