// thirdweb/dist/cjs/extensions/erc4337/account/addAdmin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addAdmin = addAdmin;
exports.isAddAdminSupported = isAddAdminSupported;
const setPermissionsForSigner_js_1 = require("../__generated__/IAccountPermissions/write/setPermissionsForSigner.js");
const common_js_1 = require("./common.js");
/**
 * Adds admin permissions for a specified address.
 * @param options - The options for the addAdmin function.
 * @returns The transaction object to be sent.
 * @example
 * ```ts
 * import { addAdmin } from 'thirdweb/extensions/erc4337';
 * import { sendTransaction } from 'thirdweb';
 *
 * const transaction = addAdmin({
 * contract,
 * account,
 * adminAddress: '0x...'
 * });
 *
 * await sendTransaction({ transaction, account });
 * ```
 * @extension ERC4337
 */
function addAdmin(options) {
    const { contract, account, adminAddress } = options;
    return (0, setPermissionsForSigner_js_1.setPermissionsForSigner)({
        contract,
        async asyncParams() {
            const { req, signature } = await (0, common_js_1.signPermissionRequest)({
                account,
                contract,
                req: await (0, common_js_1.defaultPermissionsForAdmin)({
                    target: adminAddress,
                    action: "add-admin",
                }),
            });
            return { signature, req };
        },
    });
}
/**
 * Checks if the `isAddAdminSupported` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `isAddAdminSupported` method is supported.
 * @extension ERC4337
 * @example
 * ```ts
 * import { isAddAdminSupported } from "thirdweb/extensions/erc4337";
 *
 * const supported = isAddAdminSupported(["0x..."]);
 * ```
 */
function isAddAdminSupported(availableSelectors) {
    return (0, setPermissionsForSigner_js_1.isSetPermissionsForSignerSupported)(availableSelectors);
}
