// thirdweb/dist/cjs/extensions/erc4337/account/addSessionKey.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addSessionKey = addSessionKey;
exports.isAddSessionKeySupported = isAddSessionKeySupported;
const setPermissionsForSigner_js_1 = require("../__generated__/IAccountPermissions/write/setPermissionsForSigner.js");
const common_js_1 = require("./common.js");
/**
 * Adds session key permissions for a specified address.
 * @param options - The options for the removeSessionKey function.
 * @param {Contract} options.contract - The smart account contract to add the session key to
 * @returns The transaction object to be sent.
 * @example
 * ```ts
 * import { addSessionKey } from 'thirdweb/extensions/erc4337';
 * import { sendTransaction } from 'thirdweb';
 *
 * const transaction = addSessionKey({
 * contract,
 * account,
 * sessionKeyAddress,
 * permissions: {
 *  approvedTargets: ['0x...'],
 *  nativeTokenLimitPerTransaction: 0.1, // in ETH
 *  permissionStartTimestamp: new Date(),
 *  permissionEndTimestamp: new Date(Date.now() + 1000 * 60 * 60 * 24 * 365), // 1 year from now
 * }
 * });
 *
 * await sendTransaction({ transaction, account });
 * ```
 * @extension ERC4337
 */
function addSessionKey(options) {
    const { contract, sessionKeyAddress, account, permissions } = options;
    return (0, setPermissionsForSigner_js_1.setPermissionsForSigner)({
        contract,
        async asyncParams() {
            const { req, signature } = await (0, common_js_1.signPermissionRequest)({
                account,
                contract,
                req: await (0, common_js_1.toContractPermissions)({
                    target: sessionKeyAddress,
                    permissions,
                }),
            });
            return { signature, req };
        },
    });
}
/**
 * Checks if the `isAddSessionKeySupported` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `isAddSessionKeySupported` method is supported.
 * @extension ERC4337
 * @example
 * ```ts
 * import { isAddSessionKeySupported } from "thirdweb/extensions/erc4337";
 *
 * const supported = isAddSessionKeySupported(["0x..."]);
 * ```
 */
function isAddSessionKeySupported(availableSelectors) {
    return (0, setPermissionsForSigner_js_1.isSetPermissionsForSignerSupported)(availableSelectors);
}
