// thirdweb/dist/cjs/extensions/erc4337/account/common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.signPermissionRequest = signPermissionRequest;
exports.toContractPermissions = toContractPermissions;
exports.defaultPermissionsForAdmin = defaultPermissionsForAdmin;
const addresses_js_1 = require("../../../constants/addresses.js");
const date_js_1 = require("../../../utils/date.js");
const random_js_1 = require("../../../utils/random.js");
const units_js_1 = require("../../../utils/units.js");
const types_js_1 = require("./types.js");
/**
 * @internal
 */
async function signPermissionRequest(options) {
    const { account, contract, req } = options;
    const signature = await account.signTypedData({
        domain: {
            name: "Account",
            version: "1",
            verifyingContract: contract.address,
            chainId: contract.chain.id,
        },
        primaryType: "SignerPermissionRequest",
        types: { SignerPermissionRequest: types_js_1.SignerPermissionRequest },
        message: req,
    });
    return { req, signature };
}
/**
 * @internal
 */
async function toContractPermissions(options) {
    const { target, permissions } = options;
    return {
        approvedTargets: permissions.approvedTargets === "*"
            ? [addresses_js_1.ADDRESS_ZERO]
            : permissions.approvedTargets,
        nativeTokenLimitPerTransaction: (0, units_js_1.toWei)(permissions.nativeTokenLimitPerTransaction?.toString() || "0"),
        permissionStartTimestamp: (0, date_js_1.dateToSeconds)(permissions.permissionStartTimestamp || new Date(0)),
        permissionEndTimestamp: (0, date_js_1.dateToSeconds)(permissions.permissionEndTimestamp || (0, date_js_1.tenYearsFromNow)()),
        reqValidityStartTimestamp: 0n,
        reqValidityEndTimestamp: (0, date_js_1.dateToSeconds)((0, date_js_1.tenYearsFromNow)()),
        uid: await (0, random_js_1.randomBytesHex)(),
        isAdmin: 0, // session key flag
        signer: target,
    };
}
/**
 * @internal
 */
async function defaultPermissionsForAdmin(options) {
    const { target, action } = options;
    return {
        approvedTargets: [],
        nativeTokenLimitPerTransaction: 0n,
        permissionStartTimestamp: 0n,
        permissionEndTimestamp: 0n,
        reqValidityStartTimestamp: 0n,
        reqValidityEndTimestamp: (0, date_js_1.dateToSeconds)((0, date_js_1.tenYearsFromNow)()),
        uid: await (0, random_js_1.randomBytesHex)(),
        isAdmin: action === "add-admin" ? 1 : action === "remove-admin" ? 2 : 0,
        signer: target,
    };
}
