// thirdweb/dist/cjs/extensions/erc4337/account/removeSessionKey.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.removeSessionKey = removeSessionKey;
exports.isRemoveSessionKeySupported = isRemoveSessionKeySupported;
const setPermissionsForSigner_js_1 = require("../__generated__/IAccountPermissions/write/setPermissionsForSigner.js");
const common_js_1 = require("./common.js");
/**
 * Removes session key permissions for a specified address.
 * @param options - The options for the removeSessionKey function.
 * @returns The transaction object to be sent.
 * @example
 * ```ts
 * import { removeSessionKey } from 'thirdweb/extensions/erc4337';
 * import { sendTransaction } from 'thirdweb';
 *
 * const transaction = removeSessionKey({
 * contract,
 * account,
 * sessionKeyAddress
 * });
 *
 * await sendTransaction({ transaction, account });
 * ```
 * @extension ERC4337
 */
function removeSessionKey(options) {
    const { contract, account, sessionKeyAddress } = options;
    return (0, setPermissionsForSigner_js_1.setPermissionsForSigner)({
        contract,
        async asyncParams() {
            const { req, signature } = await (0, common_js_1.signPermissionRequest)({
                account,
                contract,
                req: await (0, common_js_1.toContractPermissions)({
                    target: sessionKeyAddress,
                    permissions: {
                        approvedTargets: [],
                        nativeTokenLimitPerTransaction: 0,
                        permissionStartTimestamp: new Date(0),
                        permissionEndTimestamp: new Date(0),
                    },
                }),
            });
            return { signature, req };
        },
    });
}
/**
 * Checks if the `isRemoveSessionKeySupported` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `isRemoveSessionKeySupported` method is supported.
 * @extension ERC4337
 * @example
 * ```ts
 * import { isRemoveSessionKeySupported } from "thirdweb/extensions/erc4337";
 *
 * const supported = isRemoveSessionKeySupported(["0x..."]);
 * ```
 */
function isRemoveSessionKeySupported(availableSelectors) {
    return (0, setPermissionsForSigner_js_1.isSetPermissionsForSignerSupported)(availableSelectors);
}
