// thirdweb/dist/cjs/extensions/modules/__generated__/TransferableERC1155/read/encodeBytesOnInstall.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isEncodeBytesOnInstallSupported = isEncodeBytesOnInstallSupported;
exports.decodeEncodeBytesOnInstallResult = decodeEncodeBytesOnInstallResult;
exports.encodeBytesOnInstall = encodeBytesOnInstall;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xfec7ff72";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "bytes",
    },
];
/**
 * Checks if the `encodeBytesOnInstall` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `encodeBytesOnInstall` method is supported.
 * @modules TransferableERC1155
 * @example
 * ```ts
 * import { TransferableERC1155 } from "thirdweb/modules";
 * const supported = TransferableERC1155.isEncodeBytesOnInstallSupported(["0x..."]);
 * ```
 */
function isEncodeBytesOnInstallSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the encodeBytesOnInstall function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @modules TransferableERC1155
 * @example
 * ```ts
 * import { TransferableERC1155 } from "thirdweb/modules";
 * const result = TransferableERC1155.decodeEncodeBytesOnInstallResultResult("...");
 * ```
 */
function decodeEncodeBytesOnInstallResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "encodeBytesOnInstall" function on the contract.
 * @param options - The options for the encodeBytesOnInstall function.
 * @returns The parsed result of the function call.
 * @modules TransferableERC1155
 * @example
 * ```ts
 * import { TransferableERC1155 } from "thirdweb/modules";
 *
 * const result = await TransferableERC1155.encodeBytesOnInstall({
 *  contract,
 * });
 *
 * ```
 */
async function encodeBytesOnInstall(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
