// thirdweb/dist/cjs/wallets/in-app/core/authentication/client-scoped-storage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClientScopedStorage = void 0;
const settings_js_1 = require("../constants/settings.js");
const data = new Map();
/**
 * @internal
 */
class ClientScopedStorage {
    /**
     * @internal
     */
    constructor({ storage, clientId, ecosystemId, }) {
        Object.defineProperty(this, "key", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "storage", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.storage = storage;
        this.key = getLocalStorageKey(clientId, ecosystemId);
    }
    async getItem(key) {
        if (this.storage) {
            return this.storage.getItem(key);
        }
        return data.get(key) ?? null;
    }
    async setItem(key, value) {
        if (this.storage) {
            return this.storage.setItem(key, value);
        }
        data.set(key, value);
    }
    async removeItem(key) {
        const item = await this.getItem(key);
        if (this.storage && item) {
            this.storage.removeItem(key);
            return true;
        }
        return false;
    }
    /**
     * @internal
     */
    async getWalletConnectSessions() {
        return this.getItem((0, settings_js_1.WALLET_CONNECT_SESSIONS_LOCAL_STORAGE_NAME)(this.key));
    }
    /**
     * @internal
     */
    async saveWalletConnectSessions(stringifiedSessions) {
        await this.setItem((0, settings_js_1.WALLET_CONNECT_SESSIONS_LOCAL_STORAGE_NAME)(this.key), stringifiedSessions);
    }
    /**
     * @internal
     */
    async savePasskeyCredentialId(id) {
        await this.setItem((0, settings_js_1.PASSKEY_CREDENTIAL_ID_LOCAL_STORAGE_NAME)(this.key), id);
    }
    /**
     * @internal
     */
    async getPasskeyCredentialId() {
        return this.getItem((0, settings_js_1.PASSKEY_CREDENTIAL_ID_LOCAL_STORAGE_NAME)(this.key));
    }
    /**
     * @internal
     */
    async saveAuthCookie(cookie) {
        await this.setItem((0, settings_js_1.AUTH_TOKEN_LOCAL_STORAGE_NAME)(this.key), cookie);
    }
    /**
     * @internal
     */
    async getAuthCookie() {
        return this.getItem((0, settings_js_1.AUTH_TOKEN_LOCAL_STORAGE_NAME)(this.key));
    }
    /**
     * @internal
     */
    async removeAuthCookie() {
        return this.removeItem((0, settings_js_1.AUTH_TOKEN_LOCAL_STORAGE_NAME)(this.key));
    }
    /**
     * @internal
     */
    async saveDeviceShare(share, userId) {
        await this.saveWalletUserId(userId);
        await this.setItem((0, settings_js_1.DEVICE_SHARE_LOCAL_STORAGE_NAME)(this.key, userId), share);
    }
    /**
     * @internal
     */
    async getDeviceShare() {
        const userId = await this.getWalletUserId();
        if (userId) {
            return this.getItem((0, settings_js_1.DEVICE_SHARE_LOCAL_STORAGE_NAME)(this.key, userId));
        }
        return null;
    }
    /**
     * @internal
     */
    async removeDeviceShare() {
        const userId = await this.getWalletUserId();
        if (userId) {
            return this.removeItem((0, settings_js_1.DEVICE_SHARE_LOCAL_STORAGE_NAME)(this.key, userId));
        }
        return false;
    }
    /**
     * @internal
     */
    async getWalletUserId() {
        return this.getItem((0, settings_js_1.WALLET_USER_ID_LOCAL_STORAGE_NAME)(this.key));
    }
    /**
     * @internal
     */
    async saveWalletUserId(userId) {
        await this.setItem((0, settings_js_1.WALLET_USER_ID_LOCAL_STORAGE_NAME)(this.key), userId);
    }
    /**
     * @internal
     */
    async removeWalletUserId() {
        return this.removeItem((0, settings_js_1.WALLET_USER_ID_LOCAL_STORAGE_NAME)(this.key));
    }
    /**
     * @internal
     */
    async getGuestSessionId() {
        return this.getItem((0, settings_js_1.GUEST_SESSION_LOCAL_STORAGE_NAME)(this.key));
    }
    /**
     * @internal
     */
    async saveGuestSessionId(sessionId) {
        await this.setItem((0, settings_js_1.GUEST_SESSION_LOCAL_STORAGE_NAME)(this.key), sessionId);
    }
}
exports.ClientScopedStorage = ClientScopedStorage;
const getLocalStorageKey = (clientId, ecosystemId) => {
    return `${clientId}${ecosystemId ? `-${ecosystemId}` : ""}`;
};
