// thirdweb/dist/cjs/react/web/ui/ConnectWallet/screens/Buy/fiat/FiatStatusScreen.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OnrampStatusScreen = OnrampStatusScreen;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_icons_1 = require("@radix-ui/react-icons");
const react_query_1 = require("@tanstack/react-query");
const react_1 = require("react");
const isMobile_js_1 = require("../../../../../../../utils/web/isMobile.js");
const index_js_1 = require("../../../../../../core/design-system/index.js");
const useBuyWithFiatStatus_js_1 = require("../../../../../../core/hooks/pay/useBuyWithFiatStatus.js");
const invalidateWalletBalance_js_1 = require("../../../../../../core/providers/invalidateWalletBalance.js");
const Spacer_js_1 = require("../../../../components/Spacer.js");
const Spinner_js_1 = require("../../../../components/Spinner.js");
const StepBar_js_1 = require("../../../../components/StepBar.js");
const basic_js_1 = require("../../../../components/basic.js");
const buttons_js_1 = require("../../../../components/buttons.js");
const text_js_1 = require("../../../../components/text.js");
const AccentFailIcon_js_1 = require("../../../icons/AccentFailIcon.js");
const statusMeta_js_1 = require("../pay-transactions/statusMeta.js");
const FiatTxDetailsTable_js_1 = require("./FiatTxDetailsTable.js");
/**
 * Poll for "Buy with Fiat" status - when the on-ramp is in progress
 * - Show success screen if swap is not required and on-ramp is completed
 * - Show Failed screen if on-ramp failed
 * - call `onShowSwapFlow` if on-ramp is completed and swap is required
 */
function OnrampStatusScreen(props) {
    const queryClient = (0, react_query_1.useQueryClient)();
    const { openedWindow, onSuccess } = props;
    const statusQuery = (0, useBuyWithFiatStatus_js_1.useBuyWithFiatStatus)({
        intentId: props.intentId,
        client: props.client,
    });
    // determine UI status
    let uiStatus = "loading";
    if (statusQuery.data?.status === "ON_RAMP_TRANSFER_FAILED" ||
        statusQuery.data?.status === "PAYMENT_FAILED") {
        uiStatus = "failed";
    }
    else if (statusQuery.data?.status === "CRYPTO_SWAP_FALLBACK") {
        uiStatus = "partialSuccess";
    }
    else if (statusQuery.data?.status === "ON_RAMP_TRANSFER_COMPLETED") {
        uiStatus = "completed";
    }
    const purchaseCbCalled = (0, react_1.useRef)(false);
    (0, react_1.useEffect)(() => {
        if (purchaseCbCalled.current || !onSuccess) {
            return;
        }
        if (statusQuery.data?.status === "ON_RAMP_TRANSFER_COMPLETED") {
            purchaseCbCalled.current = true;
            onSuccess(statusQuery.data);
        }
    }, [onSuccess, statusQuery.data]);
    // close the onramp popup if onramp is completed
    (0, react_1.useEffect)(() => {
        if (!openedWindow || !statusQuery.data) {
            return;
        }
        if (statusQuery.data?.status === "CRYPTO_SWAP_REQUIRED" ||
            statusQuery.data?.status === "ON_RAMP_TRANSFER_COMPLETED") {
            openedWindow.close();
        }
    }, [statusQuery.data, openedWindow]);
    // invalidate wallet balance when onramp is completed
    const invalidatedBalance = (0, react_1.useRef)(false);
    (0, react_1.useEffect)(() => {
        if (!invalidatedBalance.current &&
            statusQuery.data?.status === "ON_RAMP_TRANSFER_COMPLETED") {
            invalidatedBalance.current = true;
            (0, invalidateWalletBalance_js_1.invalidateWalletBalance)(queryClient);
        }
    }, [statusQuery.data, queryClient]);
    // show swap flow
    (0, react_1.useEffect)(() => {
        if (statusQuery.data?.status === "CRYPTO_SWAP_REQUIRED") {
            props.onShowSwapFlow(statusQuery.data);
        }
    }, [statusQuery.data, props.onShowSwapFlow]);
    return ((0, jsx_runtime_1.jsxs)(basic_js_1.Container, { p: "lg", children: [(0, jsx_runtime_1.jsx)(basic_js_1.ModalHeader, { title: props.title, onBack: props.onBack }), props.hasTwoSteps && ((0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [(0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "lg" }), (0, jsx_runtime_1.jsx)(StepBar_js_1.StepBar, { steps: 2, currentStep: 1 }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "sm" }), (0, jsx_runtime_1.jsxs)(text_js_1.Text, { size: "xs", children: ["Step 1 of 2 - Buying ", props.quote.onRampToken.token.symbol, " with", " ", props.quote.fromCurrencyWithFees.currencySymbol] })] })), (0, jsx_runtime_1.jsx)(OnrampStatusScreenUI, { uiStatus: uiStatus, onDone: props.onDone, fiatStatus: statusQuery.data, client: props.client, transactionMode: props.transactionMode, quote: props.quote, isEmbed: props.isEmbed })] }));
}
function OnrampStatusScreenUI(props) {
    const { uiStatus } = props;
    const statusMeta = props.fiatStatus
        ? (0, statusMeta_js_1.getBuyWithFiatStatusMeta)(props.fiatStatus)
        : undefined;
    const fiatStatus = props.fiatStatus && props.fiatStatus.status !== "NOT_FOUND"
        ? props.fiatStatus
        : undefined;
    const onRampTokenQuote = props.quote.onRampToken;
    const txDetails = ((0, jsx_runtime_1.jsx)(FiatTxDetailsTable_js_1.OnRampTxDetailsTable, { client: props.client, token: fiatStatus?.source // source tx is onRamp token
            ? {
                chainId: fiatStatus.source.token.chainId,
                address: fiatStatus.source.token.tokenAddress,
                symbol: fiatStatus.source.token.symbol || "",
                amount: fiatStatus.source.amount,
            }
            : {
                chainId: onRampTokenQuote.token.chainId,
                address: onRampTokenQuote.token.tokenAddress,
                symbol: onRampTokenQuote.token.symbol,
                amount: onRampTokenQuote.amount,
            }, fiat: {
            amount: props.quote.fromCurrencyWithFees.amount,
            currencySymbol: props.quote.fromCurrencyWithFees.currencySymbol,
        }, statusMeta: fiatStatus?.source && statusMeta
            ? {
                color: statusMeta?.color,
                text: statusMeta?.status,
                txHash: fiatStatus.source.transactionHash,
            }
            : undefined }));
    return ((0, jsx_runtime_1.jsxs)(basic_js_1.Container, { children: [(0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "xl" }), uiStatus === "loading" && ((0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [(0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "md" }), (0, jsx_runtime_1.jsx)(basic_js_1.Container, { flex: "row", center: "x", children: (0, jsx_runtime_1.jsx)(Spinner_js_1.Spinner, { size: "xxl", color: "accentText" }) }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "md" }), (0, jsx_runtime_1.jsx)(text_js_1.Text, { color: "primaryText", size: "lg", center: true, children: "Buy Pending" }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "sm" }), !(0, isMobile_js_1.isMobile)() && (0, jsx_runtime_1.jsx)(text_js_1.Text, { center: true, children: "Complete the purchase in popup" }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "xxl" }), txDetails] })), uiStatus === "failed" && ((0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [(0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "md" }), (0, jsx_runtime_1.jsx)(basic_js_1.Container, { flex: "row", center: "x", children: (0, jsx_runtime_1.jsx)(AccentFailIcon_js_1.AccentFailIcon, { size: index_js_1.iconSize["3xl"] }) }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "lg" }), (0, jsx_runtime_1.jsx)(text_js_1.Text, { color: "primaryText", size: "lg", center: true, children: "Transaction Failed" }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "xxl" }), txDetails] })), uiStatus === "completed" && ((0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [(0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "md" }), (0, jsx_runtime_1.jsx)(basic_js_1.Container, { flex: "row", center: "x", color: "success", children: (0, jsx_runtime_1.jsx)(react_icons_1.CheckCircledIcon, { width: index_js_1.iconSize["3xl"], height: index_js_1.iconSize["3xl"] }) }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "md" }), (0, jsx_runtime_1.jsx)(text_js_1.Text, { color: "primaryText", size: "lg", center: true, children: "Buy Complete" }), props.fiatStatus && props.fiatStatus.status !== "NOT_FOUND" && ((0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [(0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "xxl" }), txDetails, (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "sm" })] })), !props.isEmbed && ((0, jsx_runtime_1.jsx)(buttons_js_1.Button, { variant: "accent", fullWidth: true, onClick: props.onDone, children: props.transactionMode ? "Continue Transaction" : "Done" }))] }))] }));
}
