// thirdweb/dist/cjs/extensions/modules/__generated__/MintableERC1155/read/getSaleConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isGetSaleConfigSupported = isGetSaleConfigSupported;
exports.decodeGetSaleConfigResult = decodeGetSaleConfigResult;
exports.getSaleConfig = getSaleConfig;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xcea943ee";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "address",
        name: "primarySaleRecipient",
    },
];
/**
 * Checks if the `getSaleConfig` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `getSaleConfig` method is supported.
 * @modules MintableERC1155
 * @example
 * ```ts
 * import { MintableERC1155 } from "thirdweb/modules";
 * const supported = MintableERC1155.isGetSaleConfigSupported(["0x..."]);
 * ```
 */
function isGetSaleConfigSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the getSaleConfig function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @modules MintableERC1155
 * @example
 * ```ts
 * import { MintableERC1155 } from "thirdweb/modules";
 * const result = MintableERC1155.decodeGetSaleConfigResultResult("...");
 * ```
 */
function decodeGetSaleConfigResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getSaleConfig" function on the contract.
 * @param options - The options for the getSaleConfig function.
 * @returns The parsed result of the function call.
 * @modules MintableERC1155
 * @example
 * ```ts
 * import { MintableERC1155 } from "thirdweb/modules";
 *
 * const result = await MintableERC1155.getSaleConfig({
 *  contract,
 * });
 *
 * ```
 */
async function getSaleConfig(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
