// thirdweb/dist/cjs/extensions/multicall3/__generated__/IMulticall3/read/getBlockHash.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isGetBlockHashSupported = isGetBlockHashSupported;
exports.encodeGetBlockHashParams = encodeGetBlockHashParams;
exports.encodeGetBlockHash = encodeGetBlockHash;
exports.decodeGetBlockHashResult = decodeGetBlockHashResult;
exports.getBlockHash = getBlockHash;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xee82ac5e";
const FN_INPUTS = [
    {
        type: "uint256",
        name: "blockNumber",
    },
];
const FN_OUTPUTS = [
    {
        type: "bytes32",
        name: "blockHash",
    },
];
/**
 * Checks if the `getBlockHash` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `getBlockHash` method is supported.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { isGetBlockHashSupported } from "thirdweb/extensions/multicall3";
 * const supported = isGetBlockHashSupported(["0x..."]);
 * ```
 */
function isGetBlockHashSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "getBlockHash" function.
 * @param options - The options for the getBlockHash function.
 * @returns The encoded ABI parameters.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { encodeGetBlockHashParams } from "thirdweb/extensions/multicall3";
 * const result = encodeGetBlockHashParams({
 *  blockNumber: ...,
 * });
 * ```
 */
function encodeGetBlockHashParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.blockNumber]);
}
/**
 * Encodes the "getBlockHash" function into a Hex string with its parameters.
 * @param options - The options for the getBlockHash function.
 * @returns The encoded hexadecimal string.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { encodeGetBlockHash } from "thirdweb/extensions/multicall3";
 * const result = encodeGetBlockHash({
 *  blockNumber: ...,
 * });
 * ```
 */
function encodeGetBlockHash(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeGetBlockHashParams(options).slice(2));
}
/**
 * Decodes the result of the getBlockHash function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { decodeGetBlockHashResult } from "thirdweb/extensions/multicall3";
 * const result = decodeGetBlockHashResultResult("...");
 * ```
 */
function decodeGetBlockHashResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getBlockHash" function on the contract.
 * @param options - The options for the getBlockHash function.
 * @returns The parsed result of the function call.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { getBlockHash } from "thirdweb/extensions/multicall3";
 *
 * const result = await getBlockHash({
 *  contract,
 *  blockNumber: ...,
 * });
 *
 * ```
 */
async function getBlockHash(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.blockNumber],
    });
}
