// thirdweb/dist/cjs/extensions/multicall3/__generated__/IMulticall3/read/getCurrentBlockDifficulty.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isGetCurrentBlockDifficultySupported = isGetCurrentBlockDifficultySupported;
exports.decodeGetCurrentBlockDifficultyResult = decodeGetCurrentBlockDifficultyResult;
exports.getCurrentBlockDifficulty = getCurrentBlockDifficulty;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x72425d9d";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "uint256",
        name: "difficulty",
    },
];
/**
 * Checks if the `getCurrentBlockDifficulty` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `getCurrentBlockDifficulty` method is supported.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { isGetCurrentBlockDifficultySupported } from "thirdweb/extensions/multicall3";
 * const supported = isGetCurrentBlockDifficultySupported(["0x..."]);
 * ```
 */
function isGetCurrentBlockDifficultySupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the getCurrentBlockDifficulty function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { decodeGetCurrentBlockDifficultyResult } from "thirdweb/extensions/multicall3";
 * const result = decodeGetCurrentBlockDifficultyResultResult("...");
 * ```
 */
function decodeGetCurrentBlockDifficultyResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getCurrentBlockDifficulty" function on the contract.
 * @param options - The options for the getCurrentBlockDifficulty function.
 * @returns The parsed result of the function call.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { getCurrentBlockDifficulty } from "thirdweb/extensions/multicall3";
 *
 * const result = await getCurrentBlockDifficulty({
 *  contract,
 * });
 *
 * ```
 */
async function getCurrentBlockDifficulty(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
