// thirdweb/dist/cjs/extensions/lens/__generated__/TokenHandleRegistry/read/getDefaultHandle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isGetDefaultHandleSupported = isGetDefaultHandleSupported;
exports.encodeGetDefaultHandleParams = encodeGetDefaultHandleParams;
exports.encodeGetDefaultHandle = encodeGetDefaultHandle;
exports.decodeGetDefaultHandleResult = decodeGetDefaultHandleResult;
exports.getDefaultHandle = getDefaultHandle;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xe524488d";
const FN_INPUTS = [
    {
        type: "uint256",
        name: "profileId",
    },
];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `getDefaultHandle` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `getDefaultHandle` method is supported.
 * @extension LENS
 * @example
 * ```ts
 * import { isGetDefaultHandleSupported } from "thirdweb/extensions/lens";
 * const supported = isGetDefaultHandleSupported(["0x..."]);
 * ```
 */
function isGetDefaultHandleSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "getDefaultHandle" function.
 * @param options - The options for the getDefaultHandle function.
 * @returns The encoded ABI parameters.
 * @extension LENS
 * @example
 * ```ts
 * import { encodeGetDefaultHandleParams } from "thirdweb/extensions/lens";
 * const result = encodeGetDefaultHandleParams({
 *  profileId: ...,
 * });
 * ```
 */
function encodeGetDefaultHandleParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.profileId]);
}
/**
 * Encodes the "getDefaultHandle" function into a Hex string with its parameters.
 * @param options - The options for the getDefaultHandle function.
 * @returns The encoded hexadecimal string.
 * @extension LENS
 * @example
 * ```ts
 * import { encodeGetDefaultHandle } from "thirdweb/extensions/lens";
 * const result = encodeGetDefaultHandle({
 *  profileId: ...,
 * });
 * ```
 */
function encodeGetDefaultHandle(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeGetDefaultHandleParams(options).slice(2));
}
/**
 * Decodes the result of the getDefaultHandle function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension LENS
 * @example
 * ```ts
 * import { decodeGetDefaultHandleResult } from "thirdweb/extensions/lens";
 * const result = decodeGetDefaultHandleResultResult("...");
 * ```
 */
function decodeGetDefaultHandleResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getDefaultHandle" function on the contract.
 * @param options - The options for the getDefaultHandle function.
 * @returns The parsed result of the function call.
 * @extension LENS
 * @example
 * ```ts
 * import { getDefaultHandle } from "thirdweb/extensions/lens";
 *
 * const result = await getDefaultHandle({
 *  contract,
 *  profileId: ...,
 * });
 *
 * ```
 */
async function getDefaultHandle(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.profileId],
    });
}
