// thirdweb/dist/cjs/stories/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.storyWallet = exports.storyAccount = exports.storyClient = void 0;
const accounts_1 = require("viem/accounts");
const wallet_adapter_js_1 = require("../adapters/wallet-adapter.js");
const ethereum_js_1 = require("../chains/chain-definitions/ethereum.js");
const client_js_1 = require("../client/client.js");
const private_key_js_1 = require("../wallets/private-key.js");
const clientId = process.env.STORYBOOK_CLIENT_ID;
if (!clientId) {
    throw new Error("STORYBOOK_CLIENT_ID env is not configured");
}
exports.storyClient = (0, client_js_1.createThirdwebClient)({
    clientId: clientId,
});
exports.storyAccount = (0, private_key_js_1.privateKeyToAccount)({
    client: exports.storyClient,
    privateKey: process.env.STORYBOOK_ACCOUNT_PRIVATE_KEY || (0, accounts_1.generatePrivateKey)(),
});
exports.storyWallet = (0, wallet_adapter_js_1.createWalletAdapter)({
    adaptedAccount: exports.storyAccount,
    client: exports.storyClient,
    chain: ethereum_js_1.ethereum,
    onDisconnect: () => { },
    switchChain: () => { },
});
