// thirdweb/dist/cjs/extensions/erc20/__generated__/IERC20Permit/read/DOMAIN_SEPARATOR.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isDOMAIN_SEPARATORSupported = isDOMAIN_SEPARATORSupported;
exports.decodeDOMAIN_SEPARATORResult = decodeDOMAIN_SEPARATORResult;
exports.DOMAIN_SEPARATOR = DOMAIN_SEPARATOR;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x3644e515";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "bytes32",
    },
];
/**
 * Checks if the `DOMAIN_SEPARATOR` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `DOMAIN_SEPARATOR` method is supported.
 * @extension ERC20
 * @example
 * ```ts
 * import { isDOMAIN_SEPARATORSupported } from "thirdweb/extensions/erc20";
 * const supported = isDOMAIN_SEPARATORSupported(["0x..."]);
 * ```
 */
function isDOMAIN_SEPARATORSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the DOMAIN_SEPARATOR function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC20
 * @example
 * ```ts
 * import { decodeDOMAIN_SEPARATORResult } from "thirdweb/extensions/erc20";
 * const result = decodeDOMAIN_SEPARATORResultResult("...");
 * ```
 */
function decodeDOMAIN_SEPARATORResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "DOMAIN_SEPARATOR" function on the contract.
 * @param options - The options for the DOMAIN_SEPARATOR function.
 * @returns The parsed result of the function call.
 * @extension ERC20
 * @example
 * ```ts
 * import { DOMAIN_SEPARATOR } from "thirdweb/extensions/erc20";
 *
 * const result = await DOMAIN_SEPARATOR({
 *  contract,
 * });
 *
 * ```
 */
async function DOMAIN_SEPARATOR(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
