// thirdweb/dist/cjs/utils/any-evm/compute-published-contract-deploy-info.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.computeDeploymentInfoFromContractId = computeDeploymentInfoFromContractId;
exports.computeDeploymentInfoFromMetadata = computeDeploymentInfoFromMetadata;
exports.computeDeploymentInfoFromBytecode = computeDeploymentInfoFromBytecode;
const publisher_js_1 = require("../../contract/deployment/publisher.js");
const create_2_factory_js_1 = require("../../contract/deployment/utils/create-2-factory.js");
const encodeAbiParameters_js_1 = require("../abi/encodeAbiParameters.js");
const normalizeFunctionParams_js_1 = require("../abi/normalizeFunctionParams.js");
const prefix_js_1 = require("../bytecode/prefix.js");
const get_init_bytecode_with_salt_js_1 = require("./get-init-bytecode-with-salt.js");
/**
 * @internal
 */
async function computeDeploymentInfoFromContractId(args) {
    const { client, chain, contractId, constructorParams, salt } = args;
    const contractMetadata = await (0, publisher_js_1.fetchPublishedContractMetadata)({
        client,
        contractId,
        publisher: args.publisher,
        version: args.version,
    });
    return computeDeploymentInfoFromMetadata({
        client,
        chain,
        contractMetadata,
        constructorParams,
        salt,
    });
}
/**
 * @internal
 */
async function computeDeploymentInfoFromMetadata(args) {
    return computeDeploymentInfoFromBytecode({
        client: args.client,
        chain: args.chain,
        abi: args.contractMetadata.abi,
        bytecode: args.contractMetadata.bytecode,
        constructorParams: args.constructorParams,
        salt: args.salt,
    });
}
async function computeDeploymentInfoFromBytecode(args) {
    const { client, chain, constructorParams, salt } = args;
    const create2FactoryAddress = await (0, create_2_factory_js_1.computeCreate2FactoryAddress)({
        client,
        chain,
    });
    const bytecode = (0, prefix_js_1.ensureBytecodePrefix)(args.bytecode);
    const constructorAbi = args.abi.find((abi) => abi.type === "constructor");
    const encodedArgs = (0, encodeAbiParameters_js_1.encodeAbiParameters)(constructorAbi?.inputs ?? [], (0, normalizeFunctionParams_js_1.normalizeFunctionParams)(constructorAbi, constructorParams));
    const initBytecodeWithsalt = (0, get_init_bytecode_with_salt_js_1.getInitBytecodeWithSalt)({
        bytecode,
        encodedArgs,
        salt,
    });
    return {
        bytecode,
        initBytecodeWithsalt,
        encodedArgs,
        create2FactoryAddress,
    };
}
