// thirdweb/dist/cjs/extensions/vote/__generated__/Vote/write/execute.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isExecuteSupported = isExecuteSupported;
exports.encodeExecuteParams = encodeExecuteParams;
exports.encodeExecute = encodeExecute;
exports.execute = execute;
const prepare_contract_call_js_1 = require("../../../../../transaction/prepare-contract-call.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const once_js_1 = require("../../../../../utils/promise/once.js");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x2656227d";
const FN_INPUTS = [
    {
        type: "address[]",
        name: "targets",
    },
    {
        type: "uint256[]",
        name: "values",
    },
    {
        type: "bytes[]",
        name: "calldatas",
    },
    {
        type: "bytes32",
        name: "descriptionHash",
    },
];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `execute` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `execute` method is supported.
 * @extension VOTE
 * @example
 * ```ts
 * import { isExecuteSupported } from "thirdweb/extensions/vote";
 *
 * const supported = isExecuteSupported(["0x..."]);
 * ```
 */
function isExecuteSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "execute" function.
 * @param options - The options for the execute function.
 * @returns The encoded ABI parameters.
 * @extension VOTE
 * @example
 * ```ts
 * import { encodeExecuteParams } from "thirdweb/extensions/vote";
 * const result = encodeExecuteParams({
 *  targets: ...,
 *  values: ...,
 *  calldatas: ...,
 *  descriptionHash: ...,
 * });
 * ```
 */
function encodeExecuteParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [
        options.targets,
        options.values,
        options.calldatas,
        options.descriptionHash,
    ]);
}
/**
 * Encodes the "execute" function into a Hex string with its parameters.
 * @param options - The options for the execute function.
 * @returns The encoded hexadecimal string.
 * @extension VOTE
 * @example
 * ```ts
 * import { encodeExecute } from "thirdweb/extensions/vote";
 * const result = encodeExecute({
 *  targets: ...,
 *  values: ...,
 *  calldatas: ...,
 *  descriptionHash: ...,
 * });
 * ```
 */
function encodeExecute(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeExecuteParams(options).slice(2));
}
/**
 * Prepares a transaction to call the "execute" function on the contract.
 * @param options - The options for the "execute" function.
 * @returns A prepared transaction object.
 * @extension VOTE
 * @example
 * ```ts
 * import { sendTransaction } from "thirdweb";
 * import { execute } from "thirdweb/extensions/vote";
 *
 * const transaction = execute({
 *  contract,
 *  targets: ...,
 *  values: ...,
 *  calldatas: ...,
 *  descriptionHash: ...,
 *  overrides: {
 *    ...
 *  }
 * });
 *
 * // Send the transaction
 * await sendTransaction({ transaction, account });
 * ```
 */
function execute(options) {
    const asyncOptions = (0, once_js_1.once)(async () => {
        return "asyncParams" in options ? await options.asyncParams() : options;
    });
    return (0, prepare_contract_call_js_1.prepareContractCall)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: async () => {
            const resolvedOptions = await asyncOptions();
            return [
                resolvedOptions.targets,
                resolvedOptions.values,
                resolvedOptions.calldatas,
                resolvedOptions.descriptionHash,
            ];
        },
        value: async () => (await asyncOptions()).overrides?.value,
        accessList: async () => (await asyncOptions()).overrides?.accessList,
        gas: async () => (await asyncOptions()).overrides?.gas,
        gasPrice: async () => (await asyncOptions()).overrides?.gasPrice,
        maxFeePerGas: async () => (await asyncOptions()).overrides?.maxFeePerGas,
        maxPriorityFeePerGas: async () => (await asyncOptions()).overrides?.maxPriorityFeePerGas,
        nonce: async () => (await asyncOptions()).overrides?.nonce,
        extraGas: async () => (await asyncOptions()).overrides?.extraGas,
        erc20Value: async () => (await asyncOptions()).overrides?.erc20Value,
    });
}
