// thirdweb/dist/cjs/extensions/thirdweb/__generated__/IRulesEngine/read/getScore.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isGetScoreSupported = isGetScoreSupported;
exports.encodeGetScoreParams = encodeGetScoreParams;
exports.encodeGetScore = encodeGetScore;
exports.decodeGetScoreResult = decodeGetScoreResult;
exports.getScore = getScore;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xd47875d0";
const FN_INPUTS = [
    {
        type: "address",
        name: "_tokenOwner",
    },
];
const FN_OUTPUTS = [
    {
        type: "uint256",
        name: "score",
    },
];
/**
 * Checks if the `getScore` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `getScore` method is supported.
 * @extension THIRDWEB
 * @example
 * ```ts
 * import { isGetScoreSupported } from "thirdweb/extensions/thirdweb";
 * const supported = isGetScoreSupported(["0x..."]);
 * ```
 */
function isGetScoreSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "getScore" function.
 * @param options - The options for the getScore function.
 * @returns The encoded ABI parameters.
 * @extension THIRDWEB
 * @example
 * ```ts
 * import { encodeGetScoreParams } from "thirdweb/extensions/thirdweb";
 * const result = encodeGetScoreParams({
 *  tokenOwner: ...,
 * });
 * ```
 */
function encodeGetScoreParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.tokenOwner]);
}
/**
 * Encodes the "getScore" function into a Hex string with its parameters.
 * @param options - The options for the getScore function.
 * @returns The encoded hexadecimal string.
 * @extension THIRDWEB
 * @example
 * ```ts
 * import { encodeGetScore } from "thirdweb/extensions/thirdweb";
 * const result = encodeGetScore({
 *  tokenOwner: ...,
 * });
 * ```
 */
function encodeGetScore(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeGetScoreParams(options).slice(2));
}
/**
 * Decodes the result of the getScore function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension THIRDWEB
 * @example
 * ```ts
 * import { decodeGetScoreResult } from "thirdweb/extensions/thirdweb";
 * const result = decodeGetScoreResultResult("...");
 * ```
 */
function decodeGetScoreResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getScore" function on the contract.
 * @param options - The options for the getScore function.
 * @returns The parsed result of the function call.
 * @extension THIRDWEB
 * @example
 * ```ts
 * import { getScore } from "thirdweb/extensions/thirdweb";
 *
 * const result = await getScore({
 *  contract,
 *  tokenOwner: ...,
 * });
 *
 * ```
 */
async function getScore(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.tokenOwner],
    });
}
