// thirdweb/dist/esm/extensions/modules/__generated__/BatchMetadataERC1155/read/getAllMetadataBatches.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0xe6c23512";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "tuple[]",
        components: [
            {
                type: "uint256",
                name: "startTokenIdInclusive",
            },
            {
                type: "uint256",
                name: "endTokenIdInclusive",
            },
            {
                type: "string",
                name: "baseURI",
            },
        ],
    },
];
/**
 * Checks if the `getAllMetadataBatches` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `getAllMetadataBatches` method is supported.
 * @modules BatchMetadataERC1155
 * @example
 * ```ts
 * import { BatchMetadataERC1155 } from "thirdweb/modules";
 * const supported = BatchMetadataERC1155.isGetAllMetadataBatchesSupported(["0x..."]);
 * ```
 */
export function isGetAllMetadataBatchesSupported(availableSelectors) {
    return detectMethod({
        availableSelectors,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the getAllMetadataBatches function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @modules BatchMetadataERC1155
 * @example
 * ```ts
 * import { BatchMetadataERC1155 } from "thirdweb/modules";
 * const result = BatchMetadataERC1155.decodeGetAllMetadataBatchesResultResult("...");
 * ```
 */
export function decodeGetAllMetadataBatchesResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getAllMetadataBatches" function on the contract.
 * @param options - The options for the getAllMetadataBatches function.
 * @returns The parsed result of the function call.
 * @modules BatchMetadataERC1155
 * @example
 * ```ts
 * import { BatchMetadataERC1155 } from "thirdweb/modules";
 *
 * const result = await BatchMetadataERC1155.getAllMetadataBatches({
 *  contract,
 * });
 *
 * ```
 */
export async function getAllMetadataBatches(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
