// thirdweb/dist/esm/react/web/wallets/in-app/InputSelectionUI.js
"use client";
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { ArrowRightIcon } from "@radix-ui/react-icons";
import { useState } from "react";
import { iconSize, radius, spacing, } from "../../../core/design-system/index.js";
import { Spacer } from "../../ui/components/Spacer.js";
import { IconButton } from "../../ui/components/buttons.js";
import { Input, InputContainer } from "../../ui/components/formElements.js";
import { Text } from "../../ui/components/text.js";
import { CountrySelector } from "./CountrySelector.js";
export function InputSelectionUI(props) {
    const [countryCodeInfo, setCountryCodeInfo] = useState("US +1");
    const [input, setInput] = useState("");
    const [error, setError] = useState();
    const [showError, setShowError] = useState(false);
    const handleSelect = () => {
        setShowError(true);
        if (!input || !!error) {
            return;
        }
        props.onSelect(props.format === "phone"
            ? `+${countryCodeInfo.split("+")[1]}${input}`
            : input);
    };
    const renderingError = (showError && !!error) ||
        (!input && !!props.emptyErrorMessage && showError);
    return (_jsxs("div", { style: {
            width: "100%",
        }, children: [_jsxs(InputContainer, { style: {
                    position: "relative",
                    display: "flex",
                    flexDirection: "row",
                }, "data-error": renderingError, children: [props.format === "phone" && (_jsx(CountrySelector, { countryCode: countryCodeInfo, setCountryCode: setCountryCodeInfo })), _jsx(Input, { tabIndex: -1, placeholder: props.placeholder, style: {
                            flexGrow: 1,
                            padding: `${spacing.sm} ${props.format === "phone" ? 0 : spacing.sm}`,
                        }, variant: "transparent", type: props.type, name: props.name, value: input, disabled: props.disabled, onChange: (e) => {
                            setInput(e.target.value);
                            if (props.errorMessage) {
                                setError(props.errorMessage(e.target.value));
                            }
                            else {
                                setError(undefined);
                            }
                        }, onKeyDown: (e) => {
                            if (e.key === "Enter") {
                                handleSelect();
                            }
                        } }), _jsx(IconButton, { onClick: handleSelect, disabled: props.disabled, style: {
                            padding: spacing.sm,
                            borderRadius: `0 ${radius.md} ${radius.md} 0`,
                        }, children: _jsx(ArrowRightIcon, { width: iconSize.md, height: iconSize.md }) })] }), showError && error && (_jsxs(_Fragment, { children: [_jsx(Spacer, { y: "xs" }), _jsx(Text, { color: "danger", size: "sm", children: error })] })), !(showError && error) &&
                !input &&
                props.emptyErrorMessage &&
                showError && (_jsxs(_Fragment, { children: [_jsx(Spacer, { y: "xs" }), _jsx(Text, { color: "danger", size: "sm", children: props.emptyErrorMessage })] }))] }));
}
