// thirdweb/dist/cjs/wallets/coinbase/coinbaseWebSDK.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCoinbaseWebProvider = getCoinbaseWebProvider;
exports.isCoinbaseSDKWallet = isCoinbaseSDKWallet;
exports.coinbaseSDKWalletGetCapabilities = coinbaseSDKWalletGetCapabilities;
exports.coinbaseSDKWalletSendCalls = coinbaseSDKWalletSendCalls;
exports.coinbaseSDKWalletShowCallsStatus = coinbaseSDKWalletShowCallsStatus;
exports.coinbaseSDKWalletGetCallsStatus = coinbaseSDKWalletGetCallsStatus;
exports.connectCoinbaseWalletSDK = connectCoinbaseWalletSDK;
exports.autoConnectCoinbaseWalletSDK = autoConnectCoinbaseWalletSDK;
const viem_1 = require("viem");
const chains_js_1 = require("../utils/chains.js");
const normalizeChainId_js_1 = require("../utils/normalizeChainId.js");
const utils_js_1 = require("../../chains/utils.js");
const address_js_1 = require("../../utils/address.js");
const hex_js_1 = require("../../utils/encoding/hex.js");
const parseTypedData_js_1 = require("../../utils/signatures/helpers/parseTypedData.js");
const constants_js_1 = require("../constants.js");
const defaultDappMetadata_js_1 = require("../utils/defaultDappMetadata.js");
const utils_js_2 = require("./utils.js");
// Need to keep the provider around because it keeps a single popup window connection behind the scenes
// this should be ok since all the creation options are provided at build time
let _provider;
/**
 * @internal
 */
async function getCoinbaseWebProvider(options) {
    if (!_provider) {
        let CoinbaseWalletSDK = (await Promise.resolve().then(() => require("@coinbase/wallet-sdk")))
            .default;
        // Workaround for Vite dev import errors
        // https://github.com/vitejs/vite/issues/7112
        if (typeof CoinbaseWalletSDK !== "function" &&
            typeof CoinbaseWalletSDK.default === "function") {
            CoinbaseWalletSDK = CoinbaseWalletSDK.default;
        }
        // @ts-expect-error This import error is not visible to TypeScript
        const client = new CoinbaseWalletSDK({
            appName: options?.appMetadata?.name || (0, defaultDappMetadata_js_1.getDefaultAppMetadata)().name,
            appChainIds: options?.chains
                ? options.chains.map((c) => c.id)
                : undefined,
            appLogoUrl: options?.appMetadata?.logoUrl || (0, defaultDappMetadata_js_1.getDefaultAppMetadata)().logoUrl,
        });
        const provider = client.makeWeb3Provider(options?.walletConfig);
        _provider = provider;
        return provider;
    }
    return _provider;
}
/**
 * Checks if the provided wallet is a Coinbase SDK wallet.
 *
 * @param wallet - The wallet to check.
 * @returns True if the wallet is a Coinbase SDK wallet, false otherwise.
 */
function isCoinbaseSDKWallet(wallet) {
    return wallet.id === constants_js_1.COINBASE;
}
/**
 * @internal
 */
async function coinbaseSDKWalletGetCapabilities(args) {
    const { wallet } = args;
    const account = wallet.getAccount();
    if (!account) {
        return {
            message: `Can't get capabilities, no account connected for wallet: ${wallet.id}`,
        };
    }
    const config = wallet.getConfig();
    const provider = await getCoinbaseWebProvider(config);
    try {
        return (await provider.request({
            method: "wallet_getCapabilities",
            params: [account.address],
        }));
    }
    catch (error) {
        if (/unsupport|not support/i.test(error.message)) {
            return {
                message: `${wallet.id} does not support wallet_getCapabilities, reach out to them directly to request EIP-5792 support.`,
            };
        }
        throw error;
    }
}
/**
 * @internal
 */
async function coinbaseSDKWalletSendCalls(args) {
    const { wallet, params } = args;
    const config = wallet.getConfig();
    const provider = await getCoinbaseWebProvider(config);
    try {
        return (await provider.request({
            method: "wallet_sendCalls",
            params,
        }));
    }
    catch (error) {
        if (/unsupport|not support/i.test(error.message)) {
            throw new Error(`${wallet.id} does not support wallet_sendCalls, reach out to them directly to request EIP-5792 support.`);
        }
        throw error;
    }
}
/**
 * @internal
 */
async function coinbaseSDKWalletShowCallsStatus(args) {
    const { wallet, bundleId } = args;
    const provider = await getCoinbaseWebProvider(wallet.getConfig());
    try {
        return await provider.request({
            method: "wallet_showCallsStatus",
            params: [bundleId],
        });
    }
    catch (error) {
        if (/unsupport|not support/i.test(error.message)) {
            throw new Error(`${wallet.id} does not support wallet_showCallsStatus, reach out to them directly to request EIP-5792 support.`);
        }
        throw error;
    }
}
/**
 * @internal
 */
async function coinbaseSDKWalletGetCallsStatus(args) {
    const { wallet, bundleId } = args;
    const config = wallet.getConfig();
    const provider = await getCoinbaseWebProvider(config);
    return provider.request({
        method: "wallet_getCallsStatus",
        params: [bundleId],
    });
}
function createAccount(provider, _address) {
    const address = (0, address_js_1.getAddress)(_address);
    const account = {
        address,
        async sendTransaction(tx) {
            const transactionHash = (await provider.request({
                method: "eth_sendTransaction",
                params: [
                    {
                        accessList: tx.accessList,
                        value: tx.value ? (0, hex_js_1.numberToHex)(tx.value) : undefined,
                        gas: tx.gas ? (0, hex_js_1.numberToHex)(tx.gas) : undefined,
                        from: this.address,
                        to: tx.to,
                        data: tx.data,
                    },
                ],
            }));
            return {
                transactionHash,
            };
        },
        async signMessage({ message }) {
            if (!account.address) {
                throw new Error("Provider not setup");
            }
            const messageToSign = (() => {
                if (typeof message === "string") {
                    return (0, hex_js_1.stringToHex)(message);
                }
                if (message.raw instanceof Uint8Array) {
                    return (0, hex_js_1.uint8ArrayToHex)(message.raw);
                }
                return message.raw;
            })();
            return await provider.request({
                method: "personal_sign",
                params: [messageToSign, account.address],
            });
        },
        async signTypedData(_typedData) {
            if (!account.address) {
                throw new Error("Provider not setup");
            }
            const typedData = (0, parseTypedData_js_1.parseTypedData)(_typedData);
            const { domain, message, primaryType } = typedData;
            const types = {
                EIP712Domain: (0, viem_1.getTypesForEIP712Domain)({ domain }),
                ...typedData.types,
            };
            // Need to do a runtime validation check on addresses, byte ranges, integer ranges, etc
            // as we can't statically check this with TypeScript.
            (0, viem_1.validateTypedData)({ domain, message, primaryType, types });
            const stringifiedData = (0, viem_1.serializeTypedData)({
                domain: domain ?? {},
                message,
                primaryType,
                types,
            });
            return await provider.request({
                method: "eth_signTypedData_v4",
                params: [account.address, stringifiedData],
            });
        },
        onTransactionRequested: async () => {
            // make sure to show the coinbase popup BEFORE doing any transaction preprocessing
            // otherwise the popup might get blocked in safari
            await (0, utils_js_2.showCoinbasePopup)(provider);
        },
    };
    return account;
}
function onConnect(address, chain, provider, emitter) {
    const account = createAccount(provider, address);
    async function disconnect() {
        provider.removeListener("accountsChanged", onAccountsChanged);
        provider.removeListener("chainChanged", onChainChanged);
        provider.removeListener("disconnect", onDisconnect);
        await provider.disconnect();
    }
    async function onDisconnect() {
        disconnect();
        emitter.emit("disconnect", undefined);
    }
    function onAccountsChanged(accounts) {
        if (accounts[0]) {
            const newAccount = createAccount(provider, (0, address_js_1.getAddress)(accounts[0]));
            emitter.emit("accountChanged", newAccount);
            emitter.emit("accountsChanged", accounts);
        }
        else {
            onDisconnect();
        }
    }
    function onChainChanged(newChainId) {
        const newChain = (0, utils_js_1.getCachedChain)((0, normalizeChainId_js_1.normalizeChainId)(newChainId));
        emitter.emit("chainChanged", newChain);
    }
    // subscribe to events
    provider.on("accountsChanged", onAccountsChanged);
    provider.on("chainChanged", onChainChanged);
    provider.on("disconnect", onDisconnect);
    return [
        account,
        chain,
        onDisconnect,
        (newChain) => switchChainCoinbaseWalletSDK(provider, newChain),
    ];
}
/**
 * @internal
 */
async function connectCoinbaseWalletSDK(options, emitter, provider) {
    const accounts = (await provider.request({
        method: "eth_requestAccounts",
    }));
    if (!accounts[0]) {
        throw new Error("No accounts found");
    }
    const address = (0, address_js_1.getAddress)(accounts[0]);
    const connectedChainId = (await provider.request({
        method: "eth_chainId",
    }));
    const chainId = (0, normalizeChainId_js_1.normalizeChainId)(connectedChainId);
    let chain = options.chain && options.chain.id === chainId
        ? options.chain
        : (0, utils_js_1.getCachedChain)(chainId);
    // Switch to chain if provided
    if (connectedChainId &&
        options?.chain &&
        connectedChainId !== options?.chain.id) {
        await switchChainCoinbaseWalletSDK(provider, options.chain);
        chain = options.chain;
    }
    return onConnect(address, chain, provider, emitter);
}
/**
 * @internal
 */
async function autoConnectCoinbaseWalletSDK(options, emitter, provider) {
    // connected accounts
    const addresses = (await provider.request({
        method: "eth_accounts",
    }));
    const address = addresses[0];
    if (!address) {
        throw new Error("No accounts found");
    }
    const connectedChainId = (await provider.request({
        method: "eth_chainId",
    }));
    const chainId = (0, normalizeChainId_js_1.normalizeChainId)(connectedChainId);
    const chain = options.chain && options.chain.id === chainId
        ? options.chain
        : (0, utils_js_1.getCachedChain)(chainId);
    return onConnect(address, chain, provider, emitter);
}
async function switchChainCoinbaseWalletSDK(provider, chain) {
    const chainIdHex = (0, hex_js_1.numberToHex)(chain.id);
    try {
        await provider.request({
            method: "wallet_switchEthereumChain",
            params: [{ chainId: chainIdHex }],
        });
    }
    catch (error) {
        const apiChain = await (0, utils_js_1.getChainMetadata)(chain);
        // Indicates chain is not added to provider
        // biome-ignore lint/suspicious/noExplicitAny: TODO: fix later
        if (error?.code === 4902) {
            // try to add the chain
            await provider.request({
                method: "wallet_addEthereumChain",
                params: [
                    {
                        chainId: chainIdHex,
                        chainName: apiChain.name,
                        nativeCurrency: apiChain.nativeCurrency,
                        rpcUrls: (0, chains_js_1.getValidPublicRPCUrl)(apiChain), // no client id on purpose here
                        blockExplorerUrls: apiChain.explorers?.map((x) => x.url) || [],
                    },
                ],
            });
        }
    }
}
