// thirdweb/dist/cjs/extensions/modules/ClaimableERC1155/mint.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.encodeMintParams = void 0;
exports.mint = mint;
const viem_1 = require("viem");
const addresses_js_1 = require("../../../constants/addresses.js");
const download_js_1 = require("../../../storage/download.js");
const hex_js_1 = require("../../../utils/encoding/hex.js");
const encodeBytesBeforeMintERC1155_js_1 = require("../__generated__/ClaimableERC1155/encode/encodeBytesBeforeMintERC1155.js");
const getClaimConditionByTokenId_js_1 = require("../__generated__/ClaimableERC1155/read/getClaimConditionByTokenId.js");
const mint_js_1 = require("../__generated__/ERC1155Core/write/mint.js");
/**
 * Mints ERC1155 tokens to a specified address via a ClaimableERC1155 module.
 * @param options The options for minting tokens.
 * @returns A transaction to mint tokens.
 * @example
 * ```typescript
 * import { ClaimableERC1155 } from "thirdweb/modules";
 *
 * const transaction = ClaimableERC1155.mint({
 *   contract,
 *   to: "0x...", // Address to mint tokens to
 *   quantity: 2, // Amount of tokens to mint
 * });
 *
 * // Send the transaction
 * await sendTransaction({ transaction, account });
 * ```
 * @modules ClaimableERC1155
 */
function mint(options) {
    return (0, mint_js_1.mint)({
        contract: options.contract,
        asyncParams: async () => {
            const cc = await (0, getClaimConditionByTokenId_js_1.getClaimConditionByTokenId)({
                contract: options.contract,
                tokenId: options.tokenId,
            });
            const totalPrice = cc.pricePerUnit * BigInt(options.quantity);
            const value = (0, addresses_js_1.isNativeTokenAddress)(cc.currency) ? totalPrice : 0n;
            const erc20Value = !(0, addresses_js_1.isNativeTokenAddress)(cc.currency) && cc.pricePerUnit > 0n
                ? {
                    amountWei: totalPrice,
                    tokenAddress: cc.currency,
                }
                : undefined;
            let recipientAllowlistProof = [];
            if (cc.allowlistMerkleRoot &&
                cc.allowlistMerkleRoot !== (0, hex_js_1.padHex)("0x", { size: 32 })) {
                const { fetchProofsForClaimer } = await Promise.resolve().then(() => require("../../../utils/extensions/drops/fetch-proofs-for-claimers.js"));
                const metadataUri = cc.auxData;
                if (metadataUri) {
                    // download merkle tree from metadata
                    const metadata = await (0, download_js_1.download)({
                        client: options.contract.client,
                        uri: metadataUri,
                    });
                    const metadataJson = await metadata.json();
                    const merkleTreeUri = metadataJson.merkleTreeUri;
                    // fetch proofs
                    if (merkleTreeUri) {
                        const allowlistProof = await fetchProofsForClaimer({
                            contract: options.contract,
                            claimer: options.to,
                            merkleTreeUri,
                            tokenDecimals: 18, // unused here
                            async hashEntry(options) {
                                return (0, viem_1.keccak256)((0, viem_1.encodePacked)(["address"], [(0, viem_1.getAddress)(options.entry.address)]));
                            },
                        });
                        recipientAllowlistProof = allowlistProof?.proof || [];
                    }
                }
            }
            return {
                to: (0, viem_1.getAddress)(options.to),
                tokenId: options.tokenId,
                amount: BigInt(options.quantity),
                baseURI: "",
                data: (0, encodeBytesBeforeMintERC1155_js_1.encodeBytesBeforeMintERC1155Params)({
                    params: {
                        currency: cc.currency,
                        pricePerUnit: cc.pricePerUnit,
                        recipientAllowlistProof,
                    },
                }),
                overrides: {
                    value,
                    erc20Value,
                },
            };
        },
    });
}
exports.encodeMintParams = encodeBytesBeforeMintERC1155_js_1.encodeBytesBeforeMintERC1155Params;
