// thirdweb/dist/esm/extensions/erc4337/__generated__/IAccountFactory/read/getAllAccounts.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0x08e93d0a";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "address[]",
    },
];
/**
 * Checks if the `getAllAccounts` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `getAllAccounts` method is supported.
 * @extension ERC4337
 * @example
 * ```ts
 * import { isGetAllAccountsSupported } from "thirdweb/extensions/erc4337";
 * const supported = isGetAllAccountsSupported(["0x..."]);
 * ```
 */
export function isGetAllAccountsSupported(availableSelectors) {
    return detectMethod({
        availableSelectors,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the getAllAccounts function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC4337
 * @example
 * ```ts
 * import { decodeGetAllAccountsResult } from "thirdweb/extensions/erc4337";
 * const result = decodeGetAllAccountsResultResult("...");
 * ```
 */
export function decodeGetAllAccountsResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getAllAccounts" function on the contract.
 * @param options - The options for the getAllAccounts function.
 * @returns The parsed result of the function call.
 * @extension ERC4337
 * @example
 * ```ts
 * import { getAllAccounts } from "thirdweb/extensions/erc4337";
 *
 * const result = await getAllAccounts({
 *  contract,
 * });
 *
 * ```
 */
export async function getAllAccounts(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
