// thirdweb/dist/esm/react/native/ui/components/button.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { StyleSheet, TouchableOpacity, View, } from "react-native";
import { radius, spacing } from "../../design-system/index.js";
import { RNImage } from "./RNImage.js";
import { ThemedText } from "./text.js";
export function ThemedButton(props) {
    const variant = props.variant ?? "primary";
    const { style: styleOverride, theme, ...restProps } = props;
    return (_jsx(TouchableOpacity, { activeOpacity: 0.5, style: [
            styles.button,
            {
                borderColor: variant === "secondary" ? theme.colors.borderColor : "transparent",
                borderWidth: variant === "secondary" ? 1 : 0,
                backgroundColor: variant === "secondary"
                    ? "transparent"
                    : variant === "accent"
                        ? props.theme.colors.accentButtonBg
                        : props.theme.colors.primaryButtonBg,
            },
            styleOverride,
        ], ...restProps, children: props.children }));
}
export function ThemedButtonWithIcon(props) {
    const { theme, title, icon, onPress } = props;
    return (_jsx(ThemedButton, { theme: theme, variant: "secondary", onPress: onPress, children: _jsxs(View, { style: {
                flexDirection: "row",
                gap: spacing.md,
                paddingLeft: spacing.sm,
                alignContent: "center",
                width: "100%",
            }, children: [_jsx(RNImage, { theme: theme, size: 24, data: icon, color: theme.colors.accentButtonBg }), _jsx(ThemedText, { theme: theme, children: title })] }) }));
}
const styles = StyleSheet.create({
    button: {
        flexDirection: "row",
        padding: spacing.md,
        borderRadius: radius.lg,
        justifyContent: "center",
        alignItems: "center",
    },
});
