// thirdweb/dist/esm/react/native/ui/components/text.js
import { jsx as _jsx } from "react/jsx-runtime";
import { StyleSheet, Text } from "react-native";
export function ThemedText({ style, theme, type = "default", ...rest }) {
    return (_jsx(Text, { style: [
            { color: theme.colors.primaryText },
            type === "default" ? styles.default : undefined,
            type === "title" ? styles.title : undefined,
            type === "defaultSemiBold" ? styles.defaultSemiBold : undefined,
            type === "subtitle" ? styles.subtitle : undefined,
            type === "link" ? styles.link : undefined,
            type === "subtext"
                ? { ...styles.subtext, color: theme.colors.secondaryText }
                : undefined,
            style,
        ], ...rest }));
}
const styles = StyleSheet.create({
    default: {
        fontSize: 16,
    },
    defaultSemiBold: {
        fontSize: 16,
        fontWeight: "600",
    },
    subtext: {
        fontSize: 14,
    },
    title: {
        fontSize: 24,
        fontWeight: "bold",
        lineHeight: 28,
    },
    subtitle: {
        fontSize: 18,
        fontWeight: "600",
    },
    link: {
        fontSize: 16,
    },
});
