// thirdweb/dist/cjs/wallets/in-app/native/helpers/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ROUTE_AUTH_COGNITO_ID_TOKEN_V2 = exports.ROUTE_AUTH_COGNITO_ID_TOKEN_V1 = exports.ROUTE_AUTH_ENDPOINT_CALLBACK = exports.ROUTE_AUTH_JWT_CALLBACK = exports.ROUTE_VERIFY_THIRDWEB_CLIENT_ID = exports.ROUTE_GET_USER_SHARES = exports.ROUTE_STORE_USER_SHARES = exports.ROUTE_COGNITO_IDENTITY_POOL_URL = exports.ROUTE_EMBEDDED_WALLET_DETAILS = exports.GENERATE_RECOVERY_PASSWORD_LAMBDA_FUNCTION_V2 = exports.GENERATE_RECOVERY_PASSWORD_LAMBDA_FUNCTION_V1 = exports.COGNITO_IDENTITY_POOL_ID = exports.THIRDWEB_SESSION_NONCE_HEADER = exports.AWS_REGION = exports.RECOVERY_SHARE_INDEX = exports.DEVICE_SHARE_MISSING_MESSAGE = exports.DEVICE_SHARE_INDEX = exports.AUTH_SHARE_INDEX = void 0;
const domains_js_1 = require("../../../../utils/domains.js");
const AUTH_SHARE_ID = 3;
exports.AUTH_SHARE_INDEX = AUTH_SHARE_ID - 1;
const DEVICE_SHARE_ID = 1;
exports.DEVICE_SHARE_INDEX = DEVICE_SHARE_ID - 1;
exports.DEVICE_SHARE_MISSING_MESSAGE = "Missing device share.";
const RECOVERY_SHARE_ID = 2;
exports.RECOVERY_SHARE_INDEX = RECOVERY_SHARE_ID - 1;
exports.AWS_REGION = "us-west-2";
exports.THIRDWEB_SESSION_NONCE_HEADER = "x-session-nonce";
const COGNITO_USER_POOL_ID = "us-west-2_UFwLcZIpq";
exports.COGNITO_IDENTITY_POOL_ID = "us-west-2:2ad7ab1e-f48b-48a6-adfa-ac1090689c26";
exports.GENERATE_RECOVERY_PASSWORD_LAMBDA_FUNCTION_V1 = "arn:aws:lambda:us-west-2:324457261097:function:recovery-share-password-GenerateRecoverySharePassw-bbE5ZbVAToil";
exports.GENERATE_RECOVERY_PASSWORD_LAMBDA_FUNCTION_V2 = "arn:aws:lambda:us-west-2:324457261097:function:lambda-thirdweb-auth-enc-key-prod-ThirdwebAuthEncKeyFunction";
// TODO allow overriding domain
const DOMAIN_URL_2023 = (0, domains_js_1.getThirdwebBaseUrl)("inAppWallet");
const BASE_URL_2023 = `${DOMAIN_URL_2023}/`;
const ROUTE_2023_10_20_API_BASE_PATH = `${BASE_URL_2023}api/2023-10-20`;
const ROUTE_2024_05_05_API_BASE_PATH = `${BASE_URL_2023}api/2024-05-05`;
exports.ROUTE_EMBEDDED_WALLET_DETAILS = `${ROUTE_2023_10_20_API_BASE_PATH}/embedded-wallet/embedded-wallet-user-details`;
exports.ROUTE_COGNITO_IDENTITY_POOL_URL = `cognito-idp.${exports.AWS_REGION}.amazonaws.com/${COGNITO_USER_POOL_ID}`;
exports.ROUTE_STORE_USER_SHARES = `${ROUTE_2023_10_20_API_BASE_PATH}/embedded-wallet/embedded-wallet-shares`;
exports.ROUTE_GET_USER_SHARES = `${ROUTE_2023_10_20_API_BASE_PATH}/embedded-wallet/embedded-wallet-shares`;
exports.ROUTE_VERIFY_THIRDWEB_CLIENT_ID = `${ROUTE_2023_10_20_API_BASE_PATH}/embedded-wallet/verify-thirdweb-client-id`;
exports.ROUTE_AUTH_JWT_CALLBACK = `${ROUTE_2023_10_20_API_BASE_PATH}/embedded-wallet/validate-custom-jwt`;
exports.ROUTE_AUTH_ENDPOINT_CALLBACK = `${ROUTE_2023_10_20_API_BASE_PATH}/embedded-wallet/validate-custom-auth-endpoint`;
exports.ROUTE_AUTH_COGNITO_ID_TOKEN_V1 = `${ROUTE_2023_10_20_API_BASE_PATH}/embedded-wallet/cognito-id-token`;
exports.ROUTE_AUTH_COGNITO_ID_TOKEN_V2 = `${ROUTE_2024_05_05_API_BASE_PATH}/login/web-token-exchange`;
