// thirdweb/dist/cjs/extensions/farcaster/__generated__/IIdRegistry/read/CHANGE_RECOVERY_ADDRESS_TYPEHASH.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isCHANGE_RECOVERY_ADDRESS_TYPEHASHSupported = isCHANGE_RECOVERY_ADDRESS_TYPEHASHSupported;
exports.decodeCHANGE_RECOVERY_ADDRESS_TYPEHASHResult = decodeCHANGE_RECOVERY_ADDRESS_TYPEHASHResult;
exports.CHANGE_RECOVERY_ADDRESS_TYPEHASH = CHANGE_RECOVERY_ADDRESS_TYPEHASH;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xd5bac7f3";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "bytes32",
    },
];
/**
 * Checks if the `CHANGE_RECOVERY_ADDRESS_TYPEHASH` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `CHANGE_RECOVERY_ADDRESS_TYPEHASH` method is supported.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { isCHANGE_RECOVERY_ADDRESS_TYPEHASHSupported } from "thirdweb/extensions/farcaster";
 * const supported = isCHANGE_RECOVERY_ADDRESS_TYPEHASHSupported(["0x..."]);
 * ```
 */
function isCHANGE_RECOVERY_ADDRESS_TYPEHASHSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the CHANGE_RECOVERY_ADDRESS_TYPEHASH function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { decodeCHANGE_RECOVERY_ADDRESS_TYPEHASHResult } from "thirdweb/extensions/farcaster";
 * const result = decodeCHANGE_RECOVERY_ADDRESS_TYPEHASHResultResult("...");
 * ```
 */
function decodeCHANGE_RECOVERY_ADDRESS_TYPEHASHResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "CHANGE_RECOVERY_ADDRESS_TYPEHASH" function on the contract.
 * @param options - The options for the CHANGE_RECOVERY_ADDRESS_TYPEHASH function.
 * @returns The parsed result of the function call.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { CHANGE_RECOVERY_ADDRESS_TYPEHASH } from "thirdweb/extensions/farcaster";
 *
 * const result = await CHANGE_RECOVERY_ADDRESS_TYPEHASH({
 *  contract,
 * });
 *
 * ```
 */
async function CHANGE_RECOVERY_ADDRESS_TYPEHASH(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
