// thirdweb/dist/cjs/extensions/farcaster/__generated__/IIdRegistry/read/custodyOf.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isCustodyOfSupported = isCustodyOfSupported;
exports.encodeCustodyOfParams = encodeCustodyOfParams;
exports.encodeCustodyOf = encodeCustodyOf;
exports.decodeCustodyOfResult = decodeCustodyOfResult;
exports.custodyOf = custodyOf;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x65269e47";
const FN_INPUTS = [
    {
        type: "uint256",
        name: "fid",
    },
];
const FN_OUTPUTS = [
    {
        type: "address",
        name: "owner",
    },
];
/**
 * Checks if the `custodyOf` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `custodyOf` method is supported.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { isCustodyOfSupported } from "thirdweb/extensions/farcaster";
 * const supported = isCustodyOfSupported(["0x..."]);
 * ```
 */
function isCustodyOfSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "custodyOf" function.
 * @param options - The options for the custodyOf function.
 * @returns The encoded ABI parameters.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { encodeCustodyOfParams } from "thirdweb/extensions/farcaster";
 * const result = encodeCustodyOfParams({
 *  fid: ...,
 * });
 * ```
 */
function encodeCustodyOfParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.fid]);
}
/**
 * Encodes the "custodyOf" function into a Hex string with its parameters.
 * @param options - The options for the custodyOf function.
 * @returns The encoded hexadecimal string.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { encodeCustodyOf } from "thirdweb/extensions/farcaster";
 * const result = encodeCustodyOf({
 *  fid: ...,
 * });
 * ```
 */
function encodeCustodyOf(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeCustodyOfParams(options).slice(2));
}
/**
 * Decodes the result of the custodyOf function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { decodeCustodyOfResult } from "thirdweb/extensions/farcaster";
 * const result = decodeCustodyOfResultResult("...");
 * ```
 */
function decodeCustodyOfResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "custodyOf" function on the contract.
 * @param options - The options for the custodyOf function.
 * @returns The parsed result of the function call.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { custodyOf } from "thirdweb/extensions/farcaster";
 *
 * const result = await custodyOf({
 *  contract,
 *  fid: ...,
 * });
 *
 * ```
 */
async function custodyOf(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.fid],
    });
}
