// thirdweb/dist/cjs/react/web/ui/ConnectWallet/screens/SendFunds.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SendFunds = SendFunds;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_icons_1 = require("@radix-ui/react-icons");
const react_1 = require("react");
const index_js_1 = require("../../../../core/design-system/index.js");
const useWalletBalance_js_1 = require("../../../../core/hooks/others/useWalletBalance.js");
const useActiveAccount_js_1 = require("../../../../core/hooks/wallets/useActiveAccount.js");
const useActiveWalletChain_js_1 = require("../../../../core/hooks/wallets/useActiveWalletChain.js");
const useSendToken_js_1 = require("../../../../core/hooks/wallets/useSendToken.js");
const defaultTokens_js_1 = require("../../../../core/utils/defaultTokens.js");
const Skeleton_js_1 = require("../../components/Skeleton.js");
const Spacer_js_1 = require("../../components/Spacer.js");
const Spinner_js_1 = require("../../components/Spinner.js");
const TokenIcon_js_1 = require("../../components/TokenIcon.js");
const basic_js_1 = require("../../components/basic.js");
const buttons_js_1 = require("../../components/buttons.js");
const formElements_js_1 = require("../../components/formElements.js");
const text_js_1 = require("../../components/text.js");
const elements_js_1 = require("../../design-system/elements.js");
const TokenSelector_js_1 = require("./TokenSelector.js");
const formatTokenBalance_js_1 = require("./formatTokenBalance.js");
const nativeToken_js_1 = require("./nativeToken.js");
/**
 * @internal
 */
function SendFunds(props) {
    const [screen, setScreen] = (0, react_1.useState)("base");
    const activeChain = (0, useActiveWalletChain_js_1.useActiveWalletChain)();
    const chainId = activeChain?.id;
    const { connectLocale, client } = props;
    let defaultToken = nativeToken_js_1.NATIVE_TOKEN;
    const supportedTokens = props.supportedTokens || defaultTokens_js_1.defaultTokens;
    if (
    // if we know chainId
    chainId &&
        // if there is a list of tokens for this chain
        supportedTokens[chainId] &&
        // if the list of tokens is not the default list
        supportedTokens[chainId] !== defaultTokens_js_1.defaultTokens[chainId]) {
        // use the first token in the list as default selected
        const tokensForChain = supportedTokens[chainId];
        const firstToken = tokensForChain?.[0];
        if (firstToken) {
            defaultToken = firstToken;
        }
    }
    const [token, setToken] = (0, react_1.useState)(defaultToken);
    const [receiverAddress, setReceiverAddress] = (0, react_1.useState)("");
    const [amount, setAmount] = (0, react_1.useState)("0");
    const chain = (0, useActiveWalletChain_js_1.useActiveWalletChain)();
    const tokenList = (chain?.id ? supportedTokens[chain.id] : undefined) || [];
    if (screen === "tokenSelector" && chain) {
        return ((0, jsx_runtime_1.jsx)(TokenSelector_js_1.TokenSelector, { tokenList: tokenList, onBack: () => {
                setScreen("base");
            }, onTokenSelect: (_token) => {
                setToken(_token);
                setScreen("base");
            }, chain: chain, connectLocale: connectLocale, client: client }));
    }
    return ((0, jsx_runtime_1.jsx)(SendFundsForm, { token: token, onTokenSelect: () => {
            setScreen("tokenSelector");
        }, receiverAddress: receiverAddress, setReceiverAddress: setReceiverAddress, amount: amount, setAmount: setAmount, onBack: props.onBack, client: client, connectLocale: connectLocale }));
}
/**
 * @internal
 */
function SendFundsForm(props) {
    const locale = props.connectLocale.sendFundsScreen;
    const tokenAddress = props.token && "address" in props.token ? props.token.address : undefined;
    const chain = (0, useActiveWalletChain_js_1.useActiveWalletChain)();
    const activeAccount = (0, useActiveAccount_js_1.useActiveAccount)();
    const activeChain = (0, useActiveWalletChain_js_1.useActiveWalletChain)();
    const balanceQuery = (0, useWalletBalance_js_1.useWalletBalance)({
        chain,
        tokenAddress: tokenAddress,
        address: activeAccount?.address,
        client: props.client,
    });
    const { receiverAddress, setReceiverAddress, amount, setAmount } = props;
    const sendTokenMutation = (0, useSendToken_js_1.useSendToken)(props.client);
    function getErrorMessage(error) {
        const message = error?.data?.message || error?.message;
        if (!message) {
            return locale.transactionFailed;
        }
        if (message.includes("user rejected")) {
            return locale.transactionRejected;
        }
        if (message.includes("insufficient funds")) {
            return locale.insufficientFunds;
        }
        return message;
    }
    if (!activeChain) {
        return null; // this should never happen
    }
    if (sendTokenMutation.isError) {
        return ((0, jsx_runtime_1.jsxs)(basic_js_1.Container, { p: "lg", animate: "fadein", children: [(0, jsx_runtime_1.jsx)(basic_js_1.ModalHeader, { title: locale.title, onBack: () => {
                        sendTokenMutation.reset();
                    } }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "xl" }), (0, jsx_runtime_1.jsxs)(basic_js_1.Container, { flex: "column", gap: "lg", animate: "fadein", center: "both", style: {
                        minHeight: "200px",
                    }, color: "danger", children: [(0, jsx_runtime_1.jsx)(react_icons_1.CrossCircledIcon, { width: index_js_1.iconSize.xl, height: index_js_1.iconSize.xl }), (0, jsx_runtime_1.jsx)(text_js_1.Text, { center: true, multiline: true, color: "danger", children: getErrorMessage(sendTokenMutation.error) })] })] }));
    }
    if (sendTokenMutation.isSuccess) {
        return ((0, jsx_runtime_1.jsxs)(basic_js_1.Container, { p: "lg", animate: "fadein", children: [(0, jsx_runtime_1.jsx)(basic_js_1.ModalHeader, { title: locale.title, onBack: () => {
                        sendTokenMutation.reset();
                    } }), (0, jsx_runtime_1.jsxs)(basic_js_1.Container, { flex: "column", gap: "lg", animate: "fadein", center: "both", style: {
                        minHeight: "250px",
                    }, color: "success", children: [(0, jsx_runtime_1.jsx)(react_icons_1.CheckCircledIcon, { width: index_js_1.iconSize.xl, height: index_js_1.iconSize.xl }), (0, jsx_runtime_1.jsxs)(text_js_1.Text, { color: "success", children: [" ", locale.successMessage, " "] })] })] }));
    }
    const tokenName = (props.token && "name" in props.token ? props.token.name : undefined) ||
        balanceQuery?.data?.name;
    const tokenSymbol = (props.token && "symbol" in props.token ? props.token.symbol : undefined) ||
        balanceQuery?.data?.symbol;
    return ((0, jsx_runtime_1.jsxs)(basic_js_1.Container, { p: "lg", animate: "fadein", children: [(0, jsx_runtime_1.jsx)(basic_js_1.ModalHeader, { title: locale.title, onBack: props.onBack }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "xl" }), (0, jsx_runtime_1.jsxs)("form", { onSubmit: (e) => {
                    e.preventDefault();
                }, children: [(0, jsx_runtime_1.jsx)(formElements_js_1.Label, { htmlFor: "token", color: "secondaryText", children: locale.token }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "sm" }), (0, jsx_runtime_1.jsxs)(buttons_js_1.Button, { id: "token", variant: "outline", fullWidth: true, style: {
                            justifyContent: "flex-start",
                            gap: index_js_1.spacing.sm,
                            padding: index_js_1.spacing.sm,
                        }, onClick: props.onTokenSelect, children: [(0, jsx_runtime_1.jsx)(TokenIcon_js_1.TokenIcon, { token: props.token, chain: activeChain, size: "lg", client: props.client }), (0, jsx_runtime_1.jsxs)(basic_js_1.Container, { flex: "column", gap: "xs", children: [tokenName ? ((0, jsx_runtime_1.jsx)(text_js_1.Text, { size: "sm", color: "primaryText", children: tokenName })) : ((0, jsx_runtime_1.jsx)(Skeleton_js_1.Skeleton, { height: index_js_1.fontSize.xs, width: "150px" })), balanceQuery.data ? ((0, jsx_runtime_1.jsx)(text_js_1.Text, { size: "xs", children: (0, formatTokenBalance_js_1.formatTokenBalance)(balanceQuery.data) })) : ((0, jsx_runtime_1.jsx)(Skeleton_js_1.Skeleton, { height: index_js_1.fontSize.xs, width: "100px" }))] })] }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "lg" }), (0, jsx_runtime_1.jsx)(formElements_js_1.Label, { htmlFor: "receiver", color: "secondaryText", children: locale.sendTo }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "sm" }), (0, jsx_runtime_1.jsx)(formElements_js_1.Input, { required: true, id: "receiver", placeholder: "0x... or ENS name", variant: "outline", value: receiverAddress, onChange: (e) => {
                            setReceiverAddress(e.target.value);
                        } }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "lg" }), (0, jsx_runtime_1.jsx)(formElements_js_1.Label, { htmlFor: "amount", color: "secondaryText", children: locale.amount }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "sm" }), (0, jsx_runtime_1.jsxs)(basic_js_1.Container, { relative: true, children: [(0, jsx_runtime_1.jsx)(formElements_js_1.Input, { required: true, type: "number", id: "amount", variant: "outline", value: amount, onChange: (e) => {
                                    setAmount(e.target.value);
                                } }), (0, jsx_runtime_1.jsx)(CurrencyBadge, { children: (0, jsx_runtime_1.jsxs)(text_js_1.Text, { size: "xs", children: [" ", tokenSymbol, " "] }) })] }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "xxl" }), (0, jsx_runtime_1.jsxs)(buttons_js_1.Button, { fullWidth: true, variant: "accent", type: "submit", onClick: async () => {
                            if (!receiverAddress || !amount) {
                                return;
                            }
                            await sendTokenMutation.mutateAsync({
                                receiverAddress,
                                amount,
                                tokenAddress: tokenAddress,
                            });
                        }, style: {
                            alignItems: "center",
                            gap: index_js_1.spacing.sm,
                            padding: index_js_1.spacing.md,
                        }, children: [sendTokenMutation.isPending && ((0, jsx_runtime_1.jsx)(Spinner_js_1.Spinner, { size: "sm", color: "accentButtonText" })), sendTokenMutation.isPending ? locale.sending : locale.submitButton] })] })] }));
}
const CurrencyBadge = /* @__PURE__ */ (0, elements_js_1.StyledDiv)({
    position: "absolute",
    top: "50%",
    transform: "translateY(-50%)",
    right: index_js_1.spacing.sm,
});
