// thirdweb/dist/cjs/extensions/modules/__generated__/OpenEditionMetadataERC721/write/setSharedMetadata.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isSetSharedMetadataSupported = isSetSharedMetadataSupported;
exports.encodeSetSharedMetadataParams = encodeSetSharedMetadataParams;
exports.encodeSetSharedMetadata = encodeSetSharedMetadata;
exports.setSharedMetadata = setSharedMetadata;
const prepare_contract_call_js_1 = require("../../../../../transaction/prepare-contract-call.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const once_js_1 = require("../../../../../utils/promise/once.js");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xa7d27d9d";
const FN_INPUTS = [
    {
        type: "tuple",
        name: "_metadata",
        components: [
            {
                type: "string",
                name: "name",
            },
            {
                type: "string",
                name: "description",
            },
            {
                type: "string",
                name: "imageURI",
            },
            {
                type: "string",
                name: "animationURI",
            },
        ],
    },
];
const FN_OUTPUTS = [];
/**
 * Checks if the `setSharedMetadata` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `setSharedMetadata` method is supported.
 * @modules OpenEditionMetadataERC721
 * @example
 * ```ts
 * import { OpenEditionMetadataERC721 } from "thirdweb/modules";
 *
 * const supported = OpenEditionMetadataERC721.isSetSharedMetadataSupported(["0x..."]);
 * ```
 */
function isSetSharedMetadataSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "setSharedMetadata" function.
 * @param options - The options for the setSharedMetadata function.
 * @returns The encoded ABI parameters.
 * @modules OpenEditionMetadataERC721
 * @example
 * ```ts
 * import { OpenEditionMetadataERC721 } from "thirdweb/modules";
 * const result = OpenEditionMetadataERC721.encodeSetSharedMetadataParams({
 *  metadata: ...,
 * });
 * ```
 */
function encodeSetSharedMetadataParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.metadata]);
}
/**
 * Encodes the "setSharedMetadata" function into a Hex string with its parameters.
 * @param options - The options for the setSharedMetadata function.
 * @returns The encoded hexadecimal string.
 * @modules OpenEditionMetadataERC721
 * @example
 * ```ts
 * import { OpenEditionMetadataERC721 } from "thirdweb/modules";
 * const result = OpenEditionMetadataERC721.encodeSetSharedMetadata({
 *  metadata: ...,
 * });
 * ```
 */
function encodeSetSharedMetadata(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeSetSharedMetadataParams(options).slice(2));
}
/**
 * Prepares a transaction to call the "setSharedMetadata" function on the contract.
 * @param options - The options for the "setSharedMetadata" function.
 * @returns A prepared transaction object.
 * @modules OpenEditionMetadataERC721
 * @example
 * ```ts
 * import { sendTransaction } from "thirdweb";
 * import { OpenEditionMetadataERC721 } from "thirdweb/modules";
 *
 * const transaction = OpenEditionMetadataERC721.setSharedMetadata({
 *  contract,
 *  metadata: ...,
 *  overrides: {
 *    ...
 *  }
 * });
 *
 * // Send the transaction
 * await sendTransaction({ transaction, account });
 * ```
 */
function setSharedMetadata(options) {
    const asyncOptions = (0, once_js_1.once)(async () => {
        return "asyncParams" in options ? await options.asyncParams() : options;
    });
    return (0, prepare_contract_call_js_1.prepareContractCall)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: async () => {
            const resolvedOptions = await asyncOptions();
            return [resolvedOptions.metadata];
        },
        value: async () => (await asyncOptions()).overrides?.value,
        accessList: async () => (await asyncOptions()).overrides?.accessList,
        gas: async () => (await asyncOptions()).overrides?.gas,
        gasPrice: async () => (await asyncOptions()).overrides?.gasPrice,
        maxFeePerGas: async () => (await asyncOptions()).overrides?.maxFeePerGas,
        maxPriorityFeePerGas: async () => (await asyncOptions()).overrides?.maxPriorityFeePerGas,
        nonce: async () => (await asyncOptions()).overrides?.nonce,
        extraGas: async () => (await asyncOptions()).overrides?.extraGas,
        erc20Value: async () => (await asyncOptions()).overrides?.erc20Value,
    });
}
