// thirdweb/dist/esm/exports/extensions/airdrop.js
/**
 * READ
 */
export { isClaimed, } from "../../extensions/airdrop/__generated__/Airdrop/read/isClaimed.js";
export { tokenConditionId, } from "../../extensions/airdrop/__generated__/Airdrop/read/tokenConditionId.js";
export { tokenMerkleRoot, } from "../../extensions/airdrop/__generated__/Airdrop/read/tokenMerkleRoot.js";
export { fetchProofsERC721 } from "../../utils/extensions/airdrop/fetch-proofs-erc721.js";
export { fetchProofsERC20 } from "../../utils/extensions/airdrop/fetch-proofs-erc20.js";
export { fetchProofsERC1155 } from "../../utils/extensions/airdrop/fetch-proofs-erc1155.js";
/**
 * Write
 */
export { airdropERC20, } from "../../extensions/airdrop/__generated__/Airdrop/write/airdropERC20.js";
export { airdropERC721, } from "../../extensions/airdrop/__generated__/Airdrop/write/airdropERC721.js";
export { airdropERC1155, } from "../../extensions/airdrop/__generated__/Airdrop/write/airdropERC1155.js";
export { airdropNativeToken, } from "../../extensions/airdrop/__generated__/Airdrop/write/airdropNativeToken.js";
export { setMerkleRoot, } from "../../extensions/airdrop/__generated__/Airdrop/write/setMerkleRoot.js";
export { airdropERC20WithSignature, generateAirdropSignatureERC20, } from "../../extensions/airdrop/write/airdropERC20WithSignature.js";
export { airdropERC721WithSignature, generateAirdropSignatureERC721, } from "../../extensions/airdrop/write/airdropERC721WithSignature.js";
export { airdropERC1155WithSignature, generateAirdropSignatureERC1155, } from "../../extensions/airdrop/write/airdropERC1155WithSignature.js";
export { claimERC20, } from "../../extensions/airdrop/write/claimERC20.js";
export { claimERC721, } from "../../extensions/airdrop/write/claimERC721.js";
export { claimERC1155, } from "../../extensions/airdrop/write/claimERC1155.js";
export { generateMerkleTreeInfoERC20, } from "../../extensions/airdrop/write/merkleInfoERC20.js";
export { generateMerkleTreeInfoERC721, } from "../../extensions/airdrop/write/merkleInfoERC721.js";
export { generateMerkleTreeInfoERC1155, } from "../../extensions/airdrop/write/merkleInfoERC1155.js";
export { saveSnapshot, } from "../../extensions/airdrop/write/saveSnapshot.js";
