// thirdweb/dist/esm/exports/extensions/erc4626.js
// read
export { asset } from "../../extensions/erc4626/__generated__/IERC4626/read/asset.js";
export { convertToAssets, } from "../../extensions/erc4626/__generated__/IERC4626/read/convertToAssets.js";
export { convertToShares, } from "../../extensions/erc4626/__generated__/IERC4626/read/convertToShares.js";
export { maxDeposit, } from "../../extensions/erc4626/__generated__/IERC4626/read/maxDeposit.js";
export { maxMint, } from "../../extensions/erc4626/__generated__/IERC4626/read/maxMint.js";
export { maxRedeem, } from "../../extensions/erc4626/__generated__/IERC4626/read/maxRedeem.js";
export { maxWithdraw, } from "../../extensions/erc4626/__generated__/IERC4626/read/maxWithdraw.js";
export { previewDeposit, } from "../../extensions/erc4626/__generated__/IERC4626/read/previewDeposit.js";
export { previewMint, } from "../../extensions/erc4626/__generated__/IERC4626/read/previewMint.js";
export { previewRedeem, } from "../../extensions/erc4626/__generated__/IERC4626/read/previewRedeem.js";
export { previewWithdraw, } from "../../extensions/erc4626/__generated__/IERC4626/read/previewWithdraw.js";
export { totalAssets } from "../../extensions/erc4626/__generated__/IERC4626/read/totalAssets.js";
// write
export { deposit, } from "../../extensions/erc4626/__generated__/IERC4626/write/deposit.js";
export { redeem, } from "../../extensions/erc4626/__generated__/IERC4626/write/redeem.js";
export { mint, } from "../../extensions/erc4626/__generated__/IERC4626/write/mint.js";
export { withdraw, } from "../../extensions/erc4626/__generated__/IERC4626/write/withdraw.js";
// events
export { depositEvent, } from "../../extensions/erc4626/__generated__/IERC4626/events/Deposit.js";
export { withdrawEvent, } from "../../extensions/erc4626/__generated__/IERC4626/events/Withdraw.js";
