// thirdweb/dist/esm/extensions/marketplace/__generated__/IEnglishAuctions/read/isNewWinningBid.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { encodeAbiParameters } from "../../../../../utils/abi/encodeAbiParameters.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0x2eb566bd";
const FN_INPUTS = [
    {
        type: "uint256",
        name: "_auctionId",
    },
    {
        type: "uint256",
        name: "_bidAmount",
    },
];
const FN_OUTPUTS = [
    {
        type: "bool",
    },
];
/**
 * Checks if the `isNewWinningBid` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `isNewWinningBid` method is supported.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { isIsNewWinningBidSupported } from "thirdweb/extensions/marketplace";
 * const supported = isIsNewWinningBidSupported(["0x..."]);
 * ```
 */
export function isIsNewWinningBidSupported(availableSelectors) {
    return detectMethod({
        availableSelectors,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "isNewWinningBid" function.
 * @param options - The options for the isNewWinningBid function.
 * @returns The encoded ABI parameters.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { encodeIsNewWinningBidParams } from "thirdweb/extensions/marketplace";
 * const result = encodeIsNewWinningBidParams({
 *  auctionId: ...,
 *  bidAmount: ...,
 * });
 * ```
 */
export function encodeIsNewWinningBidParams(options) {
    return encodeAbiParameters(FN_INPUTS, [options.auctionId, options.bidAmount]);
}
/**
 * Encodes the "isNewWinningBid" function into a Hex string with its parameters.
 * @param options - The options for the isNewWinningBid function.
 * @returns The encoded hexadecimal string.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { encodeIsNewWinningBid } from "thirdweb/extensions/marketplace";
 * const result = encodeIsNewWinningBid({
 *  auctionId: ...,
 *  bidAmount: ...,
 * });
 * ```
 */
export function encodeIsNewWinningBid(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (FN_SELECTOR +
        encodeIsNewWinningBidParams(options).slice(2));
}
/**
 * Decodes the result of the isNewWinningBid function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { decodeIsNewWinningBidResult } from "thirdweb/extensions/marketplace";
 * const result = decodeIsNewWinningBidResultResult("...");
 * ```
 */
export function decodeIsNewWinningBidResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "isNewWinningBid" function on the contract.
 * @param options - The options for the isNewWinningBid function.
 * @returns The parsed result of the function call.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { isNewWinningBid } from "thirdweb/extensions/marketplace";
 *
 * const result = await isNewWinningBid({
 *  contract,
 *  auctionId: ...,
 *  bidAmount: ...,
 * });
 *
 * ```
 */
export async function isNewWinningBid(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.auctionId, options.bidAmount],
    });
}
