// thirdweb/dist/cjs/exports/modules.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.uninstallModuleByProxy = exports.uninstallPublishedModule = exports.installPublishedModule = exports.isGetModuleConfigSupported = exports.getModuleConfig = exports.isUninstallModuleSupported = exports.uninstallModule = exports.transferOwnership = exports.revokeRoles = exports.renounceRoles = exports.requestOwnershipHandover = exports.renounceOwnership = exports.isInstallModuleSupported = exports.installModule = exports.grantRoles = exports.completeOwnershipHandover = exports.cancelOwnershipHandover = exports.checkModulesCompatibility = exports.rolesOf = exports.ownershipHandoverExpiresAt = exports.owner = exports.hasAllRoles = exports.hasAnyRole = exports.isGetSupportedCallbackFunctionsSupported = exports.getSupportedCallbackFunctions = exports.isGetInstalledModulesSupported = exports.getInstalledModules = exports.getOrDeployModule = exports.getDeployedModule = exports.grantMinterRole = exports.deployModularContract = exports.TransferableERC1155 = exports.TransferableERC721 = exports.TransferableERC20 = exports.RoyaltyERC1155 = exports.RoyaltyERC721 = exports.SequentialTokenIdERC1155 = exports.BatchMetadataERC1155 = exports.BatchMetadataERC721 = exports.OpenEditionMetadataERC721 = exports.ClaimableERC1155 = exports.ClaimableERC721 = exports.ClaimableERC20 = exports.MintableERC1155 = exports.MintableERC721 = exports.MintableERC20 = void 0;
// Modules
/**
 * All functions related to the MintableERC20 module.
 * @modules MintableERC20
 */
exports.MintableERC20 = require("../extensions/modules/MintableERC20/index.js");
/**
 * All functions related to the MintableERC721 module.
 * @modules MintableERC721
 */
exports.MintableERC721 = require("../extensions/modules/MintableERC721/index.js");
/**
 * All functions related to the MintableERC1155 module.
 * @modules MintableERC1155
 */
exports.MintableERC1155 = require("../extensions/modules/MintableERC1155/index.js");
/**
 * All functions related to the ClaimableERC20 module.
 * @modules ClaimableERC20
 */
exports.ClaimableERC20 = require("../extensions/modules/ClaimableERC20/index.js");
/**
 * All functions related to the ClaimableERC721 module.
 * @modules ClaimableERC721
 */
exports.ClaimableERC721 = require("../extensions/modules/ClaimableERC721/index.js");
/**
 * All functions related to the ClaimableERC1155 module.
 * @modules ClaimableERC1155
 */
exports.ClaimableERC1155 = require("../extensions/modules/ClaimableERC1155/index.js");
/**
 * All functions related to the OpenEditionMetadataERC721 module.
 * @modules OpenEditionMetadataERC721
 */
exports.OpenEditionMetadataERC721 = require("../extensions/modules/OpenEditionMetadataERC721/index.js");
/**
 * All functions related to the BatchMetadataERC721 module.
 * @modules BatchMetadataERC721
 */
exports.BatchMetadataERC721 = require("../extensions/modules/BatchMetadataERC721/index.js");
/**
 * All functions related to the BatchMetadataERC1155 module.
 * @modules BatchMetadataERC1155
 */
exports.BatchMetadataERC1155 = require("../extensions/modules/BatchMetadataERC1155/index.js");
/**
 * All functions related to the SequentialTokenIdERC1155 module.
 * @modules SequentialTokenIdERC1155
 */
exports.SequentialTokenIdERC1155 = require("../extensions/modules/SequentialTokenIdERC1155/index.js");
/**
 * All functions related to the RoyaltyERC721 module.
 * @modules RoyaltyERC721
 */
exports.RoyaltyERC721 = require("../extensions/modules/RoyaltyERC721/index.js");
/**
 * All functions related to the RoyaltyERC1155 module.
 * @modules RoyaltyERC1155
 */
exports.RoyaltyERC1155 = require("../extensions/modules/RoyaltyERC1155/index.js");
/**
 * All functions related to the TransferableERC20 module.
 * @modules TransferableERC20
 */
exports.TransferableERC20 = require("../extensions/modules/TransferableERC20/index.js");
/**
 * All functions related to the TransferableERC721 module.
 * @modules TransferableERC721
 */
exports.TransferableERC721 = require("../extensions/modules/TransferableERC721/index.js");
/**
 * All functions related to the TransferableERC1155 module.
 * @modules TransferableERC1155
 */
exports.TransferableERC1155 = require("../extensions/modules/TransferableERC1155/index.js");
// deploy
var deploy_modular_js_1 = require("../extensions/prebuilts/deploy-modular.js");
Object.defineProperty(exports, "deployModularContract", { enumerable: true, get: function () { return deploy_modular_js_1.deployModularContract; } });
// Common
var grantMinterRole_js_1 = require("../extensions/modules/common/grantMinterRole.js");
Object.defineProperty(exports, "grantMinterRole", { enumerable: true, get: function () { return grantMinterRole_js_1.grantMinterRole; } });
var getOrDeployModule_js_1 = require("../extensions/modules/common/getOrDeployModule.js");
Object.defineProperty(exports, "getDeployedModule", { enumerable: true, get: function () { return getOrDeployModule_js_1.getDeployedModule; } });
Object.defineProperty(exports, "getOrDeployModule", { enumerable: true, get: function () { return getOrDeployModule_js_1.getOrDeployModule; } });
// generated
/**
 * READ
 */
var getInstalledModules_js_1 = require("../extensions/modules/__generated__/IModularCore/read/getInstalledModules.js");
Object.defineProperty(exports, "getInstalledModules", { enumerable: true, get: function () { return getInstalledModules_js_1.getInstalledModules; } });
Object.defineProperty(exports, "isGetInstalledModulesSupported", { enumerable: true, get: function () { return getInstalledModules_js_1.isGetInstalledModulesSupported; } });
var getSupportedCallbackFunctions_js_1 = require("../extensions/modules/__generated__/IModularCore/read/getSupportedCallbackFunctions.js");
Object.defineProperty(exports, "getSupportedCallbackFunctions", { enumerable: true, get: function () { return getSupportedCallbackFunctions_js_1.getSupportedCallbackFunctions; } });
Object.defineProperty(exports, "isGetSupportedCallbackFunctionsSupported", { enumerable: true, get: function () { return getSupportedCallbackFunctions_js_1.isGetSupportedCallbackFunctionsSupported; } });
var hasAnyRole_js_1 = require("../extensions/modules/__generated__/OwnableRoles/read/hasAnyRole.js");
Object.defineProperty(exports, "hasAnyRole", { enumerable: true, get: function () { return hasAnyRole_js_1.hasAnyRole; } });
var hasAllRoles_js_1 = require("../extensions/modules/__generated__/OwnableRoles/read/hasAllRoles.js");
Object.defineProperty(exports, "hasAllRoles", { enumerable: true, get: function () { return hasAllRoles_js_1.hasAllRoles; } });
var owner_js_1 = require("../extensions/modules/__generated__/OwnableRoles/read/owner.js");
Object.defineProperty(exports, "owner", { enumerable: true, get: function () { return owner_js_1.owner; } });
var ownershipHandoverExpiresAt_js_1 = require("../extensions/modules/__generated__/OwnableRoles/read/ownershipHandoverExpiresAt.js");
Object.defineProperty(exports, "ownershipHandoverExpiresAt", { enumerable: true, get: function () { return ownershipHandoverExpiresAt_js_1.ownershipHandoverExpiresAt; } });
var rolesOf_js_1 = require("../extensions/modules/__generated__/OwnableRoles/read/rolesOf.js");
Object.defineProperty(exports, "rolesOf", { enumerable: true, get: function () { return rolesOf_js_1.rolesOf; } });
var checkModulesCompatibility_js_1 = require("../extensions/modules/common/checkModulesCompatibility.js");
Object.defineProperty(exports, "checkModulesCompatibility", { enumerable: true, get: function () { return checkModulesCompatibility_js_1.checkModulesCompatibility; } });
/**
 * Write
 */
var cancelOwnershipHandover_js_1 = require("../extensions/modules/__generated__/OwnableRoles/write/cancelOwnershipHandover.js");
Object.defineProperty(exports, "cancelOwnershipHandover", { enumerable: true, get: function () { return cancelOwnershipHandover_js_1.cancelOwnershipHandover; } });
var completeOwnershipHandover_js_1 = require("../extensions/modules/__generated__/OwnableRoles/write/completeOwnershipHandover.js");
Object.defineProperty(exports, "completeOwnershipHandover", { enumerable: true, get: function () { return completeOwnershipHandover_js_1.completeOwnershipHandover; } });
var grantRoles_js_1 = require("../extensions/modules/__generated__/OwnableRoles/write/grantRoles.js");
Object.defineProperty(exports, "grantRoles", { enumerable: true, get: function () { return grantRoles_js_1.grantRoles; } });
var installModule_js_1 = require("../extensions/modules/__generated__/IModularCore/write/installModule.js");
Object.defineProperty(exports, "installModule", { enumerable: true, get: function () { return installModule_js_1.installModule; } });
Object.defineProperty(exports, "isInstallModuleSupported", { enumerable: true, get: function () { return installModule_js_1.isInstallModuleSupported; } });
var renounceOwnership_js_1 = require("../extensions/modules/__generated__/OwnableRoles/write/renounceOwnership.js");
Object.defineProperty(exports, "renounceOwnership", { enumerable: true, get: function () { return renounceOwnership_js_1.renounceOwnership; } });
var requestOwnershipHandover_js_1 = require("../extensions/modules/__generated__/OwnableRoles/write/requestOwnershipHandover.js");
Object.defineProperty(exports, "requestOwnershipHandover", { enumerable: true, get: function () { return requestOwnershipHandover_js_1.requestOwnershipHandover; } });
var renounceRoles_js_1 = require("../extensions/modules/__generated__/OwnableRoles/write/renounceRoles.js");
Object.defineProperty(exports, "renounceRoles", { enumerable: true, get: function () { return renounceRoles_js_1.renounceRoles; } });
var revokeRoles_js_1 = require("../extensions/modules/__generated__/OwnableRoles/write/revokeRoles.js");
Object.defineProperty(exports, "revokeRoles", { enumerable: true, get: function () { return revokeRoles_js_1.revokeRoles; } });
var transferOwnership_js_1 = require("../extensions/modules/__generated__/OwnableRoles/write/transferOwnership.js");
Object.defineProperty(exports, "transferOwnership", { enumerable: true, get: function () { return transferOwnership_js_1.transferOwnership; } });
var uninstallModule_js_1 = require("../extensions/modules/__generated__/IModularCore/write/uninstallModule.js");
Object.defineProperty(exports, "uninstallModule", { enumerable: true, get: function () { return uninstallModule_js_1.uninstallModule; } });
Object.defineProperty(exports, "isUninstallModuleSupported", { enumerable: true, get: function () { return uninstallModule_js_1.isUninstallModuleSupported; } });
var getModuleConfig_js_1 = require("../extensions/modules/__generated__/IModule/read/getModuleConfig.js");
Object.defineProperty(exports, "getModuleConfig", { enumerable: true, get: function () { return getModuleConfig_js_1.getModuleConfig; } });
Object.defineProperty(exports, "isGetModuleConfigSupported", { enumerable: true, get: function () { return getModuleConfig_js_1.isGetModuleConfigSupported; } });
var installPublishedModule_js_1 = require("../extensions/modules/common/installPublishedModule.js");
Object.defineProperty(exports, "installPublishedModule", { enumerable: true, get: function () { return installPublishedModule_js_1.installPublishedModule; } });
var uninstallPublishedModule_js_1 = require("../extensions/modules/common/uninstallPublishedModule.js");
Object.defineProperty(exports, "uninstallPublishedModule", { enumerable: true, get: function () { return uninstallPublishedModule_js_1.uninstallPublishedModule; } });
var uninstallModuleByProxy_js_1 = require("../extensions/modules/common/uninstallModuleByProxy.js");
Object.defineProperty(exports, "uninstallModuleByProxy", { enumerable: true, get: function () { return uninstallModuleByProxy_js_1.uninstallModuleByProxy; } });
