// thirdweb/dist/cjs/exports/thirdweb.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fromGwei = exports.toWei = exports.toUnits = exports.toTokens = exports.toEther = exports.watchContractEvents = exports.parseEventLogs = exports.getContractEvents = exports.prepareEvent = exports.toSerializableTransaction = exports.serializeTransaction = exports.signTransaction = exports.waitForReceipt = exports.simulateTransaction = exports.sendBatchTransaction = exports.sendAndConfirmTransaction = exports.sendTransaction = exports.getGasPrice = exports.estimateGasCost = exports.estimateGas = exports.encode = exports.resolveMethod = exports.readContract = exports.prepareContractCall = exports.isBaseTransactionOptions = exports.prepareTransaction = exports.getUser = exports.getContract = exports.eth_sendRawTransaction = exports.eth_maxPriorityFeePerGas = exports.eth_getTransactionReceipt = exports.eth_getTransactionCount = exports.eth_getTransactionByHash = exports.eth_getStorageAt = exports.eth_getLogs = exports.eth_getCode = exports.eth_getBlockByNumber = exports.eth_getBlockByHash = exports.eth_getBalance = exports.eth_gasPrice = exports.eth_estimateGas = exports.eth_call = exports.eth_blockNumber = exports.watchBlockNumber = exports.getRpcClient = exports.defineChain = exports.createThirdwebClient = exports.NATIVE_TOKEN_ADDRESS = exports.ZERO_ADDRESS = exports.ADDRESS_ZERO = void 0;
exports.verifyTypedData = exports.isAddress = exports.getAddress = exports.sha256 = exports.keccak256 = exports.fromBytes = exports.bytesToString = exports.bytesToNumber = exports.bytesToBool = exports.bytesToBigInt = exports.toBytes = exports.stringToBytes = exports.numberToBytes = exports.hexToBytes = exports.boolToBytes = exports.concatHex = exports.uint8ArrayToHex = exports.toHex = exports.stringToHex = exports.padHex = exports.numberToHex = exports.isHex = exports.hexToUint8Array = exports.hexToString = exports.hexToNumber = exports.hexToBool = exports.hexToBigInt = exports.fromHex = exports.boolToHex = exports.getBuyWithCryptoTransfer = exports.getBuyWithCryptoHistory = exports.getBuyWithCryptoStatus = exports.getBuyWithCryptoQuote = void 0;
/**
 * CONSTANTS
 */
var addresses_js_1 = require("../constants/addresses.js");
/**
 * @deprecated Use {@link ZERO_ADDRESS}.
 */
Object.defineProperty(exports, "ADDRESS_ZERO", { enumerable: true, get: function () { return addresses_js_1.ZERO_ADDRESS; } });
Object.defineProperty(exports, "ZERO_ADDRESS", { enumerable: true, get: function () { return addresses_js_1.ZERO_ADDRESS; } });
Object.defineProperty(exports, "NATIVE_TOKEN_ADDRESS", { enumerable: true, get: function () { return addresses_js_1.NATIVE_TOKEN_ADDRESS; } });
/**
 * CLIENT
 */
var client_js_1 = require("../client/client.js");
Object.defineProperty(exports, "createThirdwebClient", { enumerable: true, get: function () { return client_js_1.createThirdwebClient; } });
// define chain
var utils_js_1 = require("../chains/utils.js");
Object.defineProperty(exports, "defineChain", { enumerable: true, get: function () { return utils_js_1.defineChain; } });
/**
 * RPC
 */
// rpc client
var rpc_js_1 = require("../rpc/rpc.js");
Object.defineProperty(exports, "getRpcClient", { enumerable: true, get: function () { return rpc_js_1.getRpcClient; } });
// blockNumber watcher
var watchBlockNumber_js_1 = require("../rpc/watchBlockNumber.js");
Object.defineProperty(exports, "watchBlockNumber", { enumerable: true, get: function () { return watchBlockNumber_js_1.watchBlockNumber; } });
// all the actions
var eth_blockNumber_js_1 = require("../rpc/actions/eth_blockNumber.js");
Object.defineProperty(exports, "eth_blockNumber", { enumerable: true, get: function () { return eth_blockNumber_js_1.eth_blockNumber; } });
var eth_call_js_1 = require("../rpc/actions/eth_call.js");
Object.defineProperty(exports, "eth_call", { enumerable: true, get: function () { return eth_call_js_1.eth_call; } });
var eth_estimateGas_js_1 = require("../rpc/actions/eth_estimateGas.js");
Object.defineProperty(exports, "eth_estimateGas", { enumerable: true, get: function () { return eth_estimateGas_js_1.eth_estimateGas; } });
var eth_gasPrice_js_1 = require("../rpc/actions/eth_gasPrice.js");
Object.defineProperty(exports, "eth_gasPrice", { enumerable: true, get: function () { return eth_gasPrice_js_1.eth_gasPrice; } });
var eth_getBalance_js_1 = require("../rpc/actions/eth_getBalance.js");
Object.defineProperty(exports, "eth_getBalance", { enumerable: true, get: function () { return eth_getBalance_js_1.eth_getBalance; } });
var eth_getBlockByHash_js_1 = require("../rpc/actions/eth_getBlockByHash.js");
Object.defineProperty(exports, "eth_getBlockByHash", { enumerable: true, get: function () { return eth_getBlockByHash_js_1.eth_getBlockByHash; } });
var eth_getBlockByNumber_js_1 = require("../rpc/actions/eth_getBlockByNumber.js");
Object.defineProperty(exports, "eth_getBlockByNumber", { enumerable: true, get: function () { return eth_getBlockByNumber_js_1.eth_getBlockByNumber; } });
var eth_getCode_js_1 = require("../rpc/actions/eth_getCode.js");
Object.defineProperty(exports, "eth_getCode", { enumerable: true, get: function () { return eth_getCode_js_1.eth_getCode; } });
var eth_getLogs_js_1 = require("../rpc/actions/eth_getLogs.js");
Object.defineProperty(exports, "eth_getLogs", { enumerable: true, get: function () { return eth_getLogs_js_1.eth_getLogs; } });
var eth_getStorageAt_js_1 = require("../rpc/actions/eth_getStorageAt.js");
Object.defineProperty(exports, "eth_getStorageAt", { enumerable: true, get: function () { return eth_getStorageAt_js_1.eth_getStorageAt; } });
var eth_getTransactionByHash_js_1 = require("../rpc/actions/eth_getTransactionByHash.js");
Object.defineProperty(exports, "eth_getTransactionByHash", { enumerable: true, get: function () { return eth_getTransactionByHash_js_1.eth_getTransactionByHash; } });
var eth_getTransactionCount_js_1 = require("../rpc/actions/eth_getTransactionCount.js");
Object.defineProperty(exports, "eth_getTransactionCount", { enumerable: true, get: function () { return eth_getTransactionCount_js_1.eth_getTransactionCount; } });
var eth_getTransactionReceipt_js_1 = require("../rpc/actions/eth_getTransactionReceipt.js");
Object.defineProperty(exports, "eth_getTransactionReceipt", { enumerable: true, get: function () { return eth_getTransactionReceipt_js_1.eth_getTransactionReceipt; } });
var eth_maxPriorityFeePerGas_js_1 = require("../rpc/actions/eth_maxPriorityFeePerGas.js");
Object.defineProperty(exports, "eth_maxPriorityFeePerGas", { enumerable: true, get: function () { return eth_maxPriorityFeePerGas_js_1.eth_maxPriorityFeePerGas; } });
var eth_sendRawTransaction_js_1 = require("../rpc/actions/eth_sendRawTransaction.js");
Object.defineProperty(exports, "eth_sendRawTransaction", { enumerable: true, get: function () { return eth_sendRawTransaction_js_1.eth_sendRawTransaction; } });
/**
 * CONTRACT
 */
var contract_js_1 = require("../contract/contract.js");
Object.defineProperty(exports, "getContract", { enumerable: true, get: function () { return contract_js_1.getContract; } });
/**
 * WALLETS
 */
var getUser_js_1 = require("../wallets/in-app/core/users/getUser.js");
Object.defineProperty(exports, "getUser", { enumerable: true, get: function () { return getUser_js_1.getUser; } });
/**
 * TRANSACTIONS
 */
var prepare_transaction_js_1 = require("../transaction/prepare-transaction.js");
Object.defineProperty(exports, "prepareTransaction", { enumerable: true, get: function () { return prepare_transaction_js_1.prepareTransaction; } });
var types_js_1 = require("../transaction/types.js");
Object.defineProperty(exports, "isBaseTransactionOptions", { enumerable: true, get: function () { return types_js_1.isBaseTransactionOptions; } });
var prepare_contract_call_js_1 = require("../transaction/prepare-contract-call.js");
Object.defineProperty(exports, "prepareContractCall", { enumerable: true, get: function () { return prepare_contract_call_js_1.prepareContractCall; } });
var read_contract_js_1 = require("../transaction/read-contract.js");
Object.defineProperty(exports, "readContract", { enumerable: true, get: function () { return read_contract_js_1.readContract; } });
// method resolver
var resolve_method_js_1 = require("../transaction/resolve-method.js");
Object.defineProperty(exports, "resolveMethod", { enumerable: true, get: function () { return resolve_method_js_1.resolveMethod; } });
// transaction actions
var encode_js_1 = require("../transaction/actions/encode.js");
Object.defineProperty(exports, "encode", { enumerable: true, get: function () { return encode_js_1.encode; } });
var estimate_gas_js_1 = require("../transaction/actions/estimate-gas.js");
Object.defineProperty(exports, "estimateGas", { enumerable: true, get: function () { return estimate_gas_js_1.estimateGas; } });
var estimate_gas_cost_js_1 = require("../transaction/actions/estimate-gas-cost.js");
Object.defineProperty(exports, "estimateGasCost", { enumerable: true, get: function () { return estimate_gas_cost_js_1.estimateGasCost; } });
var get_gas_price_js_1 = require("../gas/get-gas-price.js");
Object.defineProperty(exports, "getGasPrice", { enumerable: true, get: function () { return get_gas_price_js_1.getGasPrice; } });
var send_transaction_js_1 = require("../transaction/actions/send-transaction.js");
Object.defineProperty(exports, "sendTransaction", { enumerable: true, get: function () { return send_transaction_js_1.sendTransaction; } });
var send_and_confirm_transaction_js_1 = require("../transaction/actions/send-and-confirm-transaction.js");
Object.defineProperty(exports, "sendAndConfirmTransaction", { enumerable: true, get: function () { return send_and_confirm_transaction_js_1.sendAndConfirmTransaction; } });
var send_batch_transaction_js_1 = require("../transaction/actions/send-batch-transaction.js");
Object.defineProperty(exports, "sendBatchTransaction", { enumerable: true, get: function () { return send_batch_transaction_js_1.sendBatchTransaction; } });
var simulate_js_1 = require("../transaction/actions/simulate.js");
Object.defineProperty(exports, "simulateTransaction", { enumerable: true, get: function () { return simulate_js_1.simulateTransaction; } });
var wait_for_tx_receipt_js_1 = require("../transaction/actions/wait-for-tx-receipt.js");
Object.defineProperty(exports, "waitForReceipt", { enumerable: true, get: function () { return wait_for_tx_receipt_js_1.waitForReceipt; } });
var sign_transaction_js_1 = require("../transaction/actions/sign-transaction.js");
Object.defineProperty(exports, "signTransaction", { enumerable: true, get: function () { return sign_transaction_js_1.signTransaction; } });
var serialize_transaction_js_1 = require("../transaction/serialize-transaction.js");
Object.defineProperty(exports, "serializeTransaction", { enumerable: true, get: function () { return serialize_transaction_js_1.serializeTransaction; } });
var to_serializable_transaction_js_1 = require("../transaction/actions/to-serializable-transaction.js");
Object.defineProperty(exports, "toSerializableTransaction", { enumerable: true, get: function () { return to_serializable_transaction_js_1.toSerializableTransaction; } });
/**
 * EVENTS
 */
var prepare_event_js_1 = require("../event/prepare-event.js");
Object.defineProperty(exports, "prepareEvent", { enumerable: true, get: function () { return prepare_event_js_1.prepareEvent; } });
// actions
var get_events_js_1 = require("../event/actions/get-events.js");
Object.defineProperty(exports, "getContractEvents", { enumerable: true, get: function () { return get_events_js_1.getContractEvents; } });
var parse_logs_js_1 = require("../event/actions/parse-logs.js");
Object.defineProperty(exports, "parseEventLogs", { enumerable: true, get: function () { return parse_logs_js_1.parseEventLogs; } });
var watch_events_js_1 = require("../event/actions/watch-events.js");
Object.defineProperty(exports, "watchContractEvents", { enumerable: true, get: function () { return watch_events_js_1.watchContractEvents; } });
/**
 * UNITS
 */
var units_js_1 = require("../utils/units.js");
Object.defineProperty(exports, "toEther", { enumerable: true, get: function () { return units_js_1.toEther; } });
Object.defineProperty(exports, "toTokens", { enumerable: true, get: function () { return units_js_1.toTokens; } });
Object.defineProperty(exports, "toUnits", { enumerable: true, get: function () { return units_js_1.toUnits; } });
Object.defineProperty(exports, "toWei", { enumerable: true, get: function () { return units_js_1.toWei; } });
Object.defineProperty(exports, "fromGwei", { enumerable: true, get: function () { return units_js_1.fromGwei; } });
var getQuote_js_1 = require("../pay/buyWithCrypto/getQuote.js");
Object.defineProperty(exports, "getBuyWithCryptoQuote", { enumerable: true, get: function () { return getQuote_js_1.getBuyWithCryptoQuote; } });
var getStatus_js_1 = require("../pay/buyWithCrypto/getStatus.js");
Object.defineProperty(exports, "getBuyWithCryptoStatus", { enumerable: true, get: function () { return getStatus_js_1.getBuyWithCryptoStatus; } });
var getHistory_js_1 = require("../pay/buyWithCrypto/getHistory.js");
Object.defineProperty(exports, "getBuyWithCryptoHistory", { enumerable: true, get: function () { return getHistory_js_1.getBuyWithCryptoHistory; } });
var getTransfer_js_1 = require("../pay/buyWithCrypto/getTransfer.js");
Object.defineProperty(exports, "getBuyWithCryptoTransfer", { enumerable: true, get: function () { return getTransfer_js_1.getBuyWithCryptoTransfer; } });
// ------------------------------------------------
// encoding
// ------------------------------------------------
// hex
var hex_js_1 = require("../utils/encoding/hex.js");
Object.defineProperty(exports, "boolToHex", { enumerable: true, get: function () { return hex_js_1.boolToHex; } });
// from
Object.defineProperty(exports, "fromHex", { enumerable: true, get: function () { return hex_js_1.fromHex; } });
Object.defineProperty(exports, "hexToBigInt", { enumerable: true, get: function () { return hex_js_1.hexToBigInt; } });
Object.defineProperty(exports, "hexToBool", { enumerable: true, get: function () { return hex_js_1.hexToBool; } });
Object.defineProperty(exports, "hexToNumber", { enumerable: true, get: function () { return hex_js_1.hexToNumber; } });
Object.defineProperty(exports, "hexToString", { enumerable: true, get: function () { return hex_js_1.hexToString; } });
Object.defineProperty(exports, "hexToUint8Array", { enumerable: true, get: function () { return hex_js_1.hexToUint8Array; } });
// util
Object.defineProperty(exports, "isHex", { enumerable: true, get: function () { return hex_js_1.isHex; } });
Object.defineProperty(exports, "numberToHex", { enumerable: true, get: function () { return hex_js_1.numberToHex; } });
Object.defineProperty(exports, "padHex", { enumerable: true, get: function () { return hex_js_1.padHex; } });
Object.defineProperty(exports, "stringToHex", { enumerable: true, get: function () { return hex_js_1.stringToHex; } });
// to
Object.defineProperty(exports, "toHex", { enumerable: true, get: function () { return hex_js_1.toHex; } });
Object.defineProperty(exports, "uint8ArrayToHex", { enumerable: true, get: function () { return hex_js_1.uint8ArrayToHex; } });
var concat_hex_js_1 = require("../utils/encoding/helpers/concat-hex.js");
Object.defineProperty(exports, "concatHex", { enumerable: true, get: function () { return concat_hex_js_1.concatHex; } });
// bytes
// to
var to_bytes_js_1 = require("../utils/encoding/to-bytes.js");
Object.defineProperty(exports, "boolToBytes", { enumerable: true, get: function () { return to_bytes_js_1.boolToBytes; } });
Object.defineProperty(exports, "hexToBytes", { enumerable: true, get: function () { return to_bytes_js_1.hexToBytes; } });
Object.defineProperty(exports, "numberToBytes", { enumerable: true, get: function () { return to_bytes_js_1.numberToBytes; } });
Object.defineProperty(exports, "stringToBytes", { enumerable: true, get: function () { return to_bytes_js_1.stringToBytes; } });
Object.defineProperty(exports, "toBytes", { enumerable: true, get: function () { return to_bytes_js_1.toBytes; } });
// from
var from_bytes_js_1 = require("../utils/encoding/from-bytes.js");
Object.defineProperty(exports, "bytesToBigInt", { enumerable: true, get: function () { return from_bytes_js_1.bytesToBigInt; } });
Object.defineProperty(exports, "bytesToBool", { enumerable: true, get: function () { return from_bytes_js_1.bytesToBool; } });
Object.defineProperty(exports, "bytesToNumber", { enumerable: true, get: function () { return from_bytes_js_1.bytesToNumber; } });
Object.defineProperty(exports, "bytesToString", { enumerable: true, get: function () { return from_bytes_js_1.bytesToString; } });
Object.defineProperty(exports, "fromBytes", { enumerable: true, get: function () { return from_bytes_js_1.fromBytes; } });
// ------------------------------------------------
// hashing
// ------------------------------------------------
// keccak256
var keccak256_js_1 = require("../utils/hashing/keccak256.js");
Object.defineProperty(exports, "keccak256", { enumerable: true, get: function () { return keccak256_js_1.keccak256; } });
// sha256
var sha256_js_1 = require("../utils/hashing/sha256.js");
Object.defineProperty(exports, "sha256", { enumerable: true, get: function () { return sha256_js_1.sha256; } });
// ------------------------------------------------
// address
// ------------------------------------------------
var address_js_1 = require("../utils/address.js");
Object.defineProperty(exports, "getAddress", { enumerable: true, get: function () { return address_js_1.getAddress; } });
Object.defineProperty(exports, "isAddress", { enumerable: true, get: function () { return address_js_1.isAddress; } });
var verify_typed_data_js_1 = require("../auth/verify-typed-data.js");
Object.defineProperty(exports, "verifyTypedData", { enumerable: true, get: function () { return verify_typed_data_js_1.verifyTypedData; } });
