// thirdweb/dist/cjs/exports/wallets.js
"use strict";
// --- KEEEP IN SYNC with exports/wallets.native.ts ---
Object.defineProperty(exports, "__esModule", { value: true });
exports.injectedProvider = exports.DefaultWalletConnectRequestHandlers = exports.getActiveWalletConnectSessions = exports.disconnectWalletConnectSession = exports.createWalletConnectSession = exports.createWalletConnectClient = exports.createWalletAdapter = exports.getWalletInfo = exports.getAllWalletsList = exports.linkProfile = exports.getProfiles = exports.privateKeyAccount = exports.privateKeyToAccount = exports.generateAccount = exports.getWalletBalance = exports.smartWallet = exports.ecosystemWallet = exports.embeddedWallet = exports.inAppWallet = exports.walletConnect = exports.createWallet = void 0;
var create_wallet_js_1 = require("../wallets/create-wallet.js");
Object.defineProperty(exports, "createWallet", { enumerable: true, get: function () { return create_wallet_js_1.createWallet; } });
Object.defineProperty(exports, "walletConnect", { enumerable: true, get: function () { return create_wallet_js_1.walletConnect; } });
var in_app_js_1 = require("../wallets/in-app/web/in-app.js");
Object.defineProperty(exports, "inAppWallet", { enumerable: true, get: function () { return in_app_js_1.inAppWallet; } });
/**
 * @deprecated use inAppWallet instead
 */
Object.defineProperty(exports, "embeddedWallet", { enumerable: true, get: function () { return in_app_js_1.inAppWallet; } });
var ecosystem_js_1 = require("../wallets/in-app/web/ecosystem.js");
Object.defineProperty(exports, "ecosystemWallet", { enumerable: true, get: function () { return ecosystem_js_1.ecosystemWallet; } });
var smart_wallet_js_1 = require("../wallets/smart/smart-wallet.js");
Object.defineProperty(exports, "smartWallet", { enumerable: true, get: function () { return smart_wallet_js_1.smartWallet; } });
// utils
var getWalletBalance_js_1 = require("../wallets/utils/getWalletBalance.js");
Object.defineProperty(exports, "getWalletBalance", { enumerable: true, get: function () { return getWalletBalance_js_1.getWalletBalance; } });
var generateAccount_js_1 = require("../wallets/utils/generateAccount.js");
Object.defineProperty(exports, "generateAccount", { enumerable: true, get: function () { return generateAccount_js_1.generateAccount; } });
// private-key
var private_key_js_1 = require("../wallets/private-key.js");
Object.defineProperty(exports, "privateKeyToAccount", { enumerable: true, get: function () { return private_key_js_1.privateKeyToAccount; } });
/**
 * @internal
 * @deprecated - use {@link privateKeyToAccount} instead
 */
Object.defineProperty(exports, "privateKeyAccount", { enumerable: true, get: function () { return private_key_js_1.privateKeyToAccount; } });
var profiles_js_1 = require("../wallets/in-app/core/wallet/profiles.js");
Object.defineProperty(exports, "getProfiles", { enumerable: true, get: function () { return profiles_js_1.getProfiles; } });
Object.defineProperty(exports, "linkProfile", { enumerable: true, get: function () { return profiles_js_1.linkProfile; } });
var getAllWalletsList_js_1 = require("../wallets/getAllWalletsList.js");
Object.defineProperty(exports, "getAllWalletsList", { enumerable: true, get: function () { return getAllWalletsList_js_1.getAllWalletsList; } });
var getWalletInfo_js_1 = require("../wallets/__generated__/getWalletInfo.js");
Object.defineProperty(exports, "getWalletInfo", { enumerable: true, get: function () { return getWalletInfo_js_1.getWalletInfo; } });
var wallet_adapter_js_1 = require("../adapters/wallet-adapter.js");
Object.defineProperty(exports, "createWalletAdapter", { enumerable: true, get: function () { return wallet_adapter_js_1.createWalletAdapter; } });
// wallet connect
var index_js_1 = require("../wallets/wallet-connect/receiver/index.js");
Object.defineProperty(exports, "createWalletConnectClient", { enumerable: true, get: function () { return index_js_1.createWalletConnectClient; } });
Object.defineProperty(exports, "createWalletConnectSession", { enumerable: true, get: function () { return index_js_1.createWalletConnectSession; } });
Object.defineProperty(exports, "disconnectWalletConnectSession", { enumerable: true, get: function () { return index_js_1.disconnectWalletConnectSession; } });
Object.defineProperty(exports, "getActiveWalletConnectSessions", { enumerable: true, get: function () { return index_js_1.getActiveWalletConnectSessions; } });
Object.defineProperty(exports, "DefaultWalletConnectRequestHandlers", { enumerable: true, get: function () { return index_js_1.DefaultWalletConnectRequestHandlers; } });
// WEB ONLY EXPORTS
// injected
var mipdStore_js_1 = require("../wallets/injected/mipdStore.js");
Object.defineProperty(exports, "injectedProvider", { enumerable: true, get: function () { return mipdStore_js_1.injectedProvider; } });
