// thirdweb/dist/cjs/extensions/modules/__generated__/ClaimableERC20/read/getClaimCondition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isGetClaimConditionSupported = isGetClaimConditionSupported;
exports.decodeGetClaimConditionResult = decodeGetClaimConditionResult;
exports.getClaimCondition = getClaimCondition;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x30a63e11";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "tuple",
        name: "claimCondition",
        components: [
            {
                type: "uint256",
                name: "availableSupply",
            },
            {
                type: "bytes32",
                name: "allowlistMerkleRoot",
            },
            {
                type: "uint256",
                name: "pricePerUnit",
            },
            {
                type: "address",
                name: "currency",
            },
            {
                type: "uint256",
                name: "maxMintPerWallet",
            },
            {
                type: "uint48",
                name: "startTimestamp",
            },
            {
                type: "uint48",
                name: "endTimestamp",
            },
            {
                type: "string",
                name: "auxData",
            },
        ],
    },
];
/**
 * Checks if the `getClaimCondition` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `getClaimCondition` method is supported.
 * @modules ClaimableERC20
 * @example
 * ```ts
 * import { ClaimableERC20 } from "thirdweb/modules";
 * const supported = ClaimableERC20.isGetClaimConditionSupported(["0x..."]);
 * ```
 */
function isGetClaimConditionSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the getClaimCondition function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @modules ClaimableERC20
 * @example
 * ```ts
 * import { ClaimableERC20 } from "thirdweb/modules";
 * const result = ClaimableERC20.decodeGetClaimConditionResultResult("...");
 * ```
 */
function decodeGetClaimConditionResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getClaimCondition" function on the contract.
 * @param options - The options for the getClaimCondition function.
 * @returns The parsed result of the function call.
 * @modules ClaimableERC20
 * @example
 * ```ts
 * import { ClaimableERC20 } from "thirdweb/modules";
 *
 * const result = await ClaimableERC20.getClaimCondition({
 *  contract,
 * });
 *
 * ```
 */
async function getClaimCondition(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
