// thirdweb/dist/cjs/extensions/lens/__generated__/FollowNFT/read/getFollowTokenId.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isGetFollowTokenIdSupported = isGetFollowTokenIdSupported;
exports.encodeGetFollowTokenIdParams = encodeGetFollowTokenIdParams;
exports.encodeGetFollowTokenId = encodeGetFollowTokenId;
exports.decodeGetFollowTokenIdResult = decodeGetFollowTokenIdResult;
exports.getFollowTokenId = getFollowTokenId;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x11c763d6";
const FN_INPUTS = [
    {
        type: "uint256",
        name: "followerProfileId",
    },
];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `getFollowTokenId` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `getFollowTokenId` method is supported.
 * @extension LENS
 * @example
 * ```ts
 * import { isGetFollowTokenIdSupported } from "thirdweb/extensions/lens";
 * const supported = isGetFollowTokenIdSupported(["0x..."]);
 * ```
 */
function isGetFollowTokenIdSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "getFollowTokenId" function.
 * @param options - The options for the getFollowTokenId function.
 * @returns The encoded ABI parameters.
 * @extension LENS
 * @example
 * ```ts
 * import { encodeGetFollowTokenIdParams } from "thirdweb/extensions/lens";
 * const result = encodeGetFollowTokenIdParams({
 *  followerProfileId: ...,
 * });
 * ```
 */
function encodeGetFollowTokenIdParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.followerProfileId]);
}
/**
 * Encodes the "getFollowTokenId" function into a Hex string with its parameters.
 * @param options - The options for the getFollowTokenId function.
 * @returns The encoded hexadecimal string.
 * @extension LENS
 * @example
 * ```ts
 * import { encodeGetFollowTokenId } from "thirdweb/extensions/lens";
 * const result = encodeGetFollowTokenId({
 *  followerProfileId: ...,
 * });
 * ```
 */
function encodeGetFollowTokenId(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeGetFollowTokenIdParams(options).slice(2));
}
/**
 * Decodes the result of the getFollowTokenId function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension LENS
 * @example
 * ```ts
 * import { decodeGetFollowTokenIdResult } from "thirdweb/extensions/lens";
 * const result = decodeGetFollowTokenIdResultResult("...");
 * ```
 */
function decodeGetFollowTokenIdResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getFollowTokenId" function on the contract.
 * @param options - The options for the getFollowTokenId function.
 * @returns The parsed result of the function call.
 * @extension LENS
 * @example
 * ```ts
 * import { getFollowTokenId } from "thirdweb/extensions/lens";
 *
 * const result = await getFollowTokenId({
 *  contract,
 *  followerProfileId: ...,
 * });
 *
 * ```
 */
async function getFollowTokenId(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.followerProfileId],
    });
}
