// thirdweb/dist/cjs/extensions/lens/__generated__/FollowNFT/read/getFollowerCount.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isGetFollowerCountSupported = isGetFollowerCountSupported;
exports.decodeGetFollowerCountResult = decodeGetFollowerCountResult;
exports.getFollowerCount = getFollowerCount;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x7829ae4a";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `getFollowerCount` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `getFollowerCount` method is supported.
 * @extension LENS
 * @example
 * ```ts
 * import { isGetFollowerCountSupported } from "thirdweb/extensions/lens";
 * const supported = isGetFollowerCountSupported(["0x..."]);
 * ```
 */
function isGetFollowerCountSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the getFollowerCount function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension LENS
 * @example
 * ```ts
 * import { decodeGetFollowerCountResult } from "thirdweb/extensions/lens";
 * const result = decodeGetFollowerCountResultResult("...");
 * ```
 */
function decodeGetFollowerCountResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getFollowerCount" function on the contract.
 * @param options - The options for the getFollowerCount function.
 * @returns The parsed result of the function call.
 * @extension LENS
 * @example
 * ```ts
 * import { getFollowerCount } from "thirdweb/extensions/lens";
 *
 * const result = await getFollowerCount({
 *  contract,
 * });
 *
 * ```
 */
async function getFollowerCount(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
