// thirdweb/dist/cjs/extensions/modules/MintableERC721/mintWithRole.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mintWithRole = mintWithRole;
const address_js_1 = require("../../../utils/address.js");
const ipfs_js_1 = require("../../../utils/ipfs.js");
const mint_js_1 = require("../__generated__/ERC721Core/write/mint.js");
/**
 * Mints ERC721 tokens to a specified address via a MintableERC721 module.
 * @param options The options for minting tokens.
 * @returns A transaction to mint tokens.
 * @example
 * ```typescript
 * import { MintableERC721 } from "thirdweb/modules";
 *
 * const transaction = MintableERC721.mintWithRole({
 *   contract,
 *   to: "0x...", // Address to mint tokens to
 *   nfts: [
 *     {
 *       name: "My NFT",
 *       description: "This is my NFT",
 *       image: "ipfs://...",
 *     },
 *   ],
 * });
 *
 * // Send the transaction
 * await sendTransaction({ transaction, account });
 * ```
 * @modules MintableERC721
 */
function mintWithRole(options) {
    return (0, mint_js_1.mint)({
        contract: options.contract,
        asyncParams: async () => {
            const batchOfUris = await (0, ipfs_js_1.uploadOrExtractURIs)(options.nfts, options.contract.client);
            const baseURI = (0, ipfs_js_1.getBaseUriFromBatch)(batchOfUris);
            return {
                to: (0, address_js_1.getAddress)(options.to),
                amount: BigInt(options.nfts.length),
                baseURI,
                data: "0x",
            };
        },
    });
}
