// thirdweb/dist/esm/extensions/erc1155/__generated__/Zora1155/read/nextTokenId.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0x75794a3c";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `nextTokenId` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `nextTokenId` method is supported.
 * @extension ERC1155
 * @example
 * ```ts
 * import { isNextTokenIdSupported } from "thirdweb/extensions/erc1155";
 * const supported = isNextTokenIdSupported(["0x..."]);
 * ```
 */
export function isNextTokenIdSupported(availableSelectors) {
    return detectMethod({
        availableSelectors,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the nextTokenId function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC1155
 * @example
 * ```ts
 * import { decodeNextTokenIdResult } from "thirdweb/extensions/erc1155";
 * const result = decodeNextTokenIdResultResult("...");
 * ```
 */
export function decodeNextTokenIdResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "nextTokenId" function on the contract.
 * @param options - The options for the nextTokenId function.
 * @returns The parsed result of the function call.
 * @extension ERC1155
 * @example
 * ```ts
 * import { nextTokenId } from "thirdweb/extensions/erc1155";
 *
 * const result = await nextTokenId({
 *  contract,
 * });
 *
 * ```
 */
export async function nextTokenId(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
