// thirdweb/dist/esm/wallets/in-app/web/lib/auth/passkeys.js
import { client } from "@passwordless-id/webauthn";
import { webLocalStorage } from "../../../../../utils/storage/webStorage.js";
import { base64ToString, base64UrlToBase64, } from "../../../../../utils/uint8-array.js";
import { ClientScopedStorage } from "../../../core/authentication/client-scoped-storage.js";
export class PasskeyWebClient {
    isAvailable() {
        return client.isAvailable();
    }
    async register(args) {
        const { name, challenge, rp } = args;
        const registration = await client.register(name, challenge, {
            authenticatorType: "auto",
            userVerification: "required",
            domain: rp.id,
            attestation: true,
            debug: false,
        });
        const clientDataB64 = base64UrlToBase64(registration.clientData);
        const clientDataParsed = JSON.parse(base64ToString(clientDataB64));
        return {
            authenticatorData: registration.authenticatorData,
            credentialId: registration.credential.id,
            clientData: registration.clientData,
            credential: {
                publicKey: registration.credential.publicKey,
                algorithm: registration.credential.algorithm,
            },
            origin: clientDataParsed.origin,
        };
    }
    async authenticate(args) {
        const { credentialId, challenge, rp } = args;
        const result = await client.authenticate(credentialId ? [credentialId] : [], challenge, {
            authenticatorType: "auto",
            userVerification: "required",
            domain: rp.id,
        });
        const clientDataB64 = base64UrlToBase64(result.clientData);
        const clientDataParsed = JSON.parse(base64ToString(clientDataB64));
        return {
            authenticatorData: result.authenticatorData,
            credentialId: result.credentialId,
            clientData: result.clientData,
            signature: result.signature,
            origin: clientDataParsed.origin,
        };
    }
}
/**
 * Returns whether this device has a stored passkey ready to be used for sign-in
 * @param client - the thirdweb client
 * @returns whether the device has a stored passkey
 * @walletUtils
 */
export async function hasStoredPasskey(client, ecosystemId) {
    const storage = new ClientScopedStorage({
        storage: webLocalStorage, // TODO (passkey) react native variant of this fn
        clientId: client.clientId,
        ecosystemId: ecosystemId,
    });
    const credId = await storage.getPasskeyCredentialId();
    return !!credId;
}
