// thirdweb/dist/cjs/wallets/wallet-connect/receiver/request-handlers/sign-transaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleSignTransactionRequest = handleSignTransactionRequest;
const hex_js_1 = require("../../../../utils/encoding/hex.js");
const utils_js_1 = require("../utils.js");
/**
 * @internal
 */
async function handleSignTransactionRequest(options) {
    const { account, params: [transaction], } = options;
    if (!account.signTransaction) {
        throw new Error("The current account does not support signing transactions");
    }
    if (transaction.from !== undefined) {
        (0, utils_js_1.validateAccountAddress)(account, transaction.from);
    }
    return account.signTransaction({
        gas: transaction.gas ? (0, hex_js_1.hexToBigInt)(transaction.gas) : undefined,
        gasPrice: transaction.gasPrice
            ? (0, hex_js_1.hexToBigInt)(transaction.gasPrice)
            : undefined,
        value: transaction.value ? (0, hex_js_1.hexToBigInt)(transaction.value) : undefined,
        nonce: transaction.nonce ? (0, hex_js_1.hexToNumber)(transaction.nonce) : undefined,
        to: transaction.to,
        data: transaction.data,
    });
}
