// thirdweb/dist/cjs/extensions/ens/__generated__/AddressResolver/read/text.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isTextSupported = isTextSupported;
exports.encodeTextParams = encodeTextParams;
exports.encodeText = encodeText;
exports.decodeTextResult = decodeTextResult;
exports.text = text;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x59d1d43c";
const FN_INPUTS = [
    {
        type: "bytes32",
        name: "name",
    },
    {
        type: "string",
        name: "key",
    },
];
const FN_OUTPUTS = [
    {
        type: "string",
    },
];
/**
 * Checks if the `text` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `text` method is supported.
 * @extension ENS
 * @example
 * ```ts
 * import { isTextSupported } from "thirdweb/extensions/ens";
 * const supported = isTextSupported(["0x..."]);
 * ```
 */
function isTextSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "text" function.
 * @param options - The options for the text function.
 * @returns The encoded ABI parameters.
 * @extension ENS
 * @example
 * ```ts
 * import { encodeTextParams } from "thirdweb/extensions/ens";
 * const result = encodeTextParams({
 *  name: ...,
 *  key: ...,
 * });
 * ```
 */
function encodeTextParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.name, options.key]);
}
/**
 * Encodes the "text" function into a Hex string with its parameters.
 * @param options - The options for the text function.
 * @returns The encoded hexadecimal string.
 * @extension ENS
 * @example
 * ```ts
 * import { encodeText } from "thirdweb/extensions/ens";
 * const result = encodeText({
 *  name: ...,
 *  key: ...,
 * });
 * ```
 */
function encodeText(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeTextParams(options).slice(2));
}
/**
 * Decodes the result of the text function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ENS
 * @example
 * ```ts
 * import { decodeTextResult } from "thirdweb/extensions/ens";
 * const result = decodeTextResultResult("...");
 * ```
 */
function decodeTextResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "text" function on the contract.
 * @param options - The options for the text function.
 * @returns The parsed result of the function call.
 * @extension ENS
 * @example
 * ```ts
 * import { text } from "thirdweb/extensions/ens";
 *
 * const result = await text({
 *  contract,
 *  name: ...,
 *  key: ...,
 * });
 *
 * ```
 */
async function text(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.name, options.key],
    });
}
