// thirdweb/dist/esm/extensions/erc1155/__generated__/IClaimableERC1155/read/verifyClaim.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { encodeAbiParameters } from "../../../../../utils/abi/encodeAbiParameters.js";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0x4bbb1abf";
const FN_INPUTS = [
    {
        type: "address",
        name: "_claimer",
    },
    {
        type: "uint256",
        name: "_tokenId",
    },
    {
        type: "uint256",
        name: "_quantity",
    },
];
const FN_OUTPUTS = [];
/**
 * Checks if the `verifyClaim` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `verifyClaim` method is supported.
 * @extension ERC1155
 * @example
 * ```ts
 * import { isVerifyClaimSupported } from "thirdweb/extensions/erc1155";
 * const supported = isVerifyClaimSupported(["0x..."]);
 * ```
 */
export function isVerifyClaimSupported(availableSelectors) {
    return detectMethod({
        availableSelectors,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "verifyClaim" function.
 * @param options - The options for the verifyClaim function.
 * @returns The encoded ABI parameters.
 * @extension ERC1155
 * @example
 * ```ts
 * import { encodeVerifyClaimParams } from "thirdweb/extensions/erc1155";
 * const result = encodeVerifyClaimParams({
 *  claimer: ...,
 *  tokenId: ...,
 *  quantity: ...,
 * });
 * ```
 */
export function encodeVerifyClaimParams(options) {
    return encodeAbiParameters(FN_INPUTS, [
        options.claimer,
        options.tokenId,
        options.quantity,
    ]);
}
/**
 * Encodes the "verifyClaim" function into a Hex string with its parameters.
 * @param options - The options for the verifyClaim function.
 * @returns The encoded hexadecimal string.
 * @extension ERC1155
 * @example
 * ```ts
 * import { encodeVerifyClaim } from "thirdweb/extensions/erc1155";
 * const result = encodeVerifyClaim({
 *  claimer: ...,
 *  tokenId: ...,
 *  quantity: ...,
 * });
 * ```
 */
export function encodeVerifyClaim(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (FN_SELECTOR +
        encodeVerifyClaimParams(options).slice(2));
}
/**
 * Calls the "verifyClaim" function on the contract.
 * @param options - The options for the verifyClaim function.
 * @returns The parsed result of the function call.
 * @extension ERC1155
 * @example
 * ```ts
 * import { verifyClaim } from "thirdweb/extensions/erc1155";
 *
 * const result = await verifyClaim({
 *  contract,
 *  claimer: ...,
 *  tokenId: ...,
 *  quantity: ...,
 * });
 *
 * ```
 */
export async function verifyClaim(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.claimer, options.tokenId, options.quantity],
    });
}
