// thirdweb/dist/esm/react/web/ui/ConnectWallet/screens/LinkedProfilesScreen.js
"use client";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { shortenAddress } from "../../../../../utils/address.js";
import { fontSize, iconSize } from "../../../../core/design-system/index.js";
import { useProfiles } from "../../../../core/hooks/others/useProfiles.js";
import { getWalletIcon } from "../../../../core/utils/walletIcon.js";
import { LoadingScreen } from "../../../wallets/shared/LoadingScreen.js";
import { Img } from "../../components/Img.js";
import { Spacer } from "../../components/Spacer.js";
import { Container, Line, ModalHeader } from "../../components/basic.js";
import { Text } from "../../components/text.js";
import { Blobbie } from "../Blobbie.js";
import { MenuButton } from "../MenuButton.js";
import { AddUserIcon } from "../icons/AddUserIcon.js";
function getProfileDisplayName(profile) {
    switch (true) {
        case profile.type === "email" && profile.details.email !== undefined:
            return profile.details.email;
        case profile.type === "google" && profile.details.email !== undefined:
            return profile.details.email;
        case profile.type === "phone" && profile.details.phone !== undefined:
            return profile.details.phone;
        case profile.details.address !== undefined:
            return shortenAddress(profile.details.address, 6);
        case profile.type === "cognito" &&
            profile.details.email !== undefined:
            return profile.details.email;
        default:
            return profile.type.slice(0, 1).toUpperCase() + profile.type.slice(1);
    }
}
/**
 * @internal
 */
export function LinkedProfilesScreen(props) {
    const { data: connectedProfiles, isLoading } = useProfiles();
    if (isLoading) {
        return _jsx(LoadingScreen, {});
    }
    return (_jsxs(Container, { style: {
            minHeight: "300px",
        }, children: [_jsx(Container, { p: "lg", children: _jsx(ModalHeader, { title: props.locale.manageWallet.linkedProfiles, onBack: props.onBack }) }), _jsx(Line, {}), isLoading ? (_jsx(LoadingScreen, {})) : (_jsxs(Container, { scrollY: true, style: {
                    height: "300px",
                }, children: [_jsx(Spacer, { y: "md" }), _jsxs(Container, { px: "sm", children: [_jsxs(MenuButton, { onClick: () => {
                                    props.setScreen("link-profile");
                                }, style: {
                                    fontSize: fontSize.sm,
                                }, children: [_jsx(AddUserIcon, { size: iconSize.md }), _jsx(Text, { color: "primaryText", children: props.locale.manageWallet.linkProfile })] }), _jsx(Spacer, { y: "xs" }), connectedProfiles?.map((profile) => (_jsxs(MenuButton, { onClick: () => {
                                    props.setScreen("linked-profiles");
                                }, style: {
                                    fontSize: fontSize.sm,
                                    cursor: "default",
                                }, disabled // disabled until we have more data to show on a dedicated profile screen
                                : true, children: [profile.type === "wallet" && profile.details.address ? (_jsx(Container, { style: {
                                            width: "24px",
                                            height: "24px",
                                            borderRadius: "100%",
                                            overflow: "hidden",
                                        }, children: _jsx(Blobbie, { address: profile.details.address, size: 24 }) })) : (_jsx(Img, { src: getWalletIcon(profile.type), width: iconSize.md, height: iconSize.md, loading: "eager", client: props.client })), _jsx(Text, { color: "primaryText", children: getProfileDisplayName(profile) })] }, `${profile.type}-${getProfileDisplayName(profile)}`)))] }), _jsx(Spacer, { y: "md" })] }))] }));
}
