// thirdweb/dist/cjs/wallets/in-app/web/lib/auth/otp.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.verifyOtp = exports.sendOtp = void 0;
const getLoginPath_js_1 = require("../../../core/authentication/getLoginPath.js");
/**
 * @internal
 */
const sendOtp = async (args) => {
    const { client, ecosystem } = args;
    const url = (0, getLoginPath_js_1.getLoginUrl)({ client, ecosystem, authOption: args.strategy });
    const headers = {
        "Content-Type": "application/json",
        "x-client-id": client.clientId,
    };
    if (ecosystem?.id) {
        headers["x-ecosystem-id"] = ecosystem.id;
    }
    if (ecosystem?.partnerId) {
        headers["x-ecosystem-partner-id"] = ecosystem.partnerId;
    }
    const body = (() => {
        switch (args.strategy) {
            case "email":
                return {
                    email: args.email,
                };
            case "phone":
                return {
                    phone: args.phoneNumber,
                };
        }
    })();
    const response = await fetch(url, {
        method: "POST",
        headers,
        body: JSON.stringify(body),
    });
    if (!response.ok) {
        throw new Error("Failed to send verification code");
    }
    return await response.json();
};
exports.sendOtp = sendOtp;
/**
 * @internal
 */
const verifyOtp = async (args) => {
    const { client, ecosystem } = args;
    const url = (0, getLoginPath_js_1.getLoginCallbackUrl)({
        authOption: args.strategy,
        client: args.client,
        ecosystem: args.ecosystem,
    });
    const headers = {
        "Content-Type": "application/json",
        "x-client-id": client.clientId,
    };
    if (ecosystem?.id) {
        headers["x-ecosystem-id"] = ecosystem.id;
    }
    if (ecosystem?.partnerId) {
        headers["x-ecosystem-partner-id"] = ecosystem.partnerId;
    }
    const body = (() => {
        switch (args.strategy) {
            case "email":
                return {
                    email: args.email,
                    code: args.verificationCode,
                };
            case "phone":
                return {
                    phone: args.phoneNumber,
                    code: args.verificationCode,
                };
        }
    })();
    const response = await fetch(url, {
        method: "POST",
        headers,
        body: JSON.stringify(body),
    });
    if (!response.ok) {
        throw new Error("Failed to verify verification code");
    }
    return await response.json();
};
exports.verifyOtp = verifyOtp;
