// thirdweb/dist/cjs/extensions/marketplace/__generated__/IOffers/read/totalOffers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isTotalOffersSupported = isTotalOffersSupported;
exports.decodeTotalOffersResult = decodeTotalOffersResult;
exports.totalOffers = totalOffers;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xa9fd8ed1";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `totalOffers` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `totalOffers` method is supported.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { isTotalOffersSupported } from "thirdweb/extensions/marketplace";
 * const supported = isTotalOffersSupported(["0x..."]);
 * ```
 */
function isTotalOffersSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the totalOffers function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { decodeTotalOffersResult } from "thirdweb/extensions/marketplace";
 * const result = decodeTotalOffersResultResult("...");
 * ```
 */
function decodeTotalOffersResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "totalOffers" function on the contract.
 * @param options - The options for the totalOffers function.
 * @returns The parsed result of the function call.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { totalOffers } from "thirdweb/extensions/marketplace";
 *
 * const result = await totalOffers({
 *  contract,
 * });
 *
 * ```
 */
async function totalOffers(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
