// thirdweb/dist/cjs/transaction/extract-error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractError = extractError;
const viem_1 = require("viem");
const resolve_abi_js_1 = require("../contract/actions/resolve-abi.js");
const hex_js_1 = require("../utils/encoding/hex.js");
const json_js_1 = require("../utils/json.js");
/**
 * @internal
 */
async function extractError(args) {
    const { error, contract } = args;
    if (typeof error === "object") {
        // try to parse RPC error
        const errorObj = error;
        if (errorObj.data) {
            if (errorObj.data !== "0x" && (0, hex_js_1.isHex)(errorObj.data)) {
                let abi = contract?.abi;
                if (contract && !abi) {
                    abi = await (0, resolve_abi_js_1.resolveContractAbi)(contract).catch(() => undefined);
                }
                const parsedError = (0, viem_1.decodeErrorResult)({
                    data: errorObj.data,
                    abi,
                });
                return new TransactionError(`${parsedError.errorName}${parsedError.args ? ` - ${parsedError.args}` : ""}`, contract);
            }
            return new TransactionError(`Execution Reverted: ${(0, json_js_1.stringify)(errorObj)}`, contract);
        }
    }
    return error;
}
const __DEV__ = process.env.NODE_ENV !== "production";
class TransactionError extends Error {
    constructor(reason, contract) {
        let message = reason;
        if (__DEV__ && contract) {
            // show more infor in dev
            message = [
                reason,
                "",
                `contract: ${contract.address}`,
                `chainId: ${contract.chain?.id}`,
            ].join("\n");
        }
        super(message);
        Object.defineProperty(this, "contractAddress", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "chainId", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.name = "TransactionError";
        this.contractAddress = contract?.address;
        this.chainId = contract?.chain?.id;
        this.message = message;
    }
}
