// thirdweb/dist/esm/extensions/modules/__generated__/RoyaltyERC721/read/getTransferValidator.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0x098144d4";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "address",
        name: "validator",
    },
];
/**
 * Checks if the `getTransferValidator` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `getTransferValidator` method is supported.
 * @modules RoyaltyERC721
 * @example
 * ```ts
 * import { RoyaltyERC721 } from "thirdweb/modules";
 * const supported = RoyaltyERC721.isGetTransferValidatorSupported(["0x..."]);
 * ```
 */
export function isGetTransferValidatorSupported(availableSelectors) {
    return detectMethod({
        availableSelectors,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the getTransferValidator function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @modules RoyaltyERC721
 * @example
 * ```ts
 * import { RoyaltyERC721 } from "thirdweb/modules";
 * const result = RoyaltyERC721.decodeGetTransferValidatorResultResult("...");
 * ```
 */
export function decodeGetTransferValidatorResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getTransferValidator" function on the contract.
 * @param options - The options for the getTransferValidator function.
 * @returns The parsed result of the function call.
 * @modules RoyaltyERC721
 * @example
 * ```ts
 * import { RoyaltyERC721 } from "thirdweb/modules";
 *
 * const result = await RoyaltyERC721.getTransferValidator({
 *  contract,
 * });
 *
 * ```
 */
export async function getTransferValidator(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
