// thirdweb/dist/cjs/extensions/erc1155/__generated__/Zora1155/read/nextTokenId.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isNextTokenIdSupported = isNextTokenIdSupported;
exports.decodeNextTokenIdResult = decodeNextTokenIdResult;
exports.nextTokenId = nextTokenId;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x75794a3c";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `nextTokenId` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `nextTokenId` method is supported.
 * @extension ERC1155
 * @example
 * ```ts
 * import { isNextTokenIdSupported } from "thirdweb/extensions/erc1155";
 * const supported = isNextTokenIdSupported(["0x..."]);
 * ```
 */
function isNextTokenIdSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the nextTokenId function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC1155
 * @example
 * ```ts
 * import { decodeNextTokenIdResult } from "thirdweb/extensions/erc1155";
 * const result = decodeNextTokenIdResultResult("...");
 * ```
 */
function decodeNextTokenIdResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "nextTokenId" function on the contract.
 * @param options - The options for the nextTokenId function.
 * @returns The parsed result of the function call.
 * @extension ERC1155
 * @example
 * ```ts
 * import { nextTokenId } from "thirdweb/extensions/erc1155";
 *
 * const result = await nextTokenId({
 *  contract,
 * });
 *
 * ```
 */
async function nextTokenId(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
