// thirdweb/dist/esm/extensions/farcaster/write/registerFid.js
import { prepareContractCall } from "../../../transaction/prepare-contract-call.js";
import { toBigInt } from "../../../utils/bigint.js";
import { getIdGateway } from "../contracts/getIdGateway.js";
import { getRegistrationPrice } from "../read/getRegistrationPrice.js";
/**
 * Registers a Farcaster fid for the given wallet.
 * @param options - The options for registering an account.
 * @returns A prepared transaction object to register the account.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { registerFid } from "thirdweb/extensions/farcaster";
 * import { sendTransaction } from "thirdweb";
 *
 * const transaction = registerFid({
 *  client,
 * 	recoveryAddress
 * });
 *
 * await sendTransaction({ transaction, account });
 * ```
 */
export function registerFid(options) {
    const extraStorage = toBigInt(options.extraStorage ?? 0);
    if (extraStorage < 0n) {
        throw new Error(`Expected extraStorage to be greater than or equal to 0, got ${extraStorage}`);
    }
    return prepareContractCall({
        contract: getIdGateway({
            client: options.client,
            chain: options.chain,
        }),
        method: [
            "0x6d705ebb",
            [
                {
                    type: "address",
                    name: "recovery",
                },
                {
                    type: "uint256",
                    name: "extraStorage",
                },
            ],
            [
                {
                    type: "uint256",
                    name: "fid",
                },
                {
                    type: "uint256",
                    name: "overpayment",
                },
            ],
        ],
        value: async () => {
            return await getRegistrationPrice({
                client: options.client,
                chain: options.chain,
                extraStorage: extraStorage,
                disableCache: options.disableCache,
            });
        },
        params: [options.recoveryAddress, extraStorage],
    });
}
