// thirdweb/dist/cjs/extensions/thirdweb/__generated__/ITWMultichainRegistry/read/getAll.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isGetAllSupported = isGetAllSupported;
exports.encodeGetAllParams = encodeGetAllParams;
exports.encodeGetAll = encodeGetAll;
exports.decodeGetAllResult = decodeGetAllResult;
exports.getAll = getAll;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xeb077342";
const FN_INPUTS = [
    {
        type: "address",
        name: "_deployer",
    },
];
const FN_OUTPUTS = [
    {
        type: "tuple[]",
        name: "allDeployments",
        components: [
            {
                type: "address",
                name: "deploymentAddress",
            },
            {
                type: "uint256",
                name: "chainId",
            },
            {
                type: "string",
                name: "metadataURI",
            },
        ],
    },
];
/**
 * Checks if the `getAll` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `getAll` method is supported.
 * @extension THIRDWEB
 * @example
 * ```ts
 * import { isGetAllSupported } from "thirdweb/extensions/thirdweb";
 * const supported = isGetAllSupported(["0x..."]);
 * ```
 */
function isGetAllSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "getAll" function.
 * @param options - The options for the getAll function.
 * @returns The encoded ABI parameters.
 * @extension THIRDWEB
 * @example
 * ```ts
 * import { encodeGetAllParams } from "thirdweb/extensions/thirdweb";
 * const result = encodeGetAllParams({
 *  deployer: ...,
 * });
 * ```
 */
function encodeGetAllParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.deployer]);
}
/**
 * Encodes the "getAll" function into a Hex string with its parameters.
 * @param options - The options for the getAll function.
 * @returns The encoded hexadecimal string.
 * @extension THIRDWEB
 * @example
 * ```ts
 * import { encodeGetAll } from "thirdweb/extensions/thirdweb";
 * const result = encodeGetAll({
 *  deployer: ...,
 * });
 * ```
 */
function encodeGetAll(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeGetAllParams(options).slice(2));
}
/**
 * Decodes the result of the getAll function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension THIRDWEB
 * @example
 * ```ts
 * import { decodeGetAllResult } from "thirdweb/extensions/thirdweb";
 * const result = decodeGetAllResultResult("...");
 * ```
 */
function decodeGetAllResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getAll" function on the contract.
 * @param options - The options for the getAll function.
 * @returns The parsed result of the function call.
 * @extension THIRDWEB
 * @example
 * ```ts
 * import { getAll } from "thirdweb/extensions/thirdweb";
 *
 * const result = await getAll({
 *  contract,
 *  deployer: ...,
 * });
 *
 * ```
 */
async function getAll(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.deployer],
    });
}
