// thirdweb/dist/cjs/react/web/ui/ConnectWallet/screens/Buy/Stepper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StepIcon = StepIcon;
exports.Step = Step;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("@emotion/react");
const react_icons_1 = require("@radix-ui/react-icons");
const index_js_1 = require("../../../../../core/design-system/index.js");
const basic_js_1 = require("../../../components/basic.js");
const elements_js_1 = require("../../../design-system/elements.js");
function StepIcon(props) {
    return ((0, jsx_runtime_1.jsx)(basic_js_1.Container, { flex: "row", center: "both", color: props.isDone
            ? "success"
            : props.isActive
                ? "accentText"
                : "secondaryText", children: (0, jsx_runtime_1.jsx)(Circle, { children: props.isDone ? ((0, jsx_runtime_1.jsx)(react_icons_1.CheckIcon, { width: index_js_1.iconSize.sm, height: index_js_1.iconSize.sm })) : ((0, jsx_runtime_1.jsx)(PulsingDot, { "data-active": props.isActive })) }) }));
}
function Step(props) {
    return ((0, jsx_runtime_1.jsxs)(basic_js_1.Container, { flex: "row", center: "y", gap: "xs", style: {
            fontSize: index_js_1.fontSize.sm,
        }, color: props.isDone
            ? "success"
            : props.isActive
                ? "accentText"
                : "secondaryText", children: [(0, jsx_runtime_1.jsx)(StepIcon, { isDone: props.isDone, isActive: props.isActive }), props.label] }));
}
const pulseAnimation = (0, react_1.keyframes) `
0% {
  opacity: 1;
  transform: scale(0.5);
}
100% {
  opacity: 0;
  transform: scale(1.5);
}
`;
const PulsingDot = /* @__PURE__ */ (0, elements_js_1.StyledDiv)(() => {
    return {
        background: "currentColor",
        width: "9px",
        height: "9px",
        borderRadius: "50%",
        '&[data-active="true"]': {
            animation: `${pulseAnimation} 1s infinite`,
        },
    };
});
const Circle = /* @__PURE__ */ (0, elements_js_1.StyledDiv)(() => {
    return {
        border: "1px solid currentColor",
        width: "20px",
        height: "20px",
        borderRadius: "50%",
        display: "flex",
        alignItems: "center",
        justifyContent: "center",
    };
});
