// thirdweb/dist/cjs/extensions/modules/__generated__/ClaimableERC1155/read/getClaimConditionByTokenId.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isGetClaimConditionByTokenIdSupported = isGetClaimConditionByTokenIdSupported;
exports.encodeGetClaimConditionByTokenIdParams = encodeGetClaimConditionByTokenIdParams;
exports.encodeGetClaimConditionByTokenId = encodeGetClaimConditionByTokenId;
exports.decodeGetClaimConditionByTokenIdResult = decodeGetClaimConditionByTokenIdResult;
exports.getClaimConditionByTokenId = getClaimConditionByTokenId;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x29a20bf4";
const FN_INPUTS = [
    {
        type: "uint256",
        name: "tokenId",
    },
];
const FN_OUTPUTS = [
    {
        type: "tuple",
        name: "claimCondition",
        components: [
            {
                type: "uint256",
                name: "availableSupply",
            },
            {
                type: "bytes32",
                name: "allowlistMerkleRoot",
            },
            {
                type: "uint256",
                name: "pricePerUnit",
            },
            {
                type: "address",
                name: "currency",
            },
            {
                type: "uint256",
                name: "maxMintPerWallet",
            },
            {
                type: "uint48",
                name: "startTimestamp",
            },
            {
                type: "uint48",
                name: "endTimestamp",
            },
            {
                type: "string",
                name: "auxData",
            },
        ],
    },
];
/**
 * Checks if the `getClaimConditionByTokenId` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `getClaimConditionByTokenId` method is supported.
 * @modules ClaimableERC1155
 * @example
 * ```ts
 * import { ClaimableERC1155 } from "thirdweb/modules";
 * const supported = ClaimableERC1155.isGetClaimConditionByTokenIdSupported(["0x..."]);
 * ```
 */
function isGetClaimConditionByTokenIdSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "getClaimConditionByTokenId" function.
 * @param options - The options for the getClaimConditionByTokenId function.
 * @returns The encoded ABI parameters.
 * @modules ClaimableERC1155
 * @example
 * ```ts
 * import { ClaimableERC1155 } from "thirdweb/modules";
 * const result = ClaimableERC1155.encodeGetClaimConditionByTokenIdParams({
 *  tokenId: ...,
 * });
 * ```
 */
function encodeGetClaimConditionByTokenIdParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.tokenId]);
}
/**
 * Encodes the "getClaimConditionByTokenId" function into a Hex string with its parameters.
 * @param options - The options for the getClaimConditionByTokenId function.
 * @returns The encoded hexadecimal string.
 * @modules ClaimableERC1155
 * @example
 * ```ts
 * import { ClaimableERC1155 } from "thirdweb/modules";
 * const result = ClaimableERC1155.encodeGetClaimConditionByTokenId({
 *  tokenId: ...,
 * });
 * ```
 */
function encodeGetClaimConditionByTokenId(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeGetClaimConditionByTokenIdParams(options).slice(2));
}
/**
 * Decodes the result of the getClaimConditionByTokenId function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @modules ClaimableERC1155
 * @example
 * ```ts
 * import { ClaimableERC1155 } from "thirdweb/modules";
 * const result = ClaimableERC1155.decodeGetClaimConditionByTokenIdResultResult("...");
 * ```
 */
function decodeGetClaimConditionByTokenIdResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getClaimConditionByTokenId" function on the contract.
 * @param options - The options for the getClaimConditionByTokenId function.
 * @returns The parsed result of the function call.
 * @modules ClaimableERC1155
 * @example
 * ```ts
 * import { ClaimableERC1155 } from "thirdweb/modules";
 *
 * const result = await ClaimableERC1155.getClaimConditionByTokenId({
 *  contract,
 *  tokenId: ...,
 * });
 *
 * ```
 */
async function getClaimConditionByTokenId(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.tokenId],
    });
}
