// thirdweb/dist/cjs/react/native/ui/components/button.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ThemedButton = ThemedButton;
exports.ThemedButtonWithIcon = ThemedButtonWithIcon;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_native_1 = require("react-native");
const index_js_1 = require("../../design-system/index.js");
const RNImage_js_1 = require("./RNImage.js");
const text_js_1 = require("./text.js");
function ThemedButton(props) {
    const variant = props.variant ?? "primary";
    const { style: styleOverride, theme, ...restProps } = props;
    return ((0, jsx_runtime_1.jsx)(react_native_1.TouchableOpacity, { activeOpacity: 0.5, style: [
            styles.button,
            {
                borderColor: variant === "secondary" ? theme.colors.borderColor : "transparent",
                borderWidth: variant === "secondary" ? 1 : 0,
                backgroundColor: variant === "secondary"
                    ? "transparent"
                    : variant === "accent"
                        ? props.theme.colors.accentButtonBg
                        : props.theme.colors.primaryButtonBg,
            },
            styleOverride,
        ], ...restProps, children: props.children }));
}
function ThemedButtonWithIcon(props) {
    const { theme, title, icon, onPress } = props;
    return ((0, jsx_runtime_1.jsx)(ThemedButton, { theme: theme, variant: "secondary", onPress: onPress, children: (0, jsx_runtime_1.jsxs)(react_native_1.View, { style: {
                flexDirection: "row",
                gap: index_js_1.spacing.md,
                paddingLeft: index_js_1.spacing.sm,
                alignContent: "center",
                width: "100%",
            }, children: [(0, jsx_runtime_1.jsx)(RNImage_js_1.RNImage, { theme: theme, size: 24, data: icon, color: theme.colors.accentButtonBg }), (0, jsx_runtime_1.jsx)(text_js_1.ThemedText, { theme: theme, children: title })] }) }));
}
const styles = react_native_1.StyleSheet.create({
    button: {
        flexDirection: "row",
        padding: index_js_1.spacing.md,
        borderRadius: index_js_1.radius.lg,
        justifyContent: "center",
        alignItems: "center",
    },
});
