// thirdweb/dist/cjs/react/native/ui/components/Skeleton.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Skeleton = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const react_native_1 = require("react-native");
const index_js_1 = require("../../design-system/index.js");
const { width } = react_native_1.Dimensions.get("window");
const Skeleton = ({ style, color, theme, }) => {
    const translateX = (0, react_1.useRef)(new react_native_1.Animated.Value(-width));
    (0, react_1.useEffect)(() => {
        const animation = react_native_1.Animated.loop(react_native_1.Animated.sequence([
            react_native_1.Animated.timing(translateX.current, {
                toValue: width,
                duration: 2500,
                useNativeDriver: true,
            }),
            react_native_1.Animated.timing(translateX.current, {
                toValue: -width,
                duration: 0,
                useNativeDriver: true,
            }),
        ]));
        animation.start();
        return () => animation.stop();
    }, []);
    return ((0, jsx_runtime_1.jsxs)(react_native_1.View, { style: [
            styles.container,
            { backgroundColor: color || theme.colors.borderColor },
            style,
        ], children: [(0, jsx_runtime_1.jsx)(react_native_1.View, { style: styles.skeleton }), (0, jsx_runtime_1.jsx)(react_native_1.Animated.View, { style: [
                    styles.gradient,
                    {
                        transform: [{ translateX: translateX.current }],
                    },
                ], children: (0, jsx_runtime_1.jsx)(react_native_1.View, { style: styles.innerGradient }) })] }));
};
exports.Skeleton = Skeleton;
const styles = react_native_1.StyleSheet.create({
    container: {
        overflow: "hidden",
        borderRadius: index_js_1.radius.lg,
    },
    skeleton: {
        height: "100%",
        width: "100%",
    },
    gradient: {
        position: "absolute",
        top: 0,
        bottom: 0,
        width: "100%",
    },
    innerGradient: {
        width: "100%",
        height: "100%",
        backgroundColor: "rgba(255,255,255,0.5)",
        opacity: 0.3,
        transform: [{ skewX: "-20deg" }],
    },
});
