// thirdweb/dist/cjs/extensions/erc20/write/approve.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.approve = approve;
const units_js_1 = require("../../../utils/units.js");
const approve_js_1 = require("../__generated__/IERC20/write/approve.js");
/**
 * Approves the spending of tokens by a specific address.
 * @param options - The transaction options.
 * @returns A prepared transaction object.
 * @extension ERC20
 * @example
 * ```ts
 * import { approve } from "thirdweb/extensions/erc20";
 * import { sendTransaction } from "thirdweb";
 *
 * const transaction = await approve({
 *  contract,
 *  spender: "0x...",
 *  amount: 100,
 * });
 *
 * await sendTransaction({ transaction, account });
 * ```
 */
function approve(options) {
    return (0, approve_js_1.approve)({
        contract: options.contract,
        asyncParams: async () => {
            let amount;
            if ("amount" in options) {
                // if we need to parse the amount from ether to gwei then we pull in the decimals extension
                const { decimals } = await Promise.resolve().then(() => require("../read/decimals.js"));
                // if this fails we fall back to `18` decimals
                const d = await decimals(options).catch(() => 18);
                // turn ether into gwei
                amount = (0, units_js_1.toUnits)(options.amount.toString(), d);
            }
            else {
                amount = options.amountWei;
            }
            return {
                spender: options.spender,
                value: amount,
                overrides: {
                    erc20Value: {
                        amountWei: amount,
                        tokenAddress: options.contract.address,
                    },
                },
            };
        },
    });
}
