// thirdweb/dist/cjs/extensions/erc721/write/mintTo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isMintToSupported = void 0;
exports.mintTo = mintTo;
const mintTo_js_1 = require("../__generated__/IMintableERC721/write/mintTo.js");
var mintTo_js_2 = require("../__generated__/IMintableERC721/write/mintTo.js");
Object.defineProperty(exports, "isMintToSupported", { enumerable: true, get: function () { return mintTo_js_2.isMintToSupported; } });
/**
 * Mints a new ERC721 token and assigns it to the specified address.
 * If the `nft` parameter is a string, it will be used as the token URI.
 * If the `nft` parameter is a file, it will be uploaded to the storage server and the resulting URI will be used as the token URI.
 * @param options - The transaction options.
 * @returns A promise that resolves to the transaction result.
 * @extension ERC721
 * @example
 * ```ts
 * import { mintTo } from "thirdweb/extensions/erc721";
 * import { sendTransaction } from "thirdweb";
 *
 * const transaction = mintTo({
 *  contract,
 *  to: "0x...",
 *  nft: {
 *    name: "My NFT",
 *    description: "This is my NFT",
 *    image: "https://example.com/image.png",
 *  },
 * });
 *
 * await sendTransaction({ transaction, account });
 * ```
 */
function mintTo(options) {
    return (0, mintTo_js_1.mintTo)({
        contract: options.contract,
        asyncParams: async () => {
            let tokenUri;
            if (typeof options.nft === "string") {
                // if the input is already a string then we just use that
                tokenUri = options.nft;
            }
            else {
                // otherwise we need to upload the file to the storage server
                // load the upload code if we need it
                const { upload } = await Promise.resolve().then(() => require("../../../storage/upload.js"));
                tokenUri = await upload({
                    client: options.contract.client,
                    files: [options.nft],
                });
            }
            return {
                to: options.to,
                uri: tokenUri,
                overrides: options.overrides,
            };
        },
    });
}
