// thirdweb/dist/cjs/wallets/in-app/native/helpers/auth/cognitoAuth.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cognitoEmailSignUp = cognitoEmailSignUp;
exports.cognitoEmailSignIn = cognitoEmailSignIn;
exports.cognitoPhoneSignIn = cognitoPhoneSignIn;
exports.cognitoPhoneSignUp = cognitoPhoneSignUp;
const aws_amplify_1 = require("aws-amplify");
const random_js_1 = require("../../../../../utils/random.js");
const constants_js_1 = require("../constants.js");
aws_amplify_1.Amplify.configure({
    Auth: {
        region: constants_js_1.AWS_REGION,
        userPoolId: constants_js_1.COGNITO_USER_POOL_ID,
        userPoolWebClientId: constants_js_1.COGNITO_APP_CLIENT_ID,
    },
});
async function cognitoEmailSignUp(email, clientId) {
    await aws_amplify_1.Auth.signUp({
        username: `${email}:email:${clientId}`,
        password: (0, random_js_1.randomBytesHex)(30),
        attributes: {
            email,
        },
    });
    await aws_amplify_1.Auth.signOut();
}
async function cognitoEmailSignIn(email, clientId) {
    const cognitoUser = await aws_amplify_1.Auth.signIn(`${email}:email:${clientId}`);
    return cognitoUser;
}
async function cognitoPhoneSignIn(phoneNumber, clientId) {
    const cognitoUser = (await aws_amplify_1.Auth.signIn(`${phoneNumber}:sms:${clientId}`));
    return cognitoUser;
}
async function cognitoPhoneSignUp(phoneNumber, clientId) {
    await aws_amplify_1.Auth.signUp({
        username: `${phoneNumber}:sms:${clientId}`,
        password: (0, random_js_1.randomBytesHex)(30),
        attributes: {
            // ! This is a placeholder email, it will not be used for anything. We simply need this to satisfy the Cognito API.
            email: "cognito@thirdweb.com",
        },
    });
    await aws_amplify_1.Auth.signOut();
}
