// thirdweb/dist/esm/extensions/ens/__generated__/UniversalResolver/read/reverse.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { encodeAbiParameters } from "../../../../../utils/abi/encodeAbiParameters.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0xec11c823";
const FN_INPUTS = [
    {
        type: "bytes",
        name: "reverseName",
    },
];
const FN_OUTPUTS = [
    {
        type: "string",
    },
    {
        type: "address",
    },
    {
        type: "address",
    },
    {
        type: "address",
    },
];
/**
 * Checks if the `reverse` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `reverse` method is supported.
 * @extension ENS
 * @example
 * ```ts
 * import { isReverseSupported } from "thirdweb/extensions/ens";
 * const supported = isReverseSupported(["0x..."]);
 * ```
 */
export function isReverseSupported(availableSelectors) {
    return detectMethod({
        availableSelectors,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "reverse" function.
 * @param options - The options for the reverse function.
 * @returns The encoded ABI parameters.
 * @extension ENS
 * @example
 * ```ts
 * import { encodeReverseParams } from "thirdweb/extensions/ens";
 * const result = encodeReverseParams({
 *  reverseName: ...,
 * });
 * ```
 */
export function encodeReverseParams(options) {
    return encodeAbiParameters(FN_INPUTS, [options.reverseName]);
}
/**
 * Encodes the "reverse" function into a Hex string with its parameters.
 * @param options - The options for the reverse function.
 * @returns The encoded hexadecimal string.
 * @extension ENS
 * @example
 * ```ts
 * import { encodeReverse } from "thirdweb/extensions/ens";
 * const result = encodeReverse({
 *  reverseName: ...,
 * });
 * ```
 */
export function encodeReverse(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (FN_SELECTOR +
        encodeReverseParams(options).slice(2));
}
/**
 * Decodes the result of the reverse function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ENS
 * @example
 * ```ts
 * import { decodeReverseResult } from "thirdweb/extensions/ens";
 * const result = decodeReverseResultResult("...");
 * ```
 */
export function decodeReverseResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result);
}
/**
 * Calls the "reverse" function on the contract.
 * @param options - The options for the reverse function.
 * @returns The parsed result of the function call.
 * @extension ENS
 * @example
 * ```ts
 * import { reverse } from "thirdweb/extensions/ens";
 *
 * const result = await reverse({
 *  contract,
 *  reverseName: ...,
 * });
 *
 * ```
 */
export async function reverse(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.reverseName],
    });
}
