// thirdweb/dist/cjs/extensions/modules/__generated__/BatchMetadataERC721/module/install.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.module = module;
exports.install = install;
exports.encodeInstall = encodeInstall;
const getOrDeployModule_js_1 = require("../../../common/getOrDeployModule.js");
const installPublishedModule_js_1 = require("../../../common/installPublishedModule.js");
const contractId = "BatchMetadataERC721";
/**
 * Convenience function to add the BatchMetadataERC721 module as a default module on a core contract.
 * @param params - The parameters for the module.
 * @returns - The module function.
 * @example
 * ```ts
 * import { BatchMetadataERC721, deployModularContract } from "thirdweb/modules";
 *
 * const deployed = deployModularContract({
 *   client,
 *   chain,
 *   account,
 *   core: "ERC721",
 *   params: {
 *     name: "My Modular Contract",
 *   },
 *   modules: [
 *     BatchMetadataERC721.module(),
 *   ],
 * });
 * ```
 * @modules BatchMetadataERC721
 */
function module(params) {
    return async (args) => {
        // deploys if needed
        const moduleContract = await (0, getOrDeployModule_js_1.getOrDeployModule)({
            account: args.account,
            chain: args.chain,
            client: args.client,
            contractId,
            publisher: params?.publisher,
        });
        return {
            module: moduleContract.address,
            data: "0x",
        };
    };
}
/**
 * Installs the BatchMetadataERC721 module on a core contract.
 * @param options
 * @returns the transaction to install the module
 * @example
 * ```ts
 * import { BatchMetadataERC721 } from "thirdweb/modules";
 *
 * const transaction  = BatchMetadataERC721.install({
 *  contract: coreContract,
 *  account: account,
 
 * });
 *
 * await sendTransaction({
 *  transaction,
 *  account,
 * });
 * ```
 * @modules BatchMetadataERC721
 */
function install(options) {
    return (0, installPublishedModule_js_1.installPublishedModule)({
        account: options.account,
        contract: options.contract,
        moduleName: contractId,
        moduleData: "0x",
        publisher: options.params?.publisher,
    });
}
/**
 * Encodes the install data for the BatchMetadataERC721 module.
 * @param params - The parameters for the module.
 * @returns - The encoded data.
 * @modules BatchMetadataERC721
 */
function encodeInstall() {
    return "0x";
}
