// thirdweb/dist/cjs/extensions/erc20/__generated__/IVotes/read/delegates.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isDelegatesSupported = isDelegatesSupported;
exports.encodeDelegatesParams = encodeDelegatesParams;
exports.encodeDelegates = encodeDelegates;
exports.decodeDelegatesResult = decodeDelegatesResult;
exports.delegates = delegates;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x587cde1e";
const FN_INPUTS = [
    {
        type: "address",
        name: "account",
    },
];
const FN_OUTPUTS = [
    {
        type: "address",
    },
];
/**
 * Checks if the `delegates` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `delegates` method is supported.
 * @extension ERC20
 * @example
 * ```ts
 * import { isDelegatesSupported } from "thirdweb/extensions/erc20";
 * const supported = isDelegatesSupported(["0x..."]);
 * ```
 */
function isDelegatesSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "delegates" function.
 * @param options - The options for the delegates function.
 * @returns The encoded ABI parameters.
 * @extension ERC20
 * @example
 * ```ts
 * import { encodeDelegatesParams } from "thirdweb/extensions/erc20";
 * const result = encodeDelegatesParams({
 *  account: ...,
 * });
 * ```
 */
function encodeDelegatesParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.account]);
}
/**
 * Encodes the "delegates" function into a Hex string with its parameters.
 * @param options - The options for the delegates function.
 * @returns The encoded hexadecimal string.
 * @extension ERC20
 * @example
 * ```ts
 * import { encodeDelegates } from "thirdweb/extensions/erc20";
 * const result = encodeDelegates({
 *  account: ...,
 * });
 * ```
 */
function encodeDelegates(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeDelegatesParams(options).slice(2));
}
/**
 * Decodes the result of the delegates function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC20
 * @example
 * ```ts
 * import { decodeDelegatesResult } from "thirdweb/extensions/erc20";
 * const result = decodeDelegatesResultResult("...");
 * ```
 */
function decodeDelegatesResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "delegates" function on the contract.
 * @param options - The options for the delegates function.
 * @returns The parsed result of the function call.
 * @extension ERC20
 * @example
 * ```ts
 * import { delegates } from "thirdweb/extensions/erc20";
 *
 * const result = await delegates({
 *  contract,
 *  account: ...,
 * });
 *
 * ```
 */
async function delegates(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.account],
    });
}
