// thirdweb/dist/cjs/extensions/erc20/__generated__/IVotes/read/getPastTotalSupply.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isGetPastTotalSupplySupported = isGetPastTotalSupplySupported;
exports.encodeGetPastTotalSupplyParams = encodeGetPastTotalSupplyParams;
exports.encodeGetPastTotalSupply = encodeGetPastTotalSupply;
exports.decodeGetPastTotalSupplyResult = decodeGetPastTotalSupplyResult;
exports.getPastTotalSupply = getPastTotalSupply;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x8e539e8c";
const FN_INPUTS = [
    {
        type: "uint256",
        name: "blockNumber",
    },
];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `getPastTotalSupply` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `getPastTotalSupply` method is supported.
 * @extension ERC20
 * @example
 * ```ts
 * import { isGetPastTotalSupplySupported } from "thirdweb/extensions/erc20";
 * const supported = isGetPastTotalSupplySupported(["0x..."]);
 * ```
 */
function isGetPastTotalSupplySupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "getPastTotalSupply" function.
 * @param options - The options for the getPastTotalSupply function.
 * @returns The encoded ABI parameters.
 * @extension ERC20
 * @example
 * ```ts
 * import { encodeGetPastTotalSupplyParams } from "thirdweb/extensions/erc20";
 * const result = encodeGetPastTotalSupplyParams({
 *  blockNumber: ...,
 * });
 * ```
 */
function encodeGetPastTotalSupplyParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.blockNumber]);
}
/**
 * Encodes the "getPastTotalSupply" function into a Hex string with its parameters.
 * @param options - The options for the getPastTotalSupply function.
 * @returns The encoded hexadecimal string.
 * @extension ERC20
 * @example
 * ```ts
 * import { encodeGetPastTotalSupply } from "thirdweb/extensions/erc20";
 * const result = encodeGetPastTotalSupply({
 *  blockNumber: ...,
 * });
 * ```
 */
function encodeGetPastTotalSupply(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeGetPastTotalSupplyParams(options).slice(2));
}
/**
 * Decodes the result of the getPastTotalSupply function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC20
 * @example
 * ```ts
 * import { decodeGetPastTotalSupplyResult } from "thirdweb/extensions/erc20";
 * const result = decodeGetPastTotalSupplyResultResult("...");
 * ```
 */
function decodeGetPastTotalSupplyResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getPastTotalSupply" function on the contract.
 * @param options - The options for the getPastTotalSupply function.
 * @returns The parsed result of the function call.
 * @extension ERC20
 * @example
 * ```ts
 * import { getPastTotalSupply } from "thirdweb/extensions/erc20";
 *
 * const result = await getPastTotalSupply({
 *  contract,
 *  blockNumber: ...,
 * });
 *
 * ```
 */
async function getPastTotalSupply(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.blockNumber],
    });
}
