// thirdweb/dist/cjs/extensions/modules/__generated__/IModularCore/read/getInstalledModules.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isGetInstalledModulesSupported = isGetInstalledModulesSupported;
exports.decodeGetInstalledModulesResult = decodeGetInstalledModulesResult;
exports.getInstalledModules = getInstalledModules;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x3e429396";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "tuple[]",
        components: [
            {
                type: "address",
                name: "implementation",
            },
            {
                type: "tuple",
                name: "config",
                components: [
                    {
                        type: "bool",
                        name: "registerInstallationCallback",
                    },
                    {
                        type: "bytes4[]",
                        name: "requiredInterfaces",
                    },
                    {
                        type: "bytes4[]",
                        name: "supportedInterfaces",
                    },
                    {
                        type: "tuple[]",
                        name: "callbackFunctions",
                        components: [
                            {
                                type: "bytes4",
                                name: "selector",
                            },
                        ],
                    },
                    {
                        type: "tuple[]",
                        name: "fallbackFunctions",
                        components: [
                            {
                                type: "bytes4",
                                name: "selector",
                            },
                            {
                                type: "uint256",
                                name: "permissionBits",
                            },
                        ],
                    },
                ],
            },
        ],
    },
];
/**
 * Checks if the `getInstalledModules` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `getInstalledModules` method is supported.
 * @extension MODULES
 * @example
 * ```ts
 * import { isGetInstalledModulesSupported } from "thirdweb/extensions/modules";
 * const supported = isGetInstalledModulesSupported(["0x..."]);
 * ```
 */
function isGetInstalledModulesSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the getInstalledModules function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension MODULES
 * @example
 * ```ts
 * import { decodeGetInstalledModulesResult } from "thirdweb/extensions/modules";
 * const result = decodeGetInstalledModulesResultResult("...");
 * ```
 */
function decodeGetInstalledModulesResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getInstalledModules" function on the contract.
 * @param options - The options for the getInstalledModules function.
 * @returns The parsed result of the function call.
 * @extension MODULES
 * @example
 * ```ts
 * import { getInstalledModules } from "thirdweb/extensions/modules";
 *
 * const result = await getInstalledModules({
 *  contract,
 * });
 *
 * ```
 */
async function getInstalledModules(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
