// thirdweb/dist/cjs/extensions/modules/__generated__/IModularCore/read/supportsInterface.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isSupportsInterfaceSupported = isSupportsInterfaceSupported;
exports.encodeSupportsInterfaceParams = encodeSupportsInterfaceParams;
exports.encodeSupportsInterface = encodeSupportsInterface;
exports.decodeSupportsInterfaceResult = decodeSupportsInterfaceResult;
exports.supportsInterface = supportsInterface;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x01ffc9a7";
const FN_INPUTS = [
    {
        type: "bytes4",
        name: "interfaceID",
    },
];
const FN_OUTPUTS = [
    {
        type: "bool",
    },
];
/**
 * Checks if the `supportsInterface` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `supportsInterface` method is supported.
 * @extension MODULES
 * @example
 * ```ts
 * import { isSupportsInterfaceSupported } from "thirdweb/extensions/modules";
 * const supported = isSupportsInterfaceSupported(["0x..."]);
 * ```
 */
function isSupportsInterfaceSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "supportsInterface" function.
 * @param options - The options for the supportsInterface function.
 * @returns The encoded ABI parameters.
 * @extension MODULES
 * @example
 * ```ts
 * import { encodeSupportsInterfaceParams } from "thirdweb/extensions/modules";
 * const result = encodeSupportsInterfaceParams({
 *  interfaceID: ...,
 * });
 * ```
 */
function encodeSupportsInterfaceParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.interfaceID]);
}
/**
 * Encodes the "supportsInterface" function into a Hex string with its parameters.
 * @param options - The options for the supportsInterface function.
 * @returns The encoded hexadecimal string.
 * @extension MODULES
 * @example
 * ```ts
 * import { encodeSupportsInterface } from "thirdweb/extensions/modules";
 * const result = encodeSupportsInterface({
 *  interfaceID: ...,
 * });
 * ```
 */
function encodeSupportsInterface(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeSupportsInterfaceParams(options).slice(2));
}
/**
 * Decodes the result of the supportsInterface function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension MODULES
 * @example
 * ```ts
 * import { decodeSupportsInterfaceResult } from "thirdweb/extensions/modules";
 * const result = decodeSupportsInterfaceResultResult("...");
 * ```
 */
function decodeSupportsInterfaceResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "supportsInterface" function on the contract.
 * @param options - The options for the supportsInterface function.
 * @returns The parsed result of the function call.
 * @extension MODULES
 * @example
 * ```ts
 * import { supportsInterface } from "thirdweb/extensions/modules";
 *
 * const result = await supportsInterface({
 *  contract,
 *  interfaceID: ...,
 * });
 *
 * ```
 */
async function supportsInterface(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.interfaceID],
    });
}
