// thirdweb/dist/cjs/wallets/in-app/web/lib/web-connector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InAppWebConnector = void 0;
const domains_js_1 = require("../../../../utils/domains.js");
const webStorage_js_1 = require("../../../../utils/storage/webStorage.js");
const guest_js_1 = require("../../core/authentication/guest.js");
const passkeys_js_1 = require("../../core/authentication/passkeys.js");
const siwe_js_1 = require("../../core/authentication/siwe.js");
const types_js_1 = require("../../core/authentication/types.js");
const get_enclave_user_status_js_1 = require("../lib/actions/get-enclave-user-status.js");
const InAppWalletIframeCommunicator_js_1 = require("../utils/iFrameCommunication/InAppWalletIframeCommunicator.js");
const iframe_auth_js_1 = require("./auth/iframe-auth.js");
const oauth_js_1 = require("./auth/oauth.js");
const otp_js_1 = require("./auth/otp.js");
const enclave_wallet_js_1 = require("./enclave-wallet.js");
const get_auth_token_js_1 = require("./get-auth-token.js");
const iframe_wallet_js_1 = require("./iframe-wallet.js");
/**
 * @internal
 */
class InAppWebConnector {
    isClientIdLegacyPaper(clientId) {
        if (clientId.indexOf("-") > 0 && clientId.length === 36) {
            return true;
        }
        return false;
    }
    /**
     * @example
     * `const thirdwebInAppWallet = new InAppWalletSdk({ clientId: "", chain: "Goerli" });`
     * @internal
     */
    constructor({ client, onAuthSuccess, ecosystem, passkeyDomain, }) {
        Object.defineProperty(this, "client", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "ecosystem", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "querier", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "wallet", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * Used to manage the Auth state of the user.
         */
        Object.defineProperty(this, "auth", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "passkeyDomain", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        if (this.isClientIdLegacyPaper(client.clientId)) {
            throw new Error("You are using a legacy clientId. Please use the clientId found on the thirdweb dashboard settings page");
        }
        const baseUrl = (0, domains_js_1.getThirdwebBaseUrl)("inAppWallet");
        this.client = client;
        this.ecosystem = ecosystem;
        this.passkeyDomain = passkeyDomain;
        this.querier = new InAppWalletIframeCommunicator_js_1.InAppWalletIframeCommunicator({
            clientId: client.clientId,
            ecosystem,
            baseUrl,
        });
        this.auth = new iframe_auth_js_1.Auth({
            client,
            querier: this.querier,
            baseUrl,
            ecosystem,
            onAuthSuccess: async (authResult) => {
                onAuthSuccess?.(authResult);
                if (this.ecosystem &&
                    authResult.storedToken.authDetails.walletType === "sharded") {
                    // If this is an existing sharded ecosystem wallet, we'll need to migrate
                    const result = await this.querier.call({
                        procedureName: "migrateFromShardToEnclave",
                        params: {
                            storedToken: authResult.storedToken,
                        },
                    });
                    if (!result) {
                        throw new Error("Failed to migrate from sharded to enclave wallet");
                    }
                }
                await this.initializeWallet(authResult.storedToken.cookieString);
                if (!this.wallet) {
                    throw new Error("Failed to initialize wallet");
                }
                await this.wallet.postWalletSetUp({
                    ...authResult.walletDetails,
                    authToken: authResult.storedToken.cookieString,
                    walletUserId: authResult.storedToken.authDetails.userWalletId,
                });
                if (authResult.storedToken.authDetails.walletType !== "enclave") {
                    await this.querier.call({
                        procedureName: "initIframe",
                        params: {
                            partnerId: ecosystem?.partnerId,
                            ecosystemId: ecosystem?.id,
                            clientId: this.client.clientId,
                            // For enclave wallets we won't have a device share
                            deviceShareStored: "deviceShareStored" in authResult.walletDetails
                                ? authResult.walletDetails.deviceShareStored
                                : null,
                            walletUserId: authResult.storedToken.authDetails.userWalletId,
                            authCookie: authResult.storedToken.cookieString,
                        },
                    });
                }
                return {
                    user: {
                        status: types_js_1.UserWalletStatus.LOGGED_IN_WALLET_INITIALIZED,
                        authDetails: authResult.storedToken.authDetails,
                        account: await this.wallet.getAccount(),
                        walletAddress: authResult.walletDetails.walletAddress,
                    },
                };
            },
        });
    }
    async initializeWallet(authToken) {
        const storedAuthToken = await (0, get_auth_token_js_1.getAuthToken)(this.client, this.ecosystem);
        if (!authToken && storedAuthToken === null) {
            throw new Error("No auth token provided and no stored auth token found to initialize the wallet");
        }
        const user = await (0, get_enclave_user_status_js_1.getUserStatus)({
            authToken: authToken || storedAuthToken,
            client: this.client,
            ecosystem: this.ecosystem,
        });
        if (!user) {
            throw new Error("Cannot initialize wallet, no user logged in");
        }
        if (user.wallets.length === 0) {
            throw new Error("Cannot initialize wallet, this user does not have a wallet generated yet");
        }
        if (user.wallets[0].type === "enclave") {
            this.wallet = new enclave_wallet_js_1.EnclaveWallet({
                client: this.client,
                ecosystem: this.ecosystem,
                address: user.wallets[0].address,
            });
            return;
        }
        this.wallet = new iframe_wallet_js_1.IFrameWallet({
            client: this.client,
            ecosystem: this.ecosystem,
            querier: this.querier,
        });
    }
    /**
     * Gets the user if they're logged in
     * @example
     * ```js
     *  const user = await thirdwebInAppWallet.getUser();
     *  switch (user.status) {
     *     case UserWalletStatus.LOGGED_OUT: {
     *       // User is logged out, call one of the auth methods on thirdwebInAppWallet.auth to authenticate the user
     *       break;
     *     }
     *     case UserWalletStatus.LOGGED_IN_WALLET_INITIALIZED: {
     *       // user is logged in and wallet is all set up.
     *       // You have access to:
     *       user.status;
     *       user.authDetails;
     *       user.walletAddress;
     *       user.wallet;
     *       break;
     *     }
     * }
     * ```
     * @returns GetUser - an object to containing various information on the user statuses
     */
    async getUser() {
        // If we don't have a wallet yet we'll create one
        if (!this.wallet) {
            const maybeAuthToken = await (0, get_auth_token_js_1.getAuthToken)(this.client, this.ecosystem);
            if (!maybeAuthToken) {
                return { status: types_js_1.UserWalletStatus.LOGGED_OUT };
            }
            await this.initializeWallet(maybeAuthToken);
        }
        if (!this.wallet) {
            throw new Error("Wallet not initialized");
        }
        return await this.wallet.getUserWalletStatus();
    }
    getAccount() {
        if (!this.wallet) {
            throw new Error("Wallet not initialized");
        }
        return this.wallet.getAccount();
    }
    async preAuthenticate(args) {
        return (0, otp_js_1.sendOtp)({
            ...args,
            client: this.client,
            ecosystem: this.ecosystem,
        });
    }
    authenticateWithRedirect(strategy, mode, redirectUrl) {
        (0, oauth_js_1.loginWithOauthRedirect)({
            authOption: strategy,
            client: this.client,
            ecosystem: this.ecosystem,
            redirectUrl,
            mode,
        });
    }
    async loginWithAuthToken(authResult) {
        return this.auth.loginWithAuthToken(authResult);
    }
    /**
     * Authenticates the user and returns the auth token, but does not instantiate their wallet
     */
    async authenticate(args) {
        const strategy = args.strategy;
        switch (strategy) {
            case "email":
                return (0, otp_js_1.verifyOtp)({
                    ...args,
                    client: this.client,
                    ecosystem: this.ecosystem,
                });
            case "phone":
                return (0, otp_js_1.verifyOtp)({
                    ...args,
                    client: this.client,
                    ecosystem: this.ecosystem,
                });
            case "jwt":
                return this.auth.authenticateWithCustomJwt({
                    jwt: args.jwt,
                    encryptionKey: args.encryptionKey,
                });
            case "passkey": {
                return this.passkeyAuth(args);
            }
            case "auth_endpoint": {
                return this.auth.authenticateWithCustomAuthEndpoint({
                    payload: args.payload,
                    encryptionKey: args.encryptionKey,
                });
            }
            case "iframe_email_verification": {
                return this.auth.authenticateWithIframe({
                    email: args.email,
                });
            }
            case "iframe": {
                return this.auth.authenticateWithModal();
            }
            case "apple":
            case "facebook":
            case "google":
            case "telegram":
            case "farcaster":
            case "line":
            case "x":
            case "coinbase":
            case "discord": {
                return (0, oauth_js_1.loginWithOauth)({
                    authOption: strategy,
                    client: this.client,
                    ecosystem: this.ecosystem,
                    closeOpenedWindow: args.closeOpenedWindow,
                    openedWindow: args.openedWindow,
                });
            }
            case "guest": {
                return (0, guest_js_1.guestAuthenticate)({
                    client: this.client,
                    ecosystem: this.ecosystem,
                    storage: webStorage_js_1.webLocalStorage,
                });
            }
            case "wallet": {
                return (0, siwe_js_1.siweAuthenticate)({
                    ecosystem: this.ecosystem,
                    client: this.client,
                    wallet: args.wallet,
                    chain: args.chain,
                });
            }
        }
    }
    /**
     * Authenticates the user then instantiates their wallet using the resulting auth token
     */
    async connect(args) {
        const strategy = args.strategy;
        switch (strategy) {
            case "jwt": {
                return this.auth.loginWithCustomJwt({
                    jwt: args.jwt,
                    encryptionKey: args.encryptionKey,
                });
            }
            case "auth_endpoint": {
                return this.auth.loginWithCustomAuthEndpoint({
                    payload: args.payload,
                    encryptionKey: args.encryptionKey,
                });
            }
            case "iframe_email_verification": {
                return this.auth.loginWithIframe({
                    email: args.email,
                });
            }
            case "iframe": {
                return this.auth.loginWithModal();
            }
            case "passkey": {
                const authToken = await this.passkeyAuth(args);
                return this.loginWithAuthToken(authToken);
            }
            case "phone":
            case "email":
            case "wallet":
            case "apple":
            case "facebook":
            case "google":
            case "farcaster":
            case "telegram":
            case "line":
            case "x":
            case "guest":
            case "coinbase":
            case "discord": {
                const authToken = await this.authenticate(args);
                return await this.auth.loginWithAuthToken(authToken);
            }
            default:
                assertUnreachable(strategy);
        }
    }
    async logout() {
        return await this.auth.logout();
    }
    async passkeyAuth(args) {
        const { PasskeyWebClient } = await Promise.resolve().then(() => require("./auth/passkeys.js"));
        const passkeyClient = new PasskeyWebClient();
        const storage = webStorage_js_1.webLocalStorage;
        if (args.type === "sign-up") {
            return (0, passkeys_js_1.registerPasskey)({
                client: this.client,
                ecosystem: this.ecosystem,
                username: args.passkeyName,
                passkeyClient,
                storage,
                rp: {
                    id: this.passkeyDomain ?? window.location.hostname,
                    name: this.passkeyDomain ?? window.document.title,
                },
            });
        }
        return (0, passkeys_js_1.loginWithPasskey)({
            client: this.client,
            ecosystem: this.ecosystem,
            passkeyClient,
            storage,
            rp: {
                id: this.passkeyDomain ?? window.location.hostname,
                name: this.passkeyDomain ?? window.document.title,
            },
        });
    }
}
exports.InAppWebConnector = InAppWebConnector;
function assertUnreachable(x, message) {
    throw new Error(message ?? `Invalid param: ${x}`);
}
