// thirdweb/dist/cjs/extensions/erc4626/__generated__/IERC4626/read/asset.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isAssetSupported = isAssetSupported;
exports.decodeAssetResult = decodeAssetResult;
exports.asset = asset;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x38d52e0f";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "address",
        name: "assetTokenAddress",
    },
];
/**
 * Checks if the `asset` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `asset` method is supported.
 * @extension ERC4626
 * @example
 * ```ts
 * import { isAssetSupported } from "thirdweb/extensions/erc4626";
 * const supported = isAssetSupported(["0x..."]);
 * ```
 */
function isAssetSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the asset function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC4626
 * @example
 * ```ts
 * import { decodeAssetResult } from "thirdweb/extensions/erc4626";
 * const result = decodeAssetResultResult("...");
 * ```
 */
function decodeAssetResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "asset" function on the contract.
 * @param options - The options for the asset function.
 * @returns The parsed result of the function call.
 * @extension ERC4626
 * @example
 * ```ts
 * import { asset } from "thirdweb/extensions/erc4626";
 *
 * const result = await asset({
 *  contract,
 * });
 *
 * ```
 */
async function asset(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
