// thirdweb/dist/cjs/extensions/modules/__generated__/TransferableERC721/read/isTransferEnabled.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isIsTransferEnabledSupported = isIsTransferEnabledSupported;
exports.decodeIsTransferEnabledResult = decodeIsTransferEnabledResult;
exports.isTransferEnabled = isTransferEnabled;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xcca5dcb6";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "bool",
    },
];
/**
 * Checks if the `isTransferEnabled` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `isTransferEnabled` method is supported.
 * @modules TransferableERC721
 * @example
 * ```ts
 * import { TransferableERC721 } from "thirdweb/modules";
 * const supported = TransferableERC721.isIsTransferEnabledSupported(["0x..."]);
 * ```
 */
function isIsTransferEnabledSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the isTransferEnabled function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @modules TransferableERC721
 * @example
 * ```ts
 * import { TransferableERC721 } from "thirdweb/modules";
 * const result = TransferableERC721.decodeIsTransferEnabledResultResult("...");
 * ```
 */
function decodeIsTransferEnabledResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "isTransferEnabled" function on the contract.
 * @param options - The options for the isTransferEnabled function.
 * @returns The parsed result of the function call.
 * @modules TransferableERC721
 * @example
 * ```ts
 * import { TransferableERC721 } from "thirdweb/modules";
 *
 * const result = await TransferableERC721.isTransferEnabled({
 *  contract,
 * });
 *
 * ```
 */
async function isTransferEnabled(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
