// thirdweb/dist/cjs/extensions/modules/__generated__/TransferableERC721/read/isTransferEnabledFor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isIsTransferEnabledForSupported = isIsTransferEnabledForSupported;
exports.encodeIsTransferEnabledForParams = encodeIsTransferEnabledForParams;
exports.encodeIsTransferEnabledFor = encodeIsTransferEnabledFor;
exports.decodeIsTransferEnabledForResult = decodeIsTransferEnabledForResult;
exports.isTransferEnabledFor = isTransferEnabledFor;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x735d0538";
const FN_INPUTS = [
    {
        type: "address",
        name: "target",
    },
];
const FN_OUTPUTS = [
    {
        type: "bool",
    },
];
/**
 * Checks if the `isTransferEnabledFor` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `isTransferEnabledFor` method is supported.
 * @modules TransferableERC721
 * @example
 * ```ts
 * import { TransferableERC721 } from "thirdweb/modules";
 * const supported = TransferableERC721.isIsTransferEnabledForSupported(["0x..."]);
 * ```
 */
function isIsTransferEnabledForSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "isTransferEnabledFor" function.
 * @param options - The options for the isTransferEnabledFor function.
 * @returns The encoded ABI parameters.
 * @modules TransferableERC721
 * @example
 * ```ts
 * import { TransferableERC721 } from "thirdweb/modules";
 * const result = TransferableERC721.encodeIsTransferEnabledForParams({
 *  target: ...,
 * });
 * ```
 */
function encodeIsTransferEnabledForParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.target]);
}
/**
 * Encodes the "isTransferEnabledFor" function into a Hex string with its parameters.
 * @param options - The options for the isTransferEnabledFor function.
 * @returns The encoded hexadecimal string.
 * @modules TransferableERC721
 * @example
 * ```ts
 * import { TransferableERC721 } from "thirdweb/modules";
 * const result = TransferableERC721.encodeIsTransferEnabledFor({
 *  target: ...,
 * });
 * ```
 */
function encodeIsTransferEnabledFor(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeIsTransferEnabledForParams(options).slice(2));
}
/**
 * Decodes the result of the isTransferEnabledFor function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @modules TransferableERC721
 * @example
 * ```ts
 * import { TransferableERC721 } from "thirdweb/modules";
 * const result = TransferableERC721.decodeIsTransferEnabledForResultResult("...");
 * ```
 */
function decodeIsTransferEnabledForResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "isTransferEnabledFor" function on the contract.
 * @param options - The options for the isTransferEnabledFor function.
 * @returns The parsed result of the function call.
 * @modules TransferableERC721
 * @example
 * ```ts
 * import { TransferableERC721 } from "thirdweb/modules";
 *
 * const result = await TransferableERC721.isTransferEnabledFor({
 *  contract,
 *  target: ...,
 * });
 *
 * ```
 */
async function isTransferEnabledFor(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.target],
    });
}
