// thirdweb/dist/cjs/react/core/hooks/pay/useBuyWithCryptoHistory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useBuyWithCryptoHistory = useBuyWithCryptoHistory;
const react_query_1 = require("@tanstack/react-query");
const getHistory_js_1 = require("../../../../pay/buyWithCrypto/getHistory.js");
/**
 * Hook to get the "Buy with crypto" transaction history for a given wallet address.
 *
 * This hook is a React Query wrapper of the [`getBuyWithCryptoHistory`](https://portal.thirdweb.com/references/typescript/v5/getBuyWithCryptoHistory) function.
 * You can also use that function directly
 * @param params - object of type [`BuyWithCryptoHistoryParams`](https://portal.thirdweb.com/references/typescript/v5/BuyWithCryptoHistoryParams)
 * @param queryParams - options to configure the react query
 * @returns A React Query object which contains the data of type [`BuyWithCryptoHistoryData`](https://portal.thirdweb.com/references/typescript/v5/BuyWithCryptoHistoryData)
 * @example
 * ```tsx
 * import { useBuyWithCryptoHistory } from "thirdweb/react";
 *
 * function Component() {
 *  const buyWithCryptoHistory = useBuyWithCryptoHistory(params);
 *  return <div> ... </div>
 * }
 * ```
 * @buyCrypto
 */
function useBuyWithCryptoHistory(params, queryParams) {
    return (0, react_query_1.useQuery)({
        ...queryParams,
        queryKey: ["getBuyWithCryptoHistory", params],
        queryFn: () => {
            if (!params) {
                throw new Error("Swap params are required");
            }
            return (0, getHistory_js_1.getBuyWithCryptoHistory)(params);
        },
        enabled: !!params,
    });
}
