// thirdweb/dist/cjs/react/core/hooks/pay/useBuyWithFiatHistory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useBuyWithFiatHistory = useBuyWithFiatHistory;
const react_query_1 = require("@tanstack/react-query");
const getHistory_js_1 = require("../../../../pay/buyWithFiat/getHistory.js");
/**
 * Hook to get the "Buy with Fiat" transaction history for a given wallet address.
 *
 * This hook is a React Query wrapper of the [`getBuyWithFiatHistory`](https://portal.thirdweb.com/references/typescript/v5/getBuyWithFiatHistory) function.
 * You can also use that function directly
 * @param params - object of type [`BuyWithFiatHistoryParams`](https://portal.thirdweb.com/references/typescript/v5/BuyWithFiatHistoryParams)
 * @param queryParams - options to configure the react query
 * @returns A React Query object which contains the data of type [`BuyWithFiatHistoryData`](https://portal.thirdweb.com/references/typescript/v5/BuyWithFiatHistoryData)
 * @example
 * ```tsx
 * import { useBuyWithFiatHistory } from "thirdweb/react";
 *
 * function Component() {
 *  const historyQuery = useBuyWithFiatHistory(params);
 *  return <div> ... </div>
 * }
 * ```
 * @buyCrypto
 */
function useBuyWithFiatHistory(params, queryParams) {
    return (0, react_query_1.useQuery)({
        ...queryParams,
        queryKey: ["buyWithFiatHistory", params],
        queryFn: () => {
            if (!params) {
                throw new Error("params are required");
            }
            return (0, getHistory_js_1.getBuyWithFiatHistory)(params);
        },
        enabled: !!params,
    });
}
