// thirdweb/dist/cjs/react/core/hooks/pay/useBuyHistory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useBuyHistory = useBuyHistory;
const react_query_1 = require("@tanstack/react-query");
const getBuyHistory_js_1 = require("../../../../pay/getBuyHistory.js");
/**
 * Hook to get the history of Buy transactions for a given wallet - This includes both "buy with crypto" and "buy with fiat" transactions.
 *
 * This hook is a React Query wrapper of the [`getBuyHistory`](https://portal.thirdweb.com/references/typescript/v5/getBuyHistory) function.
 * You can also use that function directly
 *
 * @param params - object of type [`BuyHistoryParams`](https://portal.thirdweb.com/references/typescript/v5/BuyHistoryParams)
 * @param queryParams - options to configure the react query
 * @returns A React Query object which contains the data of type [`BuyHistoryData`](https://portal.thirdweb.com/references/typescript/v5/BuyHistoryData)
 * @example
 * ```tsx
 * import { useBuyHistory } from "thirdweb/react";
 *
 * function Component() {
 *  const buyHistoryQuery = useBuyHistory(params);
 *  return <div> ... </div>
 * }
 * ```
 * @buyCrypto
 */
function useBuyHistory(params, queryParams) {
    return (0, react_query_1.useQuery)({
        ...queryParams,
        queryKey: ["getBuyHistory", params],
        queryFn: () => {
            if (!params) {
                throw new Error("params are required");
            }
            return (0, getBuyHistory_js_1.getBuyHistory)(params);
        },
        enabled: !!params,
    });
}
