// thirdweb/dist/cjs/extensions/modules/__generated__/OwnableRoles/read/hasAnyRole.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isHasAnyRoleSupported = isHasAnyRoleSupported;
exports.encodeHasAnyRoleParams = encodeHasAnyRoleParams;
exports.encodeHasAnyRole = encodeHasAnyRole;
exports.decodeHasAnyRoleResult = decodeHasAnyRoleResult;
exports.hasAnyRole = hasAnyRole;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x514e62fc";
const FN_INPUTS = [
    {
        type: "address",
        name: "user",
    },
    {
        type: "uint256",
        name: "roles",
    },
];
const FN_OUTPUTS = [
    {
        type: "bool",
    },
];
/**
 * Checks if the `hasAnyRole` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `hasAnyRole` method is supported.
 * @extension MODULES
 * @example
 * ```ts
 * import { isHasAnyRoleSupported } from "thirdweb/extensions/modules";
 * const supported = isHasAnyRoleSupported(["0x..."]);
 * ```
 */
function isHasAnyRoleSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "hasAnyRole" function.
 * @param options - The options for the hasAnyRole function.
 * @returns The encoded ABI parameters.
 * @extension MODULES
 * @example
 * ```ts
 * import { encodeHasAnyRoleParams } from "thirdweb/extensions/modules";
 * const result = encodeHasAnyRoleParams({
 *  user: ...,
 *  roles: ...,
 * });
 * ```
 */
function encodeHasAnyRoleParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.user, options.roles]);
}
/**
 * Encodes the "hasAnyRole" function into a Hex string with its parameters.
 * @param options - The options for the hasAnyRole function.
 * @returns The encoded hexadecimal string.
 * @extension MODULES
 * @example
 * ```ts
 * import { encodeHasAnyRole } from "thirdweb/extensions/modules";
 * const result = encodeHasAnyRole({
 *  user: ...,
 *  roles: ...,
 * });
 * ```
 */
function encodeHasAnyRole(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeHasAnyRoleParams(options).slice(2));
}
/**
 * Decodes the result of the hasAnyRole function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension MODULES
 * @example
 * ```ts
 * import { decodeHasAnyRoleResult } from "thirdweb/extensions/modules";
 * const result = decodeHasAnyRoleResultResult("...");
 * ```
 */
function decodeHasAnyRoleResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "hasAnyRole" function on the contract.
 * @param options - The options for the hasAnyRole function.
 * @returns The parsed result of the function call.
 * @extension MODULES
 * @example
 * ```ts
 * import { hasAnyRole } from "thirdweb/extensions/modules";
 *
 * const result = await hasAnyRole({
 *  contract,
 *  user: ...,
 *  roles: ...,
 * });
 *
 * ```
 */
async function hasAnyRole(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.user, options.roles],
    });
}
