// thirdweb/dist/cjs/extensions/modules/__generated__/OwnableRoles/read/rolesOf.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isRolesOfSupported = isRolesOfSupported;
exports.encodeRolesOfParams = encodeRolesOfParams;
exports.encodeRolesOf = encodeRolesOf;
exports.decodeRolesOfResult = decodeRolesOfResult;
exports.rolesOf = rolesOf;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x2de94807";
const FN_INPUTS = [
    {
        type: "address",
        name: "user",
    },
];
const FN_OUTPUTS = [
    {
        type: "uint256",
        name: "roles",
    },
];
/**
 * Checks if the `rolesOf` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `rolesOf` method is supported.
 * @extension MODULES
 * @example
 * ```ts
 * import { isRolesOfSupported } from "thirdweb/extensions/modules";
 * const supported = isRolesOfSupported(["0x..."]);
 * ```
 */
function isRolesOfSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "rolesOf" function.
 * @param options - The options for the rolesOf function.
 * @returns The encoded ABI parameters.
 * @extension MODULES
 * @example
 * ```ts
 * import { encodeRolesOfParams } from "thirdweb/extensions/modules";
 * const result = encodeRolesOfParams({
 *  user: ...,
 * });
 * ```
 */
function encodeRolesOfParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.user]);
}
/**
 * Encodes the "rolesOf" function into a Hex string with its parameters.
 * @param options - The options for the rolesOf function.
 * @returns The encoded hexadecimal string.
 * @extension MODULES
 * @example
 * ```ts
 * import { encodeRolesOf } from "thirdweb/extensions/modules";
 * const result = encodeRolesOf({
 *  user: ...,
 * });
 * ```
 */
function encodeRolesOf(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeRolesOfParams(options).slice(2));
}
/**
 * Decodes the result of the rolesOf function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension MODULES
 * @example
 * ```ts
 * import { decodeRolesOfResult } from "thirdweb/extensions/modules";
 * const result = decodeRolesOfResultResult("...");
 * ```
 */
function decodeRolesOfResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "rolesOf" function on the contract.
 * @param options - The options for the rolesOf function.
 * @returns The parsed result of the function call.
 * @extension MODULES
 * @example
 * ```ts
 * import { rolesOf } from "thirdweb/extensions/modules";
 *
 * const result = await rolesOf({
 *  contract,
 *  user: ...,
 * });
 *
 * ```
 */
async function rolesOf(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.user],
    });
}
