// thirdweb/dist/esm/transaction/extract-error.js
import { decodeErrorResult } from "viem";
import { resolveContractAbi } from "../contract/actions/resolve-abi.js";
import { isHex } from "../utils/encoding/hex.js";
import { stringify } from "../utils/json.js";
/**
 * @internal
 */
export async function extractError(args) {
    const { error, contract } = args;
    if (typeof error === "object") {
        // try to parse RPC error
        const errorObj = error;
        if (errorObj.data) {
            if (errorObj.data !== "0x" && isHex(errorObj.data)) {
                let abi = contract?.abi;
                if (contract && !abi) {
                    abi = await resolveContractAbi(contract).catch(() => undefined);
                }
                const parsedError = decodeErrorResult({
                    data: errorObj.data,
                    abi,
                });
                return new TransactionError(`${parsedError.errorName}${parsedError.args ? ` - ${parsedError.args}` : ""}`, contract);
            }
            return new TransactionError(`Execution Reverted: ${stringify(errorObj)}`, contract);
        }
    }
    return error;
}
export const __DEV__ = process.env.NODE_ENV !== "production";
class TransactionError extends Error {
    constructor(reason, contract) {
        let message = reason;
        if (__DEV__ && contract) {
            // show more infor in dev
            message = [
                reason,
                "",
                `contract: ${contract.address}`,
                `chainId: ${contract.chain?.id}`,
            ].join("\n");
        }
        super(message);
        Object.defineProperty(this, "contractAddress", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "chainId", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.name = "TransactionError";
        this.contractAddress = contract?.address;
        this.chainId = contract?.chain?.id;
        this.message = message;
    }
}
