// thirdweb/dist/esm/extensions/modules/common/installPublishedModule.js
import { getOrDeployInfraForPublishedContract } from "../../../contract/deployment/utils/bootstrap.js";
import { installModule } from "../__generated__/IModularCore/write/installModule.js";
/**
 * Install a published module on a modular contract
 * @param options - The options for installing a published module
 * @returns A prepared transaction to send
 * @modules
 * @example
 * ```ts
 * import { installPublishedModule } from "thirdweb/modules";
 * const transaction = installPublishedModule({
 *  client,
 *  chain,
 *  account,
 *  contract,
 *  moduleName: "MyModule",
 *  publisherAddress: "0x...",
 * });
 * await sendTransaction({ transaction, account });
 * ```
 */
export function installPublishedModule(options) {
    const { account, contract, moduleName, constructorParams, publisher, moduleData, } = options;
    return installModule({
        contract,
        asyncParams: async () => {
            const deployedModule = await getOrDeployInfraForPublishedContract({
                chain: contract.chain,
                client: contract.client,
                account,
                contractId: moduleName,
                constructorParams,
                publisher,
            });
            return {
                moduleContract: deployedModule.implementationContract.address,
                data: moduleData || "0x",
            };
        },
    });
}
