// thirdweb/dist/esm/extensions/erc4337/account/removeAdmin.js
import { isSetPermissionsForSignerSupported, setPermissionsForSigner, } from "../__generated__/IAccountPermissions/write/setPermissionsForSigner.js";
import { defaultPermissionsForAdmin, signPermissionRequest } from "./common.js";
/**
 * Removes admin permissions for a specified address.
 * @param options - The options for the removeAdmin function.
 * @returns The transaction object to be sent.
 * @example
 * ```ts
 * import { removeAdmin } from 'thirdweb/extensions/erc4337';
 * import { sendTransaction } from 'thirdweb';
 *
 * const transaction = removeAdmin({
 *  contract,
 *  account,
 *  adminAddress: '0x...'
 * });
 *
 * await sendTransaction({ transaction, account });
 * ```
 * @extension ERC4337
 */
export function removeAdmin(options) {
    const { contract, account, adminAddress } = options;
    return setPermissionsForSigner({
        contract,
        async asyncParams() {
            const { req, signature } = await signPermissionRequest({
                account,
                contract,
                req: await defaultPermissionsForAdmin({
                    target: adminAddress,
                    action: "remove-admin",
                }),
            });
            return {
                signature,
                req,
            };
        },
    });
}
/**
 * Checks if the `isRemoveAdminSupported` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `isRemoveAdminSupported` method is supported.
 * @extension ERC4337
 * @example
 * ```ts
 * import { isRemoveAdminSupported } from "thirdweb/extensions/erc4337";
 *
 * const supported = isRemoveAdminSupported(["0x..."]);
 * ```
 */
export function isRemoveAdminSupported(availableSelectors) {
    return isSetPermissionsForSignerSupported(availableSelectors);
}
