// thirdweb/dist/esm/react/native/ui/connect/ReceiveScreen.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { StyleSheet, View } from "react-native";
import { shortenAddress } from "../../../../utils/address.js";
import { spacing } from "../../design-system/index.js";
import { Address } from "../components/Address.js";
import { Header } from "../components/Header.js";
import { WalletImage } from "../components/WalletImage.js";
import { Spacer } from "../components/spacer.js";
import { ThemedText } from "../components/text.js";
export const ReceiveScreen = (props) => {
    const { wallet, account, theme, onClose, onBack, containerType } = props;
    return (_jsxs(_Fragment, { children: [_jsx(Header, { theme: theme, onClose: onClose, onBack: onBack, containerType: containerType, title: "Receive Funds" }), _jsxs(View, { style: styles.container, children: [_jsx(WalletImage, { theme: theme, wallet: wallet, size: 80 }), _jsx(Spacer, { size: "lg" }), _jsx(View, { style: [
                            styles.addressContainer,
                            { borderColor: theme.colors.borderColor },
                        ], children: _jsx(Address, { account: account, addressOrENS: shortenAddress(account?.address, 8), theme: theme }) }), _jsx(ThemedText, { theme: theme, type: "subtext", style: { textAlign: "center" }, children: "Copy your address to send funds to this wallet" })] })] }));
};
const styles = StyleSheet.create({
    container: {
        paddingHorizontal: spacing.lg,
        paddingVertical: spacing.xxl,
        flexDirection: "column",
        gap: spacing.md,
        alignItems: "center",
        justifyContent: "center",
        flex: 1,
    },
    addressContainer: {
        width: "100%",
        flexDirection: "row",
        gap: spacing.sm,
        alignItems: "center",
        justifyContent: "center",
        borderWidth: 1,
        padding: spacing.md,
        borderRadius: spacing.lg,
    },
});
