// thirdweb/dist/esm/react/native/ui/connect/SendScreen.js
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "react/jsx-runtime";
import { useState } from "react";
import { StyleSheet, View } from "react-native";
import { useChainSymbol } from "../../../core/hooks/others/useChainQuery.js";
import { useActiveAccount } from "../../../core/hooks/wallets/useActiveAccount.js";
import { useActiveWalletChain } from "../../../core/hooks/wallets/useActiveWalletChain.js";
import { useSendToken } from "../../../core/hooks/wallets/useSendToken.js";
import { radius, spacing } from "../../design-system/index.js";
import { Header } from "../components/Header.js";
import { ThemedButton } from "../components/button.js";
import { ThemedInput } from "../components/input.js";
import { Spacer } from "../components/spacer.js";
import { ThemedSpinner } from "../components/spinner.js";
import { ThemedText } from "../components/text.js";
import { ErrorView } from "./ErrorView.js";
import { SuccessView } from "./SuccessView.js";
import { TokenListScreen, TokenRow } from "./TokenListScreen.js";
export const SendScreen = (props) => {
    const { theme, client, onClose, onBack, containerType, supportedTokens } = props;
    const [receiverAddress, setReceiverAddress] = useState("");
    const [amount, setAmount] = useState("0");
    const [selectedToken, setSelectedToken] = useState();
    const account = useActiveAccount();
    const chain = useActiveWalletChain();
    const { symbol } = useChainSymbol(chain);
    const sendMutation = useSendToken(client);
    const [screen, setScreen] = useState("base");
    const handleTokenClicked = () => {
        setScreen("tokenList");
    };
    const handleSend = async () => {
        sendMutation.mutate({
            amount,
            receiverAddress,
            tokenAddress: selectedToken?.address,
        }, {
            onSuccess() {
                setScreen("success");
            },
            onError() {
                setScreen("error");
            },
        });
    };
    if (screen === "success") {
        return (_jsxs(_Fragment, { children: [_jsx(Header, { theme: theme, onClose: onClose, onBack: () => setScreen("base"), containerType: containerType, title: "Funds Sent" }), _jsx(Spacer, { size: "xl" }), _jsx(View, { style: { flex: 1 }, children: _jsx(SuccessView, { theme: theme, title: "Transaction Successful" }) })] }));
    }
    if (screen === "error") {
        return (_jsxs(_Fragment, { children: [_jsx(Header, { theme: theme, onClose: onClose, onBack: () => setScreen("base"), containerType: containerType, title: "Error Sending Funds" }), _jsx(Spacer, { size: "xl" }), _jsx(View, { style: { flex: 1 }, children: _jsx(ErrorView, { theme: theme, title: sendMutation.error?.message || "Unknown error" }) })] }));
    }
    if (screen === "tokenList") {
        return (_jsxs(_Fragment, { children: [_jsx(Header, { theme: theme, onClose: onClose, onBack: () => setScreen("base"), containerType: containerType, title: "Token to Send" }), _jsx(Spacer, { size: "xl" }), _jsx(View, { style: { flex: 1 }, children: _jsx(TokenListScreen, { theme: theme, client: client, supportedTokens: supportedTokens, onTokenSelected: (t) => {
                            setSelectedToken(t);
                            setScreen("base");
                        } }) })] }));
    }
    return (_jsxs(_Fragment, { children: [_jsx(Header, { theme: theme, onClose: onClose, onBack: onBack, containerType: containerType, title: "Send Funds" }), _jsxs(View, { style: styles.container, children: [_jsxs(View, { style: styles.inputContainer, children: [_jsx(ThemedText, { theme: theme, type: "subtext", children: "Token" }), _jsx(View, { style: {
                                    borderColor: theme.colors.borderColor,
                                    borderWidth: 1,
                                    paddingHorizontal: spacing.md,
                                    paddingVertical: spacing.smd,
                                    borderRadius: radius.lg,
                                }, children: _jsx(TokenRow, { theme: theme, client: client, address: account?.address, chain: chain, token: selectedToken, onTokenSelected: handleTokenClicked }) })] }), _jsxs(View, { style: styles.inputContainer, children: [_jsx(ThemedText, { theme: theme, type: "subtext", children: "Send to" }), _jsx(ThemedInput, { theme: theme, onChangeText: setReceiverAddress, value: receiverAddress, placeholder: "0x... / ENS name", textContentType: "URL", keyboardType: "url", autoCapitalize: "none" })] }), _jsxs(View, { style: styles.inputContainer, children: [_jsx(ThemedText, { theme: theme, type: "subtext", children: "Amount" }), _jsx(ThemedInput, { theme: theme, onChangeText: setAmount, value: amount, inputMode: "numeric", rightView: _jsx(ThemedText, { theme: theme, type: "subtext", style: { marginRight: spacing.md }, children: selectedToken?.symbol || symbol }) })] }), _jsx(View, { style: styles.inputContainer, children: _jsx(ThemedButton, { theme: theme, variant: "accent", onPress: handleSend, disabled: sendMutation.isPending, children: sendMutation.isPending ? (_jsx(ThemedSpinner, { color: theme.colors.accentButtonText })) : (_jsx(ThemedText, { theme: theme, type: "defaultSemiBold", style: {
                                    color: theme.colors.accentButtonText,
                                }, children: "Send" })) }) })] })] }));
};
const styles = StyleSheet.create({
    container: {
        paddingHorizontal: spacing.lg,
        paddingVertical: spacing.lg,
        flexDirection: "column",
        gap: spacing.md,
    },
    inputContainer: {
        gap: spacing.sm,
    },
});
