// thirdweb/dist/esm/react/native/ui/connect/TokenListScreen.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { ScrollView, StyleSheet, TouchableOpacity, View } from "react-native";
import { formatNumber } from "../../../../utils/formatNumber.js";
import { useWalletBalance } from "../../../core/hooks/others/useWalletBalance.js";
import { useActiveAccount } from "../../../core/hooks/wallets/useActiveAccount.js";
import { useActiveWalletChain } from "../../../core/hooks/wallets/useActiveWalletChain.js";
import { defaultTokens, } from "../../../core/utils/defaultTokens.js";
import { spacing } from "../../design-system/index.js";
import { RNImage } from "../components/RNImage.js";
import { Skeleton } from "../components/Skeleton.js";
import { TokenIcon } from "../components/TokenIcon.js";
import { ThemedText } from "../components/text.js";
import { RIGHT_CHEVRON } from "../icons/svgs.js";
export const TokenListScreen = (props) => {
    const supportedTokens = props.supportedTokens || defaultTokens;
    const chain = useActiveWalletChain();
    const account = useActiveAccount();
    const tokens = chain ? supportedTokens[chain.id] || [] : [];
    return (_jsx(ScrollView, { style: { flex: 1 }, children: _jsxs(View, { style: styles.listContainer, children: [_jsx(TokenRow, { chain: chain, address: account?.address, client: props.client, theme: props.theme, onTokenSelected: props.onTokenSelected }), tokens.map((t) => {
                    return (_jsx(TokenRow, { chain: chain, client: props.client, address: account?.address, theme: props.theme, token: t, onTokenSelected: props.onTokenSelected }, t.address));
                })] }) }));
};
export const TokenRow = (props) => {
    const { token, theme, address, chain, client, onTokenSelected } = props;
    const balanceQuery = useWalletBalance({
        address,
        chain,
        client,
        tokenAddress: token?.address,
    });
    const tokenName = props.token ? props.token.name : balanceQuery.data?.name;
    const inner = (_jsxs(_Fragment, { children: [_jsx(TokenIcon, { token: token, size: 40, chain: chain, client: client, theme: theme }), _jsxs(View, { style: { flexDirection: "column", gap: spacing.xxs }, children: [_jsx(ThemedText, { theme: theme, type: "defaultSemiBold", children: tokenName }), address && (_jsx(_Fragment, { children: balanceQuery.data ? (_jsxs(ThemedText, { theme: theme, type: "subtext", children: [formatBalanceOnButton(Number(balanceQuery.data.displayValue)), " ", balanceQuery.data?.symbol] })) : (_jsx(Skeleton, { theme: theme, style: { width: 80, height: 14 } })) }))] }), props.onTokenSelected && (_jsxs(_Fragment, { children: [_jsx(View, { style: { flex: 1 } }), _jsx(RNImage, { theme: theme, size: 24, data: RIGHT_CHEVRON, color: theme.colors.secondaryIconColor })] }))] }));
    return onTokenSelected ? (_jsx(TouchableOpacity, { style: styles.tokenRowContainer, onPress: () => onTokenSelected(token), children: inner })) : (_jsx(View, { style: styles.tokenRowContainer, children: inner }));
};
function formatBalanceOnButton(num) {
    return formatNumber(num, num < 1 ? 5 : 4);
}
const styles = StyleSheet.create({
    listContainer: {
        flexDirection: "column",
        flex: 1,
        paddingHorizontal: spacing.lg,
        gap: spacing.md,
    },
    tokenRowContainer: {
        flexDirection: "row",
        gap: spacing.md,
        alignItems: "center",
    },
    emptyContainer: {
        flexDirection: "column",
        flex: 1,
        justifyContent: "center",
        alignItems: "center",
        padding: spacing.lg,
    },
});
