// thirdweb/dist/esm/react/native/ui/connect/WalletLoadingThumbnail.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useEffect, useRef } from "react";
import { Animated, Easing, StyleSheet, View } from "react-native";
import { Rect, Svg } from "react-native-svg";
import { radius } from "../../design-system/index.js";
const AnimatedRect = Animated.createAnimatedComponent(Rect);
const PADDING = 10;
const INTERNAL_PADDING = 5;
function WalletLoadingThumbnail({ theme, children, showError, imageSize, animate, roundLoader, }) {
    const spinValue = useRef(new Animated.Value(0));
    useEffect(() => {
        if (!animate)
            return;
        const animation = Animated.timing(spinValue.current, {
            toValue: 1,
            duration: 1150,
            useNativeDriver: true,
            easing: Easing.linear,
        });
        const loop = Animated.loop(animation);
        loop.start();
        return () => {
            loop.stop();
        };
    }, [animate]);
    const spin = spinValue.current.interpolate({
        inputRange: [0, 1],
        outputRange: [0, -400],
    });
    const rx = roundLoader ? imageSize / 2 : 15;
    return (_jsxs(View, { style: styles.container, children: [_jsx(Svg, { width: imageSize + PADDING, height: imageSize + PADDING, viewBox: `0 0 ${imageSize + PADDING} ${imageSize + PADDING}`, style: styles.loader, children: animate && (_jsx(AnimatedRect, { x: "2", y: "2", width: imageSize + INTERNAL_PADDING, height: imageSize + INTERNAL_PADDING, rx: rx, stroke: showError ? "transparent" : theme.colors.accentText, strokeWidth: 3, fill: "transparent", strokeDasharray: "100 300", strokeDashoffset: spin })) }), showError && (_jsx(View, { style: [
                    {
                        position: "absolute",
                        borderWidth: 3,
                        height: imageSize + INTERNAL_PADDING,
                        width: imageSize + INTERNAL_PADDING,
                        borderRadius: radius.lg,
                        borderColor: theme.colors.borderColor,
                    },
                ] })), children] }));
}
const styles = StyleSheet.create({
    container: {
        alignItems: "center",
        justifyContent: "center",
    },
    loader: {
        position: "absolute",
    },
});
export default WalletLoadingThumbnail;
