// thirdweb/dist/cjs/extensions/erc4337/__generated__/IEntryPoint/read/getDepositInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isGetDepositInfoSupported = isGetDepositInfoSupported;
exports.encodeGetDepositInfoParams = encodeGetDepositInfoParams;
exports.encodeGetDepositInfo = encodeGetDepositInfo;
exports.decodeGetDepositInfoResult = decodeGetDepositInfoResult;
exports.getDepositInfo = getDepositInfo;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x5287ce12";
const FN_INPUTS = [
    {
        type: "address",
        name: "account",
    },
];
const FN_OUTPUTS = [
    {
        type: "tuple",
        name: "info",
        components: [
            {
                type: "uint112",
                name: "deposit",
            },
            {
                type: "bool",
                name: "staked",
            },
            {
                type: "uint112",
                name: "stake",
            },
            {
                type: "uint32",
                name: "unstakeDelaySec",
            },
            {
                type: "uint48",
                name: "withdrawTime",
            },
        ],
    },
];
/**
 * Checks if the `getDepositInfo` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `getDepositInfo` method is supported.
 * @extension ERC4337
 * @example
 * ```ts
 * import { isGetDepositInfoSupported } from "thirdweb/extensions/erc4337";
 * const supported = isGetDepositInfoSupported(["0x..."]);
 * ```
 */
function isGetDepositInfoSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "getDepositInfo" function.
 * @param options - The options for the getDepositInfo function.
 * @returns The encoded ABI parameters.
 * @extension ERC4337
 * @example
 * ```ts
 * import { encodeGetDepositInfoParams } from "thirdweb/extensions/erc4337";
 * const result = encodeGetDepositInfoParams({
 *  account: ...,
 * });
 * ```
 */
function encodeGetDepositInfoParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.account]);
}
/**
 * Encodes the "getDepositInfo" function into a Hex string with its parameters.
 * @param options - The options for the getDepositInfo function.
 * @returns The encoded hexadecimal string.
 * @extension ERC4337
 * @example
 * ```ts
 * import { encodeGetDepositInfo } from "thirdweb/extensions/erc4337";
 * const result = encodeGetDepositInfo({
 *  account: ...,
 * });
 * ```
 */
function encodeGetDepositInfo(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeGetDepositInfoParams(options).slice(2));
}
/**
 * Decodes the result of the getDepositInfo function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC4337
 * @example
 * ```ts
 * import { decodeGetDepositInfoResult } from "thirdweb/extensions/erc4337";
 * const result = decodeGetDepositInfoResultResult("...");
 * ```
 */
function decodeGetDepositInfoResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getDepositInfo" function on the contract.
 * @param options - The options for the getDepositInfo function.
 * @returns The parsed result of the function call.
 * @extension ERC4337
 * @example
 * ```ts
 * import { getDepositInfo } from "thirdweb/extensions/erc4337";
 *
 * const result = await getDepositInfo({
 *  contract,
 *  account: ...,
 * });
 *
 * ```
 */
async function getDepositInfo(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.account],
    });
}
