// thirdweb/dist/esm/extensions/erc4337/__generated__/IAccountPermissions/read/getAllAdmins.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0xe9523c97";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "address[]",
        name: "admins",
    },
];
/**
 * Checks if the `getAllAdmins` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `getAllAdmins` method is supported.
 * @extension ERC4337
 * @example
 * ```ts
 * import { isGetAllAdminsSupported } from "thirdweb/extensions/erc4337";
 * const supported = isGetAllAdminsSupported(["0x..."]);
 * ```
 */
export function isGetAllAdminsSupported(availableSelectors) {
    return detectMethod({
        availableSelectors,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the getAllAdmins function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC4337
 * @example
 * ```ts
 * import { decodeGetAllAdminsResult } from "thirdweb/extensions/erc4337";
 * const result = decodeGetAllAdminsResultResult("...");
 * ```
 */
export function decodeGetAllAdminsResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getAllAdmins" function on the contract.
 * @param options - The options for the getAllAdmins function.
 * @returns The parsed result of the function call.
 * @extension ERC4337
 * @example
 * ```ts
 * import { getAllAdmins } from "thirdweb/extensions/erc4337";
 *
 * const result = await getAllAdmins({
 *  contract,
 * });
 *
 * ```
 */
export async function getAllAdmins(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
