// thirdweb/dist/esm/extensions/multicall3/__generated__/IMulticall3/read/getBlockNumber.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0x42cbb15c";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "uint256",
        name: "blockNumber",
    },
];
/**
 * Checks if the `getBlockNumber` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `getBlockNumber` method is supported.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { isGetBlockNumberSupported } from "thirdweb/extensions/multicall3";
 * const supported = isGetBlockNumberSupported(["0x..."]);
 * ```
 */
export function isGetBlockNumberSupported(availableSelectors) {
    return detectMethod({
        availableSelectors,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the getBlockNumber function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { decodeGetBlockNumberResult } from "thirdweb/extensions/multicall3";
 * const result = decodeGetBlockNumberResultResult("...");
 * ```
 */
export function decodeGetBlockNumberResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getBlockNumber" function on the contract.
 * @param options - The options for the getBlockNumber function.
 * @returns The parsed result of the function call.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { getBlockNumber } from "thirdweb/extensions/multicall3";
 *
 * const result = await getBlockNumber({
 *  contract,
 * });
 *
 * ```
 */
export async function getBlockNumber(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
