// thirdweb/dist/esm/extensions/modules/__generated__/TransferableERC20/module/install.js
import { getOrDeployModule } from "../../../common/getOrDeployModule.js";
import { installPublishedModule } from "../../../common/installPublishedModule.js";
const contractId = "TransferableERC20";
/**
 * Convenience function to add the TransferableERC20 module as a default module on a core contract.
 * @param params - The parameters for the module.
 * @returns - The module function.
 * @example
 * ```ts
 * import { TransferableERC20, deployModularContract } from "thirdweb/modules";
 *
 * const deployed = deployModularContract({
 *   client,
 *   chain,
 *   account,
 *   core: "ERC20",
 *   params: {
 *     name: "My Modular Contract",
 *   },
 *   modules: [
 *     TransferableERC20.module(),
 *   ],
 * });
 * ```
 * @modules TransferableERC20
 */
export function module(params) {
    return async (args) => {
        // deploys if needed
        const moduleContract = await getOrDeployModule({
            account: args.account,
            chain: args.chain,
            client: args.client,
            contractId,
            publisher: params?.publisher,
        });
        return {
            module: moduleContract.address,
            data: "0x",
        };
    };
}
/**
 * Installs the TransferableERC20 module on a core contract.
 * @param options
 * @returns the transaction to install the module
 * @example
 * ```ts
 * import { TransferableERC20 } from "thirdweb/modules";
 *
 * const transaction  = TransferableERC20.install({
 *  contract: coreContract,
 *  account: account,
 
 * });
 *
 * await sendTransaction({
 *  transaction,
 *  account,
 * });
 * ```
 * @modules TransferableERC20
 */
export function install(options) {
    return installPublishedModule({
        account: options.account,
        contract: options.contract,
        moduleName: contractId,
        moduleData: "0x",
        publisher: options.params?.publisher,
    });
}
/**
 * Encodes the install data for the TransferableERC20 module.
 * @param params - The parameters for the module.
 * @returns - The encoded data.
 * @modules TransferableERC20
 */
export function encodeInstall() {
    return "0x";
}
