// thirdweb/dist/esm/stories/ConnectButton/hideButtons.stories.js
import { ConnectButton } from "../../react/web/ui/ConnectWallet/ConnectButton.js";
import { storyClient } from "../utils.js";
const meta = {
    title: "Connect/ConnectButton/hide buttons",
    component: ConnectButton,
    parameters: {
        layout: "centered",
    },
    args: {
        client: storyClient,
    },
};
export const Default = {
    args: {},
};
export const HideSend = {
    args: {
        detailsModal: {
            hideSendFunds: true,
        },
    },
};
export const HideReceive = {
    args: {
        detailsModal: {
            hideReceiveFunds: true,
        },
    },
};
export const HideBuy = {
    args: {
        detailsModal: {
            hideBuyFunds: true,
        },
    },
};
export const HideSendReceive = {
    args: {
        detailsModal: {
            hideSendFunds: true,
            hideReceiveFunds: true,
        },
    },
};
export const HideSendBuy = {
    args: {
        detailsModal: {
            hideSendFunds: true,
            hideBuyFunds: true,
        },
    },
};
export const HideReceiveBuy = {
    args: {
        detailsModal: {
            hideReceiveFunds: true,
            hideBuyFunds: true,
        },
    },
};
export const HideSendReceiveBuy = {
    args: {
        detailsModal: {
            hideSendFunds: true,
            hideReceiveFunds: true,
            hideBuyFunds: true,
        },
    },
};
export default meta;
